// @farris/ui-lookup/esm5/lib/quick-select/quick-select-panel.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, EventEmitter, HostListener, Injector, Input, Output } from '@angular/core';
var LookupQuickSelectPanelComponent = /** @class */ (function () {
    function LookupQuickSelectPanelComponent(injector, cdr) {
        this.injector = injector;
        this.cdr = cdr;
        this.showMore = true;
        this.data = [];
        this.textField = '';
        this.moreClcik = new EventEmitter();
        this.itemClick = new EventEmitter();
        this.activeIndex = -1;
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    LookupQuickSelectPanelComponent.prototype.onClick = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        $event.stopPropagation();
    };
    /**
     * @return {?}
     */
    LookupQuickSelectPanelComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} items
     * @return {?}
     */
    LookupQuickSelectPanelComponent.prototype.loadData = /**
     * @param {?} items
     * @return {?}
     */
    function (items) {
        this.data = items;
        if (!this.cdr['destroyed']) {
            this.cdr.detectChanges();
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    LookupQuickSelectPanelComponent.prototype.onMoreClick = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        $event.stopPropagation();
        this.moreClcik.emit($event);
    };
    /**
     * @param {?} rowObj
     * @return {?}
     */
    LookupQuickSelectPanelComponent.prototype.formatData = /**
     * @param {?} rowObj
     * @return {?}
     */
    function (rowObj) {
        return this.formatter(rowObj);
    };
    /**
     * @param {?} $event
     * @param {?} item
     * @return {?}
     */
    LookupQuickSelectPanelComponent.prototype.onItemClick = /**
     * @param {?} $event
     * @param {?} item
     * @return {?}
     */
    function ($event, item) {
        this.itemClick.emit({ data: item, evnet: $event });
    };
    /**
     * @param {?} index
     * @return {?}
     */
    LookupQuickSelectPanelComponent.prototype.setActiveItem = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        this.activeIndex = index;
        if (!this.cdr['destroyed']) {
            this.cdr.detectChanges();
        }
    };
    LookupQuickSelectPanelComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lookup-quick-select-panel',
                    template: "<div class=\"d-flex flex-column\" style=\"width: 100%;height: 100%;padding-top: 5px\">\r\n    <!-- <div class=\"header\"></div> -->\r\n    <div class=\"list f-utils-fill f-datalist\" style=\"overflow: auto;\">\r\n        <ul class=\"list-group list-group-flush p-0\"  [class.h-100]=\"!data || !data.length\">\r\n            <li class=\"list-group-item list-group-item-action\" [class.active]=\"activeIndex === i\" *ngFor=\"let item of data; index as i\" (click)=\"onItemClick($event, item)\">\r\n                <span *ngIf=\"!formatter\">{{ item[textField] }}</span> \r\n                <span *ngIf=\"formatter\" [innerHTML]=\"formatData(item) | safe:'html'\"></span>\r\n            </li>\r\n\r\n            <li class=\"norecord\" *ngIf=\"!data || !data.length\">\r\n                \u672A\u627E\u5230\u641C\u7D22\u5185\u5BB9\r\n            </li>\r\n        </ul>\r\n    </div>\r\n    <div class=\"more\" *ngIf=\"showMore && data && data.length\">\r\n        <button class=\"btn btn-link\" (click)=\"onMoreClick($event)\">\u663E\u793A\u66F4\u591A</button>\r\n    </div>\r\n     <!-- <div class=\"footer\"></div> -->\r\n</div>"
                }] }
    ];
    /** @nocollapse */
    LookupQuickSelectPanelComponent.ctorParameters = function () { return [
        { type: Injector },
        { type: ChangeDetectorRef }
    ]; };
    LookupQuickSelectPanelComponent.propDecorators = {
        showMore: [{ type: Input }],
        data: [{ type: Input }],
        textField: [{ type: Input }],
        formatter: [{ type: Input }],
        moreClcik: [{ type: Output }],
        itemClick: [{ type: Output }],
        onClick: [{ type: HostListener, args: ['mousedown', ['$event'],] }]
    };
    return LookupQuickSelectPanelComponent;
}());
export { LookupQuickSelectPanelComponent };
if (false) {
    /** @type {?} */
    LookupQuickSelectPanelComponent.prototype.showMore;
    /** @type {?} */
    LookupQuickSelectPanelComponent.prototype.data;
    /** @type {?} */
    LookupQuickSelectPanelComponent.prototype.textField;
    /** @type {?} */
    LookupQuickSelectPanelComponent.prototype.formatter;
    /** @type {?} */
    LookupQuickSelectPanelComponent.prototype.moreClcik;
    /** @type {?} */
    LookupQuickSelectPanelComponent.prototype.itemClick;
    /** @type {?} */
    LookupQuickSelectPanelComponent.prototype.activeIndex;
    /**
     * @type {?}
     * @private
     */
    LookupQuickSelectPanelComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    LookupQuickSelectPanelComponent.prototype.cdr;
}
