// @gsp-wf/wf-process-editor/esm5/lib/views/execute-forms/execute-forms.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, HostBinding, Output, EventEmitter, Input, ViewChild, TemplateRef, ElementRef, ComponentFactoryResolver, Injector } from '@angular/core';
import { of } from 'rxjs/observable/of';
import { FormBuilder } from '@angular/forms';
import { ExecuteFormsUIState } from './execute-forms.uistate';
import { ExecuteFormsViewModel } from './execute-forms.viewmodel';
import { ExecuteFormsRepository } from './execute-forms.repository';
import { ExecuteFormModel, ActualParameter } from './entities/execute-form.entity';
import { BsModalService } from '@farris/ui-modal';
import { AddExecuteFormComponent } from './add-execute-form/add-execute-form.component';
import { FormConfigContainerComponent } from './form-config-container/form-config-container.component';
import { ParameterSettingComponent } from './parameter-setting/parameter-setting.component';
import { switchMap } from 'rxjs/operators';
var ExecuteFormsComponent = /** @class */ (function () {
    // *******************************************************
    function ExecuteFormsComponent(formBuilder, uistate, viewmodel, resolver, injector, modalService) {
        this.formBuilder = formBuilder;
        this.uistate = uistate;
        this.viewmodel = viewmodel;
        this.resolver = resolver;
        this.injector = injector;
        this.modalService = modalService;
        this.modalConfig = {
            title: '表单定义',
            width: 794,
            height: 505,
            showButtons: true
        };
        this.cls = 'farris-main-area flex-column';
        this.style = '#FFF';
        this.closeModal = new EventEmitter();
        this.submitModal = new EventEmitter();
        this.editorParams = {};
        this.tab = 'web';
        this.groupIcon = '<i class="f-icon f-icon-lookup"></i>';
        this.cols = [];
        this.showMethodTab = true;
        this.form = new ExecuteFormModel();
    }
    Object.defineProperty(ExecuteFormsComponent.prototype, "value", {
        set: /**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            if (item) {
                /** @type {?} */
                var paramArray_1 = new Array();
                item.ActualParameters.forEach((/**
                 * @param {?} a
                 * @return {?}
                 */
                function (a) {
                    /** @type {?} */
                    var para = new ActualParameter();
                    para.Name = a.Name;
                    para.Value = a.Value;
                    paramArray_1.push(para);
                }));
                this.form.ActualParameters = paramArray_1;
                this.form.Id = item.Id;
                this.form.Name = item.Name;
                this.form.ComponentId = item.ComponentId;
                this.form.formFormatId = item.formFormatId;
                this.form.formatConfigId = item.formatConfigId;
                this.form.FormType = item.FormType;
                this.form.TerminalType = item.TerminalType;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ExecuteFormsComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        // this.editCSS();
    };
    /**
     * @return {?}
     */
    ExecuteFormsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.location = this.editorParams.location;
        this.tab = this.editorParams.tab;
        if (this.location === 'userActivity') {
            this.showMethodTab = true;
        }
        else {
            this.showMethodTab = false;
        }
    };
    /**
     * @return {?}
     */
    ExecuteFormsComponent.prototype.clickOK = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var data = new ExecuteFormModel();
        this.getForm().subscribe((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            if (!re.result) {
                return;
            }
            data = re.form;
            /** @type {?} */
            var object = { value: data, parameters: null };
            _this.submitModal.emit(object);
        }));
    };
    /**
     * @return {?}
     */
    ExecuteFormsComponent.prototype.clickCancel = /**
     * @return {?}
     */
    function () {
        this.closeModal.emit();
    };
    // tabChange(data) {
    //   this.tab = data.nextId;
    // }
    // editCSS() {
    //   this.configSection.nativeElement.querySelector('div.f-section-header').classList.add('mb-0');
    // }
    // tabChange(data) {
    //   this.tab = data.nextId;
    // }
    // editCSS() {
    //   this.configSection.nativeElement.querySelector('div.f-section-header').classList.add('mb-0');
    // }
    /**
     * @private
     * @return {?}
     */
    ExecuteFormsComponent.prototype.getForm = 
    // tabChange(data) {
    //   this.tab = data.nextId;
    // }
    // editCSS() {
    //   this.configSection.nativeElement.querySelector('div.f-section-header').classList.add('mb-0');
    // }
    /**
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.form.ComponentId) {
            return of({ result: true, form: this.form });
        }
        else if (this.form.formFormatId) {
            return this.configContainer.submit().pipe(switchMap((/**
             * @param {?} re
             * @return {?}
             */
            function (re) {
                if (re.result) {
                    _this.form.formatConfigId = re.configId;
                    return of({ result: true, form: _this.form });
                }
                else {
                    return of({ result: false, form: _this.form });
                }
            })));
        }
        else {
            return of({ result: true, form: null });
        }
    };
    /**
     * 参数设置
     */
    /**
     * 参数设置
     * @return {?}
     */
    ExecuteFormsComponent.prototype.parametersSet = /**
     * 参数设置
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var compFactory = this.resolver.resolveComponentFactory(ParameterSettingComponent);
        /** @type {?} */
        var compRef = compFactory.create(this.injector);
        compRef.instance.mode = 'bpmn';
        if (this.tab === 'web') {
            if (this.form.ComponentId) {
                compRef.instance.showFormType = true;
                compRef.instance.formType = this.form.FormType;
            }
            compRef.instance.parameters = tslib_1.__spread(this.form.ActualParameters.map((/**
             * @param {?} p
             * @return {?}
             */
            function (p) { return Object.assign({}, p); })));
        }
        else {
            compRef.instance.parameters = tslib_1.__spread(this.form.ActualParameters.map((/**
             * @param {?} p
             * @return {?}
             */
            function (p) { return Object.assign({}, p); })));
        }
        /** @type {?} */
        var modalConfig = {
            title: '参数设置',
            width: 620,
            height: 400,
            showButtons: true,
            buttons: compRef.instance.modalFooter
        };
        /** @type {?} */
        var dialog = this.modalService.show(compRef, modalConfig);
        compRef.instance.submit.subscribe((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            if (_this.tab === 'web') {
                _this.form.ActualParameters = re.parameters;
                _this.form.FormType = re.formType;
            }
            else {
                _this.form.ActualParameters = re.parameters;
            }
            dialog.close();
        }));
        compRef.instance.cancel.subscribe((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            dialog.close();
        }));
    };
    /**
     * 选择表单定义
     */
    /**
     * 选择表单定义
     * @return {?}
     */
    ExecuteFormsComponent.prototype.selectExecuteForm = /**
     * 选择表单定义
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var compFactory = this.resolver.resolveComponentFactory(AddExecuteFormComponent);
        /** @type {?} */
        var compRef = compFactory.create(this.injector);
        compRef.instance.tab = this.tab;
        /** @type {?} */
        var modalConfig = {
            title: '表单定义列表',
            width: 650,
            height: 400,
            showButtons: true,
            buttons: compRef.instance.modalFooter
        };
        /** @type {?} */
        var dialog = this.modalService.show(compRef, modalConfig);
        compRef.instance.submit.subscribe((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            _this.form = re;
            dialog.close();
        }));
        compRef.instance.cancel.subscribe((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            dialog.close();
        }));
    };
    /**
     * @return {?}
     */
    ExecuteFormsComponent.prototype.clear = /**
     * @return {?}
     */
    function () {
        this.form = new ExecuteFormModel();
    };
    ExecuteFormsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-execute-forms',
                    template: "  <!-- <div class=\"h-100 farris-split-section m-2\">\r\n    <farris-tabs #tabs [value]=\"'web'\" [fill]=\"true\" (tabChange)=\"tabChange($event)\" [contentFill]=\"true\" class=\"f-utils-fill h-100\">\r\n      <farris-tab title=\"Web\u8868\u5355\" [show]=\"true\" [id]=\"'web'\" class=\"h-100\">\r\n        <ng-template #content>\r\n          <farris-section class=\"f-section-form f-section-in-mainsubcard pt-0\" [enableAccordion]=\"'default'\" [mainTitle]=\"'\u8868\u5355\u5B9A\u4E49'\"\r\n            [subTitle]=\"''\" [enableMaximize]=\"false\">\r\n            <div class=\"farris-form farris-form-controls-inline farris-form-auto f-form-lable-auto\">\r\n              <div class=\"f-section-formgroup-inputs\">\r\n                <div class=\"col-12 col-md-10\">\r\n                  <div class=\"farris-group-wrap\">\r\n                    <div class=\"form-group farris-form-group mb-0\">\r\n                      <div class=\"farris-input-wrap pl-4\">\r\n                        <input-group class=\"form-control\" style=\"width: 100%;padding:0;border:0;\" [enableClear]=\"true\"\r\n                          [showButtonWhenDisabled]=\"false\" [(ngModel)]=\"webForm.Name\" name=\"webForm\" [placeholder]=\"'\u8BF7\u9009\u62E9'\"\r\n                          [editable]=\"false\" (clickHandle)=\"selectExecuteForm()\" (clear)=\"clear()\" [groupText]=\"groupIcon\">\r\n                        </input-group>\r\n                        <div class=\"farris-feedback valid-feedback\"></div>\r\n                      </div>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n                <div class=\"col-12 col-md-2\">\r\n                  <div class=\"farris-group-wrap\">\r\n                    <div class=\"form-group farris-form-group mb-0\">\r\n                      <div class=\"farris-input-wrap\">\r\n                        <button [disabled]=\"!webForm.ComponentId && !webForm.formFormatId\" class=\"btn-sm btn-secondary btn\" \r\n                          (click)=\"parametersSet()\">\r\n                          \u8868\u5355\u53C2\u6570\r\n                        </button>\r\n                      </div>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </farris-section>\r\n          <farris-section #webConfigSection class=\"f-section-tabs f-section-in-mainsubcard pb-0 pt-1\" [enableAccordion]=\"''\" [mainTitle]=\"'\u8868\u5355\u683C\u5F0F\u8BBE\u7F6E'\"\r\n            [subTitle]=\"''\" [enableMaximize]=\"false\" [fill]=\"true\">\r\n            <lib-form-config-container #webConfigContainer *ngIf=\"webForm.formFormatId; else empty\" \r\n            [formFormatId]=\"webForm.formFormatId\" [formatConfigId]=\"webForm.formatConfigId\" [showMethodTab]=\"showMethodTab\" class=\"h-100\"></lib-form-config-container>\r\n          </farris-section>\r\n        </ng-template>\r\n      </farris-tab>\r\n      <farris-tab title=\"\u79FB\u52A8\u8868\u5355\" [id]=\"'mobile'\" class=\"h-100\">\r\n        <ng-template #content>\r\n          <farris-section class=\"f-section-form f-section-in-mainsubcard pt-0\" [enableAccordion]=\"'default'\" [mainTitle]=\"'\u8868\u5355\u5B9A\u4E49'\"\r\n            [subTitle]=\"''\" [enableMaximize]=\"false\">\r\n            <div class=\"farris-form farris-form-controls-inline farris-form-auto f-form-lable-auto\">\r\n              <div class=\"f-section-formgroup-inputs\">\r\n                <div class=\"col-12 col-md-10\">\r\n                  <div class=\"farris-group-wrap\">\r\n                    <div class=\"form-group farris-form-group mb-0\">\r\n                      <div class=\"farris-input-wrap pl-4\">\r\n                        <input-group class=\"form-control\" style=\"width: 100%;padding:0;border:0;\" [enableClear]=\"true\"\r\n                          [showButtonWhenDisabled]=\"false\" [(ngModel)]=\"mobileForm.Name\" name=\"mobileForm\" [placeholder]=\"'\u8BF7\u9009\u62E9'\"\r\n                          [editable]=\"false\" (clickHandle)=\"selectExecuteForm()\" (clear)=\"clear()\" [groupText]=\"groupIcon\">\r\n                        </input-group>\r\n                        <div class=\"farris-feedback valid-feedback\"></div>\r\n                      </div>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n                <div class=\"col-12 col-md-2\">\r\n                  <div class=\"farris-group-wrap\">\r\n                    <div class=\"form-group farris-form-group mb-0\">\r\n                      <div class=\"farris-input-wrap\">\r\n                        <button [disabled]=\"!mobileForm.ComponentId && !mobileForm.formFormatId\" class=\"btn-sm btn-secondary btn\"\r\n                          (click)=\"parametersSet()\">\r\n                          \u8868\u5355\u53C2\u6570\r\n                        </button>\r\n                      </div>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </farris-section>\r\n          <farris-section #appConfigSection class=\"f-section-tabs f-section-in-mainsubcard pb-0 pt-1\" [enableAccordion]=\"''\"\r\n            [mainTitle]=\"'\u8868\u5355\u683C\u5F0F\u8BBE\u7F6E'\" [subTitle]=\"''\" [enableMaximize]=\"false\" [fill]=\"true\">\r\n            <lib-form-config-container #appConfigContainer *ngIf=\"mobileForm.formFormatId; else empty\"\r\n              [formFormatId]=\"mobileForm.formFormatId\" [formatConfigId]=\"mobileForm.formatConfigId\" [showMethodTab]=\"showMethodTab\" class=\"h-100\">\r\n            </lib-form-config-container>\r\n          </farris-section>\r\n        </ng-template>\r\n      </farris-tab>\r\n    </farris-tabs>\r\n  </div> -->\r\n  \r\n<ng-template #iconSelectFooter>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm mx-2\" (click)=\"clickOK()\">\u786E\u5B9A</button>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm\" (click)=\"clickCancel()\">\u53D6\u6D88</button>\r\n</ng-template>\r\n\r\n<ng-template #empty>\r\n  <div class=\"f-utils-fill-flex-column h-100\">\r\n    <div class=\"f-datagrid-norecords\" style=\"height: 100%;\">\r\n      <div class=\"f-datagrid-norecords-content w-100\">\u6682\u65E0\u8868\u5355\u683C\u5F0F</div>\r\n    </div>\r\n  </div>\r\n</ng-template>\r\n\r\n<div class=\"f-page f-page-is-managelist\">\r\n  <div class=\"f-page-main mx-3\">\r\n    <div class=\"farris-form farris-form-controls-inline farris-form-auto f-form-lable-auto\">\r\n      <div class=\"f-section-formgroup-inputs\">\r\n        <div class=\"col-12 col-md-5 px-0 my-2\">\r\n          <div class=\"farris-group-wrap\">\r\n            <div class=\"form-group farris-form-group mb-0\">\r\n              <div class=\"farris-input-wrap\">\r\n                <input-group class=\"form-control\" style=\"width: 100%;padding:0;border:0;\" [enableClear]=\"true\"\r\n                  [showButtonWhenDisabled]=\"false\" [(ngModel)]=\"form.Name\" name=\"form\" [placeholder]=\"'\u8BF7\u9009\u62E9'\"\r\n                  [editable]=\"false\" (clickHandle)=\"selectExecuteForm()\" (clear)=\"clear()\" [groupText]=\"groupIcon\">\r\n                </input-group>\r\n                <div class=\"farris-feedback valid-feedback\"></div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n        <div class=\"col-12 col-md-2 my-2 px-0\">\r\n          <div class=\"farris-group-wrap\" style=\"padding-top: 2px;\">\r\n            <div class=\"form-group farris-form-group mb-0\">\r\n              <div class=\"farris-input-wrap\">\r\n                <farris-button [disabled]=\"!form.ComponentId && !form.formFormatId\" [type]=\"'link'\" (click)=\"parametersSet()\">\r\n                  \u53C2\u6570\r\n                </farris-button>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n    </div>\r\n    <lib-form-config-container #configContainer\r\n      [formFormatId]=\"form.formFormatId\" [formatConfigId]=\"form.formatConfigId\" [showMethodTab]=\"showMethodTab\"\r\n      class=\"h-100\"></lib-form-config-container>\r\n  </div>\r\n</div>\r\n\r\n\r\n\r\n\r\n",
                    providers: [ExecuteFormsUIState, ExecuteFormsViewModel, ExecuteFormsRepository],
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ExecuteFormsComponent.ctorParameters = function () { return [
        { type: FormBuilder },
        { type: ExecuteFormsUIState },
        { type: ExecuteFormsViewModel },
        { type: ComponentFactoryResolver },
        { type: Injector },
        { type: BsModalService }
    ]; };
    ExecuteFormsComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        style: [{ type: HostBinding, args: ['style.background',] }],
        closeModal: [{ type: Output }],
        submitModal: [{ type: Output }],
        value: [{ type: Input }],
        editorParams: [{ type: Input }],
        modalFooter: [{ type: ViewChild, args: ['iconSelectFooter',] }],
        configContainer: [{ type: ViewChild, args: ['configContainer',] }],
        configSection: [{ type: ViewChild, args: ['configSection', { read: ElementRef },] }]
    };
    return ExecuteFormsComponent;
}());
export { ExecuteFormsComponent };
if (false) {
    /** @type {?} */
    ExecuteFormsComponent.prototype.modalConfig;
    /** @type {?} */
    ExecuteFormsComponent.prototype.cls;
    /** @type {?} */
    ExecuteFormsComponent.prototype.style;
    /** @type {?} */
    ExecuteFormsComponent.prototype.closeModal;
    /** @type {?} */
    ExecuteFormsComponent.prototype.submitModal;
    /** @type {?} */
    ExecuteFormsComponent.prototype.editorParams;
    /** @type {?} */
    ExecuteFormsComponent.prototype.modalFooter;
    /** @type {?} */
    ExecuteFormsComponent.prototype.configContainer;
    /** @type {?} */
    ExecuteFormsComponent.prototype.configSection;
    /** @type {?} */
    ExecuteFormsComponent.prototype.form;
    /** @type {?} */
    ExecuteFormsComponent.prototype.tab;
    /** @type {?} */
    ExecuteFormsComponent.prototype.groupIcon;
    /** @type {?} */
    ExecuteFormsComponent.prototype.cols;
    /** @type {?} */
    ExecuteFormsComponent.prototype.location;
    /** @type {?} */
    ExecuteFormsComponent.prototype.showMethodTab;
    /**
     * @type {?}
     * @private
     */
    ExecuteFormsComponent.prototype.formBuilder;
    /** @type {?} */
    ExecuteFormsComponent.prototype.uistate;
    /** @type {?} */
    ExecuteFormsComponent.prototype.viewmodel;
    /**
     * @type {?}
     * @private
     */
    ExecuteFormsComponent.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    ExecuteFormsComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    ExecuteFormsComponent.prototype.modalService;
}
var ExecuteFormsConverter = /** @class */ (function () {
    function ExecuteFormsConverter() {
    }
    /**
     * @param {?} data
     * @return {?}
     */
    ExecuteFormsConverter.prototype.convertTo = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (data) {
            return data.Name;
        }
        else {
            return '（未设置）';
        }
    };
    return ExecuteFormsConverter;
}());
export { ExecuteFormsConverter };
