// @farris/ui-personnel-selector/esm5/utils/user.type.js
/**
 * @fileoverview added by tsickle
 * Generated from: utils/user.type.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function UserInfo() { }
if (false) {
    /** @type {?} */
    UserInfo.prototype.orgAllPath;
    /** @type {?} */
    UserInfo.prototype.abbreviation;
    /** @type {?} */
    UserInfo.prototype.code;
    /** @type {?} */
    UserInfo.prototype.defaultOrgId;
    /** @type {?} */
    UserInfo.prototype.defaultOrgPath;
    /** @type {?} */
    UserInfo.prototype.ext1;
    /** @type {?} */
    UserInfo.prototype.ext2;
    /** @type {?} */
    UserInfo.prototype.fullPathName;
    /** @type {?} */
    UserInfo.prototype.name;
    /** @type {?} */
    UserInfo.prototype.orgTypeId;
    /** @type {?} */
    UserInfo.prototype.secLevel;
    /** @type {?} */
    UserInfo.prototype.sortOrder;
    /** @type {?} */
    UserInfo.prototype.state;
    /** @type {?} */
    UserInfo.prototype.su;
    /** @type {?} */
    UserInfo.prototype.userId;
}
/**
 * @record
 */
export function PersonnelRequestParam() { }
if (false) {
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.pageSize;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.pageIndex;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.withAllOrg;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.notGetStopUser;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.notGetCurrentUser;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.curUser;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.getUserInfo;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.orgId;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.posId;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.orderField;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.userTypeMode;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.searchMode;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.pinyin;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.initials;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.filterId;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.groupId;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.withAllGroup;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.withPermission;
    /** @type {?|undefined} */
    PersonnelRequestParam.prototype.authOp;
}
/**
 * @record
 */
export function RfRequestParam() { }
if (false) {
    /** @type {?} */
    RfRequestParam.prototype.absOrgType;
    /** @type {?} */
    RfRequestParam.prototype.bizId;
    /** @type {?} */
    RfRequestParam.prototype.objectValue;
    /** @type {?|undefined} */
    RfRequestParam.prototype.addList;
    /** @type {?|undefined} */
    RfRequestParam.prototype.deleteList;
    /** @type {?|undefined} */
    RfRequestParam.prototype.notGetStopUser;
}
/**
 * @record
 */
export function OrgRequestParam() { }
if (false) {
    /** @type {?} */
    OrgRequestParam.prototype.absOrgType;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.layer;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.codeOrName;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.name;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.parentID;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.isSearchTree;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.getCurUser;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.notGetStopOrg;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.id;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.getAllParentOrg;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.getOrgAllPath;
    /** @type {?|undefined} */
    OrgRequestParam.prototype.getOrgAllPathInfo;
}
/**
 * @record
 */
export function PanelMode() { }
if (false) {
    /** @type {?} */
    PanelMode.prototype.showQuickMode;
    /** @type {?} */
    PanelMode.prototype.showAdvancedMode;
}
var FilterPosition = /** @class */ (function () {
    function FilterPosition() {
    }
    return FilterPosition;
}());
export { FilterPosition };
if (false) {
    /** @type {?} */
    FilterPosition.prototype.id;
    /** @type {?} */
    FilterPosition.prototype.code;
    /** @type {?} */
    FilterPosition.prototype.name;
}
var FilterOrg = /** @class */ (function () {
    function FilterOrg() {
    }
    return FilterOrg;
}());
export { FilterOrg };
if (false) {
    /** @type {?} */
    FilterOrg.prototype.id;
    /** @type {?} */
    FilterOrg.prototype.value;
    /** @type {?} */
    FilterOrg.prototype.name;
}
