// @farris/ui-lookup/esm5/lib/services/multi-selection.manager.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { FavoriteIcon, FAVORITE_FIELD_NAME } from '../lookup-displaytype';
import { cloneDeep } from 'lodash-es';
var MultiSelectionManager = /** @class */ (function () {
    function MultiSelectionManager(ins) {
        this.ins = ins;
    }
    /**
     * @param {?} e
     * @return {?}
     */
    MultiSelectionManager.prototype.onSelectedTableCellClick = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        if (e.col.field === FAVORITE_FIELD_NAME) {
            e.event.stopPropagation();
            /** @type {?} */
            var classList = e.event.target['classList'];
            if (classList.contains('f-lookup-unfavorite')) {
                /** @type {?} */
                var rid = e.row[this.ins.idField];
                this.ins.lookupSelectionSer.unSelect(rid);
                // 取消选中 主列表 收藏列表 中的数据
                if (this.ins.isTree()) {
                    /** @type {?} */
                    var tt = (/** @type {?} */ (this.ins.lookupCmpMgr.getComponentInstance('treetable')));
                    tt.unCheckedNode(rid);
                    tt.unSelectNode(rid);
                    if (this.ins.useFavorite) {
                        /** @type {?} */
                        var _tt = (/** @type {?} */ (this.ins.lookupCmpMgr.getComponentInstance('fav')));
                        if (_tt.data && _tt.data.length && _tt.findRowNode(rid)) {
                            _tt.unCheckedNode(rid);
                            _tt.unSelectNode(rid);
                        }
                    }
                }
                else {
                    /** @type {?} */
                    var dt = (/** @type {?} */ (this.ins.lookupCmpMgr.getComponentInstance()));
                    dt.unCheckRow(rid);
                    if (this.ins.useFavorite) {
                        /** @type {?} */
                        var _dt = (/** @type {?} */ (this.ins.lookupCmpMgr.getComponentInstance('fav')));
                        if (_dt.data && _dt.data.length) {
                            _dt.unCheckRow(rid);
                        }
                    }
                }
            }
        }
    };
    // 初始化已选数据列信息
    // 初始化已选数据列信息
    /**
     * @return {?}
     */
    MultiSelectionManager.prototype.initSelectedColumns = 
    // 初始化已选数据列信息
    /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var selectedColumns = [];
        if (this.ins.showSelected && !this.ins.singleSelect) {
            selectedColumns = cloneDeep(this.ins.gridOptions.columns);
            /** @type {?} */
            var favcol = selectedColumns.find((/**
             * @param {?} n
             * @return {?}
             */
            function (n) { return n.field === FAVORITE_FIELD_NAME; }));
            if (favcol) {
                favcol.formatter = (/**
                 * @return {?}
                 */
                function () {
                    return FavoriteIcon.remove;
                });
            }
            else {
                selectedColumns = selectedColumns.concat([
                    { field: FAVORITE_FIELD_NAME, width: 80, formatter: (/**
                         * @return {?}
                         */
                        function () {
                            return FavoriteIcon.remove;
                        })
                    }
                ]);
            }
        }
        return selectedColumns;
    };
    /**
     * @param {?} data
     * @return {?}
     */
    MultiSelectionManager.prototype.updateSelections = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (Array.isArray(data)) {
            this.ins.lookupSelectionSer.updateSelections(data, true);
        }
        else {
            this.ins.lookupSelectionSer.select(data);
        }
    };
    /**
     * @param {?} id
     * @return {?}
     */
    MultiSelectionManager.prototype.remove = /**
     * @param {?} id
     * @return {?}
     */
    function (id) {
        if (id) {
            this.ins.lookupSelectionSer.unSelect(id);
        }
    };
    /**
     * @return {?}
     */
    MultiSelectionManager.prototype.clear = /**
     * @return {?}
     */
    function () {
        this.ins.lookupSelectionSer.clearSelections();
    };
    /**
     * @param {?} rows
     * @return {?}
     */
    MultiSelectionManager.prototype.save = /**
     * @param {?} rows
     * @return {?}
     */
    function (rows) {
        if (this.ins.showSelected) {
            this.ins.personalConf.selections = rows;
            this.ins.personalConfigService.savePersonalConfig(this.ins.personalConf);
        }
    };
    /**
     * @return {?}
     */
    MultiSelectionManager.prototype.loadData = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var items = this.ins.personalConf ? (this.ins.personalConf.selections || []) : [];
        if (!items.length) {
            items = this.ins.lookupSelectionSer.getSelections();
        }
        this.ins.lookupSelectionSer.loadSelections(items);
    };
    return MultiSelectionManager;
}());
export { MultiSelectionManager };
if (false) {
    /**
     * @type {?}
     * @private
     */
    MultiSelectionManager.prototype.ins;
}
