// @gsp-wf/wf-process-editor/esm5/lib/views/execute-forms/parameter-setting/parameter-setting.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, ViewChild, TemplateRef } from '@angular/core';
import { ExpressionService, InjectedType } from '@gsp-svc/expression';
import { ProcessDesignerUIState } from '../../../app/process-designer.uistate';
import { Subject } from 'rxjs';
import { DatagridComponent } from '@farris/ui-datagrid';
var ParameterSettingComponent = /** @class */ (function () {
    function ParameterSettingComponent(expr, designerState) {
        this.expr = expr;
        this.designerState = designerState;
        this.idField = 'Name';
        this._mode = 'bpmn';
        this.submit = new Subject();
        this.cancel = new Subject();
        this.formTypeCollection = [
            { text: '流程嵌表单', value: '0' },
            { text: '表单嵌流程', value: '1' },
        ];
        this.cols = new Array();
        this.parameters = new Array();
        this.formType = this.formTypeCollection[0].value;
        this.showFormType = false;
    }
    Object.defineProperty(ParameterSettingComponent.prototype, "mode", {
        get: /**
         * @return {?}
         */
        function () {
            return this._mode;
        },
        set: /**
         * @param {?} m
         * @return {?}
         */
        function (m) {
            var _this = this;
            this._mode = m;
            if (m === 'bpmn') {
                this.idField = 'Name';
                this.cols = [
                    { field: 'Name', width: 50, title: '编号' },
                    {
                        field: 'Value', width: 100, title: '参数值',
                        editor: {
                            type: 'input-group',
                            options: {
                                groupText: '<i class="f-icon f-icon-lookup"></i>',
                                editable: false,
                                clickHandle: (/**
                                 * @param {?} e
                                 * @return {?}
                                 */
                                function (e) {
                                    e.formControl.setValue(e.instance.value ? e.instance.value : '');
                                    _this.showExprHelp(e);
                                }),
                                clear: (/**
                                 * @param {?} e
                                 * @return {?}
                                 */
                                function (e) {
                                    console.log('clear', e);
                                })
                            }
                        }
                    }
                ];
            }
            else {
                this.idField = 'code';
                this.cols = [
                    { field: 'code', width: 50, title: '编号' },
                    { field: 'name', width: 50, title: '名称' },
                    {
                        field: 'value', width: 100, title: '参数值',
                        editor: {
                            type: 'input-group',
                            options: {
                                groupText: '<i class="f-icon f-icon-lookup"></i>',
                                editable: false,
                                clickHandle: (/**
                                 * @param {?} e
                                 * @return {?}
                                 */
                                function (e) {
                                    e.formControl.setValue(e.instance.value ? e.instance.value : '');
                                    _this.showExprHelp(e);
                                }),
                                clear: (/**
                                 * @param {?} e
                                 * @return {?}
                                 */
                                function (e) {
                                    console.log('clear', e);
                                })
                            }
                        }
                    }
                ];
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ParameterSettingComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} e
     * @return {?}
     */
    ParameterSettingComponent.prototype.showExprHelp = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        var e_1, _a, e_2, _b, e_3, _c, e_4, _d, e_5, _e, e_6, _f;
        if (this.designerState.schemas.length > 0) {
            try {
                for (var _g = tslib_1.__values(this.designerState.schemas), _h = _g.next(); !_h.done; _h = _g.next()) {
                    var schema = _h.value;
                    this.expr.addSchema('Schema', schema);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_h && !_h.done && (_a = _g.return)) _a.call(_g);
                }
                finally { if (e_1) throw e_1.error; }
            }
            if (this.designerState.schemas[0].entityTypes && this.designerState.schemas[0].entityTypes.length > 0) {
                /** @type {?} */
                var entityType = this.designerState.schemas[0].entityTypes[0];
                this.expr.addInjectedEntity(entityType.name, 'Schema', entityType.name, InjectedType.EntityType);
            }
        }
        /** @type {?} */
        var context = new Array();
        if (this.designerState.procInstContext.length > 0) {
            try {
                for (var _j = tslib_1.__values(this.designerState.procInstContext), _k = _j.next(); !_k.done; _k = _j.next()) {
                    var variable = _k.value;
                    context.push(variable);
                    // this.expr.addContext(variable);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_k && !_k.done && (_b = _j.return)) _b.call(_j);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
        if (this.designerState.formalParameterContext.length > 0) {
            var _loop_1 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _l = tslib_1.__values(this.designerState.formalParameterContext), _m = _l.next(); !_m.done; _m = _l.next()) {
                    var variable = _m.value;
                    _loop_1(variable);
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_m && !_m.done && (_c = _l.return)) _c.call(_l);
                }
                finally { if (e_3) throw e_3.error; }
            }
        }
        if (this.designerState.participantContext.length > 0) {
            var _loop_2 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _o = tslib_1.__values(this.designerState.participantContext), _p = _o.next(); !_p.done; _p = _o.next()) {
                    var variable = _p.value;
                    _loop_2(variable);
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (_p && !_p.done && (_d = _o.return)) _d.call(_o);
                }
                finally { if (e_4) throw e_4.error; }
            }
        }
        if (this.designerState.processVariableContext.length > 0) {
            var _loop_3 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _q = tslib_1.__values(this.designerState.processVariableContext), _r = _q.next(); !_r.done; _r = _q.next()) {
                    var variable = _r.value;
                    _loop_3(variable);
                }
            }
            catch (e_5_1) { e_5 = { error: e_5_1 }; }
            finally {
                try {
                    if (_r && !_r.done && (_e = _q.return)) _e.call(_q);
                }
                finally { if (e_5) throw e_5.error; }
            }
        }
        if (context.length > 0) {
            try {
                for (var context_1 = tslib_1.__values(context), context_1_1 = context_1.next(); !context_1_1.done; context_1_1 = context_1.next()) {
                    var v = context_1_1.value;
                    this.expr.addContext(v);
                }
            }
            catch (e_6_1) { e_6 = { error: e_6_1 }; }
            finally {
                try {
                    if (context_1_1 && !context_1_1.done && (_f = context_1.return)) _f.call(context_1);
                }
                finally { if (e_6) throw e_6.error; }
            }
        }
        this.expr.addExpressionText(e.instance.value);
        this.expr.buildExpression().then((/**
         * @param {?} d
         * @return {?}
         */
        function (d) {
            if (d) {
                e.formControl.setValue(d);
                e.instance.cd.detectChanges();
            }
        }));
    };
    /**
     * @return {?}
     */
    ParameterSettingComponent.prototype.clickOK = /**
     * @return {?}
     */
    function () {
        this.grid.endCellEdit();
        this.submit.next({ parameters: this.parameters, formType: this.formType });
    };
    /**
     * @return {?}
     */
    ParameterSettingComponent.prototype.clickCancel = /**
     * @return {?}
     */
    function () {
        this.cancel.next();
    };
    ParameterSettingComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-parameter-setting',
                    template: "\r\n<div class=\"f-page f-page-is-managelist\">\r\n  <div class=\"f-page-main my-0\">\r\n    <div class=\"f-struct-wrapper f-utils-fill-flex-column\">\r\n      <farris-section class=\"f-section-grid f-section-in-managelist px-0\" [showHeader]=\"showFormType\" [mainTitle]=\"'\u53C2\u6570'\" [enableMaximize]=\"false\">\r\n        <ng-template farrisSectionToolbar>\r\n          <farris-combo-list [(ngModel)]=\"formType\" name=\"formType\" [idField]=\"'value'\" [textField]=\"'text'\"\r\n            [enableClear]=\"false\" [valueField]=\"'value'\" [data]=\"formTypeCollection\" [panelHeight]=\"'auto'\" [editable]=\"false\"\r\n            style=\"width: 130px;\">\r\n          </farris-combo-list>\r\n        </ng-template>\r\n        <farris-datagrid class=\"f-component-grid\" [columns]=\"cols\" [data]=\"parameters\" #grid\r\n          [idField]=\"idField\" [fit]=\"true\" [fitColumns]=\"true\" [editable]=\"true\" [editMode]=\"'cell'\" [pagination]=\"false\"\r\n          [showCheckbox]=\"false\" [multiSelect]=\"false\" [showLineNumber]=\"true\" [striped]=\"true\"\r\n          [virtualized]=\"true\">\r\n        </farris-datagrid>\r\n      </farris-section>\r\n    </div>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #iconSelectFooter>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm mx-2\" (click)=\"clickOK()\">\u786E\u5B9A</button>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm\" (click)=\"clickCancel()\">\u53D6\u6D88</button>\r\n</ng-template>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ParameterSettingComponent.ctorParameters = function () { return [
        { type: ExpressionService },
        { type: ProcessDesignerUIState }
    ]; };
    ParameterSettingComponent.propDecorators = {
        modalFooter: [{ type: ViewChild, args: ['iconSelectFooter',] }],
        grid: [{ type: ViewChild, args: ['grid',] }]
    };
    return ParameterSettingComponent;
}());
export { ParameterSettingComponent };
if (false) {
    /** @type {?} */
    ParameterSettingComponent.prototype.modalFooter;
    /** @type {?} */
    ParameterSettingComponent.prototype.grid;
    /** @type {?} */
    ParameterSettingComponent.prototype.idField;
    /**
     * @type {?}
     * @private
     */
    ParameterSettingComponent.prototype._mode;
    /** @type {?} */
    ParameterSettingComponent.prototype.submit;
    /** @type {?} */
    ParameterSettingComponent.prototype.cancel;
    /** @type {?} */
    ParameterSettingComponent.prototype.formTypeCollection;
    /** @type {?} */
    ParameterSettingComponent.prototype.cols;
    /** @type {?} */
    ParameterSettingComponent.prototype.parameters;
    /** @type {?} */
    ParameterSettingComponent.prototype.formType;
    /** @type {?} */
    ParameterSettingComponent.prototype.showFormType;
    /**
     * @type {?}
     * @private
     */
    ParameterSettingComponent.prototype.expr;
    /**
     * @type {?}
     * @private
     */
    ParameterSettingComponent.prototype.designerState;
}
