// @farris/extend-file-upload/esm5/lib/f-file-upload.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/f-file-upload.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FFileDropDirective, FFileSelectDirective, FFilePreviewListTypeTemplateDirective, FFileUploadDirective, FFilePreviewToolbarTemplateDirective } from './directive/index';
import { FFileUploadComponent, FFileViewImageComponent, FFilePreviewComponent, PreviewCheckboxComponent } from './components/index';
import { UploadServerService } from './model/index';
import { FFileSizePipe, FFileDatePipe } from './pipe/index';
import { FFileUploadAPIServiceToken, FFileUploadDefaultService, FFileUploaderService } from './service/index';
import { NotifyModule } from '@farris/ui-notify';
import { ModalModule } from '@farris/ui-modal';
import { DateHelperModule } from '@farris/ui-common/date';
import { MessagerModule } from '@farris/ui-messager';
import { FAreaResponseModule } from '@farris/ui-area-response';
import { UploadAndPreviewComponent } from './components/upload-and-preview/upload-and-preview.component';
import { UploadProgressComponent } from './components/upload-progress/upload-progress.component';
import { MultiDownloadInfoComponent } from './components/multi-download-info/multi-download-info.component';
import { UploadFileOrderDragDirective } from './components/upload-and-preview/file-order-drag.directive';
import { LocaleModule } from '@farris/ui-locale';
import { UploadReplaceXPipe } from './components/upload-and-preview/replaceX.pipe';
var FFileUploadModule = /** @class */ (function () {
    function FFileUploadModule() {
    }
    /**
     * @param {?=} config
     * @param {?=} uploadSer
     * @return {?}
     */
    FFileUploadModule.forRoot = /**
     * @param {?=} config
     * @param {?=} uploadSer
     * @return {?}
     */
    function (config, uploadSer) {
        return {
            ngModule: FFileUploadModule,
            providers: [
                { provide: FFileUploadAPIServiceToken, useValue: config },
                FFileUploaderService,
                {
                    provide: UploadServerService,
                    useClass: uploadSer || FFileUploadDefaultService
                }
            ]
        };
    };
    FFileUploadModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [FFileDropDirective, FFileSelectDirective, FFileUploadComponent,
                        FFileViewImageComponent, FFilePreviewComponent, FFilePreviewListTypeTemplateDirective, FFileUploadDirective,
                        PreviewCheckboxComponent, FFilePreviewToolbarTemplateDirective, FFileSizePipe, UploadAndPreviewComponent, UploadProgressComponent,
                        FFileDatePipe, MultiDownloadInfoComponent, UploadFileOrderDragDirective,
                        UploadReplaceXPipe
                    ],
                    exports: [FFileDropDirective, FFileSelectDirective, FFileUploadComponent, FFileViewImageComponent,
                        FFilePreviewComponent, FFilePreviewListTypeTemplateDirective, FFileUploadDirective, FFilePreviewToolbarTemplateDirective, FFileSizePipe,
                        UploadAndPreviewComponent, UploadProgressComponent, FFileDatePipe, UploadFileOrderDragDirective],
                    imports: [
                        CommonModule,
                        NotifyModule,
                        MessagerModule,
                        FAreaResponseModule,
                        DateHelperModule.forRoot(),
                        ModalModule.forRoot(),
                        LocaleModule.forRoot()
                    ],
                    providers: [
                        FFileUploaderService,
                        {
                            provide: UploadServerService,
                            useClass: FFileUploadDefaultService
                        }
                    ],
                    entryComponents: [
                        MultiDownloadInfoComponent
                    ]
                },] }
    ];
    return FFileUploadModule;
}());
export { FFileUploadModule };
