// @farris/command-services/esm2015/lib/grid.middleware.service.js
/*
 * @Author: aalizzwell
 * @Date: 2019-08-05 11:48:27
 * @Last Modified by: aalizzwell
 * @Last Modified time: 2019-08-06 15:01:53
 */
import { Injectable, Optional } from '@angular/core';
import { EMPTY, of } from 'rxjs';
import { switchMap } from 'rxjs/operators';
import { FrameContext, ViewModel } from '@farris/devkit';
import { FormMessageService } from './form-message.service';
import { LanguageService } from './languag.service';
import { BefRepositoryUtil } from '@farris/bef';
import { CommandService } from './command-service';
// tslint:disable: no-string-literal
/**
 * Grid中间件服务
 * @scope FrameComponent
 */
export class GridMiddlewareService {
    constructor(frameContext, msgService, languageService, viewModel, commandService) {
        this.frameContext = frameContext;
        this.msgService = msgService;
        this.languageService = languageService;
        this.viewModel = viewModel;
        this.commandService = commandService;
        this.repository = this.frameContext.repository;
        if (!this.languageService) {
            this.languageService = LanguageService.getInstance();
        }
    }
    /**
     * 页码切换前
     */
    onPageChanging() {
        if (this.isChanged) {
            return this.msgService.question(this.languageService['gridDataNotSave']).pipe(switchMap((result) => {
                if (!result) {
                    return EMPTY;
                }
                else {
                    return of(true);
                }
            }));
        }
        else {
            return of(true);
        }
    }
    /**
     * 表格过滤
     * @param commandName 命令名称
     * @param frameId frameId
     * @returns
     */
    filter(commandName, frameId) {
        const self = this;
        let filters = self.context && self.context.eventParam || [];
        if (typeof (filters) === 'string') {
            filters = JSON.parse(filters);
        }
        // 查询时重置页码为第一页
        this.viewModel.frameContext.repository.entityCollection.pageIndex = 1;
        this.viewModel.frameContext.repository.filterConditionManager.setConditions(this.viewModel.bindingPath, filters);
        return this.commandService.execute(commandName, frameId);
    }
    /**
     * 是否有未保存的变更
     */
    get isChanged() {
        const befRepository = this.repository;
        return BefRepositoryUtil.isExistUnsaveData(befRepository);
    }
}
GridMiddlewareService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
GridMiddlewareService.ctorParameters = () => [
    { type: FrameContext },
    { type: FormMessageService },
    { type: LanguageService, decorators: [{ type: Optional }] },
    { type: ViewModel },
    { type: CommandService }
];
