// @farris/mobile-devkit/esm5/lib/app/app_eventBus.js
import { Subject } from 'rxjs';
var AppEventBus = /** @class */ (function () {
    /**
     * 构造函数
     */
    function AppEventBus() {
        this.eventBus = new Subject();
        this.subscriptionsMap = new Map();
    }
    AppEventBus.prototype.triggerEvent = function (event) {
        this.eventBus.next(event);
    };
    AppEventBus.prototype.subscribe = function (componentId, func, funcError) {
        if (this.subscriptionsMap.has(componentId) === false) {
            var subscription = this.eventBus.subscribe(function (value) { return func(value); }, function (value) { return funcError(value); });
            this.subscriptionsMap.set(componentId, subscription);
        }
    };
    AppEventBus.prototype.unsubscribe = function (componentId) {
        if (this.subscriptionsMap.has(componentId) === true) {
            this.subscriptionsMap.get(componentId).unsubscribe();
            this.subscriptionsMap.delete(componentId);
        }
    };
    return AppEventBus;
}());
export { AppEventBus };
