// @farris/ui-filter-editor/esm5/lib/components/filter-tabs.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-07-10 11:44:49
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-11-30 10:18:04
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { Component, EventEmitter, Output, Input } from '@angular/core';
var FilterTabsComponent = /** @class */ (function () {
    function FilterTabsComponent() {
        this.columns = [];
        this.conditions = [];
        this.showCode = false;
        this.showSql = false;
        this.fieldEditor = { type: 'select' };
        this.selectTabChange = new EventEmitter();
        this.enableExpress = false;
        this.enableFrontExpress = false;
        this.showExpress = new EventEmitter();
        this.insertType = 'below';
        this.currentTabIndex = 0;
    }
    /**
     * @return {?}
     */
    FilterTabsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} index
     * @return {?}
     */
    FilterTabsComponent.prototype.showTab = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        this.currentTabIndex = index;
        this.selectTabChange.emit(index);
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    FilterTabsComponent.prototype.showExpressDialog = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.showExpress.emit($event);
    };
    FilterTabsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'filter-tabs',
                    template: "\n    <ul class=\"nav nav-tabs\" style=\"min-height: 38px\" *ngIf=\"showCode || showSql\">\n        <li class=\"nav-item\">\n            <a class=\"nav-link\" [class.active]=\"currentTabIndex === 0\"  (click)=\"showTab(0)\">{{ 'filterEditor.designTab' | locale }}</a>\n        </li>\n        <li class=\"nav-item\" *ngIf=\"showCode\">\n            <a class=\"nav-link\" [class.active]=\"currentTabIndex === 1\" (click)=\"showTab(1)\">{{ 'filterEditor.jsonTab' | locale }}</a>\n        </li>\n        <li class=\"nav-item\" *ngIf=\"showSql\">\n            <a class=\"nav-link\" [class.active]=\"currentTabIndex === 2\" (click)=\"showTab(2)\">{{ 'filterEditor.sqlTab' | locale }}</a>\n        </li>\n    </ul>\n\n    <ng-container *ngIf=\"currentTabIndex === 0\">\n        <filter-grid style=\"overflow: hidden;\"\n          [enableExpress]=\"enableExpress\" [enableFrontExpress]=\"enableFrontExpress\"\n          class=\"d-flex flex-column flex-fill\"  [fieldEditor]=\"fieldEditor\" [insertType]=\"insertType\"\n          [conditions]=\"conditions\" [columns]=\"columns\" (showExpress)=\"showExpressDialog($event)\"></filter-grid>\n    </ng-container>\n    <ng-container *ngIf=\"currentTabIndex === 1 && showCode\">\n        <filter-code [height]=\"height\" [conditions]=\"conditions\"></filter-code>\n    </ng-container>\n    <ng-container *ngIf=\"currentTabIndex === 2 && showSql\">\n        <filter-sql [height]=\"height\" [conditions]=\"conditions\"></filter-sql>\n    </ng-container>\n\n    ",
                    styles: [".nav-item { cursor: default;}\n        .nav-item a.active {font-weight: bold;}\n        "]
                }] }
    ];
    /** @nocollapse */
    FilterTabsComponent.ctorParameters = function () { return []; };
    FilterTabsComponent.propDecorators = {
        columns: [{ type: Input }],
        conditions: [{ type: Input }],
        height: [{ type: Input }],
        showCode: [{ type: Input }],
        showSql: [{ type: Input }],
        fieldEditor: [{ type: Input }],
        selectTabChange: [{ type: Output }],
        enableExpress: [{ type: Input }],
        enableFrontExpress: [{ type: Input }],
        showExpress: [{ type: Output }],
        insertType: [{ type: Input }]
    };
    return FilterTabsComponent;
}());
export { FilterTabsComponent };
if (false) {
    /** @type {?} */
    FilterTabsComponent.prototype.columns;
    /** @type {?} */
    FilterTabsComponent.prototype.conditions;
    /** @type {?} */
    FilterTabsComponent.prototype.height;
    /** @type {?} */
    FilterTabsComponent.prototype.showCode;
    /** @type {?} */
    FilterTabsComponent.prototype.showSql;
    /** @type {?} */
    FilterTabsComponent.prototype.fieldEditor;
    /** @type {?} */
    FilterTabsComponent.prototype.selectTabChange;
    /** @type {?} */
    FilterTabsComponent.prototype.enableExpress;
    /** @type {?} */
    FilterTabsComponent.prototype.enableFrontExpress;
    /** @type {?} */
    FilterTabsComponent.prototype.showExpress;
    /** @type {?} */
    FilterTabsComponent.prototype.insertType;
    /** @type {?} */
    FilterTabsComponent.prototype.currentTabIndex;
}
