// @farris/ui-datagrid-settings/esm2015/lib/scheme/newscheme.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ChangeDetectorRef, Component, ElementRef, EventEmitter, Input, Output, ViewChild } from '@angular/core';
export class NewSchemeComponent {
    /**
     * @param {?} cd
     */
    constructor(cd) {
        this.cd = cd;
        this.dataPermissionIds = [];
        this.closeModal = new EventEmitter();
        this.submitModal = new EventEmitter();
        this.checkAuthority = new EventEmitter();
        this.schemeInfo = {
            name: '',
            isDefault: false,
            type: 'private',
            orgids: ''
        };
        this.schemeType = 'private';
        this.noPublicAuthority = false;
        this.noOrgAuthority = false;
        this.languages = [{ code: 'zh-CHS', name: '中文简体', default: true }, { code: 'en', name: 'English' }, { code: 'zh-CHT', name: '中文繁体' }];
        this.nameMaxLength = { 'zh-CHS': 9, 'zh-CHT': 9, en: 50 };
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    close($event) {
        this.closeModal.emit();
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    submit($event) {
        this.submitModal.emit(this.schemeInfo);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onCancelClick($event) {
        this.close($event);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onSelectedOrg($event) {
        if ($event.selections && $event.selections.length) {
            /** @type {?} */
            let orgids = $event.selections.map((/**
             * @param {?} n
             * @return {?}
             */
            n => n.orgId));
            this.schemeInfo.orgids = orgids;
        }
    }
    /**
     * @param {?} $event
     * @param {?} val
     * @return {?}
     */
    onSetRangeType($event, val) {
        $event.stopPropagation();
        this.schemeType = val;
        if (val === 'public' || val === 'org') {
            this.checkAuthority.emit({ value: val, schemeInfo: this.schemeInfo });
        }
        else {
            this.schemeInfo.type = val;
        }
        // console.log($event, this.schemeInfo.type);
    }
    /**
     * @param {?=} hasAuthority
     * @return {?}
     */
    refresh(hasAuthority = { public: true, org: true }) {
        this.schemeInfo = Object.assign({}, this.schemeInfo);
        this.schemeType = this.schemeInfo.type;
        this.noPublicAuthority = !hasAuthority.public;
        this.noOrgAuthority = !hasAuthority.org;
        this.cd.detectChanges();
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    getSelectedId($event) {
        console.log($event);
        if ($event.selections && $event.selections.length) {
            this.schemeInfo.orgids = $event.selections.map((/**
             * @param {?} n
             * @return {?}
             */
            n => n.orgId));
        }
        else {
            this.schemeInfo.orgids = '';
        }
    }
    /**
     * @return {?}
     */
    clearAllOrgs() {
        this.schemeInfo.orgids = '';
    }
}
NewSchemeComponent.decorators = [
    { type: Component, args: [{
                selector: 'new-scheme',
                template: "\r\n\r\n\r\n<div class=\"f-datagrid-save-scheme d-flex flex-column\" style=\"height: 100%;\">\r\n\r\n    <div class=\"d-flex justify-content-between header\" #header >\r\n        <div class=\"title\">{{'datagrid.scheme.saveTitle' | locale}}</div>\r\n        <div class=\"close\" style=\"margin-right: 20px;\">\r\n            <span class=\"f-icon modal_close\" (click)=\"close($event)\"></span>\r\n        </div>\r\n    </div>\r\n\r\n    <div class=\"f-utils-fill  mx-3 p-3 mt-1\" >\r\n        <!-- <div class=\"input-group mb-3 scheme-manager__search-box\">\r\n            <input #schemeName type=\"text\" class=\"form-control\" maxlength=\"9\" placeholder=\"{{'datagrid.scheme.namePlaceholder' | locale}}\" style=\"height: 26px;\" name=\"name\" [(ngModel)]=\"schemeInfo.name\">\r\n        </div> -->\r\n        <div class=\" mb-3\">\r\n            <language-textbox [(ngModel)]=\"schemeInfo.name\"  [languages]=\"languages\" [maxWords]=\"nameMaxLength\"></language-textbox>\r\n        </div>\r\n            \r\n    \r\n        <div class=\"input-group mb-2\" style=\"border: 0;\">\r\n          \r\n            <div class=\"custom-control custom-radio  custom-control-inline \" style=\"align-items: center\">\r\n                <input class=\"custom-control-input\" type=\"radio\" id=\"private-scheme\"  name=\"schemeType\"  (click)=\"onSetRangeType($event, 'private')\" value=\"private\" [checked]=\"schemeType === 'private'\">\r\n                <label class=\"custom-control-label\" for=\"private-scheme\">{{'datagrid.scheme.setPrivate' | locale}}</label>\r\n                <span class=\"icon-question\" farrisTooltip [placement]=\"'right'\" [delay]=\"500\" [content]=\"'datagrid.scheme.privateTip' | locale\">?</span>\r\n            </div>\r\n            <div class=\"custom-control custom-radio  custom-control-inline \" style=\"align-items: center\">\r\n                <input class=\"custom-control-input\" type=\"radio\" id=\"public-scheme\" name=\"schemeType\" (click)=\"onSetRangeType($event, 'public')\" value=\"public\" [checked]=\"schemeType === 'public'\" [disabled]=\"noPublicAuthority\">\r\n                <label class=\"custom-control-label\" for=\"public-scheme\">{{'datagrid.scheme.setPublic' | locale}}</label>\r\n                <span class=\"icon-question\" farrisTooltip [placement]=\"'right'\" [delay]=\"500\" [content]=\"'datagrid.scheme.publicTip' | locale\">?</span>\r\n            </div>\r\n            <div class=\"custom-control custom-radio  custom-control-inline \" style=\"align-items: center\">\r\n                <input class=\"custom-control-input\" type=\"radio\" id=\"org-scheme\" name=\"schemeType\" (click)=\"onSetRangeType($event, 'org')\"  value=\"org\" [checked]=\"schemeType === 'org'\" [disabled]=\"noOrgAuthority\">\r\n                <label class=\"custom-control-label\" for=\"org-scheme\">{{'datagrid.scheme.setOrg' | locale}}</label>\r\n                <span class=\"icon-question\" farrisTooltip [placement]=\"'right'\" [delay]=\"500\" [content]=\"'datagrid.scheme.orgTip' | locale\">?</span>\r\n            </div>\r\n        </div>\r\n        <div class=\"input-group mb-2\" style=\"border: 0;\" *ngIf=\"schemeInfo.type === 'private'\">\r\n            <div class=\"custom-control custom-checkbox  custom-control-inline \" style=\"align-items: center\">\r\n                <input class=\"custom-control-input\" type=\"checkbox\" id=\"default-scheme\" name=\"isDefault\" [(ngModel)]=\"schemeInfo.isDefault\">\r\n                <label class=\"custom-control-label\" for=\"default-scheme\">{{'datagrid.scheme.setDefault' | locale}} <i class=\"scheme-org-tip\">({{'datagrid.scheme.defaultTip' | locale}})</i></label>\r\n                <!-- <span class=\"icon-question\" farrisTooltip [placement]=\"'right'\" [delay]=\"500\" [content]=\"'\u4E0B\u6B21\u8FDB\u5165\u6B64\u529F\u80FD\u65F6\uFF0C\u9ED8\u8BA4\u52A0\u8F7D\u8BE5\u65B9\u6848'\">?</span> -->\r\n            </div>\r\n        </div>\r\n\r\n        <div class=\"input-group mb-2\" style=\"border: 0;\" *ngIf=\"schemeInfo.type === 'org' && !noOrgAuthority\">\r\n            <div class=\"input-group mb-3\">\r\n                <farris-organization-selector\r\n                    [primaryField]=\"'orgId'\" [displayField]=\"'name'\"\r\n                    [absOrgType]=\"'System_organization'\"\r\n                    [layer]=\"1\" [multiSelect]=\"true\" [showTabIds]=\"'organization'\"\r\n                    [enableHierarchicalLoading]=\"false\" [cascadeCheck]=\"true\" [cascadeDown]=\"true\"\r\n                    (selectionsChange)=\"onSelectedOrg($event)\"\r\n                    [cascadeUp]=\"false\"\r\n                    [orgIds]=\"dataPermissionIds\"\r\n                    (inputClear)=\"clearAllOrgs()\" (tagRemoved)=\"getSelectedId($event)\"\r\n                    [dataPermission]=\"true\" [dataPermissionIds]=\"dataPermissionIds\" \r\n                ></farris-organization-selector>\r\n            </div>\r\n        </div>\r\n    </div>\r\n\r\n    <div class=\"p-3 mb-3\" style=\"height: 50px;text-align: right;\">\r\n        <button class=\"btn btn-secondary cancel btn-lg mr-3\" style=\"width: 70px\"\r\n            (click)=\"onCancelClick($event)\">{{'datagrid.settings.cancel' | locale}}</button>\r\n        <button class=\"btn btn-primary ok btn-lg\" style=\"width: 70px\"\r\n            (click)=\"submit($event)\">{{'datagrid.settings.ok' | locale}}</button>\r\n    </div>\r\n\r\n</div>",
                styles: [`
        .icon-question {
            width: 16px;
            border-radius: 20px;
            height: 16px;
            text-align: center;
            vertical-align: middle;
            line-height: 15px;
            font-size: 10px;
            margin-left: 10px;
            background: #EAF4FF;
            color: #2A87FF;
            cursor: pointer
        }

        .icon-question:hover {
            background: #2A87FF;
            color: #fff;
        }
        `]
            }] }
];
/** @nocollapse */
NewSchemeComponent.ctorParameters = () => [
    { type: ChangeDetectorRef }
];
NewSchemeComponent.propDecorators = {
    header: [{ type: ViewChild, args: ['header',] }],
    dataPermissionIds: [{ type: Input }],
    closeModal: [{ type: Output }],
    submitModal: [{ type: Output }],
    checkAuthority: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    NewSchemeComponent.prototype.header;
    /** @type {?} */
    NewSchemeComponent.prototype.dataPermissionIds;
    /** @type {?} */
    NewSchemeComponent.prototype.closeModal;
    /** @type {?} */
    NewSchemeComponent.prototype.submitModal;
    /** @type {?} */
    NewSchemeComponent.prototype.checkAuthority;
    /** @type {?} */
    NewSchemeComponent.prototype.schemeInfo;
    /** @type {?} */
    NewSchemeComponent.prototype.schemeType;
    /** @type {?} */
    NewSchemeComponent.prototype.noPublicAuthority;
    /** @type {?} */
    NewSchemeComponent.prototype.noOrgAuthority;
    /** @type {?} */
    NewSchemeComponent.prototype.languages;
    /** @type {?} */
    NewSchemeComponent.prototype.nameMaxLength;
    /**
     * @type {?}
     * @private
     */
    NewSchemeComponent.prototype.cd;
}
