// @farris/ui-view-change/esm2015/lib/view-change.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
/**
 * @record
 */
function ViewItems() { }
if (false) {
    /** @type {?} */
    ViewItems.prototype.items;
}
/**
 * @record
 */
function viewContainer() { }
export class ViewChangeService {
    constructor() {
        /**
         * 默认视图切换分组名称
         */
        this.defaultId = 'default';
        //当前显示视图
        this.currentViewType = {};
        /**
         * contents
         */
        this.viewContainers = {};
        this.changeCurrentViewId = new BehaviorSubject(null);
    }
    /**
     * @param {?} itemId
     * @param {?=} viewContainerId
     * @return {?}
     */
    addViewItem(itemId, viewContainerId = this.defaultId) {
        if (this.viewContainers[viewContainerId]) {
            if (this.hasElement(itemId, viewContainerId)) {
                return;
            }
            else {
                this.viewContainers[viewContainerId].items.push(itemId);
            }
        }
        else {
            this.viewContainers[viewContainerId] = (/** @type {?} */ ({
                items: []
            }));
            this.viewContainers[viewContainerId].items.push(itemId);
        }
    }
    /**
     * @param {?} itemId
     * @param {?=} viewContainerId
     * @return {?}
     */
    removeViewItem(itemId, viewContainerId = this.defaultId) {
        delete this.viewContainers[viewContainerId];
        if (this.viewContainers && this.viewContainers[viewContainerId]) {
            /** @type {?} */
            const items = this.viewContainers[viewContainerId].items.filter((/**
             * @param {?} el
             * @return {?}
             */
            el => {
                return el !== itemId;
            }));
            if (!items.length) {
                delete this.viewContainers[viewContainerId];
                return;
            }
            this.viewContainers[viewContainerId].items = items;
        }
    }
    /**
     * 存储currentpageid
     * @param {?} type
     * @param {?=} viewContainerId
     * @return {?}
     */
    updateCurrentViewId(type, viewContainerId = this.defaultId) {
        this.currentViewType[viewContainerId] = type;
        /** @type {?} */
        let changeView = {};
        changeView[viewContainerId] = type;
        this.changeCurrentViewId.next(changeView);
    }
    /**
     * @return {?}
     */
    getChangeCurrentViewId() {
        return this.changeCurrentViewId;
    }
    /**
     * @private
     * @param {?} itemId
     * @param {?} groupId
     * @return {?}
     */
    hasElement(itemId, groupId) {
        return this.viewContainers[groupId].items.some((/**
         * @param {?} item
         * @return {?}
         */
        (item) => item === itemId));
    }
}
ViewChangeService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
ViewChangeService.ctorParameters = () => [];
/** @nocollapse */ ViewChangeService.ngInjectableDef = i0.defineInjectable({ factory: function ViewChangeService_Factory() { return new ViewChangeService(); }, token: ViewChangeService, providedIn: "root" });
if (false) {
    /**
     * 默认视图切换分组名称
     * @type {?}
     * @private
     */
    ViewChangeService.prototype.defaultId;
    /** @type {?} */
    ViewChangeService.prototype.currentViewType;
    /**
     * contents
     * @type {?}
     */
    ViewChangeService.prototype.viewContainers;
    /** @type {?} */
    ViewChangeService.prototype.changeCurrentViewId;
}
