// @farris/ui-lookup/esm5/lib/advanced-layout/lookup-advanced-layout.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { EventEmitter, Component, ElementRef, Injector, Input, NgZone, Output, Renderer2, ViewChild, ViewEncapsulation } from '@angular/core';
import { DialogComponent } from '@farris/ui-dialog';
var LookupAdvancedLayoutComponent = /** @class */ (function () {
    function LookupAdvancedLayoutComponent(elRef, injector, zone, render2) {
        this.elRef = elRef;
        this.injector = injector;
        this.zone = zone;
        this.render2 = render2;
        this.title = '帮助选择器';
        this.activeTab = 'datalist';
        this.tabChanged = new EventEmitter();
        this.close = new EventEmitter();
        this.dialogRef = this.injector.get(DialogComponent, null);
    }
    /**
     * @return {?}
     */
    LookupAdvancedLayoutComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    LookupAdvancedLayoutComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    LookupAdvancedLayoutComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (changes.activeTab && !changes.activeTab.isFirstChange()) {
            this.setCurrentTab();
        }
    };
    /**
     * @return {?}
     */
    LookupAdvancedLayoutComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (this.dialogRef) {
            this.dialogRef.dragHandle = this.headerRel.nativeElement;
        }
        this.setCurrentTab();
    };
    /**
     * @private
     * @return {?}
     */
    LookupAdvancedLayoutComponent.prototype.setCurrentTab = /**
     * @private
     * @return {?}
     */
    function () {
        this.render2.setStyle(this.activetabRef.nativeElement, 'transform', "translate3d(" + (this.activeTab == 'datalist' ? 0 : 84) + "px, 0px, 0px)");
    };
    /**
     * @param {?} $event
     * @param {?} tabstr
     * @return {?}
     */
    LookupAdvancedLayoutComponent.prototype.changeTabs = /**
     * @param {?} $event
     * @param {?} tabstr
     * @return {?}
     */
    function ($event, tabstr) {
        this.activeTab = tabstr;
        this.setCurrentTab();
        this.tabChanged.emit({ tab: tabstr, $event: $event });
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    LookupAdvancedLayoutComponent.prototype.closeDialog = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (this.dialogRef) {
            $event.stopPropagation();
            this.dialogRef.close();
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    LookupAdvancedLayoutComponent.prototype.changeDialogSize = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (this.dialogRef) {
            $event.stopPropagation();
            this.dialogRef.changeDialogSize();
        }
    };
    LookupAdvancedLayoutComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lookup-advanced-layout',
                    template: "<div class=\"d-flex justify-content-between header\" #header [style.cursor]=\"dialogRef.isMax ? 'default': 'move'\">\r\n    <div class=\"title\">{{title}}</div>\r\n\r\n    <div style=\"width: 100%;\" class=\"tabs\">\r\n        <div class=\"d-flex justify-content-center\">\r\n            <div class=\"d-flex tabs-labels\" *ngIf=\"enableFav\"  (mousedown)=\"$event.stopPropagation()\">\r\n                <div class=\"active-tab\" #activetab></div>\r\n                <div class=\"tab-title d-flex\" [class.active]=\"activeTab === 'datalist'\" (click)=\"changeTabs($event, 'datalist')\" >\r\n                    {{ 'lookup.datalist' | locale: '\u6570\u636E\u5217\u8868' }}\r\n                </div>\r\n                <div class=\"tab-title d-flex\" [class.active]=\"activeTab === 'favorite'\" (click)=\"changeTabs($event, 'favorite')\">\r\n                    {{'lookup.favorites' | locale: '\u6536\u85CF\u5939'}}\r\n                </div>\r\n            </div>\r\n        </div>\r\n    </div>\r\n\r\n    <div class=\"close d-flex flex-row h-100\">\r\n        <span class=\"f-icon modal_maximize\" [class.modalrevert]=\"dialogRef.isMax\" (click)=\"changeDialogSize($event)\"></span>\r\n        <span class=\"f-icon modal_close\" (click)=\"closeDialog($event)\"></span>\r\n    </div>\r\n</div>",
                    encapsulation: ViewEncapsulation.None,
                    styles: [".lookup-advanced-layout .header{position:relative;width:100%;height:56px;border:0;cursor:move}.lookup-advanced-layout .title{display:block;white-space:nowrap;font-size:17px;line-height:56px;position:absolute}.lookup-advanced-layout .tabs{width:100%;height:32px;top:12px;font-size:14px;position:relative;line-height:32px}.lookup-advanced-layout .tabs .tabs-labels{background:#eff5ff;border-radius:17px;color:#6080ad;cursor:pointer;-webkit-box-align:center;align-items:center}.lookup-advanced-layout .tabs .tabs-labels .active{cursor:default;color:#2a87ff}.lookup-advanced-layout .tabs .tabs-labels .tab-title{border:0;z-index:2;-webkit-box-pack:center;justify-content:center;width:84px}.lookup-advanced-layout .close{line-height:56px;padding:0;z-index:55555;position:absolute;right:0;-webkit-box-pack:justify;justify-content:space-between;-webkit-box-align:center;align-items:center;width:44px}.lookup-advanced-layout .active-tab{position:absolute;width:80px;z-index:1;-webkit-transition:.18s ease-in;transition:.18s ease-in;line-height:28px;margin:2px;height:28px;background:rgba(255,255,255,.9);box-shadow:0 4px 18px 0 rgba(2,75,193,.2);border-radius:15px}"]
                }] }
    ];
    /** @nocollapse */
    LookupAdvancedLayoutComponent.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Injector },
        { type: NgZone },
        { type: Renderer2 }
    ]; };
    LookupAdvancedLayoutComponent.propDecorators = {
        title: [{ type: Input }],
        activeTab: [{ type: Input }],
        enableFav: [{ type: Input }],
        tabChanged: [{ type: Output }],
        close: [{ type: Output }],
        headerRel: [{ type: ViewChild, args: ['header',] }],
        activetabRef: [{ type: ViewChild, args: ['activetab',] }]
    };
    return LookupAdvancedLayoutComponent;
}());
export { LookupAdvancedLayoutComponent };
if (false) {
    /** @type {?} */
    LookupAdvancedLayoutComponent.prototype.title;
    /** @type {?} */
    LookupAdvancedLayoutComponent.prototype.activeTab;
    /** @type {?} */
    LookupAdvancedLayoutComponent.prototype.enableFav;
    /** @type {?} */
    LookupAdvancedLayoutComponent.prototype.tabChanged;
    /** @type {?} */
    LookupAdvancedLayoutComponent.prototype.close;
    /** @type {?} */
    LookupAdvancedLayoutComponent.prototype.headerRel;
    /** @type {?} */
    LookupAdvancedLayoutComponent.prototype.activetabRef;
    /** @type {?} */
    LookupAdvancedLayoutComponent.prototype.dialogRef;
    /** @type {?} */
    LookupAdvancedLayoutComponent.prototype.elRef;
    /**
     * @type {?}
     * @private
     */
    LookupAdvancedLayoutComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    LookupAdvancedLayoutComponent.prototype.zone;
    /**
     * @type {?}
     * @private
     */
    LookupAdvancedLayoutComponent.prototype.render2;
}
