// @gsp-wf/wf-process-editor/esm5/lib/views/organization-potential-owner/organization-potential-owner.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, Output, EventEmitter, Input, ViewChild, TemplateRef } from '@angular/core';
import { ActionOrganizationScope, ActionScope, OrganizationPotentialOwner } from './actionOrganizationScope';
var OrganizationPotentialOwnerComponent = /** @class */ (function () {
    function OrganizationPotentialOwnerComponent() {
        var _this = this;
        this.cls = 'farris-main-area flex-column';
        this.closeModal = new EventEmitter();
        this.submitModal = new EventEmitter();
        this.modalConfig = {
            title: '选项设置',
            width: 425,
            height: 360,
            showButtons: true
        };
        this.organizationLevelCollection = [{
                value: 'GLOBAL',
                text: '全员',
                orgId: 'global',
            }, {
                value: 'UNIT',
                text: '本单位',
                orgId: 'currentUnit'
            }, {
                value: 'DEPARTMENT',
                text: '本部门',
                orgId: 'currentDepartment'
            }];
        this.actionCollection = [{
                value: ActionScope.AddSignBehind, name: '加签'
            }, {
                value: ActionScope.AddSignFront, name: '征询'
            }, {
                value: ActionScope.Transfer, name: '移交'
            }, {
                value: ActionScope.Notify, name: '知会'
            }];
        this.enable = false;
        this.organizationLevel = this.organizationLevelCollection[0].value;
        this.action = this.actionCollection[0].value;
        this.includeChildren = false;
        this.actionName = (/**
         * @return {?}
         */
        function () {
            return _this.actionCollection.find((/**
             * @param {?} a
             * @return {?}
             */
            function (a) { return a.value === _this.action; })).name;
        });
    }
    Object.defineProperty(OrganizationPotentialOwnerComponent.prototype, "value", {
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            this.action = v.action;
            if (this.action === this.actionCollection[2].value) {
                if (v.enable === null || v.enable === undefined) {
                    this.enable = true;
                }
                else {
                    this.enable = v.enable;
                }
            }
            else {
                this.enable = v.enable;
            }
            if (v.scope) {
                this.organizationLevel = v.scope.organizationLevel;
                this.includeChildren = v.scope.includeChildren;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    OrganizationPotentialOwnerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    OrganizationPotentialOwnerComponent.prototype.clickOK = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var scope = new ActionOrganizationScope();
        scope.enable = this.enable;
        scope.action = this.action;
        if (scope.enable && this.organizationLevel !== this.organizationLevelCollection[0].value) {
            /** @type {?} */
            var potential = new OrganizationPotentialOwner();
            potential.actionScope = this.action;
            potential.organizationLevel = (/** @type {?} */ (this.organizationLevel));
            potential.organizationId = this.organizationLevelCollection.find((/**
             * @param {?} o
             * @return {?}
             */
            function (o) { return o.value === _this.organizationLevel; })).orgId;
            potential.organizationName = this.organizationLevelCollection.find((/**
             * @param {?} o
             * @return {?}
             */
            function (o) { return o.value === _this.organizationLevel; })).text;
            potential.includeChildren = this.includeChildren;
            potential.tag = 'variable';
            scope.scope = potential;
        }
        /** @type {?} */
        var object = { value: scope, parameters: null };
        this.submitModal.emit(object);
    };
    /**
     * @return {?}
     */
    OrganizationPotentialOwnerComponent.prototype.clickCancel = /**
     * @return {?}
     */
    function () {
        this.closeModal.emit();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    OrganizationPotentialOwnerComponent.prototype.valueChange = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        switch (event.value) {
            case this.organizationLevelCollection[0].value:
                this.includeChildren = false;
                break;
            case this.organizationLevelCollection[1].value:
                this.includeChildren = true;
                break;
            case this.organizationLevelCollection[2].value:
                this.includeChildren = false;
                break;
        }
    };
    OrganizationPotentialOwnerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-organization-potential-owner',
                    template: "<div class=\"f-page\">\r\n  <div class=\"f-utils-fill-flex-column farris-split-section m-2 h-100\">\r\n    <section class=\"f-utils-fill-flex-column h-100\">\r\n      <div class=\"f-section-header pt-2 mx-2\">\r\n        <div class=\"f-title\">\r\n          <h4 class=\"f-title-text\">{{actionName()+'\u9009\u9879'}}</h4>\r\n        </div>\r\n      </div>\r\n      <div class=\"f-utils-fill-flex-column h-100\">\r\n        <div class=\"farris-form form-inline farris-form-inline\">\r\n          <div class=\"col-12 farris-group-auto my-3 pr-5\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-text\">\u662F\u5426\u542F\u7528</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap ml-1\">\r\n                  <farris-switch [(ngModel)]=\"enable\" name=\"enable\"></farris-switch>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"col-12 farris-group-auto mt-4 pr-5\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-text\">{{actionName()+'\u8303\u56F4'}}</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap ml-1 pr-5\">\r\n                  <farris-combo-list [panelHeight]=\"100\" [(ngModel)]=\"organizationLevel\" name=\"actionScope\" [idField]=\"'value'\"\r\n                    [textField]=\"'text'\" [valueField]=\"'value'\" [data]=\"organizationLevelCollection\"\r\n                    (valueChange)=\"valueChange($event)\" [readonly]=\"enable===false\">\r\n                  </farris-combo-list>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div *ngIf=\"organizationLevel===organizationLevelCollection[2].value\" class=\"col-12 farris-group-auto my-3 pr-5\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-text\">\u5305\u542B\u4E0B\u7EA7</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap ml-1\">\r\n                  <farris-switch [(ngModel)]=\"includeChildren\" name=\"includeChildren\" [editable]=\"enable===true\"></farris-switch>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n    </section>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #iconSelectFooter>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm mx-2\" (click)=\"clickOK()\">\u786E\u5B9A</button>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm mx-2\" (click)=\"clickCancel()\">\u53D6\u6D88</button>\r\n</ng-template>\r\n",
                    styles: [".title-content{color:#000;display:inline-flex;align-items:center;line-height:1.375rem;position:relative;overflow:hidden;padding:0 0 0 1rem;min-width:390px}.title-content ::before{content:'';width:.25rem;height:1.125rem;background:#59a1ff;border-radius:1px;position:absolute;top:50%;left:0;margin-top:-.5625rem}.title-text{font-size:1rem;line-height:1.375rem;overflow:hidden;white-space:nowrap;text-overflow:ellipsis;margin-bottom:0;margin-right:10px}.title-line{flex-grow:1;border-bottom:1px dashed rgba(0,0,0,.1)}"]
                }] }
    ];
    /** @nocollapse */
    OrganizationPotentialOwnerComponent.ctorParameters = function () { return []; };
    OrganizationPotentialOwnerComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        closeModal: [{ type: Output }],
        submitModal: [{ type: Output }],
        value: [{ type: Input }],
        modalFooter: [{ type: ViewChild, args: ['iconSelectFooter',] }]
    };
    return OrganizationPotentialOwnerComponent;
}());
export { OrganizationPotentialOwnerComponent };
if (false) {
    /** @type {?} */
    OrganizationPotentialOwnerComponent.prototype.cls;
    /** @type {?} */
    OrganizationPotentialOwnerComponent.prototype.closeModal;
    /** @type {?} */
    OrganizationPotentialOwnerComponent.prototype.submitModal;
    /** @type {?} */
    OrganizationPotentialOwnerComponent.prototype.modalFooter;
    /** @type {?} */
    OrganizationPotentialOwnerComponent.prototype.modalConfig;
    /** @type {?} */
    OrganizationPotentialOwnerComponent.prototype.organizationLevelCollection;
    /** @type {?} */
    OrganizationPotentialOwnerComponent.prototype.actionCollection;
    /** @type {?} */
    OrganizationPotentialOwnerComponent.prototype.enable;
    /** @type {?} */
    OrganizationPotentialOwnerComponent.prototype.organizationLevel;
    /** @type {?} */
    OrganizationPotentialOwnerComponent.prototype.action;
    /** @type {?} */
    OrganizationPotentialOwnerComponent.prototype.includeChildren;
    /** @type {?} */
    OrganizationPotentialOwnerComponent.prototype.actionName;
}
var OrganizationPotentialOwnerConverter = /** @class */ (function () {
    function OrganizationPotentialOwnerConverter() {
    }
    /**
     * @param {?} data
     * @return {?}
     */
    OrganizationPotentialOwnerConverter.prototype.convertTo = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (data.action === ActionScope.Transfer && (data.enable === null || data.enable === undefined)) {
            return '（已设置）';
        }
        else {
            if (data.enable) {
                return '（已设置）';
            }
            else {
                return '（未设置）';
            }
        }
    };
    return OrganizationPotentialOwnerConverter;
}());
export { OrganizationPotentialOwnerConverter };
