// @farris/command-services/esm5/lib/feature-edit/feature-edit.service.js
import { Injectable, Injector } from "@angular/core";
import { FeatureEditorService } from '@farris/ui-feature-editor';
import { FormNotifyService } from "../form-notify.service";
import { map, switchMap, tap } from "rxjs/operators";
import { FormLoadingService } from "../form-loading/form-loading.service";
import { FeatureDataService } from './feature-data.service';
import { LanguageService } from '../languag.service';
import { of } from "rxjs";
var FeatureEditService = /** @class */ (function () {
    function FeatureEditService(injector, featureEditorService, featureDataService, loadingService, notifyService, languageService) {
        this.injector = injector;
        this.featureEditorService = featureEditorService;
        this.featureDataService = featureDataService;
        this.loadingService = loadingService;
        this.notifyService = notifyService;
        this.languageService = languageService;
    }
    FeatureEditService.prototype.edit = function (materialId, materialFeatureId, options) {
        var _this = this;
        if (!materialId) {
            throw new Error('[FeatureEditService]物料id不能为空！');
        }
        if (!options) {
            options = {};
        }
        if (typeof options === 'string' && options.startsWith('{') && options.endsWith('}')) {
            options = JSON.parse(options);
        }
        options.getHelpInfo = this.featureDataService.getHelpInfo;
        this.loadingService.show();
        return this.featureDataService.getFeaturesByMaterialId(materialId).pipe(tap(function (result) {
            _this.loadingService.hide();
            var returnValue = result || null;
            if (returnValue) {
                var featureInfo = JSON.parse(returnValue);
                var props = featureInfo.props || null;
                if (!props || props.length < 1) {
                    _this.notifyService.warning(_this.languageService.propsIsEmpty);
                    return;
                }
                _this.featureEditorService.show(props, options);
            }
            else {
                _this.notifyService.error(_this.languageService.propsIsEmpty);
            }
        }));
    };
    FeatureEditService.prototype.buildFeatures = function (materialId, materialFeatureId) {
        var _this = this;
        var features$ = this.featureDataService.getFeaturesByMaterialId(materialId);
        return features$.pipe(switchMap(function (featureSet) {
            if (materialFeatureId) {
                var configedFeatures$ = _this.featureDataService.getConfigedValueByFeatureId(materialFeatureId, materialFeatureId);
                return configedFeatures$.pipe(map(function (defaultConfigs) {
                    var featureTemplate = featureSet.props;
                    var features = _this.mergeFeatures(featureTemplate, defaultConfigs);
                    return features;
                }));
            }
            else {
                return of(featureSet.props);
            }
        }));
    };
    FeatureEditService.prototype.mergeFeatures = function (featureTemplate, defaultConfigs) {
        return null;
    };
    FeatureEditService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    FeatureEditService.ctorParameters = function () { return [
        { type: Injector },
        { type: FeatureEditorService },
        { type: FeatureDataService },
        { type: FormLoadingService },
        { type: FormNotifyService },
        { type: LanguageService }
    ]; };
    return FeatureEditService;
}());
export { FeatureEditService };
