// @gsp-wf/wf-process-editor/esm2015/lib/app/process-designer.viewmodel.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector, ComponentFactoryResolver } from '@angular/core';
import { ProcessDesignerUIState } from './process-designer.uistate';
import { ProcessDesignerDataService } from '../domain/process-designer.dataservice';
import { forkJoin } from 'rxjs/observable/forkJoin';
import { ProcessClrTypeID } from '../model/process-clrtypeId';
import { BpmnProcessViewModel } from './bpmn-process.viewmodel';
import { WFFlowElementsViewModel } from './wf-flowElements.viewModel';
import { NotifyService } from '@farris/ui-notify';
import { MessagerService } from '@farris/ui-messager';
import { map, switchMap } from 'rxjs/operators';
import { FlowFormInfo } from '../domain/entities/flowform-info';
import { of, Subject, from } from 'rxjs';
import { BsModalService } from '@farris/ui-modal';
import { SaveAsEditComponent } from '../views/save-as-edit/save-as-edit.component';
import { ProcessDeUtil } from '../domain/process-de-util';
import { SubProcessViewModel } from './sub-process.viewmodel';
import { AutoActivityAdapter } from './auto-activity.adapter';
import { CatchEventAdapter } from './catch-event.adapter';
import { MessageActivityAdapter } from './message-activity.adapter';
import { AuditEvent, AuditProcessDefInfo } from '../domain/entities/audit-processDef-info';
import { HttpService } from '@ecp-caf/caf-common';
import { ProcessInfo } from '../model/process-info';
import { NotifyActivityAdapter } from './notify-activity.adapter';
import { AifCreationBillActivityAdapter } from './aif-creation-bill-activity.adapter';
export class ProcessDesignerViewModel {
    /**
     * @param {?} uiState
     * @param {?} service
     * @param {?} util
     * @param {?} notify
     * @param {?} message
     * @param {?} process
     * @param {?} activity
     * @param {?} subProcess
     * @param {?} autoActAdapter
     * @param {?} catchEventAdapter
     * @param {?} messageActAdapter
     * @param {?} notifyActAdapter
     * @param {?} creationActAdapter
     * @param {?} modalService
     * @param {?} injector
     * @param {?} resolver
     * @param {?} http
     */
    constructor(uiState, service, util, notify, message, process, activity, subProcess, autoActAdapter, catchEventAdapter, messageActAdapter, notifyActAdapter, creationActAdapter, modalService, injector, resolver, http) {
        this.uiState = uiState;
        this.service = service;
        this.util = util;
        this.notify = notify;
        this.message = message;
        this.process = process;
        this.activity = activity;
        this.subProcess = subProcess;
        this.autoActAdapter = autoActAdapter;
        this.catchEventAdapter = catchEventAdapter;
        this.messageActAdapter = messageActAdapter;
        this.notifyActAdapter = notifyActAdapter;
        this.creationActAdapter = creationActAdapter;
        this.modalService = modalService;
        this.injector = injector;
        this.resolver = resolver;
        this.http = http;
    }
    // 初始加载流程属性
    /**
     * @param {?} obj
     * @return {?}
     */
    initProcessProperty(obj) {
        /** @type {?} */
        const pptData = obj.propertyData;
        /** @type {?} */
        const ffId = pptData.EntryForm;
        if (ffId) {
            pptData.EntryForm = this.uiState.flowFormInfo.flowForm;
            this.uiState.propertyConfig = this.process.setProcessConfig(obj.propertyConfig);
            this.uiState.propertyData = pptData;
        }
        else {
            this.uiState.propertyConfig = this.process.setProcessConfig(obj.propertyConfig);
            this.uiState.propertyData = pptData;
        }
    }
    /**
     * @param {?} obj
     * @return {?}
     */
    setProperty(obj) {
        // 更新属性框数据与配置
        /** @type {?} */
        const propertyData = obj.propertyData;
        switch (propertyData.ClrTypeID) {
            case ProcessClrTypeID.process:
                this.process.setProcessProperty(obj);
                break;
            case ProcessClrTypeID.userActivity:
                this.activity.setUserActivityProperty(obj);
                break;
            case ProcessClrTypeID.subProcess:
                this.subProcess.setProperty(obj);
                break;
            case ProcessClrTypeID.autoActivity:
                this.autoActAdapter.setProperty(obj);
                break;
            case ProcessClrTypeID.catchEvent:
                this.catchEventAdapter.setProperty(obj);
                break;
            case ProcessClrTypeID.operationActivity:
                this.activity.setUserActivityProperty(obj);
                break;
            case ProcessClrTypeID.messageActivity:
                this.messageActAdapter.setProperty(obj);
                break;
            case ProcessClrTypeID.notifyActivity:
                this.notifyActAdapter.setProperty(obj);
                break;
            case ProcessClrTypeID.aifCreationBillActivity:
                this.creationActAdapter.setProperty(obj);
                break;
            default:
                this.activity.setFlowElementProperty(obj);
                break;
        }
    }
    /**
     * @return {?}
     */
    updateSchemas() {
        this.uiState.schemas = [];
        if (this.uiState.flowFormInfo.formalParameters && this.uiState.flowFormInfo.formalParameters.length > 0) {
            /** @type {?} */
            const schemaParams = this.uiState.flowFormInfo.formalParameters.filter((/**
             * @param {?} p
             * @return {?}
             */
            p => p.dataType === 'CDTSchema' && p.schemaMetadataId));
            forkJoin(schemaParams.map((/**
             * @param {?} para
             * @return {?}
             */
            para => {
                return this.service.getBusinessEntitybyId(para.schemaMetadataId);
            }))).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            data => {
                if (data && data.length > 0) {
                    data.forEach((/**
                     * @param {?} d
                     * @return {?}
                     */
                    (d) => {
                        this.uiState.schemas.push(d);
                    }));
                }
            }), (/**
             * @param {?} err
             * @return {?}
             */
            err => {
                this.uiState.schemas = [];
            }));
        }
        else {
            this.uiState.schemas = [];
        }
    }
    /**
     * @param {?} formId
     * @return {?}
     */
    getFlowFormInfoById(formId) {
        /** @type {?} */
        const $flowform = new Subject();
        if (formId) {
            forkJoin(this.service.GetFlowFormbyId(formId), this.service.getParametersbyFlowFormKey(formId).pipe(map((/**
             * @param {?} params
             * @return {?}
             */
            (params) => {
                this.setFormParamCxt(params);
                return params;
            }))), this.service.getComponentsbyFlowformkey(formId), this.service.getExecuteFormbyFlowFormKey(formId), this.service.getBizActivities(formId), this.service.getCustomizeActions(formId)).subscribe((/**
             * @param {?} re
             * @return {?}
             */
            (re) => {
                /** @type {?} */
                const flowform = new FlowFormInfo();
                flowform.flowForm = re[0];
                flowform.formalParameters = re[1];
                flowform.bizComponents = re[2].filter((/**
                 * @param {?} cmp
                 * @return {?}
                 */
                cmp => cmp.ownerType !== 'Activity'));
                flowform.excuteForms = re[3];
                flowform.bizActivities = re[4];
                flowform.customizeActions = re[5];
                this.uiState.flowFormInfo = flowform;
                // this.updateCOs();
                this.updateSchemas();
                $flowform.next(flowform);
            }));
        }
        return $flowform.asObservable();
    }
    /**
     * @return {?}
     */
    getFormFormatsByBizCategory() {
        if (this.uiState.flowFormInfo.flowForm.businessObjectId) {
            /** @type {?} */
            let url = `/api/runtime/wf/v1.0/formFormat/query?param=`;
            /** @type {?} */
            let param = `{"bizCategory":"${this.uiState.flowFormInfo.flowForm.businessObjectId}"}`;
            param = encodeURIComponent(param);
            url += param;
            return this.http.get(url);
        }
        else {
            return from(new Array());
        }
    }
    /**
     * 保存流程
     * @param {?} content 流程上下文
     * @param {?} processInfo
     * @param {?} state
     * @return {?}
     */
    saveProcess(content, processInfo, state) {
        return this.service.saveProcess(content, processInfo, state);
    }
    /**
     * 发布流程
     * @param {?} content 流程上下文
     * @return {?}
     */
    deployProcess(content) {
        return this.service.deployProcess(content);
    }
    /**
     * @param {?} auditId
     * @param {?} success
     * @param {?} key
     * @return {?}
     */
    auditProcessDef(auditId, success, key) {
        /** @type {?} */
        const info = new AuditProcessDefInfo();
        info.auditId = auditId;
        info.success = success;
        info.processDefKey = key;
        this.service.auditProcessDef(info);
    }
    /**
     * 保存运行时流程
     * @param {?} content 流程上下文
     * @param {?} procDefId
     * @return {?}
     */
    saveRTProcess(content, procDefId) {
        return this.service.saveRTProcess(content, procDefId);
    }
    /**
     * @param {?} process
     * @param {?} processInfo
     * @return {?}
     */
    saveAsProcess(process, processInfo) {
        /** @type {?} */
        const compFactory = this.resolver.resolveComponentFactory(SaveAsEditComponent);
        /** @type {?} */
        const compRef = compFactory.create(this.injector);
        compRef.instance.procKey = process.DefaultProcess.Key;
        compRef.instance.procName = process.DefaultProcess.Name + '_副本';
        compRef.instance.isAdministratorConfig = this.uiState.isAdministratorConfig;
        compRef.instance.organizations = this.uiState.organizations;
        compRef.instance.accessLevel = processInfo.accessLevel;
        compRef.instance.organization = processInfo.procOrg.orgId;
        compRef.instance.organizationName = processInfo.procOrg.orgName;
        /** @type {?} */
        const modalOptions = {
            title: '另存为',
            width: 450, height: 400,
            buttons: compRef.instance.btnRef,
            showButtons: true
        };
        /** @type {?} */
        const dlg = this.modalService.show(compRef, modalOptions);
        compRef.instance.submit.subscribe((/**
         * @param {?} re
         * @return {?}
         */
        (re) => {
            process.DefaultProcess.Name = re.procName;
            /** @type {?} */
            const procInfo = new ProcessInfo();
            procInfo.accessLevel = re.accessLevel;
            procInfo.procOrg = {
                orgId: re.organization,
                orgName: re.organizationName
            };
            this.saveProcess(JSON.stringify(process), procInfo, 1).subscribe((/**
             * @param {?} r
             * @return {?}
             */
            r => {
                this.notify.success((/** @type {?} */ ({
                    title: '系统提示', msg: '另存为成功', timeout: 3000
                })));
            }), (/**
             * @param {?} err
             * @return {?}
             */
            err => {
                this.notify.error((/** @type {?} */ ({
                    title: '系统提示', msg: '另存为失败，请检查流程是否正确！', timeout: 3000
                })));
            }));
            dlg.close();
        }));
        compRef.instance.cancel.subscribe((/**
         * @param {?} r
         * @return {?}
         */
        r => {
            dlg.close();
        }));
    }
    /**
     * 同步设计器时流程定义
     * @param {?} content
     * @param {?} rtProcessInfo
     * @return {?}
     */
    syncDtProcess(content, rtProcessInfo) {
        if (content) {
            /** @type {?} */
            const process = JSON.parse(content);
            this.message.question('是否将当前版本流程定义同步到设计时流程定义包！', (/**
             * @return {?}
             */
            () => {
                this.service.syncDtProcess(content, rtProcessInfo).subscribe((/**
                 * @param {?} data
                 * @return {?}
                 */
                (data) => {
                    this.notify.success((/** @type {?} */ ({
                        title: '系统提示', msg: '同步成功', timeout: 3000
                    })));
                    this.auditProcessDef(AuditEvent.WF_ProcessDef_SyncDT, true, process.DefaultProcess.Key);
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                (error) => {
                    this.notify.error((/** @type {?} */ ({
                        title: '系统提示', msg: '同步失败，请联系管理员查看原因！', timeout: 3000
                    })));
                    this.auditProcessDef(AuditEvent.WF_ProcessDef_SyncDT, false, process.DefaultProcess.Key);
                }));
            }));
        }
    }
    /**
     * @param {?} data
     * @param {?} processInfo
     * @return {?}
     */
    beforeCompleteStrategyChange(data, processInfo) {
        /** @type {?} */
        const url = `/api/runtime/wf/v1.0/processInstances/runtimeActivityInstances?procDefId=${processInfo.rtProcDefId}&activityDefinitionId=${data.Id}`;
        return this.http.get(url).pipe(switchMap((/**
         * @param {?} re
         * @return {?}
         */
        (re) => {
            if (re && re.length > 0) {
                this.message.info('当前节点存在运行中的活动实例，不允许调整完成策略，请了解！');
                return of(false);
            }
            else {
                return of(true);
            }
        })));
    }
    /**
     * @param {?} procCategoryId
     * @return {?}
     */
    getProcessCategory(procCategoryId) {
        return this.service.GetFlowFormbyId(procCategoryId);
    }
    /**
     * @private
     * @param {?} params
     * @return {?}
     */
    setFormParamCxt(params) {
        this.uiState.formalParameterContext = [];
        if (params && params.length > 0) {
            params.forEach((/**
             * @param {?} param
             * @return {?}
             */
            param => {
                if (param.dataType === 'CDTSchema') {
                    /** @type {?} */
                    const formalParameterContext = [{
                            key: `schemaId_${param.code}`,
                            name: `schemaId_${param.name}`,
                            description: `schemaId_${param.name}`
                        }, {
                            key: `metadataId_${param.code}`,
                            name: `metadataId_${param.name}`,
                            description: `metadataId_${param.name}`
                        }];
                    this.uiState.formalParameterContext = this.uiState.formalParameterContext.concat(formalParameterContext);
                }
            }));
        }
    }
}
ProcessDesignerViewModel.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ProcessDesignerViewModel.ctorParameters = () => [
    { type: ProcessDesignerUIState },
    { type: ProcessDesignerDataService },
    { type: ProcessDeUtil },
    { type: NotifyService },
    { type: MessagerService },
    { type: BpmnProcessViewModel },
    { type: WFFlowElementsViewModel },
    { type: SubProcessViewModel },
    { type: AutoActivityAdapter },
    { type: CatchEventAdapter },
    { type: MessageActivityAdapter },
    { type: NotifyActivityAdapter },
    { type: AifCreationBillActivityAdapter },
    { type: BsModalService },
    { type: Injector },
    { type: ComponentFactoryResolver },
    { type: HttpService }
];
if (false) {
    /** @type {?} */
    ProcessDesignerViewModel.prototype.uiState;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.service;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.util;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.notify;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.message;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.process;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.activity;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.subProcess;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.autoActAdapter;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.catchEventAdapter;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.messageActAdapter;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.notifyActAdapter;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.creationActAdapter;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.http;
}
