// @farris/command-services/esm5/lib/user-settings.service.js
import { Injectable, Injector, Optional } from '@angular/core';
import { FrmI18nSettingService } from '@gsp-sys/rtf-common';
var UserSettingsService = /** @class */ (function () {
    function UserSettingsService(injector, frmI18nSettingService) {
        this.injector = injector;
        this.frmI18nSettingService = frmI18nSettingService;
    }
    Object.defineProperty(UserSettingsService.prototype, "timeZone", {
        /**
         * 用户时区字符串
         */
        get: function () {
            return this.timezone && this.timezone.id || null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(UserSettingsService.prototype, "timeZoneOffset", {
        /**
         * 时区偏移
         */
        get: function () {
            var offset = this.timezone && this.timezone.baseOffset;
            if (offset !== null && offset !== undefined) {
                return offset;
            }
            else {
                return null;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(UserSettingsService.prototype, "defaultLanguage", {
        /**
         * 默认语言
         */
        get: function () {
            return this.i18nSetting && this.i18nSetting.defaultLanguage || null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(UserSettingsService.prototype, "dateFormat", {
        /**
         * 日期格式化字符串
         * like: yyyy-MM-dd
         */
        get: function () {
            return this.i18nSetting && this.i18nSetting.userDateFormat || null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(UserSettingsService.prototype, "timeFormat", {
        /**
         * 时间格式
         * like: HH:MM:ss
         */
        get: function () {
            return this.i18nSetting && this.i18nSetting.userTimeFormat || null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(UserSettingsService.prototype, "dateTimeFormat", {
        get: function () {
            var dateFormat = this.i18nSetting && this.i18nSetting.userDateFormat || 'yyyy-MM-dd';
            var timeFormat = this.i18nSetting && this.i18nSetting.userTimeFormat || 'HH:mm:ss';
            return dateFormat + " " + timeFormat;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(UserSettingsService.prototype, "addressFormat", {
        /**
         * 地址格式
         */
        get: function () {
            return null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(UserSettingsService.prototype, "numberFormat", {
        /**
         * 数字格式
         */
        get: function () {
            if (this.i18nSetting && this.i18nSetting.userNumberFormat) {
                var userNumberFormat = this.i18nSetting.userNumberFormat;
                var numberFormat = {
                    negativeSign: userNumberFormat.negativeSign || '-',
                    numberDecimalDigits: userNumberFormat.numberDecimalDigits || 2,
                    numberDecimalSeparator: userNumberFormat.numberDecimalSeparator || '.',
                    numberGroupSeparator: userNumberFormat.numberGroupSeparator || ','
                };
                return numberFormat;
            }
            return null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(UserSettingsService.prototype, "regionCode", {
        /**
         * 区域代码
         */
        get: function () {
            return this.i18nSetting && this.i18nSetting.userRegionCode || null;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(UserSettingsService.prototype, "userSettings", {
        /**
         * 用户配置
         */
        get: function () {
            var userSettings = this.frmI18nSettingService && this.frmI18nSettingService.getSetting();
            return userSettings;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(UserSettingsService.prototype, "timezone", {
        get: function () {
            return this.i18nSetting && this.i18nSetting.timezone;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(UserSettingsService.prototype, "i18nSetting", {
        get: function () {
            return this.userSettings && this.userSettings.i18nSetting;
        },
        enumerable: true,
        configurable: true
    });
    UserSettingsService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    UserSettingsService.ctorParameters = function () { return [
        { type: Injector },
        { type: FrmI18nSettingService, decorators: [{ type: Optional }] }
    ]; };
    return UserSettingsService;
}());
export { UserSettingsService };
