// @gsp-wf/wf-process-editor/esm5/lib/views/assign-performer-config/assign-performer-config.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, Input, ViewChild, EventEmitter, TemplateRef, HostBinding } from '@angular/core';
import { FormGroup, FormControl } from '@angular/forms';
import { AssignPerformerConfig } from './assign-performer-config.entity';
import { OrganizationPotentialOwner, ActionScope } from '../organization-potential-owner/actionOrganizationScope';
var AssignPerformerConfigComponent = /** @class */ (function () {
    function AssignPerformerConfigComponent() {
        this.cls = 'farris-main-area flex-column';
        this.closeModal = new EventEmitter();
        this.submitModal = new EventEmitter();
        this.allowGlobalAssign = false;
        this.operatorCollection = [
            { text: '当前办理人', value: 'currentUser' },
            { text: '流程发起人', value: 'procInstCreator' }
        ];
        this.operator = 'currentUser';
        this.organizationLevelCollection = [
            { value: 'GLOBAL', text: '全员', orgId: 'global' },
            { value: 'UNIT', text: '本单位', orgId: 'currentUnit' },
            { value: 'DEPARTMENT', text: '本部门', orgId: 'currentDepartment' }
        ];
        this.organizationLevel = this.organizationLevelCollection[0].value;
        this.includeChildren = false;
        this.modalConfig = {
            title: '指派规则',
            width: 450,
            height: 470,
            showButtons: true
        };
    }
    Object.defineProperty(AssignPerformerConfigComponent.prototype, "value", {
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            this.assingPerformerConfigForm = new FormGroup({
                multiSelect: new FormControl(v.multiSelect),
                necessary: new FormControl(v.necessary)
            });
            this.operator = v.operator || 'currentUser';
            this.allowGlobalAssign = v.allowGlobalAssign || false;
            if (v.assignScope) {
                this.organizationLevel = v.assignScope.organizationLevel;
                this.includeChildren = v.assignScope.includeChildren;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    AssignPerformerConfigComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    AssignPerformerConfigComponent.prototype.clickOK = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.config = {
            operator: this.operator,
            multiSelect: this.assingPerformerConfigForm.get('multiSelect').value,
            necessary: this.assingPerformerConfigForm.get('necessary').value,
            allowGlobalAssign: this.allowGlobalAssign
        };
        if (this.allowGlobalAssign && this.organizationLevel !== this.organizationLevelCollection[0].value) {
            /** @type {?} */
            var potential = new OrganizationPotentialOwner();
            potential.actionScope = ActionScope.AssignUser;
            potential.organizationLevel = (/** @type {?} */ (this.organizationLevel));
            potential.organizationId = this.organizationLevelCollection.find((/**
             * @param {?} o
             * @return {?}
             */
            function (o) { return o.value === _this.organizationLevel; })).orgId;
            potential.organizationName = this.organizationLevelCollection.find((/**
             * @param {?} o
             * @return {?}
             */
            function (o) { return o.value === _this.organizationLevel; })).text;
            potential.includeChildren = this.includeChildren;
            potential.tag = 'variable';
            this.config.assignScope = potential;
        }
        /** @type {?} */
        var object = { value: this.config, parameters: null };
        this.submitModal.emit(object);
    };
    /**
     * @return {?}
     */
    AssignPerformerConfigComponent.prototype.clickCancel = /**
     * @return {?}
     */
    function () {
        this.closeModal.emit();
    };
    /**
     * @param {?} event
     * @return {?}
     */
    AssignPerformerConfigComponent.prototype.valueChange = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        switch (event.value) {
            case this.organizationLevelCollection[0].value:
                this.includeChildren = false;
                break;
            case this.organizationLevelCollection[1].value:
                this.includeChildren = true;
                break;
            case this.organizationLevelCollection[2].value:
                this.includeChildren = false;
                break;
        }
    };
    AssignPerformerConfigComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-assign-performer-config',
                    template: "<div class=\"f-page\">\r\n  <div class=\"f-utils-fill-flex-column farris-split-section m-2 h-100\">\r\n    <section class=\"f-utils-fill-flex-column h-100\">\r\n      <div class=\"f-section-header pt-2 mx-2\">\r\n        <div class=\"f-title\">\r\n          <h4 class=\"f-title-text\">\u6307\u6D3E\u89C4\u5219\u9009\u9879</h4>\r\n        </div>\r\n      </div>\r\n      <div class=\"f-utils-fill-flex-column h-100\">\r\n        <form class=\"h-100\" [formGroup]=\"assingPerformerConfigForm\">\r\n          <div class=\"f-utils-fill-flex-column justify-content-center h-100 pb-4\">\r\n            <div class=\"my-2 mx-5\">\r\n              <div class=\"farris-group-wrap\">\r\n                <div class=\"form-group\">\r\n                  <label class=\"col-form-label\">\r\n                    <span class=\"farris-label-text\">\u6307\u6D3E\u64CD\u4F5C\u4EBA</span>\r\n                  </label>\r\n                  <div class=\"farris-input-wrap mt-1\" style=\"width: 300px;\">\r\n                    <farris-combo-list [(ngModel)]=\"operator\" name=\"operator\" [ngModelOptions]=\"{standalone: true}\"\r\n                      [idField]=\"'value'\" [textField]=\"'text'\" [valueField]=\"'value'\" [data]=\"operatorCollection\"\r\n                      [panelHeight]=\"60\">\r\n                    </farris-combo-list>\r\n                    <div class=\"farris-feedback valid-feedback\"></div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n            <div class=\"my-2 mx-5\">\r\n              <div class=\"farris-input-wrap\">\r\n                <div class=\"custom-control custom-checkbox\">\r\n                  <input class=\"custom-control-input\" type=\"checkbox\" id=\"multiSelect\" formControlName=\"multiSelect\"\r\n                    [readOnly]=\"false\" />\r\n                  <label class=\"custom-control-label\" for=\"multiSelect\">\u5141\u8BB8\u9009\u62E9\u591A\u4E2A\u5F85\u9009\u7528\u6237</label>\r\n                </div>\r\n              </div>\r\n            </div>\r\n            <div class=\"my-2 mx-5\">\r\n              <div class=\"farris-input-wrap\">\r\n                <div class=\"custom-control custom-checkbox\">\r\n                  <input class=\"custom-control-input\" type=\"checkbox\" id=\"necessary\" formControlName=\"necessary\" [readOnly]=\"false\" />\r\n                  <label class=\"custom-control-label\" for=\"necessary\">\u53EA\u6709\u4E00\u4E2A\u5F85\u9009\u7528\u6237\u65F6\uFF0C\u4ECD\u9700\u624B\u5DE5\u9009\u62E9</label>\r\n                </div>\r\n              </div>\r\n            </div>\r\n            <div class=\"mx-2 mt-3\">\r\n              <div class=\"title-content\">\r\n                <h4 class=\"title-text\">\u8BBE\u5B9A\u7684\u53C2\u4E0E\u8005\u8303\u56F4\u5916\u6307\u6D3E</h4><span class=\"title-line\"></span>\r\n              </div>\r\n            </div>\r\n            <div class=\"farris-form form-inline farris-form-inline my-1\">\r\n              <div class=\"col-12 mt-3\">\r\n                <div class=\"farris-group-wrap\">\r\n                  <div class=\"form-group farris-form-group\">\r\n                    <label class=\"col-form-label\">\r\n                      <span class=\"farris-label-text\">\u662F\u5426\u542F\u7528</span>\r\n                    </label>\r\n                    <div class=\"farris-input-wrap ml-1\">\r\n                      <farris-switch [(ngModel)]=\"allowGlobalAssign\" [ngModelOptions]=\"{standalone: true}\" name=\"allowGlobalAssign\"></farris-switch>\r\n                      <div class=\"farris-feedback valid-feedback\"></div>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n              <div class=\"d-flex flex-row\">\r\n                <div class=\"col-8 my-2\">\r\n                  <div class=\"farris-group-wrap\">\r\n                    <div class=\"form-group farris-form-group\">\r\n                      <label class=\"col-form-label\">\r\n                        <span class=\"farris-label-text\">\u7EC4\u7EC7\u8303\u56F4</span>\r\n                      </label>\r\n                      <div class=\"farris-input-wrap ml-1\">\r\n                        <farris-combo-list [panelHeight]=\"100\" [(ngModel)]=\"organizationLevel\" [ngModelOptions]=\"{standalone: true}\"\r\n                          name=\"organizationLevel\" [idField]=\"'value'\" [textField]=\"'text'\" [valueField]=\"'value'\"\r\n                          [data]=\"organizationLevelCollection\" [readonly]=\"allowGlobalAssign===false\" (valueChange)=\"valueChange($event)\">\r\n                        </farris-combo-list>\r\n                        <div class=\"farris-feedback valid-feedback\"></div>\r\n                      </div>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n                <div class=\"col-4 my-2\">\r\n                  <div class=\"farris-input-wrap pt-1\">\r\n                    <div class=\"custom-control custom-checkbox mr-3\">\r\n                      <input class=\"custom-control-input\" type=\"checkbox\" [(ngModel)]=\"includeChildren\" [ngModelOptions]=\"{standalone: true}\"\r\n                        name=\"includeChildren\" id=\"includeChildren\" [disabled]=\"!(allowGlobalAssign===true && organizationLevel === organizationLevelCollection[2].value)\" />\r\n                      <label class=\"custom-control-label\" for=\"includeChildren\">\u5305\u542B\u4E0B\u7EA7</label>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </form>\r\n      </div>\r\n    </section>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #iconSelectFooter>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm px-3 mx-2\" (click)=\"clickOK()\">\u786E\u5B9A</button>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm px-3 mx-2\" (click)=\"clickCancel()\">\u53D6\u6D88</button>\r\n</ng-template>\r\n\r\n",
                    styles: [".title-content{color:#000;display:inline-flex;align-items:center;line-height:1.375rem;position:relative;overflow:hidden;padding:0 0 0 1rem;min-width:415px}.title-content ::before{content:'';width:.25rem;height:1.125rem;background:#59a1ff;border-radius:1px;position:absolute;top:50%;left:0;margin-top:-.5625rem}.title-text{font-size:1rem;line-height:1.375rem;overflow:hidden;white-space:nowrap;text-overflow:ellipsis;margin-bottom:0;margin-right:10px}.title-line{flex-grow:1;border-bottom:1px dashed rgba(0,0,0,.1)}"]
                }] }
    ];
    /** @nocollapse */
    AssignPerformerConfigComponent.ctorParameters = function () { return []; };
    AssignPerformerConfigComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        closeModal: [{ type: Output }],
        submitModal: [{ type: Output }],
        value: [{ type: Input }],
        modalFooter: [{ type: ViewChild, args: ['iconSelectFooter',] }]
    };
    return AssignPerformerConfigComponent;
}());
export { AssignPerformerConfigComponent };
if (false) {
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.cls;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.closeModal;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.submitModal;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.assingPerformerConfigForm;
    /**
     * @type {?}
     * @private
     */
    AssignPerformerConfigComponent.prototype.config;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.modalFooter;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.allowGlobalAssign;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.operatorCollection;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.operator;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.organizationLevelCollection;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.organizationLevel;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.includeChildren;
    /** @type {?} */
    AssignPerformerConfigComponent.prototype.modalConfig;
}
var AssignConfigConverter = /** @class */ (function () {
    function AssignConfigConverter() {
    }
    /**
     * @param {?} data
     * @return {?}
     */
    AssignConfigConverter.prototype.convertTo = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        return '（已设置）';
    };
    return AssignConfigConverter;
}());
export { AssignConfigConverter };
