// @farris/ui-lookup/esm5/lib/lookup-grid-options.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
var LookupLayout = {
    /** 经典，默认布局 */
    default: 'default',
    /** 帮助窗口布局支持高级查询， */
    advanced: 'advanced',
};
export { LookupLayout };
/**
 * @record
 */
export function TreeInfo() { }
if (false) {
    /**
     * 数据加载方式： all, async
     * @type {?|undefined}
     */
    TreeInfo.prototype.loadDataType;
    /**
     * 分层方式： pathcode, parentId
     * @type {?|undefined}
     */
    TreeInfo.prototype.layerType;
    /**
     * 数据中带有分级信息的字段
     * @type {?|undefined}
     */
    TreeInfo.prototype.dataField;
    /**
     * 父ID字段
     * @type {?|undefined}
     */
    TreeInfo.prototype.parentField;
    /**
     * 分级码字段
     * @type {?|undefined}
     */
    TreeInfo.prototype.pathField;
    /**
     * 级数字段
     * @type {?|undefined}
     */
    TreeInfo.prototype.layerField;
    /**
     * 是否明细字段
     * @type {?|undefined}
     */
    TreeInfo.prototype.isDetailField;
    /**
     * 只允许选择叶子节点
     * - default: 应用服务器端设置，如服务器未设置，功能同  `no`
     * - `yes`: 服务器端设置失效。仅允许选择叶子节点。
     * - `no`:  服务器端设置失效。所有节点均可选择。
     * @type {?|undefined}
     */
    TreeInfo.prototype.onlySelectLeaf;
    /**
     * 树结构数据是否已初始化完成
     * @type {?|undefined}
     */
    TreeInfo.prototype.treeDataIsInit;
}
/**
 * @record
 */
export function CustomStyleData() { }
if (false) {
    /** @type {?} */
    CustomStyleData.prototype.field;
    /** @type {?} */
    CustomStyleData.prototype.value;
    /** @type {?|undefined} */
    CustomStyleData.prototype.comparator;
    /** @type {?} */
    CustomStyleData.prototype.styles;
}
/**
 * @record
 */
export function LookupGridResult() { }
if (false) {
    /**
     * 列信息
     * @type {?|undefined}
     */
    LookupGridResult.prototype.columns;
    /**
     * 数据列表
     * @type {?|undefined}
     */
    LookupGridResult.prototype.items;
    /**
     * 查询字段
     * @type {?|undefined}
     */
    LookupGridResult.prototype.searchFields;
    /**
     * 分页信息
     * @type {?|undefined}
     */
    LookupGridResult.prototype.pageInfo;
    /**
     * 总记录数
     * @type {?|undefined}
     */
    LookupGridResult.prototype.total;
    /**
     * 废弃
     * @type {?|undefined}
     */
    LookupGridResult.prototype.mapFields;
    /**
     * 树分级信息
     * @type {?|undefined}
     */
    LookupGridResult.prototype.treeInfo;
    /** @type {?|undefined} */
    LookupGridResult.prototype.displayType;
    /** @type {?|undefined} */
    LookupGridResult.prototype.displayInfo;
    /** @type {?|undefined} */
    LookupGridResult.prototype.title;
    /** @type {?|undefined} */
    LookupGridResult.prototype.navigation;
    /** @type {?|undefined} */
    LookupGridResult.prototype.valueField;
    /** @type {?|undefined} */
    LookupGridResult.prototype.textField;
    /** @type {?|undefined} */
    LookupGridResult.prototype.idField;
    /**
     * 已选择数据列表
     * @type {?|undefined}
     */
    LookupGridResult.prototype.selectedItems;
    /**
     * 收藏数据
     * @type {?|undefined}
     */
    LookupGridResult.prototype.favoriteItems;
    /**
     * 自定义行样式
     * @type {?|undefined}
     */
    LookupGridResult.prototype.customStyles;
}
/**
 * 个性化数据
 * @record
 */
export function PersonalConfig() { }
if (false) {
    /** @type {?|undefined} */
    PersonalConfig.prototype.tabIndex;
    /** @type {?|undefined} */
    PersonalConfig.prototype.favorite;
    /** @type {?|undefined} */
    PersonalConfig.prototype.quickSelected;
    /** @type {?|undefined} */
    PersonalConfig.prototype.size;
    /** @type {?|undefined} */
    PersonalConfig.prototype.selections;
    /** @type {?|undefined} */
    PersonalConfig.prototype.cascadeStatus;
    /* Skipping unhandled member: [key: string]: any;*/
}
/**
 * @record
 */
export function NavigationOptions() { }
if (false) {
    /** @type {?} */
    NavigationOptions.prototype.displayType;
    /** @type {?} */
    NavigationOptions.prototype.width;
    /** @type {?|undefined} */
    NavigationOptions.prototype.idField;
    /** @type {?|undefined} */
    NavigationOptions.prototype.relations;
    /** @type {?|undefined} */
    NavigationOptions.prototype.hover;
    /* Skipping unhandled member: [key: string]: any;*/
}
/**
 * @record
 */
export function LookupGridOptions() { }
if (false) {
    /** @type {?|undefined} */
    LookupGridOptions.prototype.singleSelect;
    /** @type {?|undefined} */
    LookupGridOptions.prototype.idField;
    /** @type {?|undefined} */
    LookupGridOptions.prototype.uri;
    /** @type {?|undefined} */
    LookupGridOptions.prototype.showFilterBar;
    /** @type {?|undefined} */
    LookupGridOptions.prototype.pagination;
    /** @type {?|undefined} */
    LookupGridOptions.prototype.pageIndex;
    /** @type {?|undefined} */
    LookupGridOptions.prototype.pageSize;
    /** @type {?|undefined} */
    LookupGridOptions.prototype.pageList;
    /** @type {?|undefined} */
    LookupGridOptions.prototype.total;
    /** @type {?|undefined} */
    LookupGridOptions.prototype.width;
    /** @type {?|undefined} */
    LookupGridOptions.prototype.height;
    /** @type {?|undefined} */
    LookupGridOptions.prototype.hover;
}
var DisplayInfo = /** @class */ (function () {
    function DisplayInfo() {
    }
    return DisplayInfo;
}());
export { DisplayInfo };
if (false) {
    /** @type {?} */
    DisplayInfo.prototype.title;
    /** @type {?} */
    DisplayInfo.prototype.favorites;
    /** @type {?} */
    DisplayInfo.prototype.okText;
    /** @type {?} */
    DisplayInfo.prototype.cancelText;
    /** @type {?} */
    DisplayInfo.prototype.allColumns;
}
/** @type {?} */
export var lookupGridDefaults = {
    singleSelect: true,
    showFilterBar: true,
    pagination: true,
    pageIndex: 1,
    pageSize: 10,
    pageList: [10, 20, 30, 50, 100]
};
/** @type {?} */
export var displayInfoDefault = {
    title: '此处显示标题',
    favorites: '收藏夹',
    okText: '确定',
    cancelText: '取消',
    allColumns: '所有列'
};
/**
 * 帮助前事件返回类型
 * @record
 */
export function PickingResult() { }
if (false) {
    /**
     * true: 继续打开帮助；false: 不打开帮助窗口
     * @type {?}
     */
    PickingResult.prototype.showDialog;
    /**
     * 携带的一些自定义数据，由业务在服务器端帮助前事件中处理
     * @type {?|undefined}
     */
    PickingResult.prototype.data;
    /** @type {?|undefined} */
    PickingResult.prototype.message;
}
/**
 * @record
 */
export function PickedResult() { }
if (false) {
    /** @type {?} */
    PickedResult.prototype.closeDialog;
    /** @type {?|undefined} */
    PickedResult.prototype.message;
}
