// @gsp-wf/wf-process-editor/esm5/lib/views/performer/views/position-performer/position-performer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, HostBinding, ViewChild, ElementRef, NgZone } from '@angular/core';
import { HttpService } from '@ecp-caf/caf-common';
import { SysFarrisLookupComponent } from '@gsp-sys/sysmgr-ui';
import { ProcessDesignerUIState } from '../../../../app/process-designer.uistate';
import { ExprEditMode, ExpressionService, InjectedType } from '@gsp-svc/expression';
import { PositionPerformerEntity } from '../../entities/position-performer.entity';
var PositionPerformerComponent = /** @class */ (function () {
    function PositionPerformerComponent(http, designerUIState, expr, ngZone) {
        this.http = http;
        this.designerUIState = designerUIState;
        this.expr = expr;
        this.ngZone = ngZone;
        this.cls = 'd-flex flex-fill h-100';
        this.showClearButton = false;
        this.groupIcon = '<i class="f-icon f-icon-lookup"></i>';
        this.positionTypes = [
            { value: '0', text: '组织岗' },
            { value: '1', text: '通用岗' }
        ];
        this.filter = {};
        this.position = new PositionPerformerEntity();
    }
    /**
     * @return {?}
     */
    PositionPerformerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    PositionPerformerComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.ngZone.runOutsideAngular((/**
         * @return {?}
         */
        function () {
            _this.inputGroup.nativeElement.addEventListener('mouseenter', _this.onMouseEnter.bind(_this));
            _this.inputGroup.nativeElement.addEventListener('mouseleave', _this.onMouseLeave.bind(_this));
        }));
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PositionPerformerComponent.prototype.onMouseEnter = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.showClearButton = true;
        if (this.showClearButton) {
            this.toggleClearIcon(event, true);
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PositionPerformerComponent.prototype.onMouseLeave = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.showClearButton = false;
        this.toggleClearIcon(event, false);
    };
    /**
     * @private
     * @param {?} event
     * @param {?=} isShow
     * @return {?}
     */
    PositionPerformerComponent.prototype.toggleClearIcon = /**
     * @private
     * @param {?} event
     * @param {?=} isShow
     * @return {?}
     */
    function (event, isShow) {
        if (isShow === void 0) { isShow = false; }
        /** @type {?} */
        var str = isShow ? '' : 'none';
        /** @type {?} */
        var clearIcon = event.target.querySelector('.input-group-clear');
        if (clearIcon) {
            clearIcon.style.display = str;
        }
    };
    /**
     * @return {?}
     */
    PositionPerformerComponent.prototype.clear = /**
     * @return {?}
     */
    function () {
        this.position.performerId = null;
        this.position.performerName = null;
    };
    /**
     * @param {?} data
     * @return {?}
     */
    PositionPerformerComponent.prototype.positionTypeChange = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (data.value === '0') {
            this.filter = { positionType: 0 };
        }
        else {
            this.filter = { positionType: 1 };
        }
        this.clear();
        this.position.orgType = null;
        this.position.orgId = null;
    };
    /**
     * @return {?}
     */
    PositionPerformerComponent.prototype.selectOrganition = /**
     * @return {?}
     */
    function () {
        var _this = this;
        var e_1, _a;
        if (this.position.positionType === '1') {
            if (this.designerUIState.schemas.length > 0) {
                try {
                    for (var _b = tslib_1.__values(this.designerUIState.schemas), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var schema = _c.value;
                        this.expr.addSchema('Schema', schema);
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
                if (this.designerUIState.schemas[0].entityTypes && this.designerUIState.schemas[0].entityTypes.length > 0) {
                    /** @type {?} */
                    var entityType = this.designerUIState.schemas[0].entityTypes[0];
                    this.expr.addInjectedEntity(entityType.name, 'Schema', entityType.name, InjectedType.EntityType);
                }
            }
            this.expr.addExpressionText(this.position.orgId);
            this.expr.buildExpression().then((/**
             * @param {?} expr
             * @return {?}
             */
            function (expr) {
                _this.position.orgId = expr;
            }));
        }
    };
    /**
     * @return {?}
     */
    PositionPerformerComponent.prototype.selectPosition = /**
     * @return {?}
     */
    function () {
        this.posiHelp.showHelp();
    };
    /**
     * @return {?}
     */
    PositionPerformerComponent.prototype.editCondition = /**
     * @return {?}
     */
    function () {
        var _this = this;
        var e_2, _a, e_3, _b, e_4, _c, e_5, _d, e_6, _e, e_7, _f, e_8, _g;
        if (this.designerUIState.schemas.length > 0) {
            try {
                for (var _h = tslib_1.__values(this.designerUIState.schemas), _j = _h.next(); !_j.done; _j = _h.next()) {
                    var schema = _j.value;
                    this.expr.addSchema('Schema', schema);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_j && !_j.done && (_a = _h.return)) _a.call(_h);
                }
                finally { if (e_2) throw e_2.error; }
            }
            if (this.designerUIState.schemas[0].entityTypes && this.designerUIState.schemas[0].entityTypes.length > 0) {
                /** @type {?} */
                var entityType = this.designerUIState.schemas[0].entityTypes[0];
                this.expr.addInjectedEntity(entityType.name, 'Schema', entityType.name, InjectedType.EntityType);
            }
        }
        /** @type {?} */
        var context = new Array();
        if (this.designerUIState.procInstContext.length > 0) {
            try {
                for (var _k = tslib_1.__values(this.designerUIState.procInstContext), _l = _k.next(); !_l.done; _l = _k.next()) {
                    var variable = _l.value;
                    context.push(variable);
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_l && !_l.done && (_b = _k.return)) _b.call(_k);
                }
                finally { if (e_3) throw e_3.error; }
            }
        }
        if (this.designerUIState.formalParameterContext.length > 0) {
            var _loop_1 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _m = tslib_1.__values(this.designerUIState.formalParameterContext), _o = _m.next(); !_o.done; _o = _m.next()) {
                    var variable = _o.value;
                    _loop_1(variable);
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (_o && !_o.done && (_c = _m.return)) _c.call(_m);
                }
                finally { if (e_4) throw e_4.error; }
            }
        }
        if (this.designerUIState.participantContext.length > 0) {
            var _loop_2 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _p = tslib_1.__values(this.designerUIState.participantContext), _q = _p.next(); !_q.done; _q = _p.next()) {
                    var variable = _q.value;
                    _loop_2(variable);
                }
            }
            catch (e_5_1) { e_5 = { error: e_5_1 }; }
            finally {
                try {
                    if (_q && !_q.done && (_d = _p.return)) _d.call(_p);
                }
                finally { if (e_5) throw e_5.error; }
            }
        }
        if (this.designerUIState.processVariableContext.length > 0) {
            var _loop_3 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _r = tslib_1.__values(this.designerUIState.processVariableContext), _s = _r.next(); !_s.done; _s = _r.next()) {
                    var variable = _s.value;
                    _loop_3(variable);
                }
            }
            catch (e_6_1) { e_6 = { error: e_6_1 }; }
            finally {
                try {
                    if (_s && !_s.done && (_e = _r.return)) _e.call(_r);
                }
                finally { if (e_6) throw e_6.error; }
            }
        }
        if (this.designerUIState.activityComponentContext.length > 0) {
            var _loop_4 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _t = tslib_1.__values(this.designerUIState.activityComponentContext), _u = _t.next(); !_u.done; _u = _t.next()) {
                    var variable = _u.value;
                    _loop_4(variable);
                }
            }
            catch (e_7_1) { e_7 = { error: e_7_1 }; }
            finally {
                try {
                    if (_u && !_u.done && (_f = _t.return)) _f.call(_t);
                }
                finally { if (e_7) throw e_7.error; }
            }
        }
        if (context.length > 0) {
            try {
                for (var context_1 = tslib_1.__values(context), context_1_1 = context_1.next(); !context_1_1.done; context_1_1 = context_1.next()) {
                    var v = context_1_1.value;
                    this.expr.addContext(v);
                }
            }
            catch (e_8_1) { e_8 = { error: e_8_1 }; }
            finally {
                try {
                    if (context_1_1 && !context_1_1.done && (_g = context_1.return)) _g.call(context_1);
                }
                finally { if (e_8) throw e_8.error; }
            }
        }
        this.expr.setEditPattern(ExprEditMode.Conditional); // 表达式展现模式：默认简易版
        this.expr.addExpressionText(this.position.condition);
        this.expr.buildExpression().then((/**
         * @param {?} expressioninfo
         * @return {?}
         */
        function (expressioninfo) {
            _this.position.condition = expressioninfo;
        }));
    };
    /**
     * @param {?} re
     * @return {?}
     */
    PositionPerformerComponent.prototype.orgTypeChange = /**
     * @param {?} re
     * @return {?}
     */
    function (re) {
        this.position.orgType = re.data.value;
    };
    /**
     * @return {?}
     */
    PositionPerformerComponent.prototype.clearOrgType = /**
     * @return {?}
     */
    function () {
        this.position.orgType = null;
    };
    /**
     * @param {?} data
     * @return {?}
     */
    PositionPerformerComponent.prototype.positionsSelect = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (data) {
            this.position.performerId = data.id;
            this.position.performerName = data.name;
            this.position.type = 'wf.PositionParticipant';
            if (!this.position.positionType) {
                this.position.positionType = data.posType + '';
                this.filter = {
                    positionType: data.posType
                };
            }
        }
    };
    /**
     * @return {?}
     */
    PositionPerformerComponent.prototype.getPerformers = /**
     * @return {?}
     */
    function () {
        return this.position;
    };
    PositionPerformerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-position-performer',
                    template: "<div class=\"d-flex flex-fill farris-overflow-y-auto p-md-2\">\r\n  <form style=\"background: #F0F3F6; width: 100%;\">\r\n    <div class=\"d-flex m-2 flex-column w-100\">\r\n      <div class=\"d-flex flex-column mx-5 my-4\" style=\"background: #FFF;\">\r\n        <div class=\"farris-form form-inline  farris-form-inline mr-5\">\r\n          <div class=\"col-12 farris-group-auto my-3\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <!-- <span class=\"farris-label-info text-danger\">*</span> -->\r\n                  <span class=\"farris-label-text\">\u5C97\u4F4D\u7C7B\u578B</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <farris-combo-list [(ngModel)]=\"position.positionType\" name=\"positionType\" [panelHeight]=\"'auto'\" [idField]=\"'value'\"\r\n                    [textField]=\"'text'\" [data]=\"positionTypes\" (valueChange)=\"positionTypeChange($event)\"\r\n                    [panelHeight]=\"'auto'\" [enableClear]=\"false\">\r\n                  </farris-combo-list>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"col-12 farris-group-auto my-4\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-info text-danger\">*</span>\r\n                  <span class=\"farris-label-text\">\u5C97\u4F4D</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <div #inputGroup class=\"input-group\">\r\n                    <input type=\"text\" class=\"form-control\" [(ngModel)]=\"position.performerName\" name=\"positionName\" [readOnly]=\"true\" />\r\n                    <div class=\"input-group-append\">\r\n                      <span [style.display]=\"showClearButton ? '' : 'none'\" class=\"input-group-text input-group-clear\" (click)=\"clear()\">\r\n                        <i class=\"f-icon modal_close\"></i>\r\n                      </span>\r\n                      <span class=\"input-group-text\" (click)=\"selectPosition()\" style=\"background: #FFF;\">\r\n                        <i class=\"f-icon f-icon-lookup\"></i>\r\n                      </span>\r\n                    </div>\r\n                  </div>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"col-12 farris-group-auto my-4\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <!-- <span *ngIf=\"filter.positionType===1\" class=\"farris-label-info text-danger\">*</span> -->\r\n                  <span class=\"farris-label-text\">\u7EC4\u7EC7\u7C7B\u578B</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <farris-combo-list [(ngModel)]=\"position.orgType\" name=\"orgType\" [idField]=\"'value'\" [textField]=\"'text'\" [valueField]=\"'value'\"\r\n                    [data]=\"designerUIState.orgTypeCollection\" [panelHeight]=\"'auto'\" (clear)=\"clearOrgType()\"\r\n                    (selectChange)='orgTypeChange($event)' [readonly]=\"!(position.positionType==='1')\" [placeholder]=\"position.positionType==='1'?'\u8BF7\u9009\u62E9':''\">\r\n                  </farris-combo-list>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"col-12 farris-group-auto my-3\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <!-- <span *ngIf=\"filter.positionType===1\" class=\"farris-label-info text-danger\">*</span> -->\r\n                  <span class=\"farris-label-text\">\u7EC4\u7EC7</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <!-- <div class=\"input-group\">\r\n                    <input type=\"text\" class=\"form-control\" [(ngModel)]=\"position.orgId\" name=\"organization\" [readOnly]=\"true\" />\r\n                    <div class=\"input-group-append\">\r\n                      <span class=\"input-group-text\" (click)=\"selectOrganition()\" style=\"background: #FFF;\">\r\n                        <i class=\"f-icon f-icon-lookup\"></i>\r\n                      </span>\r\n                    </div>\r\n                  </div> -->\r\n                  <input-group class=\"form-control\" style=\"width: 100%;padding:0;border:0;\" [enableClear]=\"false\"\r\n                    [showButtonWhenDisabled]=\"false\" [readonly]=\"!(position.positionType==='1')\"\r\n                    [(ngModel)]=\"position.orgId\" name=\"organization\" [placeholder]=\"position.positionType==='1'?'\u8BF7\u9009\u62E9':''\"\r\n                    [editable]=\"false\" (clickHandle)=\"selectOrganition()\" [groupText]=\"groupIcon\"></input-group>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n      <div class=\"d-flex flex-column mx-5\" style=\"background: #FFF;\">\r\n        <div class=\"farris-form form-inline farris-form-inline mr-5\">\r\n          <div class=\"col-12 farris-group-auto my-3\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-text\">\u542F\u7528\u6761\u4EF6</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <div class=\"input-group\">\r\n                    <input type=\"text\" class=\"form-control\" [(ngModel)]=\"this.position.condition\" name=\"conditon\" [readOnly]=\"true\" />\r\n                    <div class=\"input-group-append\">\r\n                      <span class=\"input-group-text\" (click)=\"editCondition()\" style=\"background: #FFF;\">\r\n                        <i class=\"f-icon f-icon-lookup\"></i>\r\n                      </span>\r\n                    </div>\r\n                  </div>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n    </div>\r\n  </form>\r\n</div>\r\n\r\n<sys-farris-grid-lookup #posiHelp style=\"display: none;\" [singleSelect]=\"true\"\r\n  [pagination]=\"true\" [pageSize]=\"20\" [title]=\"'\u7CFB\u7EDF\u5C97\u4F4D'\" [sourceType]=\"'position'\"\r\n  (afterConfirm)=\"positionsSelect($event)\" [filter]=\"filter\">\r\n</sys-farris-grid-lookup>\r\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    PositionPerformerComponent.ctorParameters = function () { return [
        { type: HttpService },
        { type: ProcessDesignerUIState },
        { type: ExpressionService },
        { type: NgZone }
    ]; };
    PositionPerformerComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        posiHelp: [{ type: ViewChild, args: [SysFarrisLookupComponent,] }],
        inputGroup: [{ type: ViewChild, args: ['inputGroup',] }]
    };
    return PositionPerformerComponent;
}());
export { PositionPerformerComponent };
if (false) {
    /** @type {?} */
    PositionPerformerComponent.prototype.cls;
    /** @type {?} */
    PositionPerformerComponent.prototype.posiHelp;
    /** @type {?} */
    PositionPerformerComponent.prototype.inputGroup;
    /** @type {?} */
    PositionPerformerComponent.prototype.showClearButton;
    /** @type {?} */
    PositionPerformerComponent.prototype.groupIcon;
    /** @type {?} */
    PositionPerformerComponent.prototype.positionTypes;
    /** @type {?} */
    PositionPerformerComponent.prototype.condition;
    /** @type {?} */
    PositionPerformerComponent.prototype.filter;
    /** @type {?} */
    PositionPerformerComponent.prototype.position;
    /**
     * @type {?}
     * @private
     */
    PositionPerformerComponent.prototype.http;
    /** @type {?} */
    PositionPerformerComponent.prototype.designerUIState;
    /**
     * @type {?}
     * @private
     */
    PositionPerformerComponent.prototype.expr;
    /**
     * @type {?}
     * @private
     */
    PositionPerformerComponent.prototype.ngZone;
}
