// @farris/ui-datagrid/esm5/lib/components/body/datagrid-grouprow.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, Injector, Renderer2, ElementRef } from '@angular/core';
import { DatagridComponent } from './../../datagrid.component';
var DatagridGroupRowDirective = /** @class */ (function () {
    function DatagridGroupRowDirective(injector, render, el, dg) {
        this.injector = injector;
        this.render = render;
        this.el = el;
        this.dg = dg;
    }
    /**
     * @return {?}
     */
    DatagridGroupRowDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.dg && this.dg.groupStyler) {
            /** @type {?} */
            var styler = this.dg.groupStyler(this.groupRow);
            /** @type {?} */
            var tr = this.el.nativeElement;
            if (!tr.id) {
                tr.id = 'group-tr_' + this.groupRow.id;
            }
            this.dg.renderCustomStyle(styler, tr);
        }
    };
    DatagridGroupRowDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[group-row]',
                },] }
    ];
    /** @nocollapse */
    DatagridGroupRowDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: Renderer2 },
        { type: ElementRef },
        { type: DatagridComponent }
    ]; };
    DatagridGroupRowDirective.propDecorators = {
        groupRow: [{ type: Input, args: ['group-row',] }]
    };
    return DatagridGroupRowDirective;
}());
export { DatagridGroupRowDirective };
if (false) {
    /** @type {?} */
    DatagridGroupRowDirective.prototype.groupRow;
    /**
     * @type {?}
     * @private
     */
    DatagridGroupRowDirective.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    DatagridGroupRowDirective.prototype.render;
    /**
     * @type {?}
     * @private
     */
    DatagridGroupRowDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    DatagridGroupRowDirective.prototype.dg;
}
