// @gsp-wf/wf-process-editor/esm5/lib/views/performer/views/performer-edit/performer-edit.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, ViewChild, Input } from '@angular/core';
import { UserPerformerComponent } from '../user-performer/user-performer.component';
import { PositionPerformerComponent } from '../position-performer/position-performer.component';
import { ComponentPerformerComponent } from '../component-performer/component-performer.component';
import { ReportingLineComponent } from '../reporting-line/reporting-line.component';
import { MatrixPerformerComponent } from '../matrix-performer/matrix-performer.component';
import { HttpService } from "@ecp-caf/caf-common";
var PerformerEditComponent = /** @class */ (function () {
    function PerformerEditComponent(http) {
        this.http = http;
        this.cls = 'farris-main-area flex-column';
        this.tab = 'user';
        this.display = true;
    }
    /**
     * @return {?}
     */
    PerformerEditComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    PerformerEditComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.performer) {
            switch (this.performer.type) {
                case 'wf.UserParticipant':
                    this.tab = 'user';
                    this.usersCmp.users = [];
                    this.usersCmp.users.push(this.performer.value);
                    // 考虑到兼容历史数据，暂时不使用用户参与者实体类中的userType属性
                    if (this.performer.value.performerId && this.performer.value.performerId.charAt(0) === '{') {
                        /** @type {?} */
                        var url = '/api/runtime/wf/v1.0/reportingOrgType/hasAdminOrg';
                        this.http.get(url).subscribe((/**
                         * @param {?} data
                         * @return {?}
                         */
                        function (data) {
                            _this.usersCmp.userTypeButtonNeedShow = data;
                            _this.usersCmp.userType = 'variable';
                            _this.usersCmp.curOrgType = _this.performer.value.orgType === 'Sys' ? 'Sys' : 'Administrative_organization';
                            _this.usersCmp.condition = _this.performer.value.condition;
                            // 根据表达式中的函数名判断是否使用的是上下文变量
                            /** @type {?} */
                            var isContextVariable = _this.performer.value.performerId.indexOf("GetContextParameter") >= 0;
                            _this.usersCmp.variableOrgTypes = isContextVariable ? [
                                { name: '系统用户', value: 'Sys' }
                            ] : [
                                { name: '系统用户', value: 'Sys' },
                                { name: '行政人员', value: 'Administrative_organization' }
                            ];
                        }));
                    }
                    else {
                        this.usersCmp.userType = 'user';
                        this.usersCmp.condition = this.performer.value.condition;
                    }
                    break;
                case 'wf.PositionParticipant':
                    this.tab = 'position';
                    this.posiCmp.position = Object.assign({}, this.performer.value);
                    if (this.posiCmp.position.positionType === '0') {
                        this.posiCmp.filter = { positionType: 0 };
                    }
                    else {
                        this.posiCmp.filter = { positionType: 1 };
                    }
                    break;
                case 'wf.ReportingLineParticipant':
                    this.tab = 'report';
                    this.reportingLineCmp.setReportingLineDataEntity = this.performer.value;
                    break;
                case 'wf.ComponentParticipant':
                    this.tab = 'component';
                    this.cmpCmp.component = this.performer.value;
                    break;
                case 'wf.MatrixParticipant':
                    this.tab = 'matrix';
                    this.cmpCmp.component = this.performer.value;
                    this.matrixCmp.setMatrixDataEntity = this.performer.value;
            }
        }
    };
    /**
     * @return {?}
     */
    PerformerEditComponent.prototype.getPerformers = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var users = this.usersCmp.getPerformers();
        /** @type {?} */
        var position = this.posiCmp.getPerformers();
        /** @type {?} */
        var component = this.cmpCmp.getPerformers();
        /** @type {?} */
        var reportingLine = this.reportingLineCmp.getPerformers();
        /** @type {?} */
        var matrix = this.matrixCmp.getPerformers();
        // let performers = [];
        /** @type {?} */
        var positionPerformers = [];
        // 只包含岗位参与者
        /** @type {?} */
        var userPerformers = [];
        if (users && users.length > 0) {
            userPerformers = userPerformers.concat(users);
        }
        if (position && position.performerId) {
            positionPerformers.push(position);
        }
        return { userPerformers: userPerformers, positionPerformers: positionPerformers, component: component, reportingLine: reportingLine, matrix: matrix };
    };
    PerformerEditComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-performer-edit',
                    template: "<div class=\"flex-fill farris-split-section m-2 farris-overflow-y-auto f-utils-flex-column\">\r\n    <farris-tabs class=\"f-utils-fill h-100\" #tabs [value]=\"tab\" [fill]=\"true\" [contentFill]=\"true\">\r\n      <farris-tab title=\"\u7528\u6237\" [show]=\"true\" [id]=\"'user'\">\r\n        <ng-template #content>\r\n          <app-user-performer></app-user-performer>\r\n        </ng-template>\r\n      </farris-tab>\r\n      <farris-tab title=\"\u5C97\u4F4D\" [id]=\"'position'\">\r\n        <ng-template #content>\r\n          <app-position-performer></app-position-performer>\r\n        </ng-template>\r\n      </farris-tab>\r\n      <farris-tab title=\"\u6C47\u62A5\u5173\u7CFB\" [id]=\"'report'\">\r\n        <ng-template #content>\r\n          <app-reporting-line></app-reporting-line>\r\n        </ng-template>\r\n      </farris-tab>\r\n      <farris-tab title=\"\u77E9\u9635\" [id]=\"'matrix'\" *ngIf=\"display\">\r\n        <ng-template #content>\r\n          <app-matrix-performer></app-matrix-performer>\r\n        </ng-template>\r\n      </farris-tab>\r\n      <farris-tab title=\"\u53C2\u4E0E\u8005\u6784\u4EF6\" [id]=\"'component'\">\r\n        <ng-template #content>\r\n          <app-component-performer [component]=\"component\" style=\"background: #FFF;\"></app-component-performer>\r\n        </ng-template>\r\n      </farris-tab>\r\n    </farris-tabs>\r\n</div>\r\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    PerformerEditComponent.ctorParameters = function () { return [
        { type: HttpService }
    ]; };
    PerformerEditComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        display: [{ type: Input }],
        usersCmp: [{ type: ViewChild, args: [UserPerformerComponent,] }],
        posiCmp: [{ type: ViewChild, args: [PositionPerformerComponent,] }],
        cmpCmp: [{ type: ViewChild, args: [ComponentPerformerComponent,] }],
        reportingLineCmp: [{ type: ViewChild, args: [ReportingLineComponent,] }],
        matrixCmp: [{ type: ViewChild, args: [MatrixPerformerComponent,] }]
    };
    return PerformerEditComponent;
}());
export { PerformerEditComponent };
if (false) {
    /** @type {?} */
    PerformerEditComponent.prototype.cls;
    /** @type {?} */
    PerformerEditComponent.prototype.component;
    /** @type {?} */
    PerformerEditComponent.prototype.performer;
    /** @type {?} */
    PerformerEditComponent.prototype.tab;
    /** @type {?} */
    PerformerEditComponent.prototype.display;
    /** @type {?} */
    PerformerEditComponent.prototype.usersCmp;
    /** @type {?} */
    PerformerEditComponent.prototype.posiCmp;
    /** @type {?} */
    PerformerEditComponent.prototype.cmpCmp;
    /** @type {?} */
    PerformerEditComponent.prototype.reportingLineCmp;
    /** @type {?} */
    PerformerEditComponent.prototype.matrixCmp;
    /**
     * @type {?}
     * @private
     */
    PerformerEditComponent.prototype.http;
}
