// @farris/mobile-devkit/esm2015/lib/http/http-util.js
class HttpUtil {
    /**
     * 追加Header
     */
    static appendHeader(headers, key, value) {
        headers = Object.assign({}, headers, { [key]: value });
        return headers;
    }
    /**
     * 向RequestConfig中追加body
     */
    static appendBodyToRequestConfig(body, requestConfig) {
        if (!requestConfig) {
            requestConfig = {};
        }
        requestConfig = Object.assign({}, requestConfig, { body: body });
        return requestConfig;
    }
    /**
     * 构造AxiosReqeustConfig
     */
    static buildAxiosRequestConfig(method, url, requestConfig) {
        requestConfig = requestConfig || {};
        const axiosRequestConfig = {
            url: url,
            method: method,
            params: requestConfig.params || null,
            headers: requestConfig.headers || null,
            responseType: requestConfig.responseType || 'json',
            data: requestConfig.body || null
        };
        return axiosRequestConfig;
    }
    /**
     * 构造Http响应信息
     */
    static buildHttpResponse(axiosResponse) {
        const httpResponse = {
            body: axiosResponse.data,
            headers: axiosResponse.headers,
            status: axiosResponse.status,
            statusText: axiosResponse.statusText
        };
        return httpResponse;
    }
}
export { HttpUtil };
