// @farris/ui-html-editor/esm2015/lib/tool-tip-module/tooltip-list.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const tooltips = {
    align: {
        '': '对齐',
        center: '居中对齐',
        right: '右对齐',
        justify: '两端对齐'
    },
    background: '背景色',
    blockquote: '引用',
    bold: '加粗',
    clean: '清除格式',
    code: '代码',
    'code-block': '代码段',
    color: '颜色',
    direction: '方向',
    float: {
        center: '居中',
        full: '全浮动',
        left: '左浮动',
        right: '右浮动'
    },
    formula: '公式',
    header: {
        '': '标题',
        '1': '标题1',
        '2': '标题2'
    },
    italic: '斜体',
    image: '图片',
    indent: {
        '+1': '缩进',
        '-1': '取消缩进'
    },
    link: '链接',
    list: {
        ordered: '有序列表',
        bullet: '无序列表',
        check: '选择列表'
    },
    script: {
        sub: '下标',
        super: '上标'
    },
    strike: '删除线',
    underline: '下划线',
    video: '视频',
    undo: '撤销',
    redo: '重做',
    size: '字号',
    font: '字体',
    divider: '分割线',
    formatBrush: '格式刷',
    emoji: '表情'
};
