// @farris/command-services/esm2015/lib/dirty-checking.service.js
import { ApplicationRef, ChangeDetectorRef, Injectable, Injector } from "@angular/core";
/**
 * angular 脏值检查服务
 */
export class DirtyCheckingService {
    constructor(injector, cd, appRef) {
        this.injector = injector;
        this.cd = cd;
        this.appRef = appRef;
    }
    /**
     * Checks this view and its children.
     * @param global 是否全局检查
     */
    detectChanges(global) {
        global = (global === true || global === 'true') ? true : false;
        if (global && this.appRef) {
            this.appRef.tick();
        }
        else if (this.cd) {
            this.cd.detectChanges();
        }
    }
    /**
     * 在指定时间（ms）后触发Angular脏值检查
     * @param duration 超时时间(ms)
     * @param global 是否全局检查
     */
    detectChangesAfter(duration, global) {
        global = (global === true || global === 'true') ? true : false;
        if (isNaN(duration)) {
            return;
        }
        if (global && this.appRef) {
            window.setTimeout(() => {
                this.appRef.tick();
            }, duration);
        }
        else if (this.cd) {
            window.setTimeout(() => {
                this.cd.detectChanges();
            }, duration);
        }
    }
}
DirtyCheckingService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
DirtyCheckingService.ctorParameters = () => [
    { type: Injector },
    { type: ChangeDetectorRef },
    { type: ApplicationRef }
];
