// @gsp-wf/wf-process-editor/esm2015/lib/views/entryForm/entry-form-editor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, Input, ViewChild, EventEmitter, TemplateRef } from '@angular/core';
import { HttpService } from '@ecp-caf/caf-common';
import { DatagridComponent } from '@farris/ui-datagrid';
import { MessagerService } from '@farris/ui-messager';
export class EntryFormEditorComponent {
    /**
     * @param {?} http
     * @param {?} message
     */
    constructor(http, message) {
        this.http = http;
        this.message = message;
        this.closeModal = new EventEmitter();
        this.submitModal = new EventEmitter();
        this.editorParams = {};
        this.modalConfig = {
            title: '流程分类',
            width: 540,
            height: 520,
            showButtons: true
        };
        this.columns = [
            { title: '编号', field: 'code', width: 100, sortable: true },
            { title: '名称', field: 'name', width: 100, sortable: true },
            { title: '类型', field: 'base', width: 60, formatter: { type: 'boolean', options: { trueText: '基础', falseText: '扩展' } } }
        ];
        this.data = [];
    }
    /**
     * @param {?} data
     * @return {?}
     */
    set value(data) {
        if (data) {
            this.flowFormId = data.id;
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.getBaseAndExtendFlowForm();
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
    }
    /**
     * @return {?}
     */
    clickOK() {
        if (!this.grid.selectedRow) {
            this.message.info('尚未选择流程分类');
            return;
        }
        /** @type {?} */
        const object = { value: this.grid.selectedRow.data, parameters: null };
        this.submitModal.emit(object);
    }
    /**
     * @return {?}
     */
    clickCancel() {
        this.closeModal.emit();
    }
    /**
     * @private
     * @return {?}
     */
    getBaseAndExtendFlowForm() {
        if (this.flowFormId) {
            /** @type {?} */
            const url = `/api/runtime/wf/v1.0/flowForm/${this.flowFormId}/baseAndExtend`;
            this.http.get(url).subscribe((/**
             * @param {?} re
             * @return {?}
             */
            (re) => {
                if (re && re.length > 0) {
                    this.data = re;
                    setTimeout((/**
                     * @return {?}
                     */
                    () => {
                        this.grid.selectRow(this.flowFormId, true);
                    }), 0);
                }
                else {
                    this.message.info('流程分类已经不存在');
                }
            }));
        }
    }
}
EntryFormEditorComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-entry-form-editor',
                template: "<div class=\"f-page\">\r\n  <div class=\"f-page-main f-utils-fill-flex-column h-100\">\r\n    <div class=\"f-utils-fill-flex-column farris-overflow-y-auto h-100\">\r\n      <farris-datagrid #grid [data]=\"data\" [columns]=\"columns\" [idField]=\"'id'\" [fit]=\"true\" [fitColumns]=\"true\"\r\n         [pagination]=\"false\" [showCheckbox]=\"false\" [multiSelect]=\"false\" [showLineNumber]=\"true\" [lineNumberTitle]=\"'\u5E8F\u53F7'\" \r\n         [lineNumberWidth]=\"50\">\r\n      </farris-datagrid>\r\n    </div>\r\n  </div>\r\n</div>\r\n\r\n\r\n<ng-template #iconSelectFooter>\r\n    <button type=\"button\" class=\"btn btn-primary btn-sm px-3 mx-2\" (click)=\"clickOK()\">\u91CD\u65B0\u7ED1\u5B9A</button>\r\n    <button type=\"button\" class=\"btn btn-secondary btn-sm px-3\" (click)=\"clickCancel()\">\u53D6\u6D88</button>\r\n</ng-template>\r\n",
                styles: [""]
            }] }
];
/** @nocollapse */
EntryFormEditorComponent.ctorParameters = () => [
    { type: HttpService },
    { type: MessagerService }
];
EntryFormEditorComponent.propDecorators = {
    closeModal: [{ type: Output }],
    submitModal: [{ type: Output }],
    value: [{ type: Input }],
    modalFooter: [{ type: ViewChild, args: ['iconSelectFooter',] }],
    grid: [{ type: ViewChild, args: ['grid',] }],
    editorParams: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    EntryFormEditorComponent.prototype.closeModal;
    /** @type {?} */
    EntryFormEditorComponent.prototype.submitModal;
    /** @type {?} */
    EntryFormEditorComponent.prototype.modalFooter;
    /** @type {?} */
    EntryFormEditorComponent.prototype.grid;
    /** @type {?} */
    EntryFormEditorComponent.prototype.editorParams;
    /**
     * @type {?}
     * @private
     */
    EntryFormEditorComponent.prototype.flowFormId;
    /** @type {?} */
    EntryFormEditorComponent.prototype.modalConfig;
    /** @type {?} */
    EntryFormEditorComponent.prototype.columns;
    /** @type {?} */
    EntryFormEditorComponent.prototype.data;
    /**
     * @type {?}
     * @private
     */
    EntryFormEditorComponent.prototype.http;
    /**
     * @type {?}
     * @private
     */
    EntryFormEditorComponent.prototype.message;
}
export class FlowFormConverter {
    constructor() { }
    /**
     * @param {?} data
     * @return {?}
     */
    convertTo(data) {
        if (data) {
            return data.name;
        }
    }
}
