// @gsp-wf/wf-process-editor/esm5/lib/views/save-as-edit/save-as-edit.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ViewChild, TemplateRef } from '@angular/core';
import { of } from 'rxjs';
import { LookupGridComponent } from '@farris/ui-lookup';
var SaveAsEditComponent = /** @class */ (function () {
    function SaveAsEditComponent() {
        var _this = this;
        this.submit = new EventEmitter();
        this.cancel = new EventEmitter();
        this.accessLevel = 'public';
        this.isAdministratorConfig = false;
        this.items = new Array();
        this.accessLevelCollection = [
            { value: 'public', text: '公有' },
            { value: 'private', text: '私有' }
        ];
        this.columns = [
            { title: '编号', field: 'code', width: '100', searchField: 'code' },
            { title: '名称', field: 'name', width: '100', searchField: 'name' },
        ];
        this.dictPicked = (/**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            _this.organizationName = data.name;
            _this.organization = data.id;
            return of(true);
        });
    }
    Object.defineProperty(SaveAsEditComponent.prototype, "organizations", {
        get: /**
         * @return {?}
         */
        function () {
            return this._organizations;
        },
        set: /**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            this._organizations = data;
            if (this._organizations && this._organizations.length > 0) {
                this.items = this.makeTreeByParent(this._organizations, 'parentId', 'id');
            }
            else {
                this.items = new Array();
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    SaveAsEditComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    SaveAsEditComponent.prototype.saveAsOK = /**
     * @return {?}
     */
    function () {
        if (this.validator()) {
            this.submit.next({
                procName: this.procName,
                accessLevel: this.accessLevel,
                organization: this.organization,
                organizationName: this.organizationName
            });
        }
    };
    /**
     * @return {?}
     */
    SaveAsEditComponent.prototype.saveAsCancel = /**
     * @return {?}
     */
    function () {
        this.cancel.emit();
    };
    /**
     * @return {?}
     */
    SaveAsEditComponent.prototype.validator = /**
     * @return {?}
     */
    function () {
        return true;
    };
    /**
     * @param {?} data
     * @return {?}
     */
    SaveAsEditComponent.prototype.search = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        /** @type {?} */
        var tt = (/** @type {?} */ (this.lookUp.componentRef.instance));
        tt.searchHandle.search(data.field, data.value, 'client');
        // if (data.field === '*') {
        //   this.organizations = this.organizations.filter(o => o.code.indexOf(data.value) > -1 || o.name.indexOf(data.value) > -1);
        // } else {
        //   this.organizations = this.organizations.filter(o => o[data.field].indexOf(data.value) > -1);
        // }
    };
    /** 按照parent码构造树 */
    /**
     * 按照parent码构造树
     * @private
     * @param {?} datas
     * @param {?} parentField
     * @param {?} keyField
     * @return {?}
     */
    SaveAsEditComponent.prototype.makeTreeByParent = /**
     * 按照parent码构造树
     * @private
     * @param {?} datas
     * @param {?} parentField
     * @param {?} keyField
     * @return {?}
     */
    function (datas, parentField, keyField) {
        var _this = this;
        // 上級爲空，或者沒有上級的
        /** @type {?} */
        var pathLayerOnes = datas.filter((/**
         * @param {?} x
         * @return {?}
         */
        function (x) {
            if (!x[parentField]) {
                return true;
            }
            /** @type {?} */
            var p = datas.find((/**
             * @param {?} parent
             * @return {?}
             */
            function (parent) { return parent[keyField] === x[parentField]; }));
            if (!p) {
                return true;
            }
        }));
        /** @type {?} */
        var treedata = [];
        pathLayerOnes.forEach((/**
         * @param {?} x
         * @return {?}
         */
        function (x) {
            /** @type {?} */
            var entity = {};
            entity.data = x;
            entity.children = [];
            treedata.push(entity);
            // this.root = x['id'];
        }));
        treedata.forEach((/**
         * @param {?} item
         * @return {?}
         */
        function (item) {
            _this.findChildrenByParent(item, datas, parentField, keyField);
        }));
        return treedata;
    };
    /** 找到子节点 */
    /**
     * 找到子节点
     * @private
     * @param {?} parent
     * @param {?} datas
     * @param {?} parentField
     * @param {?} keyField
     * @return {?}
     */
    SaveAsEditComponent.prototype.findChildrenByParent = /**
     * 找到子节点
     * @private
     * @param {?} parent
     * @param {?} datas
     * @param {?} parentField
     * @param {?} keyField
     * @return {?}
     */
    function (parent, datas, parentField, keyField) {
        var _this = this;
        /** @type {?} */
        var childList = datas.filter((/**
         * @param {?} childitem
         * @return {?}
         */
        function (childitem) {
            if (childitem[parentField] === parent.data[keyField]) {
                return true;
            }
        }));
        if (childList && childList.length > 0) {
            childList.forEach((/**
             * @param {?} child
             * @return {?}
             */
            function (child) {
                /** @type {?} */
                var childEntity = {};
                childEntity.data = child;
                childEntity.children = [];
                parent.children.push(childEntity);
                _this.findChildrenByParent(childEntity, datas, parentField, keyField);
            }));
        }
    };
    SaveAsEditComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-save-as-edit',
                    template: "<!-- <div class=\"f-page\">\r\n  <div class=\"f-page-main\">\r\n    <div class=\"d-flex m-2 py-2 farris-row\">\r\n      <div class=\"w-100 px-4\">\r\n        <label class=\"col-form-label\">\r\n          <span class=\"farris-label-text\">\u6D41\u7A0B\u540D\u79F0</span>\r\n        </label>\r\n        <div class=\"farris-input-wrap\">\r\n          <input type=\"text\" class=\"form-control\" [(ngModel)]=\"procName\" name=\"procName\" />\r\n          <div class=\"farris-feedback valid-feedback\"></div>\r\n        </div>\r\n      </div>\r\n    </div>\r\n    <div class=\"d-flex m-2 py-2 farris-row\">\r\n      <div class=\"px-4 w-100\">\r\n        <label class=\"col-form-label\">\r\n          <span class=\"farris-label-text\">\u6D41\u7A0BKey\u503C</span>\r\n        </label>\r\n        <div class=\"farris-input-wrap\">\r\n          <input type=\"text\" class=\"form-control\" [(ngModel)]=\"procKey\" name=\"procKey\" [readonly]=\"true\" />\r\n          <div class=\"farris-feedback valid-feedback\"></div>\r\n        </div>\r\n      </div>\r\n    </div>\r\n  </div>\r\n</div> -->\r\n\r\n<div class=\"f-page f-page-card f-page-is-mainsubcard\">\r\n  <section class=\"f-page-main\">\r\n    <!--\u4E0D\u786E\u8BA4---------------------------\u5361\u7247\u7EC4\u4EF6\u4E0D\u540C\u7684Section\u662F\u5728\u4E0D\u540C\u7684\u7EC4\u4EF6\u91CC\uFF0C\u8FD8\u662F\u5728\u540C\u4E00\u4E2A\u5361\u7247\u7EC4\u4EF6\u91CC-->\r\n    <!--\u8868\u5355\u533A\u57DF-->\r\n    <!--\u9700\u8981\u5D4C\u5957\u7684\u5C42---------------------------IDE\u8F6C\u5316\u6210\u7EC4\u4EF6----------->\r\n    <div class=\"f-struct-wrapper\">\r\n      <farris-section class=\"f-section-form f-section-in-mainsubcard\" [enableAccordion]=\"''\" [mainTitle]=\"'\u57FA\u7840\u5C5E\u6027'\"\r\n        [subTitle]=\"''\" [enableMaximize]=\"false\">\r\n\r\n        <!--\u8F93\u5165\u63A7\u4EF6 \u6CA1\u6709\u5206\u7EC4-->\r\n        <div class=\"f-form-layout farris-form farris-form-controls-inline pr-4 m-2\">\r\n          <div class=\"col-12 px-0 my-2\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-info text-danger\">*</span>\r\n                  <span class=\"farris-label-text\">\u6D41\u7A0B\u540D\u79F0</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <input type=\"text\" class=\"form-control\" [(ngModel)]=\"procName\" name=\"procName\" />\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"col-12 px-0 my-2\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-text\">\u6D41\u7A0BKey\u503C</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <input type=\"text\" class=\"form-control\" [(ngModel)]=\"procKey\" name=\"procKey\" [readonly]=\"true\" />\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"col-12 px-0 my-2\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-text\">\u7EF4\u62A4\u7EC4\u7EC7</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <farris-lookup-grid #lookup [(ngModel)]=\"organizationName\" name=\"orgName\"\r\n                    [columns]=\"columns\" [items]=\"items\" [singleSelect]=\"true\" [displayType]=\"'TreeList'\"\r\n                    [idField]=\"'id'\" [valueField]=\"'id'\" [textField]=\"'name'\" [dictPicked]=\"dictPicked\"\r\n                    [enableClear]=\"false\" (search)=\"search($event)\" [remoteSearch]=\"false\" [title]=\"'\u7EC4\u7EC7'\"\r\n                    [readonly]=\"!isAdministratorConfig\">\r\n                  </farris-lookup-grid>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"col-12 px-0 my-2\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-info text-danger\">*</span>\r\n                  <span class=\"farris-label-text\">\u8BBF\u95EE\u7EA7\u522B</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <farris-combo-list [(ngModel)]=\"accessLevel\" name=\"accessLevel\"\r\n                    [ngModelOptions]=\"{standalone: true}\" [idField]=\"'value'\" [textField]=\"'text'\"\r\n                    [valueField]=\"'value'\" [data]=\"accessLevelCollection\" [panelHeight]=\"'auto'\" [enableClear]=\"false\">\r\n                  </farris-combo-list>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </farris-section>\r\n    </div>\r\n  </section>\r\n</div>\r\n\r\n<ng-template #buttonRef>\r\n  <button type=\"button\" class=\"btn btn-sm btn-primary mx-2\" (click)=\"saveAsOK()\">\u786E\u5B9A</button>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm mx-2\" (click)=\"saveAsCancel()\">\u53D6\u6D88</button>\r\n</ng-template>\r\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    SaveAsEditComponent.ctorParameters = function () { return []; };
    SaveAsEditComponent.propDecorators = {
        btnRef: [{ type: ViewChild, args: ['buttonRef',] }],
        procName: [{ type: Input }],
        procKey: [{ type: Input }],
        lookUp: [{ type: ViewChild, args: ['lookup',] }],
        submit: [{ type: Output }],
        cancel: [{ type: Output }]
    };
    return SaveAsEditComponent;
}());
export { SaveAsEditComponent };
if (false) {
    /** @type {?} */
    SaveAsEditComponent.prototype.btnRef;
    /** @type {?} */
    SaveAsEditComponent.prototype.procName;
    /** @type {?} */
    SaveAsEditComponent.prototype.procKey;
    /** @type {?} */
    SaveAsEditComponent.prototype.lookUp;
    /** @type {?} */
    SaveAsEditComponent.prototype.submit;
    /** @type {?} */
    SaveAsEditComponent.prototype.cancel;
    /** @type {?} */
    SaveAsEditComponent.prototype.accessLevel;
    /** @type {?} */
    SaveAsEditComponent.prototype.organization;
    /** @type {?} */
    SaveAsEditComponent.prototype.organizationName;
    /** @type {?} */
    SaveAsEditComponent.prototype.isAdministratorConfig;
    /**
     * @type {?}
     * @private
     */
    SaveAsEditComponent.prototype._organizations;
    /** @type {?} */
    SaveAsEditComponent.prototype.items;
    /** @type {?} */
    SaveAsEditComponent.prototype.accessLevelCollection;
    /** @type {?} */
    SaveAsEditComponent.prototype.columns;
    /** @type {?} */
    SaveAsEditComponent.prototype.dictPicked;
}
