// @farris/mobile-devkit/esm5/lib/http/http-client.js
import { from, of } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';
import axios from 'axios';
import { HttpUtil } from './http-util';
/**
 * HttpClient
 */
var HttpClient = /** @class */ (function () {
    /**
     * 构造函数
     */
    function HttpClient() {
        this.axiosInstance = axios.create();
    }
    /**
     * 发送GET请求
     */
    HttpClient.prototype.get = function (url, requestConfig) {
        return this.request('GET', url, requestConfig);
    };
    /**
     * 发送POST请求
     */
    HttpClient.prototype.post = function (url, body, requestConfig) {
        requestConfig = HttpUtil.appendBodyToRequestConfig(body, requestConfig);
        return this.request('POST', url, requestConfig);
    };
    /**
     * 发送PUT请求
     */
    HttpClient.prototype.put = function (url, body, requestConfig) {
        requestConfig = HttpUtil.appendBodyToRequestConfig(body, requestConfig);
        return this.request('PUT', url, requestConfig);
    };
    /**
     * 发送PATCH请求
     */
    HttpClient.prototype.patch = function (url, body, requestConfig) {
        requestConfig = HttpUtil.appendBodyToRequestConfig(body, requestConfig);
        return this.request('PATCH', url, requestConfig);
    };
    /**
     * 发送DELETE请求
     */
    HttpClient.prototype.delete = function (url, requestConfig) {
        return this.request('DELETE', url, requestConfig);
    };
    /**
     * 发送请求
     */
    HttpClient.prototype.request = function (method, url, requestConfig) {
        var _this = this;
        var request$ = of(true).pipe(switchMap(function () {
            var axiosRequestConfig = HttpUtil.buildAxiosRequestConfig(method, url, requestConfig);
            return from(_this.axiosInstance.request(axiosRequestConfig));
        }));
        return request$.pipe(map(function (axiosResponse) {
            var httpResponse = HttpUtil.buildHttpResponse(axiosResponse);
            return requestConfig.observe === 'response' ? httpResponse : axiosResponse.data;
        }));
    };
    return HttpClient;
}());
export { HttpClient };
