// @farris/ui-common/esm5/date/lib/date-helper.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { DateTimeHelperService } from './date-helper.service';
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
var DateHelperModule = /** @class */ (function () {
    function DateHelperModule() {
    }
    /**
     * @return {?}
     */
    DateHelperModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: DateHelperModule,
            providers: [
                DateTimeHelperService
            ]
        };
    };
    DateHelperModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [],
                    imports: [CommonModule],
                    exports: [],
                    providers: [],
                },] }
    ];
    return DateHelperModule;
}());
export { DateHelperModule };
