// @gsp-wf/wf-process-editor/esm2015/lib/model/resourceKeys.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const ResourceKeys = {
    // 发出动作
    new: 'new',
    open: 'open',
    updateProperty: 'updateProperty',
    refresh: 'refresh',
    loadBizActivity: 'loadBizActivity',
    // 接受消息
    wf_initProperty: 'initProperty',
    wf_save: 'save',
    wf_saveAs: 'saveAs',
    wf_sync: 'wf_sync',
    wf_note: 'wf_note',
    wf_deploy: 'release',
    wf_closeDesigner: 'closeDesigner',
    getActivities: 'getActivities',
    getEndEvents: 'getEndEvents',
    wf_showTip: 'showTip',
    /**
     * 属性名称
     */
    EntryForm: 'EntryForm',
    ConditionExpression: 'ConditionExpression',
    ParticipantSchemes: 'ParticipantSchemes',
    ApprovalItems: 'ApprovalItems',
    AssignStrategy: 'AssignStrategy',
    OperatorConfig: 'OperatorConfig',
    CompleteStrategy: 'CompleteStrategy',
    SignRule: 'SignRule',
    Retrievable: 'Retrievable',
    AddSignFrontAble: 'AddSignFrontAble',
    AddSignBehindAble: 'AddSignBehindAble',
    NotifyAble: 'NotifyAble',
    DisplayTerminal: 'DisplayTerminal',
    ExecuteForms: 'ExecuteForms',
    sendBackedTask: 'sendBackedTask',
    SelectTargetAble: 'SelectTargetAble',
    JumpScope: 'JumpScope',
    BackWardRule: 'BackWardRule',
    Tools: 'Tools',
    SelectBranch: 'SelectBranch',
    Default: 'Default',
    allowPreAssign: 'allowPreAssign',
    startMode: 'startMode',
    transferScope: 'transferScope',
    eventDefinitions: 'EventDefinitions',
    urgeHandleRule: 'urgeHandleRule',
    // 子流程
    callActivitySubProcess: 'callActivitySubProcess',
    wf_Component: 'wf_Component',
    wf_ExecuteForms: 'wf_ExecuteForms',
    wf_ExecuteForms_PC: 'wf_ExecuteForms_PC',
    wf_ExecuteForms_Mobile: 'wf_ExecuteForms_Mobile',
    wf_TitleExpression: 'wf_TitleExpression',
    wf_CustomizeAction: 'wf_CustomizeAction',
    timeLimitControl: 'timeLimitControl',
    anonymousSet: 'anonymousSet',
    processVariables: 'processVariables',
    variableSettings: 'variableSettings',
    messageNotice: 'messageNotice',
    approvalPivotConfig: 'approvalPivotConfig',
    // state
    complete: 'complete',
    // 单据流编号
    aifProcessCode: 'aifProcessCode',
    // isRelied
    isRelied: 'isRelied',
};
export { ResourceKeys };
