// @edp-bif/runtime-api/esm5/lib/entity/form-rest/FormBillCreationProcessExecutorArgs.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/form-rest/FormBillCreationProcessExecutorArgs.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BpmnModelJsonHelper } from "@edp-aif/common-api";
import { FormCreationRuleExecutorArgs } from "@edp-aif/runtime-api";
import { ManualCreationMode } from "./ManualCreationMode";
var FormBillCreationProcessExecutorArgs = /** @class */ (function (_super) {
    tslib_1.__extends(FormBillCreationProcessExecutorArgs, _super);
    function FormBillCreationProcessExecutorArgs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    FormBillCreationProcessExecutorArgs.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.billCreationProcessId = null;
        this.creationRuleInstanceId = null;
    };
    /**
     * @return {?}
     */
    FormBillCreationProcessExecutorArgs.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "billCreationProcessId", this.billCreationProcessId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "creationRuleInstanceId", this.creationRuleInstanceId);
        BpmnModelJsonHelper.SetAttributeValue(obj, "manualCreationMode", this.manualCreationMode);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    FormBillCreationProcessExecutorArgs.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.billCreationProcessId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "billCreationProcessId");
        this.creationRuleInstanceId = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleInstanceId");
        this.manualCreationMode = ManualCreationMode[BpmnModelJsonHelper.GetAttributeValueAsString(obj, "creationRuleInstanceId")];
    };
    return FormBillCreationProcessExecutorArgs;
}(FormCreationRuleExecutorArgs));
export { FormBillCreationProcessExecutorArgs };
if (false) {
    /** @type {?} */
    FormBillCreationProcessExecutorArgs.prototype.billCreationProcessId;
    /** @type {?} */
    FormBillCreationProcessExecutorArgs.prototype.creationRuleInstanceId;
    /** @type {?} */
    FormBillCreationProcessExecutorArgs.prototype.manualCreationMode;
}
