// @farris/ui-filter/esm5/lib/entity/conditionvalue/flexibledatevalue.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from '../controltype';
import { EventEmitter } from '@angular/core';
import { DateTimeHelperService } from '@farris/ui-common/date';
var FlexibleDateValue = /** @class */ (function () {
    function FlexibleDateValue(value, single) {
        if (value === void 0) { value = { startValue: null, endValue: null, value: null }; }
        if (single === void 0) { single = true; }
        this.clear$ = new EventEmitter();
        this.startValue = value.startValue;
        this.endValue = value.endValue;
        this.value = value.value;
        this.single = single;
    }
    /**
     * @return {?}
     */
    FlexibleDateValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.FlexibleDate;
    };
    /**
     * @return {?}
     */
    FlexibleDateValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        if (this.single) {
            return !this.value;
        }
        else {
            return (this.startValue == null && this.endValue == null);
        }
    };
    /**
     * @return {?}
     */
    FlexibleDateValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.startValue = undefined;
        this.endValue = undefined;
        this.value = undefined;
        this.clear$.emit();
    };
    /**
     * @param {?} value
     * @return {?}
     */
    FlexibleDateValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        throw new Error("Method not implemented.");
    };
    /**
     * @return {?}
     */
    FlexibleDateValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        return {
            begin: this.startValue,
            end: this.endValue,
            value: this.value
        };
    };
    /**
     * @param {?=} ctrl
     * @return {?}
     */
    FlexibleDateValue.prototype.getValueText = /**
     * @param {?=} ctrl
     * @return {?}
     */
    function (ctrl) {
        if (this.single) {
            if (this.value == undefined || (!this.value && typeof (this.value) != 'undefined')) {
                return this.value;
            }
            else {
                return '' + this.value;
            }
        }
        else {
            if (this.startValue == undefined && this.endValue == undefined) {
                return '';
            }
            else {
                if (ctrl && ctrl.format) {
                    /** @type {?} */
                    var dateHlp = new DateTimeHelperService();
                    if (this.startValue == undefined) {
                        return ('~' + dateHlp.formatTo(this.endValue, ctrl.format));
                    }
                    else if (this.endValue == undefined) {
                        return (dateHlp.formatTo(this.startValue, ctrl.format) + '~');
                    }
                    else {
                        return dateHlp.formatTo(this.startValue, ctrl.format) + '~' + dateHlp.formatTo(this.endValue, ctrl.format);
                    }
                }
                else {
                    if (this.startValue == undefined) {
                        return ('~' + this.endValue);
                    }
                    else if (this.endValue == undefined) {
                        return (this.startValue + '~');
                    }
                    else {
                        return (this.startValue + '~' + this.endValue);
                    }
                }
            }
        }
    };
    return FlexibleDateValue;
}());
export { FlexibleDateValue };
if (false) {
    /** @type {?} */
    FlexibleDateValue.prototype.startValue;
    /** @type {?} */
    FlexibleDateValue.prototype.endValue;
    /** @type {?} */
    FlexibleDateValue.prototype.value;
    /** @type {?} */
    FlexibleDateValue.prototype.single;
    /** @type {?} */
    FlexibleDateValue.prototype.clear$;
}
