// @gsp-wf/wf-process-editor/esm5/lib/views/title-expression/title-expression.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Output, EventEmitter, Input, ViewChild, TemplateRef } from '@angular/core';
import { ExpressionService, ExpressioneditorComponent } from '@gsp-svc/expression';
import { ProcessDesignerUIState } from '../../app/process-designer.uistate';
var TitleExpressionComponent = /** @class */ (function () {
    function TitleExpressionComponent(expr, designerState) {
        this.expr = expr;
        this.designerState = designerState;
        this.closeModal = new EventEmitter();
        this.submitModal = new EventEmitter();
        this.editorParams = {};
        this.contextEntities = [];
        this.schema = {};
        this.injectName = '';
        this.modalConfig = {
            title: '任务描述',
            width: 1140,
            height: 520,
            showButtons: true
        };
        if (this.designerState.schemas[0]) {
            this.schema = this.designerState.schemas[0];
            if (this.designerState.schemas[0].entityTypes && this.designerState.schemas[0].entityTypes.length > 0) {
                this.injectName = this.designerState.schemas[0].entityTypes[0].name;
            }
        }
    }
    /**
     * @return {?}
     */
    TitleExpressionComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var e_1, _a, e_2, _b, e_3, _c, e_4, _d;
        // this.contextEntities = this.contextEntities.concat(this.designerState.procInstContext)
        //   .concat(this.designerState.formalParameterContext)
        //   .concat(this.designerState.participantContext)
        //   .concat(this.designerState.processVariableContext);
        this.contextEntities = new Array();
        if (this.designerState.procInstContext.length > 0) {
            try {
                for (var _e = tslib_1.__values(this.designerState.procInstContext), _f = _e.next(); !_f.done; _f = _e.next()) {
                    var variable = _f.value;
                    this.contextEntities = this.contextEntities.concat([variable]);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_f && !_f.done && (_a = _e.return)) _a.call(_e);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        if (this.designerState.formalParameterContext.length > 0) {
            var _loop_1 = function (variable) {
                if (!this_1.contextEntities.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    this_1.contextEntities = this_1.contextEntities.concat([variable]);
                }
            };
            var this_1 = this;
            try {
                for (var _g = tslib_1.__values(this.designerState.formalParameterContext), _h = _g.next(); !_h.done; _h = _g.next()) {
                    var variable = _h.value;
                    _loop_1(variable);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_h && !_h.done && (_b = _g.return)) _b.call(_g);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
        if (this.designerState.participantContext.length > 0) {
            var _loop_2 = function (variable) {
                if (!this_2.contextEntities.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    this_2.contextEntities = this_2.contextEntities.concat([variable]);
                }
            };
            var this_2 = this;
            try {
                for (var _j = tslib_1.__values(this.designerState.participantContext), _k = _j.next(); !_k.done; _k = _j.next()) {
                    var variable = _k.value;
                    _loop_2(variable);
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_k && !_k.done && (_c = _j.return)) _c.call(_j);
                }
                finally { if (e_3) throw e_3.error; }
            }
        }
        if (this.designerState.processVariableContext.length > 0) {
            var _loop_3 = function (variable) {
                if (!this_3.contextEntities.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    this_3.contextEntities = this_3.contextEntities.concat([variable]);
                }
            };
            var this_3 = this;
            try {
                for (var _l = tslib_1.__values(this.designerState.processVariableContext), _m = _l.next(); !_m.done; _m = _l.next()) {
                    var variable = _m.value;
                    _loop_3(variable);
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (_m && !_m.done && (_d = _l.return)) _d.call(_l);
                }
                finally { if (e_4) throw e_4.error; }
            }
        }
    };
    /**
     * @return {?}
     */
    TitleExpressionComponent.prototype.clickOK = /**
     * @return {?}
     */
    function () {
        this.value = this.convertToValue(this.editor.getExpr());
        /** @type {?} */
        var object = { value: this.value, parameters: null };
        this.submitModal.emit(object);
    };
    /**
     * @return {?}
     */
    TitleExpressionComponent.prototype.clickCancel = /**
     * @return {?}
     */
    function () {
        this.closeModal.emit();
    };
    /**
     * @private
     * @param {?} val
     * @return {?}
     */
    TitleExpressionComponent.prototype.convertToValue = /**
     * @private
     * @param {?} val
     * @return {?}
     */
    function (val) {
        if (val) {
            if (val.substring(9, val.length - 2)) {
                return val;
            }
            else {
                return '';
            }
        }
        else {
            return '';
        }
    };
    TitleExpressionComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-title-expression',
                    template: "<!-- <div class=\"f-page\">\r\n  <div class=\"f-utils-fill-flex-column farris-split-section m-2\">\r\n    <textarea [(ngModel)]=\"value\" name=\"condition\" class=\"form-control k-textarea h-100 w-100\"></textarea>\r\n  </div>\r\n</div> -->\r\n<app-expressioneditor \r\n  [isPanel]=\"true\" [contextEntities]=\"contextEntities\" [text]=\"value\" [customSchemaName]=\"'Schema'\"\r\n  [customSchema]=\"schema\" [injectName]=\"injectName\"\r\n  [injectSchemaName]=\"'Schema'\" [injectNameInSchema]=\"injectName\">\r\n</app-expressioneditor>\r\n\r\n<ng-template #iconSelectFooter>\r\n  <div class=\"btns\">\r\n    <!-- <button type=\"button\" class=\"btn btn-primary btn-sm mr-2\" (click)=\"clickOpen()\">\u7F16\u8F91\u5668</button> -->\r\n    <button type=\"button\" class=\"btn btn-primary btn-sm mr-2\" (click)=\"clickOK()\">\u786E\u5B9A</button>\r\n    <button type=\"button\" class=\"btn btn-secondary btn-sm\" (click)=\"clickCancel()\">\u53D6\u6D88</button>\r\n  </div>\r\n</ng-template>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    TitleExpressionComponent.ctorParameters = function () { return [
        { type: ExpressionService },
        { type: ProcessDesignerUIState }
    ]; };
    TitleExpressionComponent.propDecorators = {
        closeModal: [{ type: Output }],
        submitModal: [{ type: Output }],
        value: [{ type: Input }],
        modalFooter: [{ type: ViewChild, args: ['iconSelectFooter',] }],
        editorParams: [{ type: Input }],
        editor: [{ type: ViewChild, args: [ExpressioneditorComponent,] }]
    };
    return TitleExpressionComponent;
}());
export { TitleExpressionComponent };
if (false) {
    /** @type {?} */
    TitleExpressionComponent.prototype.closeModal;
    /** @type {?} */
    TitleExpressionComponent.prototype.submitModal;
    /** @type {?} */
    TitleExpressionComponent.prototype.value;
    /** @type {?} */
    TitleExpressionComponent.prototype.modalFooter;
    /** @type {?} */
    TitleExpressionComponent.prototype.editorParams;
    /** @type {?} */
    TitleExpressionComponent.prototype.editor;
    /** @type {?} */
    TitleExpressionComponent.prototype.contextEntities;
    /** @type {?} */
    TitleExpressionComponent.prototype.schema;
    /** @type {?} */
    TitleExpressionComponent.prototype.injectName;
    /** @type {?} */
    TitleExpressionComponent.prototype.modalConfig;
    /**
     * @type {?}
     * @private
     */
    TitleExpressionComponent.prototype.expr;
    /** @type {?} */
    TitleExpressionComponent.prototype.designerState;
}
var TitleExpressionConverter = /** @class */ (function () {
    function TitleExpressionConverter() {
    }
    /**
     * @param {?} data
     * @return {?}
     */
    TitleExpressionConverter.prototype.convertTo = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (data) {
            return data;
        }
        else {
            return '（未设置）';
        }
    };
    return TitleExpressionConverter;
}());
export { TitleExpressionConverter };
