// @farris/component-querycondition/esm2015/lib/entity/conditionvalue/single-date-value.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/conditionvalue/single-date-value.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
export class SingleDateValue {
    /**
     * @param {?} value
     * @return {?}
     */
    setOriginalValue(value) {
        this.dateValue = value.formatted;
    }
    /**
     * @return {?}
     */
    getOriginalValue() {
        return this.dateValue;
    }
    /**
     * @return {?}
     */
    getControlType() {
        return ControlType.SingleDate;
    }
    /**
     * @param {?=} date
     */
    constructor(date = '') {
        this.dateValue = date;
    }
    /**
     * @return {?}
     */
    isEmpty() {
        return !this.dateValue;
    }
    /**
     * @return {?}
     */
    clearValue() {
        this.dateValue = undefined;
    }
}
if (false) {
    /** @type {?} */
    SingleDateValue.prototype.dateValue;
}
