// @gsp-wf/wf-process-editor/esm2015/lib/views/note/note.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, TemplateRef } from '@angular/core';
import { Subject } from 'rxjs';
export class NoteComponent {
    constructor() {
        this.submit = new Subject();
        this.cancel = new Subject();
        this.placeholder = '请输入流程变更记录';
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    clickOK() {
        this.submit.next(this.record);
    }
    /**
     * @return {?}
     */
    clickCancel() {
        this.cancel.next();
    }
}
NoteComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-note',
                template: "<div class=\"f-page\">\r\n  <div class=\"f-page-main f-utils-fill-flex-column h-100\">\r\n    <div class=\"f-utils-fill-flex-column farris-overflow-y-auto h-100\">\r\n      <textarea class=\"form-control content-textarea\" style=\"height: 100%;\" [(ngModel)]=\"record\" [placeholder]=\"placeholder\"></textarea>\r\n    </div>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #footer>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm mx-2\" (click)=\"clickOK()\">\u786E\u5B9A</button>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm\" (click)=\"clickCancel()\">\u53D6\u6D88</button>\r\n</ng-template>\r\n",
                styles: [""]
            }] }
];
/** @nocollapse */
NoteComponent.ctorParameters = () => [];
NoteComponent.propDecorators = {
    footer: [{ type: ViewChild, args: ['footer',] }]
};
if (false) {
    /** @type {?} */
    NoteComponent.prototype.footer;
    /** @type {?} */
    NoteComponent.prototype.record;
    /** @type {?} */
    NoteComponent.prototype.submit;
    /** @type {?} */
    NoteComponent.prototype.cancel;
    /** @type {?} */
    NoteComponent.prototype.placeholder;
}
