// @gsp-wf/wf-process-editor/esm2015/lib/views/time-limit-control/auto-handle-operation/auto-handle-operation.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, Input, Output, TemplateRef, ViewChild } from '@angular/core';
import { AutoHandle } from "../time-limit-control.entity";
import { MessagerService } from '@farris/ui-messager';
import { ProcessDesignerUIState } from '../../../app/process-designer.uistate';
export class AutoHandleOperationComponent {
    /**
     * @param {?} message
     * @param {?} uiState
     */
    constructor(message, uiState) {
        this.message = message;
        this.uiState = uiState;
        this.placeholder = '';
        this.operation = new AutoHandle('Pass');
        this.submitEmitter = new EventEmitter();
        this.closeEmitter = new EventEmitter();
        this.autoHandleType = [
        // { value: 'autoPass', text: '自动通过' },
        // { value: 'autoBack', text: '自动驳回' }
        ];
        for (const item of this.uiState.propertyData.ApprovalItems) {
            this.autoHandleType.push({ value: item.Value, text: item.Text });
        }
    }
    /**
     * @param {?} info
     * @return {?}
     */
    set value(info) {
        this.operation = Object.assign({}, info);
        if (this.operation.autoHandleType === undefined) {
            /** @type {?} */
            const autoHandle = new AutoHandle(this.autoHandleType[0].value);
            this.operation.autoHandleType = autoHandle.autoHandleType;
            this.operation.message = autoHandle.message;
        }
    }
    /**
     * @return {?}
     */
    get value() {
        return this.operation;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    clickOK() {
        if (this.validator()) {
            this.submitEmitter.emit(this.operation);
        }
    }
    /**
     * @return {?}
     */
    clickCancel() {
        this.closeEmitter.emit();
    }
    /**
     * @private
     * @return {?}
     */
    validator() {
        if (!this.operation.message) {
            this.message.info('自动办理意见不允许为空！');
            return false;
        }
        if (!this.operation.autoHandleType) {
            this.message.info('自动办理不允许为空！');
            return false;
        }
        return true;
    }
}
AutoHandleOperationComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-auto-handle-operation',
                template: "<div class=\"f-page\">\r\n  <div class=\"f-utils-fill-flex-column farris-split-section m-2 h-100\">\r\n    <section class=\"f-utils-fill-flex-column h-100\">\r\n      <div class=\"f-section-header pt-2 mx-2\">\r\n        <div class=\"f-title\">\r\n          <h4 class=\"f-title-text\">\u81EA\u52A8\u529E\u7406\u8BBE\u7F6E</h4>\r\n        </div>\r\n      </div>\r\n      <div class=\"f-utils-fill-flex-column h-100\">\r\n        <perfect-scrollbar [config]=\"{}\">\r\n          <form class=\"h-100\">\r\n            <div class=\"farris-form form-inline farris-form-inline\">\r\n              <div class=\"farris-group-wrap\">\r\n                <div class=\"form-group farris-form-group\">\r\n                  <label class=\"col-form-label\">\r\n                    <span class=\"farris-label-info text-danger\">*</span>\r\n                    <span class=\"farris-label-text\">\u81EA\u52A8\u529E\u7406\u65B9\u5F0F</span>\r\n                  </label>\r\n                  <div class=\"farris-input-wrap ml-1\">\r\n                    <farris-combo-list [(ngModel)]=\"operation.autoHandleType\" name=\"cycleTimeUnit\" [idField]=\"'value'\" [textField]=\"'text'\"\r\n                                       [valueField]=\"'value'\" [data]=\"autoHandleType\" [panelHeight]=\"'auto'\"\r\n                                       [editable]=\"false\" [enableClear]=\"false\">\r\n                    </farris-combo-list>\r\n                    <div class=\"farris-feedback valid-feedback\"></div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n              <div class=\"farris-group-wrap my-3\">\r\n                <div class=\"form-group farris-form-group\">\r\n                  <label class=\"col-form-label\">\r\n                    <span class=\"farris-label-info text-danger\">*</span>\r\n                    <span class=\"farris-label-text\">\u81EA\u52A8\u529E\u7406\u610F\u89C1</span>\r\n                  </label>\r\n                  <div class=\"farris-input-wrap ml-1\">\r\n                    <textarea class=\"form-control\" type=\"text\" value=\"message\" [(ngModel)]=\"operation.message\" name=\"message\"\r\n                              [placeholder]=\"placeholder\" style=\"height: 100px;\"></textarea>\r\n                    <div class=\"farris-feedback valid-feedback\"></div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </form>\r\n        </perfect-scrollbar>\r\n      </div>\r\n    </section>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #footer>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm mx-2\" (click)=\"clickOK()\">\u786E\u5B9A</button>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm mx-2\" (click)=\"clickCancel()\">\u53D6\u6D88</button>\r\n</ng-template>\r\n",
                styles: [""]
            }] }
];
/** @nocollapse */
AutoHandleOperationComponent.ctorParameters = () => [
    { type: MessagerService },
    { type: ProcessDesignerUIState }
];
AutoHandleOperationComponent.propDecorators = {
    footer: [{ type: ViewChild, args: ['footer',] }],
    value: [{ type: Input }],
    submitEmitter: [{ type: Output }],
    closeEmitter: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    AutoHandleOperationComponent.prototype.footer;
    /** @type {?} */
    AutoHandleOperationComponent.prototype.placeholder;
    /** @type {?} */
    AutoHandleOperationComponent.prototype.operation;
    /** @type {?} */
    AutoHandleOperationComponent.prototype.submitEmitter;
    /** @type {?} */
    AutoHandleOperationComponent.prototype.closeEmitter;
    /** @type {?} */
    AutoHandleOperationComponent.prototype.autoHandleType;
    /**
     * @type {?}
     * @private
     */
    AutoHandleOperationComponent.prototype.message;
    /** @type {?} */
    AutoHandleOperationComponent.prototype.uiState;
}
