// @farris/kendo-binding/esm5/lib/directives/component-ref/discussion-editor-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { DiscussionEditorComponent } from '@farris/discussion-group';
var DiscussionEditorComponentRefDirective = /** @class */ (function () {
    function DiscussionEditorComponentRefDirective(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    DiscussionEditorComponentRefDirective.prototype.ngOnInit = function () {
        var id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    };
    DiscussionEditorComponentRefDirective.prototype.ngOnDestroy = function () {
    };
    DiscussionEditorComponentRefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[discussionEditorRef]'
                },] }
    ];
    /** @nocollapse */
    DiscussionEditorComponentRefDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: DiscussionEditorComponent },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    return DiscussionEditorComponentRefDirective;
}());
export { DiscussionEditorComponentRefDirective };
