// @farris/kendo-binding/esm5/lib/directives/component-ref/list-view-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { ListViewComponent } from "@farris/ui-list-view";
var ListViewComponentRefDirective = /** @class */ (function () {
    function ListViewComponentRefDirective(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ListViewComponentRefDirective.prototype.ngOnInit = function () {
        var id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    };
    ListViewComponentRefDirective.prototype.ngOnDestroy = function () {
    };
    ListViewComponentRefDirective.prototype.ngAfterViewInit = function () {
    };
    ListViewComponentRefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[listViewRef]'
                },] }
    ];
    /** @nocollapse */
    ListViewComponentRefDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: ListViewComponent },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    return ListViewComponentRefDirective;
}());
export { ListViewComponentRefDirective };
