// @farris/kendo-binding/esm5/lib/directives/component-ref/lookup-grid-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { LookupGridComponent } from '@farris/ui-lookup';
var LookupGridComponentRefDirective = /** @class */ (function () {
    function LookupGridComponentRefDirective(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    LookupGridComponentRefDirective.prototype.ngOnInit = function () {
        var id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    };
    LookupGridComponentRefDirective.prototype.ngOnDestroy = function () {
    };
    LookupGridComponentRefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[lookupGridRef]'
                },] }
    ];
    /** @nocollapse */
    LookupGridComponentRefDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: LookupGridComponent },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    return LookupGridComponentRefDirective;
}());
export { LookupGridComponentRefDirective };
