// @farris/kendo-binding/esm5/lib/directives/component-ref/multi-select-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { MultiSelectComponent } from '@farris/ui-multi-select';
var MultiSelectComponentRefDirective = /** @class */ (function () {
    function MultiSelectComponentRefDirective(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    MultiSelectComponentRefDirective.prototype.ngOnInit = function () {
        var id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    };
    MultiSelectComponentRefDirective.prototype.ngOnDestroy = function () {
    };
    MultiSelectComponentRefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[multiSelectRef]'
                },] }
    ];
    /** @nocollapse */
    MultiSelectComponentRefDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: MultiSelectComponent },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    return MultiSelectComponentRefDirective;
}());
export { MultiSelectComponentRefDirective };
