// @farris/kendo-binding/esm5/lib/directives/component-ref/scrollspy-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { ScrollspyComponent } from '@farris/ui-scrollspy';
var ScrollspyComponentRefDirective = /** @class */ (function () {
    function ScrollspyComponentRefDirective(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ScrollspyComponentRefDirective.prototype.ngOnInit = function () {
        var id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    };
    ScrollspyComponentRefDirective.prototype.ngOnDestroy = function () {
    };
    ScrollspyComponentRefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[scrollspyRef]'
                },] }
    ];
    /** @nocollapse */
    ScrollspyComponentRefDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: ScrollspyComponent },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    return ScrollspyComponentRefDirective;
}());
export { ScrollspyComponentRefDirective };
