// @farris/command-services/esm2015/lib/error/exception_strategy_factory.js
import { Optional, Injectable } from '@angular/core';
import { MessagerService } from '@farris/ui-messager';
import { LanguageService } from '../languag.service';
import { InfoExceptionStrategy, WarningExceptionStrategy, ErrorExceptionStrategy, FatalExceptionStrategy } from './exception_strategy';
export class ExceptionFactory {
    /**
     * 构造函数
     * @param messagerService 消息服务
     * @param languageService 多语言服务
     */
    constructor(messagerService, languageService) {
        this.messagerService = messagerService;
        this.languageService = languageService;
    }
    /**
     * 获取工厂实例
     * @param messagerService 消息服务
     * @param languageService 多语言服务
     */
    static getInstance(messagerService, languageService) {
        if (!this.exceptionFactory) {
            this.exceptionFactory = new ExceptionFactory(messagerService, languageService);
        }
        return this.exceptionFactory;
    }
    /**
     * 获取异常处理策略
     * @param exceptionLevel 异常等级
     */
    getExceptionHandleStrategy(exceptionLevel) {
        let exceptionStrategy;
        switch (exceptionLevel) {
            case 0 /* Info */:
                exceptionStrategy = new InfoExceptionStrategy(this.messagerService, this.languageService);
                break;
            case 1 /* Warning */:
                exceptionStrategy = new WarningExceptionStrategy(this.messagerService, this.languageService);
                break;
            case 2 /* Error */:
                exceptionStrategy = new ErrorExceptionStrategy(this.messagerService, this.languageService);
                break;
            case 3 /* Fatal */:
                exceptionStrategy = new FatalExceptionStrategy(this.messagerService, this.languageService);
                break;
            default:
                // 默认异常等级是错误
                exceptionStrategy = new ErrorExceptionStrategy(this.messagerService, this.languageService);
                break;
        }
        return exceptionStrategy;
    }
}
/**
 * 异常工厂实例
 */
ExceptionFactory.exceptionFactory = null;
ExceptionFactory.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ExceptionFactory.ctorParameters = () => [
    { type: MessagerService },
    { type: LanguageService, decorators: [{ type: Optional }] }
];
