// @gsp-wf/wf-task-handler/esm2015/lib/entity/add-sign.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BasePayload } from "./base-payload";
/**
 * 加签模拟接口返回
 */
export class AddSignSimulateResponse {
}
if (false) {
    /**
     * 组织
     * @type {?}
     */
    AddSignSimulateResponse.prototype.organization;
    /**
     * 多人会签方式
     * @type {?}
     */
    AddSignSimulateResponse.prototype.multiInstanceType;
    /**
     * 是否允许再加签
     * @type {?}
     */
    AddSignSimulateResponse.prototype.showSignToSign;
}
/** @enum {string} */
const MultiInstanceType = {
    /**
     * 抢占
     */
    Preempt: 'Preempt',
    /**
     * 并行
     */
    Parallel: 'Parallel',
    /**
     * 串行
     */
    Sequential: 'Sequential',
};
export { MultiInstanceType };
/**
 * 加签提交的payload
 */
export class AddSignPayload extends BasePayload {
}
if (false) {
    /**
     * 加签方式：前加签、后加签、并加签（当前会签）、征询/协办
     * @type {?}
     */
    AddSignPayload.prototype.actionCode;
    /**
     * 多人会签方式：抢占 Preempt、并行会签 Parallel、串行会签Sequential
     * @type {?}
     */
    AddSignPayload.prototype.multiInstanceType;
    /**
     * 会签策略：全员通过allPass、一人通过onePass
     * @type {?}
     */
    AddSignPayload.prototype.counterSignRule;
    /**
     * 办理规则
     * all所有用户都办理
     * compliance符合规则往下流转
     * @type {?}
     */
    AddSignPayload.prototype.completedRule;
    /**
     * 加签人员
     * @type {?}
     */
    AddSignPayload.prototype.signAssignees;
    /**
     * 加签节点名称
     * @type {?}
     */
    AddSignPayload.prototype.activityName;
    /**
     * 允许再加签
     * @type {?}
     */
    AddSignPayload.prototype.allowSignToSign;
}
