// @farris/kendo-binding/esm5/lib/directives/component-ref/combo-lookup-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { ComboLookupComponent } from '@farris/ui-combo-lookup';
var ComboLookupComponentRefDirective = /** @class */ (function () {
    function ComboLookupComponentRefDirective(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ComboLookupComponentRefDirective.prototype.ngOnInit = function () {
        var id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    };
    ComboLookupComponentRefDirective.prototype.ngOnDestroy = function () {
    };
    ComboLookupComponentRefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[comboLookupRef]'
                },] }
    ];
    /** @nocollapse */
    ComboLookupComponentRefDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: ComboLookupComponent },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    return ComboLookupComponentRefDirective;
}());
export { ComboLookupComponentRefDirective };
