// @farris/kendo-binding/esm5/lib/directives/component-ref/datagrid-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { DatagridComponent } from '@farris/ui-datagrid';
var DatagridComponentRefDirective = /** @class */ (function () {
    function DatagridComponentRefDirective(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    DatagridComponentRefDirective.prototype.ngOnInit = function () {
        var id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    };
    DatagridComponentRefDirective.prototype.ngOnDestroy = function () {
    };
    DatagridComponentRefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[datagridRef]'
                },] }
    ];
    /** @nocollapse */
    DatagridComponentRefDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: DatagridComponent },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    return DatagridComponentRefDirective;
}());
export { DatagridComponentRefDirective };
