// @farris/kendo-binding/esm5/lib/directives/component-ref/datepicker-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { FarrisDatepickerComponent } from '@farris/ui-datepicker';
var DatepickerComponentRefDirective = /** @class */ (function () {
    function DatepickerComponentRefDirective(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    DatepickerComponentRefDirective.prototype.ngOnInit = function () {
        var id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    };
    DatepickerComponentRefDirective.prototype.ngOnDestroy = function () {
    };
    DatepickerComponentRefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[datepickerRef]'
                },] }
    ];
    /** @nocollapse */
    DatepickerComponentRefDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: FarrisDatepickerComponent },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    return DatepickerComponentRefDirective;
}());
export { DatepickerComponentRefDirective };
