// @farris/kendo-binding/esm5/lib/directives/component-ref/discussion-list-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { DiscussionListComponent } from '@farris/discussion-group';
var DiscussionListComponentRefDirective = /** @class */ (function () {
    function DiscussionListComponentRefDirective(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    DiscussionListComponentRefDirective.prototype.ngOnInit = function () {
        var id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    };
    DiscussionListComponentRefDirective.prototype.ngOnDestroy = function () {
    };
    DiscussionListComponentRefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[discussionListRef]'
                },] }
    ];
    /** @nocollapse */
    DiscussionListComponentRefDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: DiscussionListComponent },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    return DiscussionListComponentRefDirective;
}());
export { DiscussionListComponentRefDirective };
