// @farris/kendo-binding/esm5/lib/directives/component-ref/list-nav-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { ListNavComponent } from '@farris/ui-list-nav';
var ListNavComponentRefDirective = /** @class */ (function () {
    function ListNavComponentRefDirective(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    //#region angular lifecycle
    ListNavComponentRefDirective.prototype.ngOnInit = function () {
        var id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    };
    ListNavComponentRefDirective.prototype.ngOnDestroy = function () {
    };
    ListNavComponentRefDirective.prototype.ngAfterViewInit = function () {
    };
    ListNavComponentRefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[listNavRef]'
                },] }
    ];
    /** @nocollapse */
    ListNavComponentRefDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: ListNavComponent },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    return ListNavComponentRefDirective;
}());
export { ListNavComponentRefDirective };
