// @farris/kendo-binding/esm5/lib/directives/component-ref/nav-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { ListNavComponent } from '@farris/ui-list-nav';
var NavComponentRefDirective = /** @class */ (function () {
    function NavComponentRefDirective(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    NavComponentRefDirective.prototype.ngOnInit = function () {
        var id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    };
    NavComponentRefDirective.prototype.ngOnDestroy = function () {
    };
    NavComponentRefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[navRef]'
                },] }
    ];
    /** @nocollapse */
    NavComponentRefDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: ListNavComponent },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    return NavComponentRefDirective;
}());
export { NavComponentRefDirective };
