// @farris/kendo-binding/esm5/lib/directives/component-ref/number-spinner-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { NumberSpinnerComponent } from "@farris/ui-number-spinner";
var NumberSpinnerComponentRefDirective = /** @class */ (function () {
    function NumberSpinnerComponentRefDirective(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    NumberSpinnerComponentRefDirective.prototype.ngOnInit = function () {
        var id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    };
    NumberSpinnerComponentRefDirective.prototype.ngOnDestroy = function () {
    };
    NumberSpinnerComponentRefDirective.prototype.ngAfterViewInit = function () {
    };
    NumberSpinnerComponentRefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[numberSpinnerRef]'
                },] }
    ];
    /** @nocollapse */
    NumberSpinnerComponentRefDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: NumberSpinnerComponent },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    return NumberSpinnerComponentRefDirective;
}());
export { NumberSpinnerComponentRefDirective };
