// @farris/kendo-binding/esm5/lib/directives/component-ref/time-picker-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { TimePickerComponent } from '@farris/ui-time-picker';
var TimePickerComponentRefDirective = /** @class */ (function () {
    function TimePickerComponentRefDirective(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    TimePickerComponentRefDirective.prototype.ngOnInit = function () {
        var id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    };
    TimePickerComponentRefDirective.prototype.ngOnDestroy = function () {
    };
    TimePickerComponentRefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[timePickerRef]'
                },] }
    ];
    /** @nocollapse */
    TimePickerComponentRefDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: TimePickerComponent },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    return TimePickerComponentRefDirective;
}());
export { TimePickerComponentRefDirective };
