// @farris/mobile-devkit/esm5/lib/view-model/view_model_context.js
import { Injector, InjectFlags } from '../core/index';
import { AppContext } from '../app/index';
/**
 * 视图模型上下文
 */
var ViewModelContext = /** @class */ (function () {
    function ViewModelContext() {
    }
    ViewModelContext.prototype.init = function (viewModel) {
        this.innerViewModel = viewModel;
        this.id = this.innerViewModel.id;
        this.appContext = viewModel.injector.get(AppContext);
        this.regToTree();
        this.regToAppContext();
    };
    ViewModelContext.prototype.regToTree = function () {
        var parentInjector = this.innerViewModel.injector.get(Injector, null, InjectFlags.SkipSelf);
        if (parentInjector) {
            this.parent = parentInjector.get(ViewModelContext, null);
            this.root = this.parent ? this.parent.root : this;
        }
        else {
            this.parent = null;
            this.root = this;
        }
    };
    ViewModelContext.prototype.regToAppContext = function () {
        this.appContext.regViewModelContext(this);
    };
    Object.defineProperty(ViewModelContext.prototype, "viewModel", {
        get: function () {
            return this.innerViewModel;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ViewModelContext.prototype, "injector", {
        get: function () {
            return this.innerViewModel.injector;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ViewModelContext.prototype, "repository", {
        get: function () {
            return this.innerViewModel.repository;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ViewModelContext.prototype, "bindingData", {
        get: function () {
            return this.innerViewModel.bindingData;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ViewModelContext.prototype, "uiState", {
        get: function () {
            return this.innerViewModel.uiState;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ViewModelContext.prototype, "stateMachine", {
        get: function () {
            return this.innerViewModel.stateMachine;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ViewModelContext.prototype, "form", {
        get: function () {
            return this.innerViewModel.form;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ViewModelContext.prototype, "commandBus", {
        get: function () {
            return this.innerViewModel.commandBus;
        },
        enumerable: true,
        configurable: true
    });
    return ViewModelContext;
}());
export { ViewModelContext };
