// @gsp-wf/wf-process-editor/esm5/lib/views/activity-component/activity-component.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, HostBinding, Output, EventEmitter, TemplateRef, Input, ViewChild, ComponentFactoryResolver, Injector } from '@angular/core';
import { ActivityComponentRepository } from './activity-component.repository';
import { ActivityComponentUIState } from './activity-component.uistate';
import { ActivityComponentViewModel } from './activity-component.viewmodel';
import { FormBuilder } from '@angular/forms';
import { BizComponentEntity } from '../../domain/entities/biz-component.entity';
import { ActualParameter } from '../../domain/entities/actual-parameter';
import { ProcessDesignerUIState } from '../../app/process-designer.uistate';
import { ExprEditMode, ExpressionService, InjectedType } from '@gsp-svc/expression';
import { SelectMetadataRTComponent } from '@gsp-lcm/metadatart-selector';
import { SessionService } from '@ecp-caf/caf-common';
import { BsModalService } from '@farris/ui-modal';
import { CommonComponentUiRTService } from '@gsp-cmp/ccmp-ui';
import { DatagridComponent } from '@farris/ui-datagrid';
import { of } from 'rxjs';
import { EditorTypes } from '@farris/ui-datagrid-editors';
import { MessagerService } from '@farris/ui-messager';
var ActivityComponentComponent = /** @class */ (function () {
    function ActivityComponentComponent(uistate, viewModel, formBuilder, designerState, expr, resolver, injector, sessionSrc, modalService, componentService, repository, message) {
        this.uistate = uistate;
        this.viewModel = viewModel;
        this.formBuilder = formBuilder;
        this.designerState = designerState;
        this.expr = expr;
        this.resolver = resolver;
        this.injector = injector;
        this.sessionSrc = sessionSrc;
        this.modalService = modalService;
        this.componentService = componentService;
        this.repository = repository;
        this.message = message;
        this.cls = 'farris-main-area flex-column';
        this.closeModal = new EventEmitter();
        this.submitModal = new EventEmitter();
        this.editorParams = {};
        this.modalConfig = {
            title: '构件',
            width: 1000,
            height: 530,
            showButtons: true
        };
        this.columns = [];
        this.paramColumns = [];
        this.paramAfterEdit = (/**
         * @param {?} rowIndex
         * @param {?} rowData
         * @param {?=} column
         * @return {?}
         */
        function (rowIndex, rowData, column) {
            return of(true);
        });
    }
    Object.defineProperty(ActivityComponentComponent.prototype, "value", {
        set: /**
         * @param {?} cpts
         * @return {?}
         */
        function (cpts) {
            var e_1, _a, e_2, _b;
            if (cpts && cpts.length > 0) {
                try {
                    for (var cpts_1 = tslib_1.__values(cpts), cpts_1_1 = cpts_1.next(); !cpts_1_1.done; cpts_1_1 = cpts_1.next()) {
                        var cpt = cpts_1_1.value;
                        /** @type {?} */
                        var bizComponent = new BizComponentEntity(cpt.Name || cpt.name, cpt.MetadataId || cpt.metadataId, cpt.MethodCode || cpt.methodCode);
                        bizComponent.executionTime = cpt.Time || cpt.executionTime,
                            bizComponent.id = cpt.Id || cpt.id;
                        bizComponent.isForCompensation = cpt.isForCompensation;
                        bizComponent.actualParameters = [];
                        if ((cpt.ActualParameters && cpt.ActualParameters.length > 0) || (cpt.actualParameters && cpt.actualParameters.length > 0)) {
                            /** @type {?} */
                            var actualParameters = cpt.ActualParameters || cpt.actualParameters;
                            try {
                                for (var actualParameters_1 = tslib_1.__values(actualParameters), actualParameters_1_1 = actualParameters_1.next(); !actualParameters_1_1.done; actualParameters_1_1 = actualParameters_1.next()) {
                                    var para = actualParameters_1_1.value;
                                    /** @type {?} */
                                    var actualParameter = new ActualParameter(para.Name || para.code, para.Name || para.name, para.Value || para.value);
                                    bizComponent.actualParameters.push(actualParameter);
                                }
                            }
                            catch (e_2_1) { e_2 = { error: e_2_1 }; }
                            finally {
                                try {
                                    if (actualParameters_1_1 && !actualParameters_1_1.done && (_b = actualParameters_1.return)) _b.call(actualParameters_1);
                                }
                                finally { if (e_2) throw e_2.error; }
                            }
                        }
                        this.uistate.components.push(bizComponent);
                    }
                }
                catch (e_1_1) { e_1 = { error: e_1_1 }; }
                finally {
                    try {
                        if (cpts_1_1 && !cpts_1_1.done && (_a = cpts_1.return)) _a.call(cpts_1);
                    }
                    finally { if (e_1) throw e_1.error; }
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    ActivityComponentComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.location = this.editorParams.location;
        if (this.location === 'activity') {
            this.uistate.executionCollection = this.uistate.activityCollection;
            this.columns = [
                {
                    field: 'name', width: 100, title: '构件名称'
                },
                { field: 'metadataId', width: 100, title: '构件ID' },
                { field: 'methodCode', width: 100, title: '构件方法' },
                {
                    field: 'executionTime', width: 100, title: '执行时机',
                    editor: {
                        type: EditorTypes.COMBOLIST,
                        options: { data: this.uistate.executionCollection, valueField: 'value', textField: 'text', idField: 'value', panelHeight: 'auto' }
                    },
                    formatter: (/**
                     * @param {?} v
                     * @param {?} d
                     * @param {?} i
                     * @return {?}
                     */
                    function (v, d, i) {
                        if (v) {
                            return _this.uistate.executionCollection.find((/**
                             * @param {?} item
                             * @return {?}
                             */
                            function (item) { return item.value === v; })).text;
                        }
                        else {
                            return '';
                        }
                    }),
                    validators: [
                        { type: 'required', messager: '必填' }
                    ]
                },
                {
                    field: 'isForCompensation', width: 50, title: '补偿操作',
                    editor: { type: EditorTypes.SWITCH, options: {} },
                    formatter: { type: 'boolean', options: { trueText: '是', falseText: '否' } }
                }
            ];
        }
        else if (this.location === 'autoActivity') {
            this.columns = [
                {
                    field: 'name', width: 100, title: '构件名称'
                },
                { field: 'metadataId', width: 100, title: '构件ID' },
                { field: 'methodCode', width: 100, title: '构件方法' },
                {
                    field: 'isForCompensation', width: 50, title: '补偿操作',
                    editor: { type: EditorTypes.SWITCH, options: {} },
                    formatter: { type: 'boolean', options: { trueText: '是', falseText: '否' } }
                }
            ];
        }
        else {
            this.columns = [
                {
                    field: 'name', width: 100, title: '构件名称'
                },
                { field: 'metadataId', width: 100, title: '构件ID' },
                { field: 'methodCode', width: 100, title: '构件方法' },
                {
                    field: 'executionTime', width: 100, title: '执行时机',
                    editor: {
                        type: EditorTypes.COMBOLIST,
                        options: { data: this.uistate.executionCollection, valueField: 'value', textField: 'text', idField: 'value', panelHeight: 'auto' }
                    },
                    formatter: (/**
                     * @param {?} v
                     * @param {?} d
                     * @param {?} i
                     * @return {?}
                     */
                    function (v, d, i) {
                        if (v) {
                            return _this.uistate.executionCollection.find((/**
                             * @param {?} item
                             * @return {?}
                             */
                            function (item) { return item.value === v; })).text;
                        }
                        else {
                            return '';
                        }
                    }),
                    validators: [
                        { type: 'required', messager: '必填' }
                    ]
                }
            ];
        }
        if (this.uistate.components && this.uistate.components.length > 0) {
            this.cmplist.selectRow(this.uistate.components[0].id, true);
        }
        this.paramColumns = [
            { field: 'code', width: 50, title: '编号' },
            { field: 'name', width: 50, title: '名称' },
            {
                field: 'value', width: 100, title: '参数值',
                editor: {
                    type: 'input-group',
                    options: {
                        groupText: '<i class="f-icon f-icon-lookup"></i>',
                        editable: false,
                        clickHandle: (/**
                         * @param {?} e
                         * @return {?}
                         */
                        function (e) {
                            e.formControl.setValue(e.instance.value ? e.instance.value : '');
                            _this.showExprHelp(e);
                        }),
                        clear: (/**
                         * @param {?} e
                         * @return {?}
                         */
                        function (e) {
                            console.log('clear', e);
                        })
                    }
                }
            }
        ];
        // 选中当前行
        this.repository.subject.subscribe((/**
         * @param {?} cmp
         * @return {?}
         */
        function (cmp) {
            _this.cmplist.loadData(_this.uistate.components);
            if (cmp) {
                _this.cmplist.selectRow(cmp.id, true);
            }
            else {
                _this.uistate.curComponent = null;
            }
        }));
    };
    /**
     * @return {?}
     */
    ActivityComponentComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        if (this.uistate.components && this.uistate.components.length > 0) {
            this.cmplist.selectRow(this.uistate.components[0].id, true);
        }
    };
    /**
     * @return {?}
     */
    ActivityComponentComponent.prototype.deleteRow = /**
     * @return {?}
     */
    function () {
        this.closeCellEdit();
        this.viewModel.deleteRow();
    };
    /**
     * @return {?}
     */
    ActivityComponentComponent.prototype.moveUp = /**
     * @return {?}
     */
    function () {
        if (!this.uistate.curComponent) {
            this.message.info('尚未选中构件！');
            return;
        }
        /** @type {?} */
        var curCmpId = this.uistate.curComponent.id;
        /** @type {?} */
        var i = this.uistate.components.findIndex((/**
         * @param {?} c
         * @return {?}
         */
        function (c) { return c.id === curCmpId; }));
        if (i === 0) {
            return;
        }
        this.repository.swapArray(this.uistate.components, i, i - 1);
        this.cmplist.loadData(this.uistate.components);
        this.cmplist.selectRow(curCmpId, true);
    };
    /**
     * @return {?}
     */
    ActivityComponentComponent.prototype.moveDown = /**
     * @return {?}
     */
    function () {
        if (!this.uistate.curComponent) {
            this.message.info('尚未选中构件！');
            return;
        }
        /** @type {?} */
        var curCmpId = this.uistate.curComponent.id;
        /** @type {?} */
        var i = this.uistate.components.findIndex((/**
         * @param {?} c
         * @return {?}
         */
        function (c) { return c.id === curCmpId; }));
        if (i === this.uistate.components.length - 1) {
            return;
        }
        this.repository.swapArray(this.uistate.components, i, i + 1);
        this.cmplist.loadData(this.uistate.components);
        this.cmplist.selectRow(curCmpId, true);
    };
    /**
     * @return {?}
     */
    ActivityComponentComponent.prototype.clickOK = /**
     * @return {?}
     */
    function () {
        this.closeCellEdit();
        if (this.validator()) {
            /** @type {?} */
            var object = { value: this.uistate.components, parameters: null };
            this.submitModal.emit(object);
        }
    };
    /**
     * @return {?}
     */
    ActivityComponentComponent.prototype.clickCancel = /**
     * @return {?}
     */
    function () {
        this.closeModal.emit();
    };
    /**
     * @return {?}
     */
    ActivityComponentComponent.prototype.componentSelect = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var compFactory = this.resolver.resolveComponentFactory(SelectMetadataRTComponent);
        /** @type {?} */
        var compRef = compFactory.create(this.injector);
        compRef.instance.typeList = 'WebServiceComponent'; // 目前只能根据元数据类型过滤，后续添加标签过滤
        compRef.instance.selectedMetadata.subscribe((/**
         * @param {?} $event
         * @return {?}
         */
        function ($event) {
            try {
                /** @type {?} */
                var meta = $event.metadata;
                _this.openComponentSelectorNew(meta.id, _this.sessionSrc.getUserSessionId());
            }
            catch (_a) {
                console.log('$event', $event);
                throw new Error('获取解析构件出错' + $event);
            }
        }));
        /** @type {?} */
        var modalConfig = {
            title: '构件选择',
            width: 980,
            height: 548,
            showButtons: false
        };
        /** @type {?} */
        var dialog = this.modalService.show(compRef, modalConfig);
        compRef.instance.closeModalEmitter.subscribe((/**
         * @param {?} $event
         * @return {?}
         */
        function ($event) { return dialog.close(); }));
    };
    /**
     * 取数构件帮助new
     */
    /**
     * 取数构件帮助new
     * @param {?} metadataId
     * @param {?} sessionId
     * @return {?}
     */
    ActivityComponentComponent.prototype.openComponentSelectorNew = /**
     * 取数构件帮助new
     * @param {?} metadataId
     * @param {?} sessionId
     * @return {?}
     */
    function (metadataId, sessionId) {
        var _this = this;
        this.componentService.openSelector(metadataId, sessionId).subscribe((/**
         * @param {?} selectorResult
         * @return {?}
         */
        function (selectorResult) {
            selectorResult.confirmEvent.subscribe((/**
             * @param {?} result
             * @return {?}
             */
            function (result) {
                _this.viewModel.componentSelect(result);
            }));
        }));
    };
    /**
     * @param {?} e
     * @return {?}
     */
    ActivityComponentComponent.prototype.showExprHelp = /**
     * @param {?} e
     * @return {?}
     */
    function (e) {
        var e_3, _a, e_4, _b, e_5, _c, e_6, _d, e_7, _e, e_8, _f, e_9, _g;
        if (this.designerState.schemas.length > 0) {
            try {
                for (var _h = tslib_1.__values(this.designerState.schemas), _j = _h.next(); !_j.done; _j = _h.next()) {
                    var schema = _j.value;
                    this.expr.addSchema('Schema', schema);
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_j && !_j.done && (_a = _h.return)) _a.call(_h);
                }
                finally { if (e_3) throw e_3.error; }
            }
            if (this.designerState.schemas[0].entityTypes && this.designerState.schemas[0].entityTypes.length > 0) {
                /** @type {?} */
                var entityType = this.designerState.schemas[0].entityTypes[0];
                this.expr.addInjectedEntity(entityType.name, 'Schema', entityType.name, InjectedType.EntityType);
            }
        }
        /** @type {?} */
        var context = new Array();
        if (this.designerState.procInstContext.length > 0) {
            try {
                for (var _k = tslib_1.__values(this.designerState.procInstContext), _l = _k.next(); !_l.done; _l = _k.next()) {
                    var variable = _l.value;
                    context.push(variable);
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (_l && !_l.done && (_b = _k.return)) _b.call(_k);
                }
                finally { if (e_4) throw e_4.error; }
            }
        }
        if (this.designerState.formalParameterContext.length > 0) {
            var _loop_1 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _m = tslib_1.__values(this.designerState.formalParameterContext), _o = _m.next(); !_o.done; _o = _m.next()) {
                    var variable = _o.value;
                    _loop_1(variable);
                }
            }
            catch (e_5_1) { e_5 = { error: e_5_1 }; }
            finally {
                try {
                    if (_o && !_o.done && (_c = _m.return)) _c.call(_m);
                }
                finally { if (e_5) throw e_5.error; }
            }
        }
        if (this.designerState.participantContext.length > 0) {
            var _loop_2 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _p = tslib_1.__values(this.designerState.participantContext), _q = _p.next(); !_q.done; _q = _p.next()) {
                    var variable = _q.value;
                    _loop_2(variable);
                }
            }
            catch (e_6_1) { e_6 = { error: e_6_1 }; }
            finally {
                try {
                    if (_q && !_q.done && (_d = _p.return)) _d.call(_p);
                }
                finally { if (e_6) throw e_6.error; }
            }
        }
        if (this.location === 'activity') {
            var _loop_3 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _r = tslib_1.__values(this.designerState.activityComponentContext), _s = _r.next(); !_s.done; _s = _r.next()) {
                    var variable = _s.value;
                    _loop_3(variable);
                }
            }
            catch (e_7_1) { e_7 = { error: e_7_1 }; }
            finally {
                try {
                    if (_s && !_s.done && (_e = _r.return)) _e.call(_r);
                }
                finally { if (e_7) throw e_7.error; }
            }
        }
        if (this.designerState.processVariableContext.length > 0) {
            var _loop_4 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _t = tslib_1.__values(this.designerState.processVariableContext), _u = _t.next(); !_u.done; _u = _t.next()) {
                    var variable = _u.value;
                    _loop_4(variable);
                }
            }
            catch (e_8_1) { e_8 = { error: e_8_1 }; }
            finally {
                try {
                    if (_u && !_u.done && (_f = _t.return)) _f.call(_t);
                }
                finally { if (e_8) throw e_8.error; }
            }
        }
        if (context.length > 0) {
            try {
                for (var context_1 = tslib_1.__values(context), context_1_1 = context_1.next(); !context_1_1.done; context_1_1 = context_1.next()) {
                    var v = context_1_1.value;
                    this.expr.addContext(v);
                }
            }
            catch (e_9_1) { e_9 = { error: e_9_1 }; }
            finally {
                try {
                    if (context_1_1 && !context_1_1.done && (_g = context_1.return)) _g.call(context_1);
                }
                finally { if (e_9) throw e_9.error; }
            }
        }
        this.expr.setEditPattern(ExprEditMode.Advanced); // 默认打开高级表达式编辑器
        this.expr.addExpressionText(e.instance.value);
        this.expr.buildExpression().then((/**
         * @param {?} expressioninfo
         * @return {?}
         */
        function (expressioninfo) {
            e.formControl.setValue(expressioninfo);
            e.instance.cd.detectChanges();
        }));
    };
    /**
     * @param {?} data
     * @return {?}
     */
    ActivityComponentComponent.prototype.selectionChange = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.uistate.curComponent = data.data;
    };
    /**
     * @private
     * @return {?}
     */
    ActivityComponentComponent.prototype.closeCellEdit = /**
     * @private
     * @return {?}
     */
    function () {
        this.cmplist.endCellEdit();
        if (this.paramlist) {
            this.paramlist.endCellEdit();
        }
    };
    /**
     * @private
     * @return {?}
     */
    ActivityComponentComponent.prototype.validator = /**
     * @private
     * @return {?}
     */
    function () {
        var e_10, _a;
        try {
            for (var _b = tslib_1.__values(this.uistate.components), _c = _b.next(); !_c.done; _c = _b.next()) {
                var c = _c.value;
                if (!c.executionTime && this.location !== 'autoActivity') {
                    this.message.warning('构件执行时机不允许为空！');
                    return false;
                }
            }
        }
        catch (e_10_1) { e_10 = { error: e_10_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_10) throw e_10.error; }
        }
        return true;
    };
    ActivityComponentComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-activity-component',
                    template: "\r\n<div class=\"f-page\">\r\n  <div class=\"f-page-header\">\r\n    <nav class=\"f-page-header-base\">\r\n      <div class=\"f-title\">\r\n        <span class=\"f-title-icon f-text-orna-bill\">\r\n          <i class=\"f-icon f-icon-page-title-administer\"></i>\r\n        </span>\r\n        <h4 class=\"f-title-text\">\u6784\u4EF6\u5217\u8868</h4>\r\n      </div>\r\n      <div class=\"btn-toolbar btn-group-lg\">\r\n        <button class=\"btn btn-primary mr-2\" (click)=\"componentSelect()\">\u65B0\u589E</button>\r\n        <button class=\"btn btn-secondary mr-2\" (click)=\"deleteRow()\">\u5220\u9664</button>\r\n        <button class=\"btn btn-secondary mr-2\" (click)=\"moveUp()\">\u4E0A\u79FB</button>\r\n        <button class=\"btn btn-secondary\" (click)=\"moveDown()\">\u4E0B\u79FB</button>\r\n      </div>\r\n    </nav>\r\n  </div>\r\n  <div class=\"f-page-main f-utils-fill-flex-column h-100\">\r\n    <div class=\"f-utils-fill-flex-column farris-overflow-y-auto h-100\">\r\n      <div [style.height.%]=\"uistate.curComponent?45:100\">\r\n        <farris-datagrid #cmplist [data]=\"uistate.components\" [idField]=\"'id'\" [columns]=\"columns\" [fit]=\"true\"\r\n          [fitColumns]=\"true\" [editable]=\"true\" [editMode]=\"'cell'\" [pagination]=\"false\" [showCheckbox]=\"false\"\r\n          [multiSelect]=\"false\" (selectChanged)=\"selectionChange($event)\" [closeOtherGridEditor]=\"true\">\r\n        </farris-datagrid>\r\n      </div>\r\n      <ng-container *ngIf=\"uistate.curComponent\">\r\n        <farris-tabs class=\"f-utils-fill\" [value]=\"'parameters'\" [contentFill]=\"true\" [autoTitleWidth]=\"true\">\r\n          <farris-tab [title]=\"'\u53C2\u6570\u5217\u8868'\" [id]=\"'parameters'\">\r\n            <ng-template #content>\r\n              <div class=\"f-utils-fill-flex-column h-100\">\r\n                <farris-datagrid #paramlist [data]=\"uistate.curComponent.actualParameters\" \r\n                  [columns]=\"paramColumns\" [idField]=\"'code'\" [fit]=\"true\" [fitColumns]=\"true\" \r\n                  [editable]=\"true\" [editMode]=\"'cell'\" [pagination]=\"false\"\r\n                  [showCheckbox]=\"false\" [multiSelect]=\"false\" [afterEdit]=\"paramAfterEdit\" [closeOtherGridEditor]=\"true\">\r\n                </farris-datagrid>\r\n              </div>\r\n            </ng-template>\r\n          </farris-tab>\r\n        </farris-tabs>\r\n      </ng-container>\r\n    </div>\r\n  </div>\r\n</div>\r\n\r\n\r\n<ng-template #iconSelectFooter>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm px-3 mx-2\" (click)=\"clickOK()\">\u786E\u5B9A</button>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm px-3\" (click)=\"clickCancel()\">\u53D6\u6D88</button>\r\n</ng-template>\r\n\r\n",
                    providers: [ActivityComponentRepository, ActivityComponentUIState, ActivityComponentViewModel],
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ActivityComponentComponent.ctorParameters = function () { return [
        { type: ActivityComponentUIState },
        { type: ActivityComponentViewModel },
        { type: FormBuilder },
        { type: ProcessDesignerUIState },
        { type: ExpressionService },
        { type: ComponentFactoryResolver },
        { type: Injector },
        { type: SessionService },
        { type: BsModalService },
        { type: CommonComponentUiRTService },
        { type: ActivityComponentRepository },
        { type: MessagerService }
    ]; };
    ActivityComponentComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        closeModal: [{ type: Output }],
        submitModal: [{ type: Output }],
        editorParams: [{ type: Input }],
        modalFooter: [{ type: ViewChild, args: ['iconSelectFooter',] }],
        value: [{ type: Input }],
        cmplist: [{ type: ViewChild, args: ['cmplist',] }],
        paramlist: [{ type: ViewChild, args: ['paramlist',] }]
    };
    return ActivityComponentComponent;
}());
export { ActivityComponentComponent };
if (false) {
    /** @type {?} */
    ActivityComponentComponent.prototype.cls;
    /** @type {?} */
    ActivityComponentComponent.prototype.closeModal;
    /** @type {?} */
    ActivityComponentComponent.prototype.submitModal;
    /** @type {?} */
    ActivityComponentComponent.prototype.editorParams;
    /** @type {?} */
    ActivityComponentComponent.prototype.modalFooter;
    /** @type {?} */
    ActivityComponentComponent.prototype.modalConfig;
    /** @type {?} */
    ActivityComponentComponent.prototype.cmplist;
    /** @type {?} */
    ActivityComponentComponent.prototype.paramlist;
    /** @type {?} */
    ActivityComponentComponent.prototype.location;
    /** @type {?} */
    ActivityComponentComponent.prototype.columns;
    /** @type {?} */
    ActivityComponentComponent.prototype.paramColumns;
    /** @type {?} */
    ActivityComponentComponent.prototype.paramAfterEdit;
    /** @type {?} */
    ActivityComponentComponent.prototype.uistate;
    /** @type {?} */
    ActivityComponentComponent.prototype.viewModel;
    /**
     * @type {?}
     * @private
     */
    ActivityComponentComponent.prototype.formBuilder;
    /**
     * @type {?}
     * @private
     */
    ActivityComponentComponent.prototype.designerState;
    /**
     * @type {?}
     * @private
     */
    ActivityComponentComponent.prototype.expr;
    /**
     * @type {?}
     * @private
     */
    ActivityComponentComponent.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    ActivityComponentComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    ActivityComponentComponent.prototype.sessionSrc;
    /**
     * @type {?}
     * @private
     */
    ActivityComponentComponent.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    ActivityComponentComponent.prototype.componentService;
    /**
     * @type {?}
     * @private
     */
    ActivityComponentComponent.prototype.repository;
    /**
     * @type {?}
     * @private
     */
    ActivityComponentComponent.prototype.message;
}
var ActivityComponentConverter = /** @class */ (function () {
    function ActivityComponentConverter() {
    }
    /**
     * @param {?} data
     * @return {?}
     */
    ActivityComponentConverter.prototype.convertTo = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (data && data.length > 0) {
            return '（已设置）';
        }
        else {
            return '（未设置）';
        }
    };
    return ActivityComponentConverter;
}());
export { ActivityComponentConverter };
