// @farris/kendo-binding/esm2015/lib/directives/component-ref/combo-lookup-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { ComboLookupComponent } from '@farris/ui-combo-lookup';
export class ComboLookupComponentRefDirective {
    constructor(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ngOnInit() {
        const id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    }
    ngOnDestroy() {
    }
}
ComboLookupComponentRefDirective.decorators = [
    { type: Directive, args: [{
                selector: '[comboLookupRef]'
            },] }
];
/** @nocollapse */
ComboLookupComponentRefDirective.ctorParameters = () => [
    { type: Injector },
    { type: ComboLookupComponent },
    { type: ElementRef },
    { type: FrameContext }
];
