// @gsp-wf/wf-process-editor/esm2015/lib/views/activity-component/activity-component.viewmodel.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { ActivityComponentUIState } from './activity-component.uistate';
import { ActivityComponentRepository } from './activity-component.repository';
import { MessagerService } from '@farris/ui-messager';
export class ActivityComponentViewModel {
    /**
     * @param {?} uistate
     * @param {?} repository
     * @param {?} messager
     */
    constructor(uistate, repository, messager) {
        this.uistate = uistate;
        this.repository = repository;
        this.messager = messager;
    }
    /**
     * @return {?}
     */
    deleteRow() {
        if (this.uistate.curComponent) {
            if (this.uistate.curComponent.executionTime === 'LOADDATA') {
                this.messager.info('取数构件不允许删除，请了解！');
            }
            else {
                this.repository.removeComponent(this.uistate.curComponent.id);
            }
        }
        else {
            return;
        }
    }
    /**
     * @param {?} cpt
     * @return {?}
     */
    componentSelect(cpt) {
        if (cpt) {
            this.repository.addComponent(cpt);
        }
    }
    /**
     * @param {?} flowFormId
     * @param {?} bizActId
     * @return {?}
     */
    loadComponents(flowFormId, bizActId) {
        this.repository.loadComponents(flowFormId, bizActId);
    }
}
ActivityComponentViewModel.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ActivityComponentViewModel.ctorParameters = () => [
    { type: ActivityComponentUIState },
    { type: ActivityComponentRepository },
    { type: MessagerService }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ActivityComponentViewModel.prototype.uistate;
    /**
     * @type {?}
     * @private
     */
    ActivityComponentViewModel.prototype.repository;
    /**
     * @type {?}
     * @private
     */
    ActivityComponentViewModel.prototype.messager;
}
