// @farris/kendo-binding/esm5/lib/directives/lookup-enable-extend-load-method.directive.js
import { Directive, Optional, Self, Input } from '@angular/core';
import { FrameContext } from '@farris/devkit';
import { LookupGridComponent } from '@farris/ui-lookup';
// tslint:disable-next-line: directive-selector
var LookupEnableExtendLoadMethodDirective = /** @class */ (function () {
    function LookupEnableExtendLoadMethodDirective(frameContext, lookup) {
        this.frameContext = frameContext;
        this.lookup = lookup;
    }
    LookupEnableExtendLoadMethodDirective.prototype.ngOnInit = function () {
        if (!this.frameContext || !this.lookup || this.enableExtendLoadMethod !== true) {
            return;
        }
        var befApiUrl = this.frameContext.repository.apiUri;
        var helpMetadataId = this.lookup.uri;
        var enableKey = helpMetadataId + "@" + befApiUrl;
        this.frameContext.setParam(enableKey, true);
    };
    LookupEnableExtendLoadMethodDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[enableExtendLoadMethod]'
                },] }
    ];
    /** @nocollapse */
    LookupEnableExtendLoadMethodDirective.ctorParameters = function () { return [
        { type: FrameContext, decorators: [{ type: Optional }] },
        { type: LookupGridComponent, decorators: [{ type: Optional }, { type: Self }] }
    ]; };
    LookupEnableExtendLoadMethodDirective.propDecorators = {
        enableExtendLoadMethod: [{ type: Input }]
    };
    return LookupEnableExtendLoadMethodDirective;
}());
export { LookupEnableExtendLoadMethodDirective };
