// @farris/ui-html-editor/esm5/lib/html-editor.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { HtmlEditorComponent } from './html-editor.component';
import { defaultModules } from './html-editor-defaults';
import { QUILL_CONFIG_TOKEN } from './html-editor.interfaces';
var ɵ0 = { modules: defaultModules };
var FarrisHtmlEditorModule = /** @class */ (function () {
    function FarrisHtmlEditorModule() {
    }
    /**
     * @param {?=} config
     * @return {?}
     */
    FarrisHtmlEditorModule.forRoot = /**
     * @param {?=} config
     * @return {?}
     */
    function (config) {
        return {
            ngModule: FarrisHtmlEditorModule,
            providers: [
                {
                    provide: QUILL_CONFIG_TOKEN,
                    // tslint:disable-next-line:only-arrow-functions
                    useValue: config || { modules: defaultModules }
                }
            ]
        };
    };
    FarrisHtmlEditorModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [HtmlEditorComponent],
                    imports: [],
                    exports: [HtmlEditorComponent],
                    providers: [
                        {
                            provide: QUILL_CONFIG_TOKEN,
                            useValue: ɵ0
                        }
                    ]
                },] }
    ];
    return FarrisHtmlEditorModule;
}());
export { FarrisHtmlEditorModule };
export { ɵ0 };
