// @farris/ui-time-picker/esm2015/lib/services/time-picker.locale.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { LocaleService } from '@farris/ui-locale';
import * as i0 from "@angular/core";
import * as i1 from "@farris/ui-locale";
export class TimePickerLocaleService {
    /**
     * @param {?} localeService
     */
    constructor(localeService) {
        this.localeService = localeService;
        this.localeConfig = {
            'ZH_CN': {
                placeholder: '请选择时间'
            },
            'EN_US': {
                placeholder: 'Please select a time'
            }
        };
    }
    /**
     * @return {?}
     */
    getLocaleConfig() {
        /** @type {?} */
        const locale = this.localeService.getValue('locale');
        if (locale && this.localeConfig[locale]) {
            return this.localeConfig[locale];
        }
        else {
            return this.localeConfig.ZH_CN;
        }
    }
}
TimePickerLocaleService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
TimePickerLocaleService.ctorParameters = () => [
    { type: LocaleService }
];
/** @nocollapse */ TimePickerLocaleService.ngInjectableDef = i0.defineInjectable({ factory: function TimePickerLocaleService_Factory() { return new TimePickerLocaleService(i0.inject(i1.LocaleService)); }, token: TimePickerLocaleService, providedIn: "root" });
if (false) {
    /**
     * @type {?}
     * @private
     */
    TimePickerLocaleService.prototype.localeConfig;
    /**
     * @type {?}
     * @private
     */
    TimePickerLocaleService.prototype.localeService;
}
