// @gsp-wf/wf-process-editor/esm5/lib/views/anonymous-set/anonymous-set.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, EventEmitter, Output, Input, ViewChild, TemplateRef, Injector } from '@angular/core';
import { AnonymousSet } from './anonymous-set.entity';
var AnonymousSetComponent = /** @class */ (function () {
    function AnonymousSetComponent(injector) {
        this.injector = injector;
        this.cls = 'farris-main-area flex-column';
        this.closeModal = new EventEmitter();
        this.submitModal = new EventEmitter();
        this.anonymousSet = new AnonymousSet();
        this.modalConfig = {
            title: '匿名选项',
            width: 450,
            height: 360,
            showButtons: true
        };
    }
    Object.defineProperty(AnonymousSetComponent.prototype, "value", {
        set: /**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            this.anonymousSet.isAnonymity = data.isAnonymity || false;
            this.anonymousSet.anonymous = data.anonymous || '审批人';
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    AnonymousSetComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    AnonymousSetComponent.prototype.clickOK = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var object = { value: this.anonymousSet, parameters: null };
        this.submitModal.emit(object);
    };
    /**
     * @return {?}
     */
    AnonymousSetComponent.prototype.clickCancel = /**
     * @return {?}
     */
    function () {
        this.closeModal.emit();
    };
    AnonymousSetComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-anonymous-set',
                    template: "<div class=\"f-page\">\r\n  <div class=\"f-utils-fill-flex-column farris-split-section m-2 h-100\">\r\n    <section class=\"f-utils-fill-flex-column h-100\">\r\n      <div class=\"f-section-header pt-2 mx-2\">\r\n        <div class=\"f-title\">\r\n          <h4 class=\"f-title-text\">\u533F\u540D\u8BBE\u7F6E</h4>\r\n        </div>\r\n      </div>\r\n      <div class=\"f-utils-fill-flex-column h-100\">\r\n        <div class=\"farris-form form-inline farris-form-inline\">\r\n          <div class=\"col-12 farris-group-auto my-3 pr-5\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-text\">\u662F\u5426\u542F\u7528</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap ml-1\">\r\n                  <farris-switch [(ngModel)]=\"anonymousSet.isAnonymity\" name=\"isAnonymity\"></farris-switch>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"col-12 farris-group-auto mt-4 pr-5\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-text\">\u533F\u540D\u4EE3\u79F0</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap ml-1\">\r\n                  <textarea class=\"form-control\" type=\"text\" value=\"anonymous\" [(ngModel)]=\"anonymousSet.anonymous\" name=\"anonymous\"\r\n                  style=\"height: 50px;\" [readOnly]=\"!anonymousSet.isAnonymity\"></textarea>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n    </section>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #iconSelectFooter>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm mx-2\" (click)=\"clickOK()\">\u786E\u5B9A</button>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm mx-2\" (click)=\"clickCancel()\">\u53D6\u6D88</button>\r\n</ng-template>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    AnonymousSetComponent.ctorParameters = function () { return [
        { type: Injector }
    ]; };
    AnonymousSetComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        closeModal: [{ type: Output }],
        submitModal: [{ type: Output }],
        value: [{ type: Input }],
        modalFooter: [{ type: ViewChild, args: ['iconSelectFooter',] }]
    };
    return AnonymousSetComponent;
}());
export { AnonymousSetComponent };
if (false) {
    /** @type {?} */
    AnonymousSetComponent.prototype.cls;
    /** @type {?} */
    AnonymousSetComponent.prototype.closeModal;
    /** @type {?} */
    AnonymousSetComponent.prototype.submitModal;
    /** @type {?} */
    AnonymousSetComponent.prototype.modalFooter;
    /** @type {?} */
    AnonymousSetComponent.prototype.anonymousSet;
    /** @type {?} */
    AnonymousSetComponent.prototype.modalConfig;
    /**
     * @type {?}
     * @private
     */
    AnonymousSetComponent.prototype.injector;
}
var AnonymousSetConverter = /** @class */ (function () {
    function AnonymousSetConverter() {
    }
    /**
     * @param {?} data
     * @return {?}
     */
    AnonymousSetConverter.prototype.convertTo = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (data && data.isAnonymity) {
            return '（已设置）';
        }
        else {
            return '（未设置）';
        }
    };
    return AnonymousSetConverter;
}());
export { AnonymousSetConverter };
