// @farris/kendo-binding/esm2015/lib/directives/component-ref/list-view-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { ListViewComponent } from "@farris/ui-list-view";
export class ListViewComponentRefDirective {
    constructor(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ngOnInit() {
        const id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    }
    ngOnDestroy() {
    }
    ngAfterViewInit() {
    }
}
ListViewComponentRefDirective.decorators = [
    { type: Directive, args: [{
                selector: '[listViewRef]'
            },] }
];
/** @nocollapse */
ListViewComponentRefDirective.ctorParameters = () => [
    { type: Injector },
    { type: ListViewComponent },
    { type: ElementRef },
    { type: FrameContext }
];
