// @farris/kendo-binding/esm2015/lib/directives/component-ref/number-spinner-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { NumberSpinnerComponent } from "@farris/ui-number-spinner";
export class NumberSpinnerComponentRefDirective {
    constructor(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ngOnInit() {
        const id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    }
    ngOnDestroy() {
    }
    ngAfterViewInit() {
    }
}
NumberSpinnerComponentRefDirective.decorators = [
    { type: Directive, args: [{
                selector: '[numberSpinnerRef]'
            },] }
];
/** @nocollapse */
NumberSpinnerComponentRefDirective.ctorParameters = () => [
    { type: Injector },
    { type: NumberSpinnerComponent },
    { type: ElementRef },
    { type: FrameContext }
];
