// @gsp-wf/wf-process-editor/esm5/lib/views/variable-setting/views/constant-edit/constant-edit.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, TemplateRef, ViewChild, Input, Output, EventEmitter } from '@angular/core';
var ConstantEditComponent = /** @class */ (function () {
    function ConstantEditComponent() {
        this.closeEmitter = new EventEmitter();
        this.submitEmitter = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ConstantEditComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    ConstantEditComponent.prototype.clickOK = /**
     * @return {?}
     */
    function () {
        this.submitEmitter.emit(this.value);
    };
    /**
     * @return {?}
     */
    ConstantEditComponent.prototype.clickCancel = /**
     * @return {?}
     */
    function () {
        this.closeEmitter.emit();
    };
    ConstantEditComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-constant-edit',
                    template: "<div class=\"f-page\">\r\n  <div class=\"f-utils-fill-flex-column farris-split-section m-2 h-100\">\r\n    <section class=\"f-utils-fill-flex-column h-100\">\r\n      <textarea class=\"form-control\" type=\"text\" value=\"message\" [(ngModel)]=\"value\"\r\n        name=\"message\" style=\"height: 100%;\"></textarea>\r\n      <div class=\"farris-feedback valid-feedback\"></div>        \r\n    </section>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #footer>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm mx-2\" (click)=\"clickOK()\">\u786E\u5B9A</button>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm\" (click)=\"clickCancel()\">\u53D6\u6D88</button>\r\n</ng-template>",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    ConstantEditComponent.ctorParameters = function () { return []; };
    ConstantEditComponent.propDecorators = {
        footer: [{ type: ViewChild, args: ['footer',] }],
        closeEmitter: [{ type: Output }],
        submitEmitter: [{ type: Output }],
        value: [{ type: Input }]
    };
    return ConstantEditComponent;
}());
export { ConstantEditComponent };
if (false) {
    /** @type {?} */
    ConstantEditComponent.prototype.footer;
    /** @type {?} */
    ConstantEditComponent.prototype.closeEmitter;
    /** @type {?} */
    ConstantEditComponent.prototype.submitEmitter;
    /** @type {?} */
    ConstantEditComponent.prototype.value;
}
