// @gsp-wf/wf-task-handler/esm2015/public-api.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * Public API Surface of wf-task-handler
 */
export { WfTaskHandlerService } from './lib/wf-task-handler.service';
export { WfTaskHandlerModule } from './lib/wf-task-handler.module';
export { TaskExecuteComponent } from './lib/components/task-execute/task-execute.component';
export { TaskAssigninfoComponent } from './lib/components/task-assigninfo/task-assigninfo.component';
export { TaskService } from './lib/service/task.service';
export { I18nService } from './lib/service/i18n/i18n.service';
export { TranslatePipe } from './lib/service/i18n/translate.pipe';
export { ProcessSelectorComponent } from './lib/submit/process-selector/process-selector.component';
export { AssigneeInfoComponent } from './lib/components/assignee-info/assignee-info.component';
export { TaskCommentComponent } from './lib/components/task-comment/task-comment.component';
