// @gsp-wf/wf-task-handler/esm5/lib/components/task-comment/task-comment.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild, forwardRef, Output, EventEmitter, TemplateRef } from '@angular/core';
import { DialogComponent } from '@farris/ui-dialog';
import { NG_VALUE_ACCESSOR } from '@angular/forms';
import { FDropdownDirective } from '@farris/ui-dropdown';
import { UploadDialogService } from '@gsp-svc/formdoc-upload';
import { MessagerService } from '@farris/ui-messager';
import { FileUploadEntity, GspDocMetadata, GspDocUploadEntity } from '../../entity/attachment';
import { SignHereComponent } from '@gsp-wf/wf-sign';
import { AvatarComponent } from '@farris/ui-avatar';
import { DownloadService } from '@gsp-svc/file-load';
import { TaskService } from '../../service/task.service';
import { CommonComment } from '../../entity/comment';
import { forkJoin, of } from 'rxjs';
import { HttpService } from '@ecp-caf/caf-common';
import { DatagridComponent } from '@farris/ui-datagrid';
import { I18nService } from '../../service/i18n/i18n.service';
import { EditorTypes } from '@farris/ui-datagrid-editors';
import { LoadingService } from '@farris/ui-loading';
import { ConfigCode } from '../../entity/task-center-config';
import { NotifyService } from '@farris/ui-notify';
var TaskCommentComponent = /** @class */ (function () {
    function TaskCommentComponent(taskService, httpSvc, uploadDialog, messageService, i18nService, loadingService, notify) {
        this.taskService = taskService;
        this.httpSvc = httpSvc;
        this.uploadDialog = uploadDialog;
        this.messageService = messageService;
        this.i18nService = i18nService;
        this.loadingService = loadingService;
        this.notify = notify;
        this.showAttachment = true; //是否显示附件上传，组件控制
        //是否显示附件上传，组件控制
        this.showSign = true; //是否显示签名，组件控制（不同动作可能会不显示）
        //是否显示签名，组件控制（不同动作可能会不显示）
        this.showCommonComment = true; //是否显示常用意见
        this.placeholder = '';
        this.textAreaHeight = 120; //文字框的高度
        this.attachment = new EventEmitter(); //附件
        //附件
        this.signature = new EventEmitter(); //上传签名之后，发送签名id
        //常用意见删除
        this._value = '';
        this.attachmentList = [];
        this.ifShowAttachment = false; //任务中心配置，是否显示附件上传
        //任务中心配置，是否显示附件上传
        this.ifShowSign = false; //任务中心配置，是否显示手写签名
        //临时保存上传的图片
        this.downloadService = new DownloadService();
        this.ifUseSign = false; //是否使用手写
        //是否使用手写
        this.ifUseImg = true; //是否使用上传图片
        //是否使用上传图片
        this.ifEnableSign = true; //是否默认启用签名
        //是否默认启用签名
        this.ifUploadImg = false; //是否上传过图片
        //是否上传过图片
        this.ifCanEditSign = true; //是否允许编辑签名
        //原始的签名id
        this.commentList = [];
        this.originCommentList = [];
        this.commentColumns = [];
        this.onModelChange = function (obj) { };
        this.onModelTouched = function (obj) { };
        /**
         * 系统预置的不允许编辑删除
         * @param data data.rowData为实体数据
         * @return
         */
        this.beforeEdit = function (data) {
            if (data.rowData.userId == '*') {
                return of(false);
            }
            return of(true);
        };
    }
    Object.defineProperty(TaskCommentComponent.prototype, "TaskCenterConfig", {
        set: 
        //文字框的高度
        /**
         * @param {?} configList
         * @return {?}
         */
        function (configList) {
            /** @type {?} */
            var attachment = configList.find(function (a) { return a.configCode == ConfigCode.If_Show_Attachment; });
            /** @type {?} */
            var sign = configList.find(function (a) { return a.configCode == ConfigCode.If_Show_Signature; });
            /** @type {?} */
            var enableSign = configList.find(function (a) { return a.configCode == ConfigCode.If_Use_Sign_By_Default; });
            /** @type {?} */
            var editSign = configList.find(function (a) { return a.configCode === ConfigCode.If_Can_Edit_Signature; });
            if (((/** @type {?} */ (window))).gspframeworkService.common) {
                this.curUserId = ((/** @type {?} */ (window))).gspframeworkService.common.userInfos.get().userId;
            }
            else {
                this.curUserId = ((/** @type {?} */ (window))).parent.gspframeworkService.common.userInfos.get().userId;
            }
            attachment && attachment.currentValue === 'true' && (this.ifShowAttachment = true);
            sign && sign.currentValue === 'true' && (this.ifShowSign = true);
            enableSign && enableSign.currentValue === 'false' && (this.ifEnableSign = false);
            editSign && editSign.currentValue === 'false' && (this.ifCanEditSign = false);
            if (this.ifShowSign && this.ifEnableSign) {
                this.getUserSign(this.curUserId);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(TaskCommentComponent.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this._value;
        },
        set: /**
         * @param {?} v
         * @return {?}
         */
        function (v) {
            if (this._value !== v) {
                this._value = v;
                this.onModelChange(v);
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    TaskCommentComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.commentColumns = [{
                field: 'message',
                width: 100,
                title: this.i18nService.getValue('comment.static.content'),
                editor: { type: EditorTypes.TEXTBOX, options: {} },
            }, {
                field: 'userId',
                width: 20,
                title: this.i18nService.getValue('comment.static.type'),
                template: this.commentTypeCell
            }, {
                field: 'operate',
                width: 20,
                title: this.i18nService.getValue('comment.static.operate'),
                template: this.commentOperateCell
            }];
        // 获取常用意见列表
        this.loadComments();
    };
    /**
     * @return {?}
     */
    TaskCommentComponent.prototype.onEditComment = /**
     * @return {?}
     */
    function () {
        this.dropDownEl.close();
        this.commentDialog.show();
    };
    /**
     * @return {?}
     */
    TaskCommentComponent.prototype.loadComments = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.taskService.getCommentList().subscribe(function (data) {
            _this.commentList = data;
            _this.originCommentList = tslib_1.__spread(data);
        });
    };
    /**
     * ControlValueAccessor必须实现的方法
     */
    /**
     * ControlValueAccessor必须实现的方法
     * @param {?} value
     * @return {?}
     */
    TaskCommentComponent.prototype.writeValue = /**
     * ControlValueAccessor必须实现的方法
     * @param {?} value
     * @return {?}
     */
    function (value) {
        if (value !== this.value) {
            this.value = value;
        }
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    TaskCommentComponent.prototype.registerOnChange = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onModelChange = fn;
    };
    /**
     * @param {?} fn
     * @return {?}
     */
    TaskCommentComponent.prototype.registerOnTouched = /**
     * @param {?} fn
     * @return {?}
     */
    function (fn) {
        this.onModelTouched = fn;
    };
    /**
     * 选择一条常用意见后事件
     */
    /**
     * 选择一条常用意见后事件
     * @param {?} data
     * @return {?}
     */
    TaskCommentComponent.prototype.buttonClick = /**
     * 选择一条常用意见后事件
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.dropDownEl.close();
        this.value = data;
    };
    /**
     * 上传附件
     */
    /**
     * 上传附件
     * @return {?}
     */
    TaskCommentComponent.prototype.uploadAttachment = /**
     * 上传附件
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var isRepeat = false;
        this.uploadDialog.uploadFile(this.processInstanceId, 'wf_attachment').then(function (re) {
            var e_1, _a, e_2, _b;
            try {
                for (var _c = tslib_1.__values(((/** @type {?} */ (re)))), _d = _c.next(); !_d.done; _d = _c.next()) {
                    var i = _d.value;
                    isRepeat = false;
                    try {
                        for (var _e = tslib_1.__values(_this.attachmentList), _f = _e.next(); !_f.done; _f = _e.next()) {
                            var attach = _f.value;
                            if (attach.fileName === i.fileName) {
                                _this.messageService.info('包含重复文件' + i.fileName);
                                isRepeat = true;
                                break;
                            }
                        }
                    }
                    catch (e_2_1) { e_2 = { error: e_2_1 }; }
                    finally {
                        try {
                            if (_f && !_f.done && (_b = _e.return)) _b.call(_e);
                        }
                        finally { if (e_2) throw e_2.error; }
                    }
                    if (!isRepeat) {
                        _this.attachmentList.push(i);
                    }
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_d && !_d.done && (_a = _c.return)) _a.call(_c);
                }
                finally { if (e_1) throw e_1.error; }
            }
            _this.attachment.emit(_this.attachmentList);
            //执行上传附件
            /** @type {?} */
            var files = new FileUploadEntity();
            files.rootId = "wf_attachment";
            files.metadataIdList = [];
            _this.attachmentList.forEach(function (a) {
                files.metadataIdList.push(a.metadataId);
            });
            files.metadataIdList.length && _this.taskService.uploadAttachment(files).subscribe(function () { });
        });
    };
    /**
     * 删除附件
     */
    /**
     * 删除附件
     * @param {?} attachment
     * @return {?}
     */
    TaskCommentComponent.prototype.attachmentDelete = /**
     * 删除附件
     * @param {?} attachment
     * @return {?}
     */
    function (attachment) {
        /** @type {?} */
        var gspDocDeleteEntity = {
            rootId: "wf_attachment",
            metadataId: attachment.metadataId,
            mode: "Formal"
        };
        /** @type {?} */
        var index = this.attachmentList.indexOf(attachment);
        if (index > -1) {
            this.attachmentList.splice(index, 1);
        }
        this.attachment.emit(this.attachmentList);
        this.taskService.deleteAttachment(gspDocDeleteEntity).subscribe(function () { });
    };
    /**
     * 获取不同格式附件的缩略图
     */
    /**
     * 获取不同格式附件的缩略图
     * @param {?} attachmentInfo
     * @return {?}
     */
    TaskCommentComponent.prototype.imageUrl = /**
     * 获取不同格式附件的缩略图
     * @param {?} attachmentInfo
     * @return {?}
     */
    function (attachmentInfo) {
        /** @type {?} */
        var url = '/platform/runtime/wf/webapp/mobiletaskcenter/assets/accessory/';
        return url + this.taskService.createItem(attachmentInfo) + '.svg';
    };
    /**
     * 展示手写签名dialog
     */
    /**
     * 展示手写签名dialog
     * @return {?}
     */
    TaskCommentComponent.prototype.showSignDialog = /**
     * 展示手写签名dialog
     * @return {?}
     */
    function () {
        this.signDialog.show();
        if (!this.ifEnableSign) {
            this.getUserSign(this.curUserId);
        }
    };
    /**
   * 点击删除签名
   */
    /**
     * 点击删除签名
     * @return {?}
     */
    TaskCommentComponent.prototype.clickToDeleteImage = /**
     * 点击删除签名
     * @return {?}
     */
    function () {
        this.signatureImage = '';
        this.signature.emit(null);
        /** @type {?} */
        var load = this.loadingService.show();
        this.taskService.addTaskCenterConfig(ConfigCode.If_Use_Sign_By_Default, 'false').subscribe(function (re) {
            load.close();
        }, function () {
            load.close();
        });
    };
    /**
     * @return {?}
     */
    TaskCommentComponent.prototype.clickToUseSign = /**
     * @return {?}
     */
    function () {
        this.ifUseImg = false;
        this.ifUseSign = true;
        /** @type {?} */
        var canvas = this.signaturePad.canvasDiv.nativeElement.querySelector('canvas');
        canvas.height = 252;
        canvas.width = 665;
    };
    /**
     * @return {?}
     */
    TaskCommentComponent.prototype.clickToUploadImg = /**
     * @return {?}
     */
    function () {
        this.farrisAvatar.onClick();
    };
    /**
     * @return {?}
     */
    TaskCommentComponent.prototype.clickToConfirm = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var image;
        if (this.signaturePad && this.signaturePad.isDraw) { //如果签过名
            image = this.signaturePad.confirm();
        }
        else if (this.ifUploadImg) { //如果上传过图片
            image = this.tmpSignatureImage;
        }
        if (image) {
            this.uploadSignatureImg(image);
        }
        else {
            if (this.tmpSignatureImage) {
                this.signatureImage = this.tmpSignatureImage;
                this.signature.emit(this.originSignId);
                if (!this.ifEnableSign) {
                    /** @type {?} */
                    var load_1 = this.loadingService.show();
                    this.taskService.addTaskCenterConfig(ConfigCode.If_Use_Sign_By_Default, 'true').subscribe(function (re) {
                        load_1.close();
                    }, function () {
                        load_1.close();
                    });
                }
            }
            this.signDialog.close();
        }
    };
    /**
     * @return {?}
     */
    TaskCommentComponent.prototype.clickToReset = /**
     * @return {?}
     */
    function () {
        this.ifUseSign = false;
        this.ifUseImg = true;
        this.signatureImage = '';
        this.tmpSignatureImage = '';
        this.signature.emit('');
    };
    /**
     * @return {?}
     */
    TaskCommentComponent.prototype.clickToCancel = /**
     * @return {?}
     */
    function () {
        this.signDialog.close();
    };
    /**
     * 上传图片后事件
     * event.base64 是图片信息,event.name是带后缀的图片名，event.type如image/jpeg
     */
    /**
     * 上传图片后事件
     * event.base64 是图片信息,event.name是带后缀的图片名，event.type如image/jpeg
     * @param {?} event
     * @return {?}
     */
    TaskCommentComponent.prototype.imgChange = /**
     * 上传图片后事件
     * event.base64 是图片信息,event.name是带后缀的图片名，event.type如image/jpeg
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.ifUseImg = true;
        this.ifUseSign = false;
        this.ifUploadImg = true;
        /** @type {?} */
        var image = event.base64;
        this.tmpSignatureImage = image;
        this.signatureImage = image;
    };
    /**
     * @private
     * @param {?} image
     * @return {?}
     */
    TaskCommentComponent.prototype.uploadSignatureImg = /**
     * @private
     * @param {?} image
     * @return {?}
     */
    function (image) {
        var _this = this;
        /** @type {?} */
        var uploadSignature = new GspDocUploadEntity();
        uploadSignature.fileContent = image.split(',')[1];
        uploadSignature.filePath = this.curUserId;
        uploadSignature.mode = "Formal";
        uploadSignature.metadata = new GspDocMetadata();
        uploadSignature.metadata.id = this.taskService.createNM();
        uploadSignature.metadata.fileName = this.taskService.createNM() + ".png";
        uploadSignature.metadata.rootId = "taskcentersignature";
        uploadSignature.metadata.docType = "png";
        /** @type {?} */
        var load = this.loadingService.show();
        this.taskService.uploadSignature(uploadSignature).subscribe(function (signId) {
            load.close();
            _this.originSignId = signId;
            _this.signature.emit(signId);
            _this.signatureImage = image;
            _this.signDialog.close();
        }, function (error) {
            load.close();
            _this.messageService.error('上传签名失败');
        });
    };
    /**
     * 获取用户的最近一次签名
     */
    /**
     * 获取用户的最近一次签名
     * @private
     * @param {?} userId
     * @return {?}
     */
    TaskCommentComponent.prototype.getUserSign = /**
     * 获取用户的最近一次签名
     * @private
     * @param {?} userId
     * @return {?}
     */
    function (userId) {
        var _this = this;
        /** @type {?} */
        var load = this.loadingService.show();
        this.taskService.getUserSign(userId).subscribe(function (result) {
            load.close();
            if (result && result.length) {
                /** @type {?} */
                var sign;
                if (result.length == 1) {
                    sign = result[0];
                }
                else {
                    /** @type {?} */
                    var list = result.filter(function (a) { return a && a.lastModifiedTime; });
                    sign = list.reduce(function (a, b) {
                        return a.lastModifiedTime > b.lastModifiedTime ? a : b;
                    });
                }
                _this.signature.emit(sign.id);
                _this.signatureImage = _this.downloadService.getDownloadUrl(sign.id, 'taskcentersignature');
                _this.tmpSignatureImage = _this.signatureImage;
            }
        }, function (error) {
            console.error(error);
            load.close();
        });
    };
    /**
     * 新增一条常用意见
     */
    /**
     * 新增一条常用意见
     * @return {?}
     */
    TaskCommentComponent.prototype.add = /**
     * 新增一条常用意见
     * @return {?}
     */
    function () {
        /** @type {?} */
        var newEntity = CommonComment.newEntity();
        newEntity.sortOrder = this.commentList.length + 1;
        this.commentList = this.commentList.concat(newEntity);
    };
    /**
     * 删除一条意见 data.rowData
     */
    /**
     * 删除一条意见 data.rowData
     * @param {?} data
     * @return {?}
     */
    TaskCommentComponent.prototype.clickToDeleteComment = /**
     * 删除一条意见 data.rowData
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (data.rowData.userId !== '*') {
            this.commentList = this.commentList.filter(function (c) { return c.id != data.rowData.id; });
            this.dataGrid.loadData(this.commentList);
        }
    };
    /**
     * 常用意见保存
     */
    /**
     * 常用意见保存
     * @return {?}
     */
    TaskCommentComponent.prototype.save = /**
     * 常用意见保存
     * @return {?}
     */
    function () {
        var _this = this;
        this.dataGrid.endCellEdit();
        this.httpSave().subscribe(function (r) {
            _this.loadComments();
            _this.commentDialog.close();
        });
    };
    /**
     * 调用接口执行新增删除、修改
     */
    /**
     * 调用接口执行新增删除、修改
     * @private
     * @return {?}
     */
    TaskCommentComponent.prototype.httpSave = /**
     * 调用接口执行新增删除、修改
     * @private
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var url = "/api/runtime/wf/v1.0/commonComments";
        /** @type {?} */
        var originCommentIds = this.originCommentList.filter(function (o) { return o.userId !== '*'; }).map(function (d) { return d.id; });
        //删除的
        /** @type {?} */
        var deleteIds = originCommentIds.filter(function (id) { return _this.commentList.findIndex(function (c) { return c.id === id; }) < 0; });
        return forkJoin(deleteIds.map(function (id) { return _this.httpSvc.delete(url + ("/" + id), null); })
            .concat(this.commentList.map(function (add) { return _this.httpSvc.post(url, add); })));
    };
    /**
   * web端打开附件预览
   */
    /**
     * web端打开附件预览
     * @param {?} metadataId
     * @return {?}
     */
    TaskCommentComponent.prototype.attachmentPreview = /**
     * web端打开附件预览
     * @param {?} metadataId
     * @return {?}
     */
    function (metadataId) {
        /** @type {?} */
        var options = {
            appType: "app",
            funcId: "",
            tabName: "文档在线预览",
            appId: "6ccb476b-f2ba-df8f-6f47-d37d100c56e2",
            appEntrance: "docviewertab",
            tabId: new Date().getTime(),
            entityParams: {
                metadataIdList: [metadataId],
                rootId: "wf_attachment",
                metadataId: null,
                viewOptions: {
                    showHeader: true,
                    showPaging: true,
                    showThumbnails: true,
                    showDownload: true,
                    showSearch: true,
                    showZoom: true,
                    showPrint: true,
                    showSecondaryToolBar: true,
                },
            },
        };
        ((/** @type {?} */ (window))).gspframeworkService.rtf.func.openMenu(options);
    };
    TaskCommentComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wf-comment',
                    template: "<div\r\n  class=\"fa-approve-comment fa-approve-common-gutter\"\r\n  style=\"margin-bottom: 20px\"\r\n>\r\n  <!-- \u8F93\u5165\u6846 -->\r\n  <div class=\"fa-approve-comment--content\">\r\n    <textarea\r\n      class=\"form-control content-textarea\"\r\n      style=\"font-size: 13px\"\r\n      [(ngModel)]=\"value\"\r\n      [attr.placeholder]=\"placeholder\"\r\n      [style.height.px]=\"textAreaHeight\"\r\n    ></textarea>\r\n  </div>\r\n  <!-- \u8F93\u5165\u6846\u4E0B\u65B9\u4E00\u6392\u6309\u94AE -->\r\n  <div class=\"fa-approve-comment--footer\">\r\n    <div\r\n      *ngIf=\"showAttachment || showSign || showCommonComment\"\r\n      class=\"approve-comment--text-btns\"\r\n    >\r\n      <!-- \u5E38\u7528\u610F\u89C1 -->\r\n      <div *ngIf=\"showCommonComment\" class=\"text-btns-item\">\r\n        <span class=\"f-icon f-icon-message\"></span>\r\n        <div fDropdown #dropdownEl=\"fDropdown\">\r\n          <span fDropdownToggle>\r\n            {{ \"comment.static.inputLabel\" | translate }}\r\n          </span>\r\n          <div class=\"dropdown-menu\" fDropdownMenu>\r\n            <li\r\n              class=\"dropdown-item\"\r\n              *ngFor=\"let comment of commentList\"\r\n              (click)=\"buttonClick(comment.message)\"\r\n            >\r\n              <span>{{ comment.message }}</span>\r\n            </li>\r\n            <div class=\"dropdown-divider\"></div>\r\n            <!-- \u610F\u89C1\u7BA1\u7406 -->\r\n            <div class=\"dropdown-item\" (click)=\"onEditComment()\">\r\n              <span class=\"f-icon f-icon-home-setup\"></span>\r\n              <span>{{ \"comment.static.manage\" | translate }}</span>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </div>\r\n      <!-- \u4E0A\u4F20\u9644\u4EF6 -->\r\n      <div\r\n        class=\"text-btns-item\"\r\n        *ngIf=\"ifShowAttachment && showAttachment\"\r\n        (click)=\"uploadAttachment()\"\r\n        title=\"{{ 'comment.attachment.tips' | translate }}\"\r\n      >\r\n        <span class=\"f-icon f-icon-attachment\"></span>\r\n        <span class=\"text-btns-item--text\">\r\n          {{ \"comment.static.attachmentUpload\" | translate }}\r\n        </span>\r\n      </div>\r\n      <!-- \u624B\u5199\u7B7E\u540D -->\r\n      <div\r\n        class=\"text-btns-item\"\r\n        *ngIf=\"showSign && ifShowSign && ifCanEditSign\"\r\n        title=\"{{ 'comment.signature.tips' | translate }}\"\r\n        (click)=\"showSignDialog()\"\r\n      >\r\n        <span class=\"f-icon f-icon-edit\"></span>\r\n        <span>{{ \"comment.signature.title\" | translate }}</span>\r\n      </div>\r\n      <!-- \u624B\u5199\u7B7E\u540D\u9884\u89C8 -->\r\n      <div\r\n        *ngIf=\"showSign && ifShowSign && signatureImage\"\r\n        class=\"signatureImage\"\r\n      >\r\n        <img [src]=\"signatureImage\" style=\"width: 70px; height: 26px\" />\r\n        <div *ngIf=\"ifCanEditSign\"\r\n          class=\"attachment-delete\"\r\n          (click)=\"clickToDeleteImage()\"\r\n          title=\"{{ 'comment.signature.disable' | translate }}\"\r\n        >\r\n          <span class=\"f-icon f-icon-x\"></span>\r\n        </div>\r\n      </div>\r\n    </div>\r\n    <!-- \u9644\u4EF6\u9884\u89C8 -->\r\n    <div *ngIf=\"attachmentList.length > 0\">\r\n      <li *ngFor=\"let attachment of attachmentList\">\r\n        <div class=\"wf-docFile\">\r\n          <div><img class=\"wf-docFileIcon\" [src]=\"imageUrl(attachment)\" /></div>\r\n          <div class=\"wf-docFileInfo\">\r\n            <div class=\"attachment-list\">\r\n              <a>{{ attachment.fileName }}</a>\r\n            </div>\r\n            <div\r\n              class=\"attachment-delete\"\r\n              (click)=\"attachmentDelete(attachment)\"\r\n            >\r\n              <span class=\"f-icon f-icon-close-outline\"></span>\r\n            </div>\r\n          </div>\r\n        </div>\r\n      </li>\r\n    </div>\r\n  </div>\r\n</div>\r\n\r\n<!-- \u5E38\u7528\u610F\u89C1\u7EF4\u62A4 -->\r\n<farris-dialog\r\n  #commentDialog\r\n  [buttons]=\"formButtonRef\"\r\n  [width]=\"740\"\r\n  [height]=\"440\"\r\n  [title]=\"'comment.static.title' | translate\"\r\n>\r\n  <div class=\"task-center-comment-datagrid-div\">\r\n    <farris-datagrid\r\n      #dataGrid\r\n      [data]=\"commentList\"\r\n      [columns]=\"commentColumns\"\r\n      [editable]=\"true\"\r\n      [fit]=\"true\"\r\n      [editMode]=\"'cell'\"\r\n      [pagination]=\"false\"\r\n      [fitColumns]=\"true\"\r\n      [clickToEdit]=\"true\"\r\n      [beforeEdit]=\"beforeEdit\"\r\n    >\r\n    </farris-datagrid>\r\n  </div>\r\n  <div class=\"task-comment-add task-comment-add-common\" (click)=\"add()\">\r\n    <span class=\"f-icon f-icon-plus mr-1\"></span>\r\n    <span>{{ \"comment.static.clickToAdd\" | translate }}</span>\r\n  </div>\r\n</farris-dialog>\r\n\r\n<!-- \u5E38\u7528\u610F\u89C1\u7EF4\u62A4\u6309\u94AE -->\r\n<ng-template #formButtonRef>\r\n  <button class=\"btn btn-primary mr-2\" (click)=\"save()\">\r\n    {{ \"comment.static.confirm\" | translate }}\r\n  </button>\r\n  <button class=\"btn btn-secondary\" (click)=\"commentDialog.close()\">\r\n    {{ \"comment.static.cancel\" | translate }}\r\n  </button>\r\n</ng-template>\r\n\r\n<!-- \u7B7E\u540D\u5F39\u6846 -->\r\n<farris-dialog\r\n  #signDialog\r\n  [buttons]=\"signButtons\"\r\n  [width]=\"710\"\r\n  [height]=\"420\"\r\n  [title]=\"'comment.signature.dialogTitle' | translate\"\r\n  [resizable]=\"false\"\r\n  [showMaxButton]=\"false\"\r\n>\r\n  <wf-sign-here\r\n    #wfSignHere\r\n    [title]=\"''\"\r\n    [ngStyle]=\"!ifUseSign ? { display: 'none' } : null\"\r\n  ></wf-sign-here>\r\n  <div *ngIf=\"ifUseImg && tmpSignatureImage\" class=\"signature-img-default\">\r\n    <img [src]=\"tmpSignatureImage\" style=\"height: 252px\" />\r\n  </div>\r\n  <div *ngIf=\"ifUseImg && !tmpSignatureImage\" class=\"signature-img-div\">\r\n    <div class=\"signature-img-main\">\r\n      <div style=\"margin-bottom: 28px\">\r\n        <img\r\n          src=\"/platform/runtime/common/web/@gsp-wf/wf-task-handler/assets/img/signIcon.svg?v=20210623\"\r\n        />\r\n      </div>\r\n      <div class=\"signature-img-text\">\r\n        <span>{{ \"comment.signature.note1\" | translate }}</span>\r\n        <span class=\"signature-text-action\" (click)=\"clickToUseSign()\">{{\r\n          \"comment.signature.sign\" | translate\r\n        }}</span>\r\n        <span>{{ \"comment.signature.note2\" | translate }}</span>\r\n        <span class=\"signature-text-action\" (click)=\"clickToUploadImg()\">{{\r\n          \"comment.signature.upload\" | translate\r\n        }}</span>\r\n        <span>{{ \"comment.signature.note3\" | translate }}</span>\r\n      </div>\r\n    </div>\r\n  </div>\r\n</farris-dialog>\r\n\r\n<!-- \u624B\u5199\u7B7E\u540D\u6309\u94AE -->\r\n<ng-template #signButtons>\r\n  <button class=\"btn btn-primary mr-2\" (click)=\"clickToConfirm()\">\r\n    {{ \"comment.signature.ok\" | translate }}\r\n  </button>\r\n  <button class=\"btn btn-secondary mr-2\" (click)=\"clickToReset()\">\r\n    {{ \"comment.signature.reset\" | translate }}\r\n  </button>\r\n  <button class=\"btn btn-secondary\" (click)=\"clickToCancel()\">\r\n    {{ \"comment.static.cancel\" | translate }}\r\n  </button>\r\n</ng-template>\r\n\r\n<!-- \u4F7F\u7528\u5934\u50CF\u7EC4\u4EF6\u505A\u56FE\u7247\u4E0A\u4F20 -->\r\n<farris-avatar\r\n  #farrisAvatar\r\n  (imgChange)=\"imgChange($event)\"\r\n  style=\"display: none\"\r\n></farris-avatar>\r\n\r\n<!-- \u5E38\u7528\u610F\u89C1\u7C7B\u522B -->\r\n<ng-template #commentTypeCell let-ctx>\r\n  <div>\r\n    <span\r\n      *ngIf=\"ctx.rowData.userId === '*'\"\r\n      class=\"task-comment-type task-comment-type-blue\"\r\n      >{{ \"comment.static.public\" | translate }}</span\r\n    >\r\n    <span\r\n      *ngIf=\"ctx.rowData.userId !== '*'\"\r\n      class=\"task-comment-type task-comment-type-yellow\"\r\n      >{{ \"comment.static.private\" | translate }}</span\r\n    >\r\n  </div>\r\n</ng-template>\r\n\r\n<!-- \u5E38\u7528\u610F\u89C1\u64CD\u4F5C -->\r\n<ng-template #commentOperateCell let-ctx>\r\n  <div\r\n    *ngIf=\"ctx.rowData.userId !== '*'\"\r\n    class=\"btn-link\"\r\n    style=\"padding: 5px 0\"\r\n    (click)=\"clickToDeleteComment(ctx)\"\r\n  >\r\n    {{ \"comment.static.delete\" | translate }}\r\n  </div>\r\n</ng-template>\r\n",
                    providers: [{
                            provide: NG_VALUE_ACCESSOR,
                            useExisting: forwardRef(function () { return TaskCommentComponent; }),
                            multi: true
                        },],
                    styles: [":host ::ng-deep .dropdown .btn{padding:0}:host ::ng-deep .dropdown .btn-secondary{box-shadow:none}:host ::ng-deep .dropdown .btn-secondary:focus{box-shadow:none}:host ::ng-deep .dropdown .btn-secondary:hover{color:#388fff}.task-center-comment-datagrid-div{display:flex;flex-grow:0;overflow:hidden;height:295px;padding:0 24px}.task-comment-type{padding:2px 6px;border-radius:4px;font-size:12px}.task-comment-type-blue{background:#eef5ff;border:0 solid #aed1ff;color:#5ea4ff}.task-comment-type-yellow{background:#fef5ec;border:0 solid #fad0a1;color:#f5a144}.task-comment-add{display:flex;justify-content:center;align-items:center;margin:0 24px;cursor:pointer;overflow:hidden;font-size:12px;text-align:center;padding:6px;border-radius:4px}.task-comment-add-common{color:#a8adb8;background:#f4f6fc;border:1px dashed #d2d7e2}.task-comment-add:hover{background:rgba(238,245,255,.8);border:1px solid rgba(82,157,255,.4)}.signatureImage:hover .attachment-delete{display:block}.fa-approval-operate-list{display:flex;flex-direction:row;align-items:center}.fa-approval-operate-list .approval-operate--label{position:relative;margin-right:8px;padding:3px 16px;border:1px solid rgba(0,0,0,.15);border-radius:2px;overflow:hidden;cursor:pointer;margin-bottom:0;color:#595959}.fa-approval-operate-list .approval-operate--label .label-title{font-size:14px;line-height:20px;white-space:nowrap}.fa-approval-operate-list .approval-operate--label .label-tip{position:absolute;display:block;right:0;bottom:0;width:14px;height:14px;border:7px solid #dadada;border-top:7px solid transparent;border-left:7px solid transparent}.fa-approval-operate-list .approval-operate--label .label-icon{position:absolute;display:block;right:-3px;bottom:-2px;font-size:12px;color:#fff}.fa-approval-operate-list .approval-operate--label .label-icon::before{content:\"\\e118\"}.fa-approval-operate-list .approval-operate--radio{display:none}.fa-approval-operate-list .approval-operate--radio:checked~.approval-operate--label{color:#388fff;border-color:#388fff;background:rgba(23,145,255,.05)}.fa-approval-operate-list .approval-operate--radio:checked~.approval-operate--label .label-tip{border-right-color:#388fff;border-bottom-color:#388fff}.fa-dropdown-button{border-color:#f7f7f7;font-size:12px;color:#667580;background:#f7f7f7}.fa-btn-like-dropdown{display:flex;align-items:center}.fa-btn-like-dropdown .like-dropdown--text{margin:0 4px 0 0}.fa-btn-like-dropdown .like-dropdown--icon{color:rgba(0,0,0,.25)}.fa-approve-common-gutter{margin:0 0 12px}.fa-approve-comment .approve-comment--title{display:inline-block;color:#667580;border-radius:4px 4px 0 0;height:26px;padding:0 9px;position:relative;background:#f7f7f7;cursor:pointer}.fa-approve-comment .approve-comment--title:after{content:\"\";height:100%;position:absolute;right:-10px;top:0;bottom:0;border-right:5px solid transparent;border-left:5px solid #f7f7f7;border-bottom:13px solid #f7f7f7;border-top:13px solid transparent}.fa-approve-comment .approve-comment--title .fa-btn-like-dropdown{font-size:12px;padding:2px 0}.fa-approve-comment .fa-approve-comment--content .content-textarea{border-color:#dcdcdc;border-radius:0 0 2px 2px}.fa-approve-comment .fa-approve-comment--footer{font-size:13px;color:#667580;background:#f7f7f7;align-items:center;padding:5px 0}.fa-approve-comment .approve-comment--text-btns{display:flex;flex-direction:row;padding:0 5px}.fa-approve-comment .approve-comment--text-btns .text-btns-item{display:flex;flex-direction:row;align-items:center;font-size:12px;line-height:18px;color:#667580;margin-right:20px;cursor:pointer}.fa-approve-comment .approve-comment--text-btns .text-btns-item:hover{color:#388fff}.fa-approve-comment .approve-comment--text-btns .text-btns-item .f-icon{margin-right:4px}.fa-approveal-links::after{content:\".\";display:block;height:0;clear:both;visibility:hidden}.fa-approveal-links .approveal-links--link{font-size:14px;position:relative;padding:0 16px 0 0;display:flex;align-items:center;float:left}.fa-approveal-links .approveal-links--link .link-text{color:rgba(0,0,0,.65);margin:0 8px 0 0}.fa-approveal-links .approveal-links--link+.approveal-links--link{padding-left:16px}.fa-approveal-links .approveal-links--link+.approveal-links--link::before{content:\"\";position:absolute;width:1px;height:80%;left:0;top:10%;background:#dbdbdb}.fa-btn-group-lg .btn{padding-top:6px;padding-bottom:6px;min-width:100px;text-align:center}.fa-approve-record{padding:0 0 16px;position:relative}.fa-approve-record .fa-approve-record--timeline{position:absolute;width:34px;top:0;bottom:0;z-index:100}.fa-approve-record .fa-approve-record--timeline::after{content:\"\";position:absolute;width:1px;height:100%;overflow:hidden;left:50%;top:0;z-index:101;border-left:1px dashed #dcdcdc}.fa-approve-record .fa-approve-record--timeline .fa-approve-star{position:relative;z-index:103;margin:20px auto}.fa-approve-record.fa-state-running .fa-approve-record--timeline::after{top:20px}.fa-approve-record:last-child .fa-approve-record--timeline::after{height:20px}.fa-approve-record .fa-approve-record--content{margin:0 0 0 34px;padding:16px 14px;background:#f9f9f9;border-radius:2px}.fa-approve-record .fa-approve-record--content .approve-record--header{margin:0 0 12px;display:flex;align-items:center}.fa-approve-record .fa-approve-record--content .approve-record--header .header--title{margin-bottom:0;color:rgba(0,0,0,.85)}.fa-approve-record .fa-approve-record--content .approve-record--header>*{margin:0 19px 0 0}.fa-approve-record .approve-record--list-item{display:flex;flex-direction:row}.fa-approve-record .approve-record--list-item+.approve-record--list-item{padding-top:8px}.fa-approve-record .approve-record--list-item .list-item--auxiliary{width:30px;margin:0 10px 0 0}.fa-approve-record .approve-record--list-item .list-item--main{flex-grow:1;flex-shrink:1;flex-basis:auto;padding:6px 0 0;background:#f9f9f9}.fa-approve-record .approve-record--list-item .list-item--content{padding:20px 0 0}.fa-approve-record .approve-record--list-item .list-item--content .list-item--info{font-size:12px;margin-bottom:0}.fa-approve-record .approve-record--list-item .list-item--content .list-item--info .info-label{color:rgba(0,0,0,.65)}.fa-approve-record .approve-record--list-item .list-item--content .list-item--info .info-text{color:rgba(0,0,0,.85)}.fa-approve-record .approve-record--list-item .list-item--header{display:flex;align-items:center}.fa-approve-record .approve-record--list-item .list-item--header>*{margin:0 16px 0 0}.fa-approve-record .approve-record--list-item .list-item--header .list-item--info{margin-bottom:0;font-size:12px}.fa-approve-record .approve-record--list-item .list-item--header .list-item--info .info-label{color:rgba(0,0,0,.35)}.fa-approve-record .approve-record--list-item .list-item--header .list-item--info .info-text{color:rgba(0,0,0,.65)}.fa-approve-record .approve-record--list-item .list-item--name{color:#4297fa}.fa-approve-record .approve-record--avatar{width:30px;height:30px;border-radius:15px}.fa-approve-round-bg{margin:4px auto 0;width:30px;height:30px;background:#59a1ff;border-radius:18px;line-height:30px;text-align:center;white-space:nowrap;font-size:12px;overflow:hidden}.fa-approve-round-bg-text{color:#fff}.fa-approve-star{width:14px;height:14px;display:block;background:rgba(56,143,255,.4);border-radius:7px;position:relative}.fa-approve-star::after{content:\"\";width:8px;height:8px;position:absolute;top:50%;left:50%;margin:-4px 0 0 -4px;border-radius:4px;background:#388fff}.fa-approve-star.bg-warning{background:rgba(245,156,36,.4)!important}.fa-approve-star.bg-warning::after{background:#f59c24}.fa-approve-badge{padding-top:4px;padding-bottom:4px;border-radius:10px}.fa-approve-badge-light{font-size:13px;line-height:20px;padding:0 8px;color:#5f637a;border:1px solid #e6e6e6;background:#f6faff}.fa-approve-link-border{cursor:pointer;display:inline-block;border:1px solid #60a5ff;border-radius:13px;font-size:12px;line-height:16px;color:#388fff;padding:3px 10px;background:#ebf5ff}.fa-approve-link-border:hover{background:#2881ff;color:#fff}.fm-toolbar{flex-direction:row;display:flex}.wf-docFile{flex-shrink:1;flex-grow:1;flex-basis:0;display:flex;flex-direction:row;align-items:flex-start;border-top:1px solid #f8f9fa;border-left:1px solid #f8f9fa}.wf-docFileIcon{margin:0 0 0 4px;width:20px;height:20px;line-height:20px;text-align:center;overflow:hidden;white-space:nowrap;padding:2px}.wf-docFileInfo{flex-shrink:1;flex-grow:1;flex-basis:0;padding:2px 6px 0 9px;margin-left:0}.wf-docFileInfo:hover .attachment-delete{display:block}.wf-docFileSize{font-size:11px;color:#888}.wf-doc-delete{margin:auto;font-size:18px;color:#f8f9fa}li{list-style:none}.attachment-delete{float:right;cursor:pointer;display:none;text-align:right;padding-right:10px}.attachment-delete:hover{color:#4297fa}.attachment-list{float:left;font-size:12px}.signature-img-div{display:flex;justify-content:center;align-items:center;height:100%;background:#fafcfd;border:1px solid #d8dfed;border-radius:8px;margin:0 10px}.signature-img-main{display:flex;flex-direction:column;justify-content:center;align-items:center}.signature-img-text{font-size:13px;color:rgba(106,111,121,.85)}.signature-text-action{color:#529efe;padding:0 2px;cursor:pointer}.signature-img-default{display:flex;justify-content:center;align-items:center;height:inherit}"]
                }] }
    ];
    /** @nocollapse */
    TaskCommentComponent.ctorParameters = function () { return [
        { type: TaskService },
        { type: HttpService },
        { type: UploadDialogService },
        { type: MessagerService },
        { type: I18nService },
        { type: LoadingService },
        { type: NotifyService }
    ]; };
    TaskCommentComponent.propDecorators = {
        showAttachment: [{ type: Input }],
        showSign: [{ type: Input }],
        showCommonComment: [{ type: Input }],
        comment: [{ type: Input }],
        placeholder: [{ type: Input }],
        processInstanceId: [{ type: Input }],
        textAreaHeight: [{ type: Input }],
        TaskCenterConfig: [{ type: Input }],
        attachment: [{ type: Output }],
        signature: [{ type: Output }],
        commentDialog: [{ type: ViewChild, args: ['commentDialog',] }],
        signDialog: [{ type: ViewChild, args: ['signDialog',] }],
        dropDownEl: [{ type: ViewChild, args: ['dropdownEl',] }],
        signaturePad: [{ type: ViewChild, args: ['wfSignHere',] }],
        farrisAvatar: [{ type: ViewChild, args: ['farrisAvatar',] }],
        dataGrid: [{ type: ViewChild, args: ['dataGrid',] }],
        commentTypeCell: [{ type: ViewChild, args: ['commentTypeCell',] }],
        commentOperateCell: [{ type: ViewChild, args: ['commentOperateCell',] }]
    };
    return TaskCommentComponent;
}());
export { TaskCommentComponent };
if (false) {
    /** @type {?} */
    TaskCommentComponent.prototype.showAttachment;
    /** @type {?} */
    TaskCommentComponent.prototype.showSign;
    /** @type {?} */
    TaskCommentComponent.prototype.showCommonComment;
    /** @type {?} */
    TaskCommentComponent.prototype.comment;
    /** @type {?} */
    TaskCommentComponent.prototype.placeholder;
    /** @type {?} */
    TaskCommentComponent.prototype.processInstanceId;
    /** @type {?} */
    TaskCommentComponent.prototype.textAreaHeight;
    /** @type {?} */
    TaskCommentComponent.prototype.attachment;
    /** @type {?} */
    TaskCommentComponent.prototype.signature;
    /** @type {?} */
    TaskCommentComponent.prototype.commentDialog;
    /** @type {?} */
    TaskCommentComponent.prototype.signDialog;
    /** @type {?} */
    TaskCommentComponent.prototype.dropDownEl;
    /** @type {?} */
    TaskCommentComponent.prototype.signaturePad;
    /** @type {?} */
    TaskCommentComponent.prototype.farrisAvatar;
    /** @type {?} */
    TaskCommentComponent.prototype.dataGrid;
    /** @type {?} */
    TaskCommentComponent.prototype.commentTypeCell;
    /** @type {?} */
    TaskCommentComponent.prototype.commentOperateCell;
    /**
     * @type {?}
     * @private
     */
    TaskCommentComponent.prototype._value;
    /** @type {?} */
    TaskCommentComponent.prototype.attachmentList;
    /** @type {?} */
    TaskCommentComponent.prototype.ifShowAttachment;
    /** @type {?} */
    TaskCommentComponent.prototype.ifShowSign;
    /** @type {?} */
    TaskCommentComponent.prototype.curUserId;
    /** @type {?} */
    TaskCommentComponent.prototype.signatureImage;
    /** @type {?} */
    TaskCommentComponent.prototype.tmpSignatureImage;
    /** @type {?} */
    TaskCommentComponent.prototype.downloadService;
    /** @type {?} */
    TaskCommentComponent.prototype.ifUseSign;
    /** @type {?} */
    TaskCommentComponent.prototype.ifUseImg;
    /** @type {?} */
    TaskCommentComponent.prototype.ifEnableSign;
    /** @type {?} */
    TaskCommentComponent.prototype.ifUploadImg;
    /** @type {?} */
    TaskCommentComponent.prototype.ifCanEditSign;
    /** @type {?} */
    TaskCommentComponent.prototype.originSignId;
    /** @type {?} */
    TaskCommentComponent.prototype.commentList;
    /** @type {?} */
    TaskCommentComponent.prototype.originCommentList;
    /** @type {?} */
    TaskCommentComponent.prototype.commentColumns;
    /** @type {?} */
    TaskCommentComponent.prototype.onModelChange;
    /** @type {?} */
    TaskCommentComponent.prototype.onModelTouched;
    /**
     * 系统预置的不允许编辑删除
     * \@param data data.rowData为实体数据
     * \@return
     * @type {?}
     */
    TaskCommentComponent.prototype.beforeEdit;
    /**
     * @type {?}
     * @private
     */
    TaskCommentComponent.prototype.taskService;
    /**
     * @type {?}
     * @private
     */
    TaskCommentComponent.prototype.httpSvc;
    /**
     * @type {?}
     * @private
     */
    TaskCommentComponent.prototype.uploadDialog;
    /**
     * @type {?}
     * @private
     */
    TaskCommentComponent.prototype.messageService;
    /**
     * @type {?}
     * @private
     */
    TaskCommentComponent.prototype.i18nService;
    /**
     * @type {?}
     * @private
     */
    TaskCommentComponent.prototype.loadingService;
    /**
     * @type {?}
     * @private
     */
    TaskCommentComponent.prototype.notify;
}
