// @farris/ui-datagrid/esm5/lib/components/header/header-style.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, ElementRef, Renderer2, Injector, Input } from '@angular/core';
var HeaderStyleDirective = /** @class */ (function () {
    function HeaderStyleDirective(el, render, inject) {
        this.el = el;
        this.render = render;
        this.inject = inject;
    }
    /**
     * @return {?}
     */
    HeaderStyleDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.renderCss2Element();
    };
    /**
     * @private
     * @return {?}
     */
    HeaderStyleDirective.prototype.renderCss2Element = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.column && this.column.headerStyler) {
            /** @type {?} */
            var styler = this.column.headerStyler;
            /** @type {?} */
            var dom = this.el.nativeElement;
            /** @type {?} */
            var st = typeof this.column.headerStyler;
            if (st === 'string') {
                this.render.addClass(this.el.nativeElement, (/** @type {?} */ (styler)));
            }
            else if (st === 'function') {
                /** @type {?} */
                var r = ((/** @type {?} */ (this.column.headerStyler)))(this.column);
                if (r.cls) {
                    this.render.addClass(this.el.nativeElement, r.cls);
                }
                if (r.style) {
                    this.renderStyle(dom, r.style);
                }
            }
            else if (st === 'object' && Object.keys(st).length) {
                this.renderStyle(dom, styler);
            }
        }
    };
    /**
     * @private
     * @param {?} targetEl
     * @param {?} styleObj
     * @return {?}
     */
    HeaderStyleDirective.prototype.renderStyle = /**
     * @private
     * @param {?} targetEl
     * @param {?} styleObj
     * @return {?}
     */
    function (targetEl, styleObj) {
        var _this = this;
        /** @type {?} */
        var cssKeys = Object.keys(styleObj);
        cssKeys.forEach((/**
         * @param {?} k
         * @return {?}
         */
        function (k) {
            _this.render.setStyle(targetEl, k, styleObj[k]);
        }));
    };
    HeaderStyleDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[header-style]',
                    exportAs: 'HeaderStyler'
                },] }
    ];
    /** @nocollapse */
    HeaderStyleDirective.ctorParameters = function () { return [
        { type: ElementRef },
        { type: Renderer2 },
        { type: Injector }
    ]; };
    HeaderStyleDirective.propDecorators = {
        column: [{ type: Input, args: ['header-style',] }]
    };
    return HeaderStyleDirective;
}());
export { HeaderStyleDirective };
if (false) {
    /** @type {?} */
    HeaderStyleDirective.prototype.column;
    /**
     * @type {?}
     * @private
     */
    HeaderStyleDirective.prototype.el;
    /**
     * @type {?}
     * @private
     */
    HeaderStyleDirective.prototype.render;
    /**
     * @type {?}
     * @private
     */
    HeaderStyleDirective.prototype.inject;
}
