// @farris/ui-datagrid-settings/esm5/lib/datagrid-view-columns.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, ViewChild, Injector, EventEmitter, Output, ChangeDetectorRef } from '@angular/core';
import { MultiSelectComponent } from '@farris/ui-multi-select';
var DatagridViewColumnsComponent = /** @class */ (function () {
    function DatagridViewColumnsComponent(injector, cd) {
        this.injector = injector;
        this.cd = cd;
        this.idField = 'field';
        this.valueField = 'field';
        this.textField = 'title';
        this._selectedIds = [];
        this.treeData = [];
        this.selectedIdsChange = new EventEmitter();
    }
    Object.defineProperty(DatagridViewColumnsComponent.prototype, "selectedIds", {
        get: /**
         * @return {?}
         */
        function () {
            return this._selectedIds;
        },
        set: /**
         * @param {?} vals
         * @return {?}
         */
        function (vals) {
            this._selectedIds = vals;
            this.selectedIdsChange.emit(vals);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    DatagridViewColumnsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @param {?} changes
     * @return {?}
     */
    DatagridViewColumnsComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    DatagridViewColumnsComponent.prototype.onBeforeMove = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
    };
    /**
     * @param {?} $event
     * @param {?} column
     * @return {?}
     */
    DatagridViewColumnsComponent.prototype.hiddenColumn = /**
     * @param {?} $event
     * @param {?} column
     * @return {?}
     */
    function ($event, column) {
        $event.stopPropagation();
        this.ms.moveToLeft(column);
    };
    DatagridViewColumnsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'view-columns',
                    template: "<multi-select [dataSource]=\"treeData\" [idField]=\"idField\" [valueField]=\"valueField\"\r\n[textField]=\"textField\" [displayType]=\"'Tree'\" style=\"height: 100%\" [height]=\"0\" [showCheckbox]=\"false\"\r\n[(selectedId)]=\"selectedIds\" [leftTitle]=\"'datagrid.settings.canchoose'|locale\" \r\n[rightTitle]=\"'datagrid.settings.choosed'|locale\" #ms (beforeMove)=\"onBeforeMove($event)\">\r\n    <ng-template farrisTemplate=\"text\" let-data let-side=\"side\" let-selected=\"selected\">\r\n        <ng-container *ngIf=\"side === 'L'\">\r\n            {{ data[textField] }}\r\n        </ng-container>\r\n        <ng-container *ngIf=\"side === 'R' \">\r\n\r\n            {{ data[textField] }} \r\n\r\n\r\n            <span class=\"sort-btns\" >\r\n                <a href=\"javascript: void(0)\" class=\" p-1\" (click)=\"hiddenColumn($event, data)\">\r\n                    <span class=\"f-icon f-icon-close\"></span>\r\n                </a>\r\n            </span>\r\n\r\n        </ng-container>\r\n    </ng-template>\r\n</multi-select>"
                }] }
    ];
    /** @nocollapse */
    DatagridViewColumnsComponent.ctorParameters = function () { return [
        { type: Injector },
        { type: ChangeDetectorRef }
    ]; };
    DatagridViewColumnsComponent.propDecorators = {
        idField: [{ type: Input }],
        valueField: [{ type: Input }],
        textField: [{ type: Input }],
        selectedIds: [{ type: Input }],
        treeData: [{ type: Input }],
        ms: [{ type: ViewChild, args: ['ms',] }],
        selectedIdsChange: [{ type: Output }]
    };
    return DatagridViewColumnsComponent;
}());
export { DatagridViewColumnsComponent };
if (false) {
    /** @type {?} */
    DatagridViewColumnsComponent.prototype.idField;
    /** @type {?} */
    DatagridViewColumnsComponent.prototype.valueField;
    /** @type {?} */
    DatagridViewColumnsComponent.prototype.textField;
    /**
     * @type {?}
     * @private
     */
    DatagridViewColumnsComponent.prototype._selectedIds;
    /** @type {?} */
    DatagridViewColumnsComponent.prototype.treeData;
    /** @type {?} */
    DatagridViewColumnsComponent.prototype.ms;
    /** @type {?} */
    DatagridViewColumnsComponent.prototype.selectedIdsChange;
    /**
     * @type {?}
     * @private
     */
    DatagridViewColumnsComponent.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    DatagridViewColumnsComponent.prototype.cd;
}
