// @gsp-wf/wf-process-editor/esm2015/lib/views/expressionEditor/expression-editor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, Input, EventEmitter, ViewChild, TemplateRef } from '@angular/core';
import { ExpressionService, ExpressioneditorComponent, ComputeRuleEditorComponent } from '@gsp-svc/expression';
import { ProcessDesignerUIState } from '../../app/process-designer.uistate';
export class ExpressionEditorComponent {
    /**
     * @param {?} expr
     * @param {?} designerState
     */
    constructor(expr, designerState) {
        this.expr = expr;
        this.designerState = designerState;
        this.closeModal = new EventEmitter();
        this.submitModal = new EventEmitter();
        this.editorParams = {};
        this.contextEntities = [];
        this.schema = {};
        this.injectName = '';
        this.modalConfig = {
            title: '条件编辑',
            width: 1140,
            height: 520,
            showButtons: true
        };
        if (this.designerState.schemas[0]) {
            this.schema = this.designerState.schemas[0];
            if (this.designerState.schemas[0].entityTypes && this.designerState.schemas[0].entityTypes.length > 0) {
                this.injectName = this.designerState.schemas[0].entityTypes[0].name;
            }
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.contextEntities = new Array();
        if (this.designerState.procInstContext.length > 0) {
            for (const variable of this.designerState.procInstContext) {
                this.contextEntities = this.contextEntities.concat([variable]);
            }
        }
        if (this.designerState.formalParameterContext.length > 0) {
            for (const variable of this.designerState.formalParameterContext) {
                if (!this.contextEntities.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                c => c.key === variable.key))) {
                    this.contextEntities = this.contextEntities.concat([variable]);
                }
            }
        }
        if (this.designerState.participantContext.length > 0) {
            for (const variable of this.designerState.participantContext) {
                if (!this.contextEntities.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                c => c.key === variable.key))) {
                    this.contextEntities = this.contextEntities.concat([variable]);
                }
            }
        }
        if (this.designerState.activityComponentContext.length > 0) {
            for (const variable of this.designerState.activityComponentContext) {
                if (!this.contextEntities.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                c => c.key === variable.key))) {
                    this.contextEntities = this.contextEntities.concat([variable]);
                }
            }
        }
        if (this.designerState.processVariableContext.length > 0) {
            for (const variable of this.designerState.processVariableContext) {
                if (!this.contextEntities.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                c => c.key === variable.key))) {
                    this.contextEntities = this.contextEntities.concat([variable]);
                }
            }
        }
    }
    /**
     * @return {?}
     */
    clickOK() {
        /** @type {?} */
        const result = this.compute.getExpr();
        this.value = this.convertToValue(result.text);
        /** @type {?} */
        const object = { value: this.value, parameters: null };
        this.submitModal.emit(object);
    }
    /**
     * 切换表达式类型（简易表达式<--->高级表达式）
     * @return {?}
     */
    changeType() {
        this.compute.changeType();
    }
    /**
     * @return {?}
     */
    clickCancel() {
        this.closeModal.emit();
    }
    /**
     * @private
     * @param {?} val
     * @return {?}
     */
    convertToValue(val) {
        if (val) {
            if (val.substring(9, val.length - 2)) {
                return val;
            }
            else {
                return '';
            }
        }
        else {
            return '';
        }
    }
}
ExpressionEditorComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-expression-editor',
                template: "<app-compute-rule-editor #compute\r\n  [text]=\"value\"\r\n  [advancedContexts]=\"contextEntities\"\r\n  [injectName]=\"injectName\"\r\n  [injectSchemaName]=\"'Schema'\"\r\n  [injectNameInSchema]=\"injectName\"\r\n  [schema]=\"schema\"\r\n  [schemaName]=\"'Schema'\"   >\r\n</app-compute-rule-editor>\r\n\r\n<ng-template #iconSelectFooter>\r\n  <div class=\"btns\">\r\n    <button type=\"button\" class=\"btn btn-primary btn-sm mr-2\" (click)=\"clickOK()\">\u786E\u5B9A</button>\r\n    <button type=\"button\" class=\"btn btn-secondary btn-sm mr-2\" (click)=\"changeType()\">\u5207\u6362\u7C7B\u578B</button>\r\n    <button type=\"button\" class=\"btn btn-secondary btn-sm\" (click)=\"clickCancel()\">\u53D6\u6D88</button>\r\n  </div>\r\n</ng-template>\r\n",
                styles: [""]
            }] }
];
/** @nocollapse */
ExpressionEditorComponent.ctorParameters = () => [
    { type: ExpressionService },
    { type: ProcessDesignerUIState }
];
ExpressionEditorComponent.propDecorators = {
    closeModal: [{ type: Output }],
    submitModal: [{ type: Output }],
    value: [{ type: Input }],
    modalFooter: [{ type: ViewChild, args: ['iconSelectFooter',] }],
    editorParams: [{ type: Input }],
    editor: [{ type: ViewChild, args: [ExpressioneditorComponent,] }],
    compute: [{ type: ViewChild, args: ['compute',] }]
};
if (false) {
    /** @type {?} */
    ExpressionEditorComponent.prototype.closeModal;
    /** @type {?} */
    ExpressionEditorComponent.prototype.submitModal;
    /** @type {?} */
    ExpressionEditorComponent.prototype.value;
    /** @type {?} */
    ExpressionEditorComponent.prototype.modalFooter;
    /** @type {?} */
    ExpressionEditorComponent.prototype.editorParams;
    /** @type {?} */
    ExpressionEditorComponent.prototype.editor;
    /** @type {?} */
    ExpressionEditorComponent.prototype.compute;
    /** @type {?} */
    ExpressionEditorComponent.prototype.contextEntities;
    /** @type {?} */
    ExpressionEditorComponent.prototype.schema;
    /** @type {?} */
    ExpressionEditorComponent.prototype.injectName;
    /** @type {?} */
    ExpressionEditorComponent.prototype.modalConfig;
    /**
     * @type {?}
     * @private
     */
    ExpressionEditorComponent.prototype.expr;
    /** @type {?} */
    ExpressionEditorComponent.prototype.designerState;
}
export class ExpressionEditorConverter {
    constructor() { }
    /**
     * @param {?} data
     * @return {?}
     */
    convertTo(data) {
        if (data) {
            return data;
        }
        else {
            return '（未设置）';
        }
    }
}
