// @farris/command-services/esm5/lib/data-services/save-data.service.js
import { Injectable } from '@angular/core';
import { FrameContext } from '@farris/devkit';
import { FormLoadingService } from '../form-loading/form-loading.service';
import { FormNotifyService } from '../form-notify.service';
import { LanguageService } from '../languag.service';
import { FormMessageService } from '../form-message.service';
/**
 * 删除服务
 */
var SaveDataService = /** @class */ (function () {
    /**
     * 构造函数
     * @param frameContext 组件上下文
     */
    function SaveDataService(frameContext) {
        this.frameContext = frameContext;
        this.notifyService = this.frameContext.injector.get(FormNotifyService, null);
        this.messageService = this.frameContext.injector.get(FormMessageService, null);
        this.loadingService = this.frameContext.injector.get(FormLoadingService, null);
        this.languageService = this.frameContext.injector.get(LanguageService, null);
        this.befRepository = this.frameContext.repository;
    }
    SaveDataService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    SaveDataService.ctorParameters = function () { return [
        { type: FrameContext }
    ]; };
    return SaveDataService;
}());
export { SaveDataService };
