// @farris/command-services/esm5/lib/data-services/sub-list-data.service.js
import { Injectable, Optional } from '@angular/core';
import { empty, of, EMPTY } from 'rxjs';
import { tap, switchMap, concatMap } from 'rxjs/operators';
import { Repository, ViewModel } from '@farris/devkit';
import { FormLoadingService } from '../form-loading/form-loading.service';
import { FormNotifyService } from '../form-notify.service';
import { LanguageService } from '../languag.service';
import { FormMessageService } from '../form-message.service';
import { FormErrorService } from '../error/form-error.service';
/**
 * 子列表取数服务
 */
var SubListDataService = /** @class */ (function () {
    /**
     * 构造函数
     */
    function SubListDataService(msgService, repository, loadingService, viewModel, languageService, formNotifyService, formErrorService) {
        this.msgService = msgService;
        this.repository = repository;
        this.loadingService = loadingService;
        this.viewModel = viewModel;
        this.languageService = languageService;
        this.formNotifyService = formNotifyService;
        this.formErrorService = formErrorService;
        if (!languageService) {
            this.languageService = LanguageService.getInstance();
        }
        this.viewModel = viewModel;
    }
    /**
     * 新增
     */
    SubListDataService.prototype.add = function () {
        var _this = this;
        if (this.messagePipe) {
            this.messagePipe.next({ messageType: 'add' });
        }
        var path = this.getPath();
        var loadingTimerId = this.loadingService.showLoadingWithDelay(500);
        var append$ = this.repository.appendByPath(path);
        return append$.pipe(tap(function () {
            _this.loadingService.hideDelayLoading(loadingTimerId);
        }, function (error) {
            _this.loadingService.hideDelayLoading(loadingTimerId);
            _this.formErrorService.exception(_this.languageService.addFailed, error);
        }));
    };
    /**
     * 在指定位置插入
     * @param position 位置
     */
    SubListDataService.prototype.insert = function (position) {
        var _this = this;
        if (position === void 0) { position = -1; }
        if (this.messagePipe) {
            this.messagePipe.next({ messageType: 'insert' });
        }
        var path = this.getPath();
        var loadingTimerId = this.loadingService.showLoadingWithDelay(500);
        var append$ = this.repository.insertByPath(path, position);
        return append$.pipe(tap(function () {
            _this.loadingService.hideDelayLoading(loadingTimerId);
        }, function (error) {
            _this.loadingService.hideDelayLoading(loadingTimerId);
            _this.formErrorService.exception(_this.languageService.addFailed, error);
        }));
    };
    /**
     * 删除子表数据
     */
    SubListDataService.prototype.remove = function (id, successMsg) {
        var msg = successMsg ? successMsg : '';
        return this.innerRemove(id, false, msg);
    };
    /**
     * 删除子表数据（不确认）
     */
    SubListDataService.prototype.removeWithoutConfirm = function (id) {
        return this.innerRemove(id, true, '');
    };
    /**
     * 批量删除子表
     * @param ids ids
     * @param successMsg 自定义提示信息
     */
    SubListDataService.prototype.removeChildrenByIds = function (ids, successMsg) {
        var _this = this;
        if (!ids) {
            this.formNotifyService.warning(this.languageService['plsSelectDeleteData'], { hideTitle: true });
            return EMPTY;
        }
        var action$ = this.msgService.confirm(this.languageService.confirmDeletion);
        return action$.pipe(concatMap(function (result) {
            if (!result) {
                return EMPTY;
            }
            var path = _this.getPath();
            var loadingTimerId = _this.loadingService.showLoadingWithDelay(500);
            var remove$ = _this.repository.batchRemoveByPath(path, ids);
            return remove$.pipe(tap(function () {
                _this.loadingService.hideDelayLoading(loadingTimerId);
                if (successMsg && successMsg.trim()) {
                    _this.formNotifyService.success(successMsg, { hideTitle: true });
                }
                else {
                    _this.formNotifyService.success(_this.languageService.deleteSuccess, { hideTitle: true });
                }
                // this.formNotifyService.success(this.languageService.deleteSuccess, { hideTitle: true });
            }, function (error) {
                _this.loadingService.hideDelayLoading(loadingTimerId);
                _this.formErrorService.exception(_this.languageService.deleteFailed, error);
            }));
        }));
    };
    /**
     * 删除
     */
    SubListDataService.prototype.innerRemove = function (id, isConfirm, successMsg) {
        var _this = this;
        if (this.messagePipe) {
            this.messagePipe.next({ messageType: 'remove' });
        }
        if (!id) {
            this.formNotifyService.warning(this.languageService['plsSelectDeleteData'], { hideTitle: true });
            return empty();
        }
        var action$;
        if (isConfirm === false) {
            action$ = this.msgService.question(this.languageService.confirmDeletion);
        }
        else {
            action$ = of(true);
        }
        return action$.pipe(concatMap(function (result) {
            if (!result) {
                return empty();
            }
            var path = _this.getPath();
            var loadingTimerId = _this.loadingService.showLoadingWithDelay(500);
            var remove$ = _this.repository.removeByPath(path, id);
            return remove$.pipe(tap(function () {
                _this.loadingService.hideDelayLoading(loadingTimerId);
                if (successMsg && successMsg.trim()) {
                    _this.formNotifyService.success(successMsg, { hideTitle: true });
                }
                else {
                    _this.formNotifyService.success(_this.languageService.deleteSuccess, { hideTitle: true });
                }
                // this.formNotifyService.success(this.languageService.deleteSuccess, { hideTitle: true });
            }, function (error) {
                _this.loadingService.hideDelayLoading(loadingTimerId);
                _this.formErrorService.exception(_this.languageService.deleteFailed, error);
            }));
        }));
    };
    /**
     * 删除并保存
     * @param id id
     * @param successMsg 自定义提示信息
     */
    SubListDataService.prototype.removeAndSave = function (id, successMsg) {
        var _this = this;
        if (this.messagePipe) {
            this.messagePipe.next({ messageType: 'removeAndSave' });
        }
        if (!id) {
            this.formNotifyService.warning(this.languageService.plsSelectDeleteData, { hideTitle: true });
            return empty();
        }
        var action$ = this.msgService.confirm(this.languageService.confirmDeletion);
        return action$.pipe(concatMap(function (result) {
            if (!result) {
                return empty();
            }
            // 删除子表数据
            var path = _this.getPath();
            var loadingTimerId = _this.loadingService.showLoadingWithDelay(500);
            var remove$ = _this.repository.removeByPath(path, id);
            return remove$.pipe(
            // 执行主表保存
            switchMap(function () {
                var fid = _this.viewModel.bindingData.list.currentId;
                return _this.repository.applyChangesById(fid);
            }), 
            // 隐藏加载中
            tap(function () {
                _this.loadingService.hideDelayLoading(loadingTimerId);
                if (successMsg && successMsg.trim()) {
                    _this.formNotifyService.success(successMsg, { hideTitle: true });
                }
                else {
                    _this.formNotifyService.success(_this.languageService.deleteSuccess, { hideTitle: true });
                }
                // this.formNotifyService.success(this.languageService.deleteSuccess, { hideTitle: true });
            }, function (error) {
                _this.loadingService.hideDelayLoading(loadingTimerId);
                _this.formErrorService.exception(_this.languageService.deleteFailed, error);
            }));
        }));
    };
    /**
     * 上移或下移数据
     * @param direction 方向，up | down
     * @param field 排序字段
     * @param targets 要移动的数据
     * @description
     * 使用该方法时排序字段必须有值，否则无法排序
     */
    SubListDataService.prototype.move = function (direction, field, targets) {
        var _this = this;
        if (!targets) {
            return;
        }
        if (typeof targets === 'string') {
            targets = targets.split(',').filter(function (p) { return p; });
        }
        var bindingPath = this.viewModel && this.viewModel.bindingPath || null;
        if (!bindingPath) {
            return;
        }
        var bindingList = this.viewModel.bindingData.getList();
        if (!bindingList || bindingList.length < 1) {
            return;
        }
        targets.forEach(function (target) {
            var index = bindingList.getIndexById(target);
            if (index === -1) {
                return;
            }
            var item = bindingList.findById(target);
            var position = item.getValue(field);
            var base = direction === 'up' ? -1 : 1;
            // 待交换的行信息
            var exchangeRowIndex = index + base;
            if (exchangeRowIndex < 0 || exchangeRowIndex > bindingList.length) {
                // 第一行无法上移，最后一行无法下移
                return;
            }
            var exchangeRowId = bindingList.getIdByIndex(exchangeRowIndex);
            var exchangeRow = bindingList.findById(exchangeRowId);
            var exchangeRowPosition = exchangeRow[field];
            // 移动行和交换行都没有排序
            if (_this.isNullOrEmpty(position) && _this.isNullOrEmpty(exchangeRowPosition)) {
                return;
            }
            bindingList.swapById(target, exchangeRowId);
            exchangeRow.setValue(field, position, true, true);
            item.setValue(field, exchangeRowPosition, true, true);
        });
    };
    SubListDataService.prototype.isNullOrEmpty = function (value) {
        return value === '' || value === null || value === undefined;
    };
    /**
     * 获取完整路径
     * @todo：强识别到从表这一级
     * fixed by justin: 根据bindingPath，如果是从从表，需要获取主表数据id和从表数据id
     */
    SubListDataService.prototype.getPath = function () {
        var bindingPath = this.viewModel.bindingPath;
        var rid = this.viewModel.bindingData.list.currentId; // root表数据id
        var path = '/' + rid;
        var subPaths = bindingPath.split('/');
        if (subPaths.length > 0) {
            // eg:bindingPath形如/edus/grades,split后是['', 'edus', 'grades']
            // 因此index从1开始
            for (var index = 1; index < subPaths.length - 1; index++) {
                var subPath = subPaths[index];
                var subData = this.viewModel.bindingData[subPath];
                if (!subData || !subData.currentId) {
                    this.formNotifyService.warning(this.languageService['plsSelectDetailFormData'], { hideTitle: true });
                    throw Error("\u83B7\u53D6\u5B50\u8868\u5B8C\u6574\u8DEF\u5F84\u51FA\u9519\uFF0C\u627E\u4E0D\u5230" + subData + "\u5BF9\u5E94\u7684\u5B50\u8868\uFF0C\u6216\u5BF9\u5E94\u5B50\u8868\u6CA1\u6709\u5F53\u524D\u884C\u3002");
                }
                path += "/" + subPath + "/" + subData.currentId;
            }
        }
        path += '/' + subPaths[subPaths.length - 1];
        return path;
    };
    Object.defineProperty(SubListDataService.prototype, "messagePipe", {
        /**
         * 获取根组件appContext
         */
        get: function () {
            if (this.viewModel && this.viewModel.frameContext) {
                var appContext = this.viewModel.frameContext.getFormAppContext() || null;
                if (appContext) {
                    return appContext.messagePipe || null;
                }
            }
            return null;
        },
        enumerable: true,
        configurable: true
    });
    SubListDataService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    SubListDataService.ctorParameters = function () { return [
        { type: FormMessageService },
        { type: Repository },
        { type: FormLoadingService },
        { type: ViewModel },
        { type: LanguageService, decorators: [{ type: Optional }] },
        { type: FormNotifyService },
        { type: FormErrorService }
    ]; };
    return SubListDataService;
}());
export { SubListDataService };
