// @gsp-svc/formdoc-upload/esm5/lib/uploaddialog/uploaddialog.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FarrisDialogModule } from '@farris/ui-dialog';
import { ButtonsModule } from '@progress/kendo-angular-buttons';
import { MessagerModule } from '@farris/ui-messager';
import { BsModalService } from '@farris/ui-modal';
import { UploadModule } from '../upload/upload.module';
import { UploadDialogComponent } from './uploaddialog/uploaddialog.component';
import { HttpService } from '@ecp-caf/caf-common';
import { UploadService } from '../upload/upload.service';
import { UploadDialogService } from '../uploaddialog.service';
import { HttpClientModule } from '@angular/common/http';
import { NotifyModule } from '@farris/ui-notify';
import { FileUploadComponent } from '../upload/file-upload/file-upload.component';
var UploadDialogMoudle = /** @class */ (function () {
    function UploadDialogMoudle() {
    }
    UploadDialogMoudle.decorators = [
        { type: NgModule, args: [{
                    declarations: [UploadDialogComponent],
                    imports: [
                        CommonModule,
                        FarrisDialogModule,
                        MessagerModule.forRoot(),
                        UploadModule,
                        ButtonsModule,
                        NotifyModule.forRoot(),
                        HttpClientModule
                    ],
                    exports: [UploadDialogComponent, FileUploadComponent],
                    entryComponents: [UploadDialogComponent, FileUploadComponent],
                    providers: [BsModalService, HttpService, UploadDialogService, UploadService]
                },] }
    ];
    return UploadDialogMoudle;
}());
export { UploadDialogMoudle };
