// @farris/ui-button/esm2015/lib/button.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, EventEmitter, Output } from '@angular/core';
export class FarrisButtonComponent {
    constructor() {
        this.size = 'small';
        this.type = 'primary';
        this.click = new EventEmitter();
    }
    /**
     * @param {?} e
     * @return {?}
     */
    clickEvent(e) {
        e.stopPropagation();
        if (this.disabled) {
            return;
        }
        this.click.emit(e);
    }
}
FarrisButtonComponent.decorators = [
    { type: Component, args: [{
                selector: 'farris-button',
                template: `
    <button [class]="(size === 'large'?'btn-lg ':'btn-sm ')+
    'btn '+ (type?'btn-'+type+' ':'')"
    [class.disabled]="disabled" (click)="clickEvent($event)">
      <ng-content></ng-content>
    </button>
  `
            }] }
];
/** @nocollapse */
FarrisButtonComponent.ctorParameters = () => [];
FarrisButtonComponent.propDecorators = {
    text: [{ type: Input }],
    size: [{ type: Input }],
    type: [{ type: Input }],
    disabled: [{ type: Input }],
    click: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    FarrisButtonComponent.prototype.text;
    /** @type {?} */
    FarrisButtonComponent.prototype.size;
    /** @type {?} */
    FarrisButtonComponent.prototype.type;
    /** @type {?} */
    FarrisButtonComponent.prototype.disabled;
    /** @type {?} */
    FarrisButtonComponent.prototype.click;
}
