// @farris/kendo-binding/esm2015/lib/directives/component-ref/discussion-list-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { DiscussionListComponent } from '@farris/discussion-group';
export class DiscussionListComponentRefDirective {
    constructor(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ngOnInit() {
        const id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    }
    ngOnDestroy() {
    }
}
DiscussionListComponentRefDirective.decorators = [
    { type: Directive, args: [{
                selector: '[discussionListRef]'
            },] }
];
/** @nocollapse */
DiscussionListComponentRefDirective.ctorParameters = () => [
    { type: Injector },
    { type: DiscussionListComponent },
    { type: ElementRef },
    { type: FrameContext }
];
