// @farris/kendo-binding/esm2015/lib/directives/component-ref/nav-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { ListNavComponent } from '@farris/ui-list-nav';
export class NavComponentRefDirective {
    constructor(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ngOnInit() {
        const id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    }
    ngOnDestroy() {
    }
}
NavComponentRefDirective.decorators = [
    { type: Directive, args: [{
                selector: '[navRef]'
            },] }
];
/** @nocollapse */
NavComponentRefDirective.ctorParameters = () => [
    { type: Injector },
    { type: ListNavComponent },
    { type: ElementRef },
    { type: FrameContext }
];
