// @farris/extend-page-modal/esm5/lib/page-modal.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/page-modal.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { PerfectScrollbarModule } from '@farris/ui-perfect-scrollbar';
import { AngularDraggableModule } from '@farris/ui-draggable';
//import { ModalModule } from '@farris/ui-modal';
import { PositioningService } from '@farris/ui-modal/positioning';
import { ComponentLoaderFactory } from '@farris/ui-modal/component-loader';
import { FAreaResponseModule } from '@farris/ui-area-response';
import { FEPageModalComponent } from './page-modal.component';
import { FEModalBackdropComponent } from './page-modal-backdrop.component';
import { FEPageModalService } from './page-modal.service';
import { FEPageModalSidebarComponent } from './page-sidebar.component';
var FEPageModalModule = /** @class */ (function () {
    function FEPageModalModule() {
    }
    /**
     * @return {?}
     */
    FEPageModalModule.forRoot = /**
     * @return {?}
     */
    function () {
        return {
            ngModule: FEPageModalModule,
            providers: [
                ComponentLoaderFactory,
                FEPageModalService,
                PositioningService
            ]
        };
    };
    FEPageModalModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [FEPageModalComponent, FEModalBackdropComponent, FEPageModalSidebarComponent],
                    imports: [
                        CommonModule,
                        PerfectScrollbarModule,
                        AngularDraggableModule,
                        FAreaResponseModule
                    ],
                    exports: [FEModalBackdropComponent],
                    entryComponents: [FEPageModalComponent, FEModalBackdropComponent, FEPageModalSidebarComponent]
                },] }
    ];
    return FEPageModalModule;
}());
export { FEPageModalModule };
