// @farris/ui-view-change/esm5/lib/view-change.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import * as i0 from "@angular/core";
/**
 * @record
 */
function ViewItems() { }
if (false) {
    /** @type {?} */
    ViewItems.prototype.items;
}
/**
 * @record
 */
function viewContainer() { }
var ViewChangeService = /** @class */ (function () {
    function ViewChangeService() {
        /**
         * 默认视图切换分组名称
         */
        this.defaultId = 'default';
        //当前显示视图
        this.currentViewType = {};
        /**
         * contents
         */
        this.viewContainers = {};
        this.changeCurrentViewId = new BehaviorSubject(null);
    }
    /**
     * @param {?} itemId
     * @param {?=} viewContainerId
     * @return {?}
     */
    ViewChangeService.prototype.addViewItem = /**
     * @param {?} itemId
     * @param {?=} viewContainerId
     * @return {?}
     */
    function (itemId, viewContainerId) {
        if (viewContainerId === void 0) { viewContainerId = this.defaultId; }
        if (this.viewContainers[viewContainerId]) {
            if (this.hasElement(itemId, viewContainerId)) {
                return;
            }
            else {
                this.viewContainers[viewContainerId].items.push(itemId);
            }
        }
        else {
            this.viewContainers[viewContainerId] = (/** @type {?} */ ({
                items: []
            }));
            this.viewContainers[viewContainerId].items.push(itemId);
        }
    };
    /**
     * @param {?} itemId
     * @param {?=} viewContainerId
     * @return {?}
     */
    ViewChangeService.prototype.removeViewItem = /**
     * @param {?} itemId
     * @param {?=} viewContainerId
     * @return {?}
     */
    function (itemId, viewContainerId) {
        if (viewContainerId === void 0) { viewContainerId = this.defaultId; }
        delete this.viewContainers[viewContainerId];
        if (this.viewContainers && this.viewContainers[viewContainerId]) {
            /** @type {?} */
            var items = this.viewContainers[viewContainerId].items.filter((/**
             * @param {?} el
             * @return {?}
             */
            function (el) {
                return el !== itemId;
            }));
            if (!items.length) {
                delete this.viewContainers[viewContainerId];
                return;
            }
            this.viewContainers[viewContainerId].items = items;
        }
    };
    /**存储currentpageid */
    /**
     * 存储currentpageid
     * @param {?} type
     * @param {?=} viewContainerId
     * @return {?}
     */
    ViewChangeService.prototype.updateCurrentViewId = /**
     * 存储currentpageid
     * @param {?} type
     * @param {?=} viewContainerId
     * @return {?}
     */
    function (type, viewContainerId) {
        if (viewContainerId === void 0) { viewContainerId = this.defaultId; }
        this.currentViewType[viewContainerId] = type;
        /** @type {?} */
        var changeView = {};
        changeView[viewContainerId] = type;
        this.changeCurrentViewId.next(changeView);
    };
    /**
     * @return {?}
     */
    ViewChangeService.prototype.getChangeCurrentViewId = /**
     * @return {?}
     */
    function () {
        return this.changeCurrentViewId;
    };
    /**
     * @private
     * @param {?} itemId
     * @param {?} groupId
     * @return {?}
     */
    ViewChangeService.prototype.hasElement = /**
     * @private
     * @param {?} itemId
     * @param {?} groupId
     * @return {?}
     */
    function (itemId, groupId) {
        return this.viewContainers[groupId].items.some((/**
         * @param {?} item
         * @return {?}
         */
        function (item) { return item === itemId; }));
    };
    ViewChangeService.decorators = [
        { type: Injectable, args: [{
                    providedIn: 'root'
                },] }
    ];
    /** @nocollapse */
    ViewChangeService.ctorParameters = function () { return []; };
    /** @nocollapse */ ViewChangeService.ngInjectableDef = i0.defineInjectable({ factory: function ViewChangeService_Factory() { return new ViewChangeService(); }, token: ViewChangeService, providedIn: "root" });
    return ViewChangeService;
}());
export { ViewChangeService };
if (false) {
    /**
     * 默认视图切换分组名称
     * @type {?}
     * @private
     */
    ViewChangeService.prototype.defaultId;
    /** @type {?} */
    ViewChangeService.prototype.currentViewType;
    /**
     * contents
     * @type {?}
     */
    ViewChangeService.prototype.viewContainers;
    /** @type {?} */
    ViewChangeService.prototype.changeCurrentViewId;
}
