// @farris/kendo-binding/esm2015/lib/directives/component-ref/datepicker-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { FarrisDatepickerComponent } from '@farris/ui-datepicker';
export class DatepickerComponentRefDirective {
    constructor(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ngOnInit() {
        const id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    }
    ngOnDestroy() {
    }
}
DatepickerComponentRefDirective.decorators = [
    { type: Directive, args: [{
                selector: '[datepickerRef]'
            },] }
];
/** @nocollapse */
DatepickerComponentRefDirective.ctorParameters = () => [
    { type: Injector },
    { type: FarrisDatepickerComponent },
    { type: ElementRef },
    { type: FrameContext }
];
