// @farris/kendo-binding/esm2015/lib/directives/component-ref/section-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { FarrisSectionComponent } from '@farris/ui-section';
export class SectionComponentRefDirective {
    constructor(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ngOnInit() {
        const id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    }
    ngOnDestroy() {
    }
}
SectionComponentRefDirective.decorators = [
    { type: Directive, args: [{
                selector: '[sectionRef]'
            },] }
];
/** @nocollapse */
SectionComponentRefDirective.ctorParameters = () => [
    { type: Injector },
    { type: FarrisSectionComponent },
    { type: ElementRef },
    { type: FrameContext }
];
