// @gsp-wf/wf-process-editor/esm2015/lib/views/message-notice/message-notice.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ProcessDeUtil } from '../../domain/process-de-util';
export class MessageNotice {
    constructor() {
        this.id = ProcessDeUtil.GenerateElementId();
        this.subject = '消息通知';
        // this.sendType = 'email';
        this.content = '{"expr":"\'您有一条消息通知。\'","sexpr":""}';
    }
}
if (false) {
    /** @type {?} */
    MessageNotice.prototype.id;
    /**
     * 消息内容
     * @type {?}
     */
    MessageNotice.prototype.content;
    /**
     * 消息发送时机
     * @type {?}
     */
    MessageNotice.prototype.sendTime;
    /**
     * 消息发送方式
     * @type {?}
     */
    MessageNotice.prototype.sendType;
    /**
     * 消息主题
     * @type {?}
     */
    MessageNotice.prototype.subject;
    /**
     * 消息接收人
     * @type {?}
     */
    MessageNotice.prototype.receiver;
    /**
     * 消息接收人显示
     * @type {?}
     */
    MessageNotice.prototype.receiverText;
}
/** @type {?} */
export const processSendTime = [
    { value: 'PROCESSINSTANCE_COMPLETED', text: '流程完成' },
    { value: 'PROCESSINSTANCE_ABORTED', text: '流程终止' },
    { value: 'PROCESSINSTANCE_ROLLBACK', text: '流程驳回' }
];
/** @type {?} */
export const activitySendTime = [
    { value: 'WORKITEM_CREATED', text: '任务创建' },
    { value: 'ACTIVITY_COMPLETED', text: '活动完成' },
    { value: 'ACTIVITY_ABORTED', text: '活动终止' },
];
/** @type {?} */
export const sendTypes = [
    { value: 'email', name: '邮件' },
    { value: 'sms', name: '短信' },
    { value: 'cloudplus', name: '云+' }
];
