// @gsp-wf/wf-process-editor/esm2015/lib/views/customize-action/views/customize-action-list/customize-action-list.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, ViewChild, TemplateRef, Output, EventEmitter } from '@angular/core';
import { ProcessDesignerUIState } from '../../../../app/process-designer.uistate';
import { DatagridComponent } from '@farris/ui-datagrid';
import { CustomizeAciton } from '../../entities/customize-action.entity';
export class CustomizeActionListComponent {
    /**
     * @param {?} designerUIState
     */
    constructor(designerUIState) {
        this.designerUIState = designerUIState;
        this.submit = new EventEmitter();
        this.cancel = new EventEmitter();
        this.terminalList = [
            { value: 'app', text: '移动' },
            { value: 'web', text: 'web' },
            { value: 'web,app', text: 'web,移动' }
        ];
        this.cols = [
            { field: 'code', width: 100, title: '编号' },
            { field: 'name', width: 100, title: '名称' },
            {
                field: 'terminal', width: 100, title: '适用终端',
                formatter: {
                    type: 'enum',
                    options: { valueField: 'value', textField: 'text', data: this.terminalList }
                },
            },
            {
                field: 'isAll', width: 100, title: '是否默认',
                formatter: { type: 'boolean', options: { trueText: '是', falseText: '否' } }
            }
        ];
        this.customActionList = this.designerUIState.flowFormInfo.customizeActions || [];
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    clickOK() {
        /** @type {?} */
        const selections = [];
        /** @type {?} */
        const checkeds = this.grid.checkeds;
        if (checkeds && checkeds.length > 0) {
            checkeds.forEach((/**
             * @param {?} action
             * @return {?}
             */
            action => {
                /** @type {?} */
                const item = new CustomizeAciton(action.data.id, action.data.code, action.data.name);
                item.terminalType = action.data.terminal;
                // item.enable = true;
                item.actualParameters = [];
                if (action.data.actualParameters && action.data.actualParameters.length > 0) {
                    for (const p of action.data.actualParameters) {
                        /** @type {?} */
                        const param = (/** @type {?} */ ({
                            Name: p.code,
                            Value: p.value,
                            ValueText: p.name,
                            Tag: 'expression'
                        }));
                        item.actualParameters.push(param);
                    }
                }
                selections.push(item);
            }));
        }
        this.submit.emit(selections);
    }
    /**
     * @return {?}
     */
    clickCancel() {
        this.cancel.emit(true);
    }
}
CustomizeActionListComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-customize-action-list',
                template: "<div class=\"f-utils-fill-flex-column farris-overflow-y-auto h-100 m-2\">\r\n  <farris-datagrid #grid [columns]=\"cols\" [data]=\"customActionList\" [idField]=\"'id'\" [fit]=\"true\"\r\n    [fitColumns]=\"true\" [pagination]='false' [multiSelect]=\"true\" [showCheckbox]=\"true\" [multiSelect]=\"true\"\r\n    [checkOnSelect]=\"true\" [selectOnCheck]=\"true\">\r\n  </farris-datagrid>\r\n</div>\r\n\r\n<ng-template #footer>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm px-3 mx-2\" (click)=\"clickOK()\">\u786E\u5B9A</button>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm px-3\" (click)=\"clickCancel()\">\u53D6\u6D88</button>\r\n</ng-template>",
                styles: [""]
            }] }
];
/** @nocollapse */
CustomizeActionListComponent.ctorParameters = () => [
    { type: ProcessDesignerUIState }
];
CustomizeActionListComponent.propDecorators = {
    grid: [{ type: ViewChild, args: ['grid',] }],
    footer: [{ type: ViewChild, args: ['footer',] }],
    submit: [{ type: Output }],
    cancel: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    CustomizeActionListComponent.prototype.grid;
    /** @type {?} */
    CustomizeActionListComponent.prototype.footer;
    /** @type {?} */
    CustomizeActionListComponent.prototype.submit;
    /** @type {?} */
    CustomizeActionListComponent.prototype.cancel;
    /** @type {?} */
    CustomizeActionListComponent.prototype.customActionList;
    /**
     * @type {?}
     * @private
     */
    CustomizeActionListComponent.prototype.terminalList;
    /** @type {?} */
    CustomizeActionListComponent.prototype.cols;
    /**
     * @type {?}
     * @private
     */
    CustomizeActionListComponent.prototype.designerUIState;
}
