// @farris/ui-enum-editor/esm2015/lib/enum-editor-footer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, EventEmitter, HostBinding, Output, Input } from '@angular/core';
import { EnumEditorComponent } from './enum-editor.component';
export class EnumEditorFooterComponent {
    /**
     * @param {?} grid
     */
    constructor(grid) {
        this.grid = grid;
        this.cls = 'f-filter-footer';
        this.showSortBtns = true;
        this.addFilter = new EventEmitter();
        this.clearFilter = new EventEmitter();
        this.moveTop = new EventEmitter();
        this.movePrev = new EventEmitter();
        this.moveNext = new EventEmitter();
        this.moveBottom = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @return {?}
     */
    onAddFilter() {
        this.addFilter.emit();
    }
    /**
     * @return {?}
     */
    onClear() {
        this.clearFilter.emit();
    }
    /**
     * @return {?}
     */
    onMoveTop() {
        this.moveTop.emit();
    }
    /**
     * @return {?}
     */
    onMovePrev() {
        this.movePrev.emit();
    }
    /**
     * @return {?}
     */
    onMoveNext() {
        this.moveNext.emit();
    }
    /**
     * @return {?}
     */
    onMoveBottom() {
        this.moveBottom.emit();
    }
    /**
     * @return {?}
     */
    canUse() {
        if (this.grid.currentEnumData) {
            return this.grid.data.length > 1;
        }
        return false;
    }
    /**
     * @return {?}
     */
    canMoveUp() {
        if (this.canUse()) {
            return this.grid.currentEnumData.index > 0;
        }
        return false;
    }
    /**
     * @return {?}
     */
    canMoveDown() {
        if (this.canUse()) {
            return this.grid.currentEnumData.index < this.grid.data.length - 1;
        }
        return false;
    }
}
EnumEditorFooterComponent.decorators = [
    { type: Component, args: [{
                selector: 'enum-editor-footer',
                template: `
    <div class="footer-container">
        <button type="button" (click)="onAddFilter()"
            class="k-button k-button-icontext k-flat "><span class="k-icon k-i-add"></span> {{ 'enumEditor.addWhere' | locale }}</button>
        <button type="button" (click)="onClear()" [disabled]="!grid.data.length"
            class="k-button k-button-icontext k-flat "><span class="k-icon k-i-delete"></span> {{ 'enumEditor.clear' | locale }}</button>
        <ng-container *ngIf="showSortBtns">
            <button type="button" [disabled]="!canMoveUp()" (click)="onMoveTop()"
            class="k-button k-button-icontext k-flat "><span class="k-icon k-i-arrow-end-up"></span>
            {{'enumEditor.moveTop' | locale }}</button>
        <button type="button" [disabled]="!canMoveUp()" (click)="onMovePrev()"
            class="k-button k-button-icontext k-flat "><span class="k-icon k-i-arrow-60-up"></span>
            {{'enumEditor.moveUp' | locale }}</button>
        <button type="button" [disabled]="!canMoveDown()" (click)="onMoveNext()"
            class="k-button k-button-icontext k-flat "><span class="k-icon k-i-arrow-60-down"></span>
            {{'enumEditor.moveDown' | locale }}</button>
        <button type="button" [disabled]="!canMoveDown()" (click)="onMoveBottom()"
            class="k-button k-button-icontext k-flat"><span class="k-icon k-i-arrow-end-down"></span>
            {{'enumEditor.moveBottom' | locale }}</button>
            </ng-container>
    </div>
    `
            }] }
];
/** @nocollapse */
EnumEditorFooterComponent.ctorParameters = () => [
    { type: EnumEditorComponent }
];
EnumEditorFooterComponent.propDecorators = {
    cls: [{ type: HostBinding, args: ['class',] }],
    showSortBtns: [{ type: Input }],
    addFilter: [{ type: Output }],
    clearFilter: [{ type: Output }],
    moveTop: [{ type: Output }],
    movePrev: [{ type: Output }],
    moveNext: [{ type: Output }],
    moveBottom: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    EnumEditorFooterComponent.prototype.cls;
    /** @type {?} */
    EnumEditorFooterComponent.prototype.showSortBtns;
    /** @type {?} */
    EnumEditorFooterComponent.prototype.addFilter;
    /** @type {?} */
    EnumEditorFooterComponent.prototype.clearFilter;
    /** @type {?} */
    EnumEditorFooterComponent.prototype.moveTop;
    /** @type {?} */
    EnumEditorFooterComponent.prototype.movePrev;
    /** @type {?} */
    EnumEditorFooterComponent.prototype.moveNext;
    /** @type {?} */
    EnumEditorFooterComponent.prototype.moveBottom;
    /** @type {?} */
    EnumEditorFooterComponent.prototype.grid;
}
