// @gsp-wf/wf-approval-logs/esm5/lib/wf-approval-logs.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, HostBinding } from '@angular/core';
var WfApprovalLogsComponent = /** @class */ (function () {
    function WfApprovalLogsComponent() {
        // 流程实例id
        this.ifForecast = false; // 是否预测日志
        // 是否预测日志
        this.cls = ''; // 弹框显示时，增加样式
        // 弹框显示时，增加样式
        this.ShowHeader = false; //是否显示日志头部
        //是否显示日志头部
        this.ShowViewProcess = false; //是否显示查看流程按钮
    }
    Object.defineProperty(WfApprovalLogsComponent.prototype, "DataId", {
        set: 
        //日志容器
        /**
         * @param {?} dataId
         * @return {?}
         */
        function (dataId) {
            if (dataId) {
                this.dataId = dataId;
                this.getTaskLogs();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WfApprovalLogsComponent.prototype, "ProcInstId", {
        set: /**
         * @param {?} procInstId
         * @return {?}
         */
        function (procInstId) {
            if (procInstId) {
                this.procInstId = procInstId;
                this.getTaskLogs();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WfApprovalLogsComponent.prototype, "ActivityDefinitionId", {
        set: /**
         * @param {?} actDefId
         * @return {?}
         */
        function (actDefId) {
            if (actDefId) {
                this.getTaskLogs(actDefId);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WfApprovalLogsComponent.prototype, "IfForecast", {
        set: /**
         * @param {?} ifForecast
         * @return {?}
         */
        function (ifForecast) {
            if (ifForecast) {
                this.ifForecast = ifForecast == 'true' ? true : false;
                this.getTaskLogs();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WfApprovalLogsComponent.prototype, "Type", {
        set: /**
         * @param {?} type
         * @return {?}
         */
        function (type) {
            if (type === 'process') { // type = 'process'查看流程日志；type = 'activity'，查看节点日志
                this.getTaskLogs();
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(WfApprovalLogsComponent.prototype, "processForecastPayload", {
        set: /**
         * @param {?} payload
         * @return {?}
         */
        function (payload) {
            if (payload) {
                this.ifForecast = true;
                this.getTaskLogs('', payload);
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    WfApprovalLogsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        // this.getTaskLogs();
    };
    /**
     * @return {?}
     */
    WfApprovalLogsComponent.prototype.fill = /**
     * @return {?}
     */
    function () {
        this.cls = 'd-flex flex-fill h-100 f-utils-absolute-all p-4';
    };
    /**
     * @private
     * @param {?=} actDefId
     * @param {?=} wfForecastPayload
     * @return {?}
     */
    WfApprovalLogsComponent.prototype.getTaskLogs = /**
     * @private
     * @param {?=} actDefId
     * @param {?=} wfForecastPayload
     * @return {?}
     */
    function (actDefId, wfForecastPayload) {
        var _this = this;
        if (this.dataId || this.procInstId || wfForecastPayload) {
            System.import('/platform/runtime/wf/webapp/mobiletaskcenter/taskComment.js?v=20211022').then(function () {
                if (_this.dataId) {
                    _this.logContainer = window.document.querySelector('.dataId' + _this.dataId);
                }
                else {
                    _this.logContainer = window.document.querySelector('.procInstId' + _this.procInstId);
                }
                /** @type {?} */
                var payload = {
                    bizInstId: _this.dataId,
                    processInstanceId: _this.procInstId,
                    activityDefinitionId: actDefId,
                    container: _this.logContainer,
                    terminal: 'web',
                    typeId: 'wf',
                    ifForecast: _this.ifForecast,
                    showHeader: _this.ShowHeader,
                    showViewProcess: _this.ShowViewProcess,
                    isIncludeHistory: false,
                    wfForecastPayload: wfForecastPayload
                };
                ((/** @type {?} */ (window))).gspTaskCenterBizLog.ApprovalLogUtil.getWebLogsHtml(payload).then(function (result) { });
            });
        }
    };
    WfApprovalLogsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'wf-approval-logs',
                    template: "<div\r\n  [class]=\"dataId ? 'dataId' + dataId : 'procInstId' + procInstId\"\r\n  style=\"width: 100%; height: 100%; overflow: auto\"\r\n></div>\r\n",
                    styles: [".approval-log-background{width:100%;background:#f9f9f9;padding:16px 14px;border-radius:4px}.log-activity-name{font-size:14px}.receive-time{font-size:12px;-webkit-columns:#565656;-moz-columns:#565656;columns:#565656;padding-left:40px;white-space:nowrap;color:#a1a1a1}.log-creator-name{font-size:14px;color:#4297fa;text-align:left}.log-action-name-do{font-size:12px;color:#fff;margin-left:40px;margin-right:16px;border-radius:10px;padding:2px 8px;background:#59a1ff;white-space:nowrap;float:right}.log-action-name-todo{font-size:12px;color:#fff;margin-left:40px;margin-right:16px;border-radius:10px;padding:2px 8px;background:#f59c24;white-space:nowrap;float:right}.log-action-name-undo{font-size:12px;color:#fff;margin-left:40px;margin-right:16px;border-radius:10px;padding:2px 8px;background:#5ac1c3;white-space:nowrap;float:right}.work-item-state-done{font-size:13px;border-radius:2px;padding:0 10px;background:#f6faff;margin-left:20px;color:#5f637a;border:1px solid #e6e6e6}.timeline-item{position:relative;padding:0 0 14px 10px;margin:0;font-size:14px;list-style:none}.timeline-item-tail{position:absolute;top:.75em;left:7px;height:100%;border-left:2px dashed #e8e8e8}.timeline-item-head-blue{position:absolute;width:16px;height:16px;background-color:#388fff;border:4px solid #bdd9ff;border-radius:100px;left:0;top:10px}.timeline-item-head-green{position:absolute;width:16px;height:16px;background-color:#5ac1c3;border:4px solid #7fffd4;border-radius:100px;left:0;top:10px}.timeline-item-head-yellow{position:absolute;width:16px;height:16px;background-color:#f59c24;border:4px solid #fbd7a7;border-radius:100px;left:0;top:10px}.timeline-item-content{position:relative;top:-6px;margin:0 0 0 18px}.user-avatar{height:30px;width:30px;border-radius:50%}li{list-style:none}.wf-docFile{flex-shrink:1;flex-grow:1;flex-basis:0;display:flex;flex-direction:row;align-items:flex-start;border-top:1px solid #f8f9fa;border-left:1px solid #f8f9fa}.wf-docFileIcon{width:20px;height:20px;line-height:20px;text-align:center;overflow:hidden;white-space:nowrap;padding:2px}.wf-docFileInfo{flex-shrink:1;flex-grow:1;flex-basis:0;padding:2px 6px 0 0;margin-left:0}.wf-docFileSize{font-size:11px;color:#888}.wf-doc-delete{margin:auto;font-size:18px;color:#f8f9fa}.attachment-delete:hover{color:#4297fa}.approval-log-message{font-size:14px;padding:10px 0 10px 40px}.approval-log-signature{padding-left:40px;display:flex;justify-content:flex-end}.approval-log-user-name{display:flex;align-items:center;justify-content:space-between;padding-top:14px}.log-attachment-array{padding-top:10px;padding-left:40px}.log-attachment-file-name{font-size:12px;color:#4297fa;float:left;cursor:pointer;margin-left:5px}"]
                }] }
    ];
    /** @nocollapse */
    WfApprovalLogsComponent.ctorParameters = function () { return []; };
    WfApprovalLogsComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        ShowHeader: [{ type: Input }],
        ShowViewProcess: [{ type: Input }],
        DataId: [{ type: Input }],
        ProcInstId: [{ type: Input }],
        ActivityDefinitionId: [{ type: Input }],
        IfForecast: [{ type: Input }],
        Type: [{ type: Input }],
        processForecastPayload: [{ type: Input }]
    };
    return WfApprovalLogsComponent;
}());
export { WfApprovalLogsComponent };
if (false) {
    /** @type {?} */
    WfApprovalLogsComponent.prototype.dataId;
    /** @type {?} */
    WfApprovalLogsComponent.prototype.procInstId;
    /** @type {?} */
    WfApprovalLogsComponent.prototype.ifForecast;
    /** @type {?} */
    WfApprovalLogsComponent.prototype.cls;
    /** @type {?} */
    WfApprovalLogsComponent.prototype.ShowHeader;
    /** @type {?} */
    WfApprovalLogsComponent.prototype.ShowViewProcess;
    /** @type {?} */
    WfApprovalLogsComponent.prototype.logContainer;
}
