// @gsp-wf/wf-process-editor/esm5/lib/app/process-designer.viewmodel.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable, Injector, ComponentFactoryResolver } from '@angular/core';
import { ProcessDesignerUIState } from './process-designer.uistate';
import { ProcessDesignerDataService } from '../domain/process-designer.dataservice';
import { forkJoin } from 'rxjs/observable/forkJoin';
import { ProcessClrTypeID } from '../model/process-clrtypeId';
import { BpmnProcessViewModel } from './bpmn-process.viewmodel';
import { WFFlowElementsViewModel } from './wf-flowElements.viewModel';
import { NotifyService } from '@farris/ui-notify';
import { MessagerService } from '@farris/ui-messager';
import { map, switchMap } from 'rxjs/operators';
import { FlowFormInfo } from '../domain/entities/flowform-info';
import { of, Subject, from } from 'rxjs';
import { BsModalService } from '@farris/ui-modal';
import { SaveAsEditComponent } from '../views/save-as-edit/save-as-edit.component';
import { ProcessDeUtil } from '../domain/process-de-util';
import { SubProcessViewModel } from './sub-process.viewmodel';
import { AutoActivityAdapter } from './auto-activity.adapter';
import { CatchEventAdapter } from './catch-event.adapter';
import { MessageActivityAdapter } from './message-activity.adapter';
import { AuditEvent, AuditProcessDefInfo } from '../domain/entities/audit-processDef-info';
import { HttpService } from '@ecp-caf/caf-common';
import { ProcessInfo } from '../model/process-info';
import { NotifyActivityAdapter } from './notify-activity.adapter';
import { AifCreationBillActivityAdapter } from './aif-creation-bill-activity.adapter';
var ProcessDesignerViewModel = /** @class */ (function () {
    function ProcessDesignerViewModel(uiState, service, util, notify, message, process, activity, subProcess, autoActAdapter, catchEventAdapter, messageActAdapter, notifyActAdapter, creationActAdapter, modalService, injector, resolver, http) {
        this.uiState = uiState;
        this.service = service;
        this.util = util;
        this.notify = notify;
        this.message = message;
        this.process = process;
        this.activity = activity;
        this.subProcess = subProcess;
        this.autoActAdapter = autoActAdapter;
        this.catchEventAdapter = catchEventAdapter;
        this.messageActAdapter = messageActAdapter;
        this.notifyActAdapter = notifyActAdapter;
        this.creationActAdapter = creationActAdapter;
        this.modalService = modalService;
        this.injector = injector;
        this.resolver = resolver;
        this.http = http;
    }
    // 初始加载流程属性
    // 初始加载流程属性
    /**
     * @param {?} obj
     * @return {?}
     */
    ProcessDesignerViewModel.prototype.initProcessProperty = 
    // 初始加载流程属性
    /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        /** @type {?} */
        var pptData = obj.propertyData;
        /** @type {?} */
        var ffId = pptData.EntryForm;
        if (ffId) {
            pptData.EntryForm = this.uiState.flowFormInfo.flowForm;
            this.uiState.propertyConfig = this.process.setProcessConfig(obj.propertyConfig);
            this.uiState.propertyData = pptData;
        }
        else {
            this.uiState.propertyConfig = this.process.setProcessConfig(obj.propertyConfig);
            this.uiState.propertyData = pptData;
        }
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    ProcessDesignerViewModel.prototype.setProperty = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        // 更新属性框数据与配置
        /** @type {?} */
        var propertyData = obj.propertyData;
        switch (propertyData.ClrTypeID) {
            case ProcessClrTypeID.process:
                this.process.setProcessProperty(obj);
                break;
            case ProcessClrTypeID.userActivity:
                this.activity.setUserActivityProperty(obj);
                break;
            case ProcessClrTypeID.subProcess:
                this.subProcess.setProperty(obj);
                break;
            case ProcessClrTypeID.autoActivity:
                this.autoActAdapter.setProperty(obj);
                break;
            case ProcessClrTypeID.catchEvent:
                this.catchEventAdapter.setProperty(obj);
                break;
            case ProcessClrTypeID.operationActivity:
                this.activity.setUserActivityProperty(obj);
                break;
            case ProcessClrTypeID.messageActivity:
                this.messageActAdapter.setProperty(obj);
                break;
            case ProcessClrTypeID.notifyActivity:
                this.notifyActAdapter.setProperty(obj);
                break;
            case ProcessClrTypeID.aifCreationBillActivity:
                this.creationActAdapter.setProperty(obj);
                break;
            default:
                this.activity.setFlowElementProperty(obj);
                break;
        }
    };
    /**
     * @return {?}
     */
    ProcessDesignerViewModel.prototype.updateSchemas = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.uiState.schemas = [];
        if (this.uiState.flowFormInfo.formalParameters && this.uiState.flowFormInfo.formalParameters.length > 0) {
            /** @type {?} */
            var schemaParams = this.uiState.flowFormInfo.formalParameters.filter((/**
             * @param {?} p
             * @return {?}
             */
            function (p) { return p.dataType === 'CDTSchema' && p.schemaMetadataId; }));
            forkJoin(schemaParams.map((/**
             * @param {?} para
             * @return {?}
             */
            function (para) {
                return _this.service.getBusinessEntitybyId(para.schemaMetadataId);
            }))).subscribe((/**
             * @param {?} data
             * @return {?}
             */
            function (data) {
                if (data && data.length > 0) {
                    data.forEach((/**
                     * @param {?} d
                     * @return {?}
                     */
                    function (d) {
                        _this.uiState.schemas.push(d);
                    }));
                }
            }), (/**
             * @param {?} err
             * @return {?}
             */
            function (err) {
                _this.uiState.schemas = [];
            }));
        }
        else {
            this.uiState.schemas = [];
        }
    };
    /**
     * @param {?} formId
     * @return {?}
     */
    ProcessDesignerViewModel.prototype.getFlowFormInfoById = /**
     * @param {?} formId
     * @return {?}
     */
    function (formId) {
        var _this = this;
        /** @type {?} */
        var $flowform = new Subject();
        if (formId) {
            forkJoin(this.service.GetFlowFormbyId(formId), this.service.getParametersbyFlowFormKey(formId).pipe(map((/**
             * @param {?} params
             * @return {?}
             */
            function (params) {
                _this.setFormParamCxt(params);
                return params;
            }))), this.service.getComponentsbyFlowformkey(formId), this.service.getExecuteFormbyFlowFormKey(formId), this.service.getBizActivities(formId), this.service.getCustomizeActions(formId)).subscribe((/**
             * @param {?} re
             * @return {?}
             */
            function (re) {
                /** @type {?} */
                var flowform = new FlowFormInfo();
                flowform.flowForm = re[0];
                flowform.formalParameters = re[1];
                flowform.bizComponents = re[2].filter((/**
                 * @param {?} cmp
                 * @return {?}
                 */
                function (cmp) { return cmp.ownerType !== 'Activity'; }));
                flowform.excuteForms = re[3];
                flowform.bizActivities = re[4];
                flowform.customizeActions = re[5];
                _this.uiState.flowFormInfo = flowform;
                // this.updateCOs();
                _this.updateSchemas();
                $flowform.next(flowform);
            }));
        }
        return $flowform.asObservable();
    };
    /**
     * @return {?}
     */
    ProcessDesignerViewModel.prototype.getFormFormatsByBizCategory = /**
     * @return {?}
     */
    function () {
        if (this.uiState.flowFormInfo.flowForm.businessObjectId) {
            /** @type {?} */
            var url = "/api/runtime/wf/v1.0/formFormat/query?param=";
            /** @type {?} */
            var param = "{\"bizCategory\":\"" + this.uiState.flowFormInfo.flowForm.businessObjectId + "\"}";
            param = encodeURIComponent(param);
            url += param;
            return this.http.get(url);
        }
        else {
            return from(new Array());
        }
    };
    /**
     * 保存流程
     * @param content 流程上下文
     */
    /**
     * 保存流程
     * @param {?} content 流程上下文
     * @param {?} processInfo
     * @param {?} state
     * @return {?}
     */
    ProcessDesignerViewModel.prototype.saveProcess = /**
     * 保存流程
     * @param {?} content 流程上下文
     * @param {?} processInfo
     * @param {?} state
     * @return {?}
     */
    function (content, processInfo, state) {
        return this.service.saveProcess(content, processInfo, state);
    };
    /**
     * 发布流程
     * @param content 流程上下文
     */
    /**
     * 发布流程
     * @param {?} content 流程上下文
     * @return {?}
     */
    ProcessDesignerViewModel.prototype.deployProcess = /**
     * 发布流程
     * @param {?} content 流程上下文
     * @return {?}
     */
    function (content) {
        return this.service.deployProcess(content);
    };
    /**
     * @param {?} auditId
     * @param {?} success
     * @param {?} key
     * @return {?}
     */
    ProcessDesignerViewModel.prototype.auditProcessDef = /**
     * @param {?} auditId
     * @param {?} success
     * @param {?} key
     * @return {?}
     */
    function (auditId, success, key) {
        /** @type {?} */
        var info = new AuditProcessDefInfo();
        info.auditId = auditId;
        info.success = success;
        info.processDefKey = key;
        this.service.auditProcessDef(info);
    };
    /**
     * 保存运行时流程
     * @param content 流程上下文
     */
    /**
     * 保存运行时流程
     * @param {?} content 流程上下文
     * @param {?} procDefId
     * @return {?}
     */
    ProcessDesignerViewModel.prototype.saveRTProcess = /**
     * 保存运行时流程
     * @param {?} content 流程上下文
     * @param {?} procDefId
     * @return {?}
     */
    function (content, procDefId) {
        return this.service.saveRTProcess(content, procDefId);
    };
    /**
     * @param {?} process
     * @param {?} processInfo
     * @return {?}
     */
    ProcessDesignerViewModel.prototype.saveAsProcess = /**
     * @param {?} process
     * @param {?} processInfo
     * @return {?}
     */
    function (process, processInfo) {
        var _this = this;
        /** @type {?} */
        var compFactory = this.resolver.resolveComponentFactory(SaveAsEditComponent);
        /** @type {?} */
        var compRef = compFactory.create(this.injector);
        compRef.instance.procKey = process.DefaultProcess.Key;
        compRef.instance.procName = process.DefaultProcess.Name + '_副本';
        compRef.instance.isAdministratorConfig = this.uiState.isAdministratorConfig;
        compRef.instance.organizations = this.uiState.organizations;
        compRef.instance.accessLevel = processInfo.accessLevel;
        compRef.instance.organization = processInfo.procOrg.orgId;
        compRef.instance.organizationName = processInfo.procOrg.orgName;
        /** @type {?} */
        var modalOptions = {
            title: '另存为',
            width: 450, height: 400,
            buttons: compRef.instance.btnRef,
            showButtons: true
        };
        /** @type {?} */
        var dlg = this.modalService.show(compRef, modalOptions);
        compRef.instance.submit.subscribe((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            process.DefaultProcess.Name = re.procName;
            /** @type {?} */
            var procInfo = new ProcessInfo();
            procInfo.accessLevel = re.accessLevel;
            procInfo.procOrg = {
                orgId: re.organization,
                orgName: re.organizationName
            };
            _this.saveProcess(JSON.stringify(process), procInfo, 1).subscribe((/**
             * @param {?} r
             * @return {?}
             */
            function (r) {
                _this.notify.success((/** @type {?} */ ({
                    title: '系统提示', msg: '另存为成功', timeout: 3000
                })));
            }), (/**
             * @param {?} err
             * @return {?}
             */
            function (err) {
                _this.notify.error((/** @type {?} */ ({
                    title: '系统提示', msg: '另存为失败，请检查流程是否正确！', timeout: 3000
                })));
            }));
            dlg.close();
        }));
        compRef.instance.cancel.subscribe((/**
         * @param {?} r
         * @return {?}
         */
        function (r) {
            dlg.close();
        }));
    };
    /**
     * 同步设计器时流程定义
     */
    /**
     * 同步设计器时流程定义
     * @param {?} content
     * @param {?} rtProcessInfo
     * @return {?}
     */
    ProcessDesignerViewModel.prototype.syncDtProcess = /**
     * 同步设计器时流程定义
     * @param {?} content
     * @param {?} rtProcessInfo
     * @return {?}
     */
    function (content, rtProcessInfo) {
        var _this = this;
        if (content) {
            /** @type {?} */
            var process_1 = JSON.parse(content);
            this.message.question('是否将当前版本流程定义同步到设计时流程定义包！', (/**
             * @return {?}
             */
            function () {
                _this.service.syncDtProcess(content, rtProcessInfo).subscribe((/**
                 * @param {?} data
                 * @return {?}
                 */
                function (data) {
                    _this.notify.success((/** @type {?} */ ({
                        title: '系统提示', msg: '同步成功', timeout: 3000
                    })));
                    _this.auditProcessDef(AuditEvent.WF_ProcessDef_SyncDT, true, process_1.DefaultProcess.Key);
                }), (/**
                 * @param {?} error
                 * @return {?}
                 */
                function (error) {
                    _this.notify.error((/** @type {?} */ ({
                        title: '系统提示', msg: '同步失败，请联系管理员查看原因！', timeout: 3000
                    })));
                    _this.auditProcessDef(AuditEvent.WF_ProcessDef_SyncDT, false, process_1.DefaultProcess.Key);
                }));
            }));
        }
    };
    /**
     * @param {?} data
     * @param {?} processInfo
     * @return {?}
     */
    ProcessDesignerViewModel.prototype.beforeCompleteStrategyChange = /**
     * @param {?} data
     * @param {?} processInfo
     * @return {?}
     */
    function (data, processInfo) {
        var _this = this;
        /** @type {?} */
        var url = "/api/runtime/wf/v1.0/processInstances/runtimeActivityInstances?procDefId=" + processInfo.rtProcDefId + "&activityDefinitionId=" + data.Id;
        return this.http.get(url).pipe(switchMap((/**
         * @param {?} re
         * @return {?}
         */
        function (re) {
            if (re && re.length > 0) {
                _this.message.info('当前节点存在运行中的活动实例，不允许调整完成策略，请了解！');
                return of(false);
            }
            else {
                return of(true);
            }
        })));
    };
    /**
     * @param {?} procCategoryId
     * @return {?}
     */
    ProcessDesignerViewModel.prototype.getProcessCategory = /**
     * @param {?} procCategoryId
     * @return {?}
     */
    function (procCategoryId) {
        return this.service.GetFlowFormbyId(procCategoryId);
    };
    /**
     * @private
     * @param {?} params
     * @return {?}
     */
    ProcessDesignerViewModel.prototype.setFormParamCxt = /**
     * @private
     * @param {?} params
     * @return {?}
     */
    function (params) {
        var _this = this;
        this.uiState.formalParameterContext = [];
        if (params && params.length > 0) {
            params.forEach((/**
             * @param {?} param
             * @return {?}
             */
            function (param) {
                if (param.dataType === 'CDTSchema') {
                    /** @type {?} */
                    var formalParameterContext = [{
                            key: "schemaId_" + param.code,
                            name: "schemaId_" + param.name,
                            description: "schemaId_" + param.name
                        }, {
                            key: "metadataId_" + param.code,
                            name: "metadataId_" + param.name,
                            description: "metadataId_" + param.name
                        }];
                    _this.uiState.formalParameterContext = _this.uiState.formalParameterContext.concat(formalParameterContext);
                }
            }));
        }
    };
    ProcessDesignerViewModel.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ProcessDesignerViewModel.ctorParameters = function () { return [
        { type: ProcessDesignerUIState },
        { type: ProcessDesignerDataService },
        { type: ProcessDeUtil },
        { type: NotifyService },
        { type: MessagerService },
        { type: BpmnProcessViewModel },
        { type: WFFlowElementsViewModel },
        { type: SubProcessViewModel },
        { type: AutoActivityAdapter },
        { type: CatchEventAdapter },
        { type: MessageActivityAdapter },
        { type: NotifyActivityAdapter },
        { type: AifCreationBillActivityAdapter },
        { type: BsModalService },
        { type: Injector },
        { type: ComponentFactoryResolver },
        { type: HttpService }
    ]; };
    return ProcessDesignerViewModel;
}());
export { ProcessDesignerViewModel };
if (false) {
    /** @type {?} */
    ProcessDesignerViewModel.prototype.uiState;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.service;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.util;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.notify;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.message;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.process;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.activity;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.subProcess;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.autoActAdapter;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.catchEventAdapter;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.messageActAdapter;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.notifyActAdapter;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.creationActAdapter;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.modalService;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.injector;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.resolver;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerViewModel.prototype.http;
}
