// @gsp-wf/wf-process-editor/esm5/lib/views/message-notice/message-notice-edit/message-notice-edit.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input, ViewChild, TemplateRef, Output, EventEmitter } from '@angular/core';
import { MessageNotice, activitySendTime, processSendTime } from '../message-notice';
import { MessagerService } from '@farris/ui-messager';
import { ProcessDesignerUIState } from '../../../app/process-designer.uistate';
import { ExpressionService, InjectedType } from "@gsp-svc/expression";
var MessageNoticeEditComponent = /** @class */ (function () {
    function MessageNoticeEditComponent(message, uistate, expr) {
        this.message = message;
        this.uistate = uistate;
        this.expr = expr;
        this.groupIcon2 = '<i class="f-icon f-icon-lookup"></i>';
        this.exprAble = false;
        this.closeEmitter = new EventEmitter();
        this.submitEmitter = new EventEmitter();
        this._location = 'activity';
        this.messageNotice = new MessageNotice();
        this.typeGroup = this.uistate.messageChannels.map((/**
         * @param {?} m
         * @return {?}
         */
        function (m) { return ({ value: m.code, name: m.name }); }));
        this.receiverVariables = [
            { value: 'workItemParticipant', name: '任务办理人' }
        ];
        this.sendTimes = [];
        this.bindingData = {};
    }
    Object.defineProperty(MessageNoticeEditComponent.prototype, "location", {
        get: /**
         * @return {?}
         */
        function () {
            return this._location;
        },
        set: /**
         * @param {?} l
         * @return {?}
         */
        function (l) {
            this._location = l;
            if (this.location !== 'activity') {
                this.receiverVariables = [{ value: 'procInstCreator', name: '流程发起人' }];
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MessageNoticeEditComponent.prototype, "value", {
        get: /**
         * @return {?}
         */
        function () {
            return this.messageNotice;
        },
        set: /**
         * @param {?} m
         * @return {?}
         */
        function (m) {
            var e_1, _a, e_2, _b;
            this.messageNotice = Object.assign({}, m);
            if (this.messageNotice.receiver && this.messageNotice.receiver.length > 0) {
                /** @type {?} */
                var array = this.messageNotice.receiver.split(',');
                if (array.indexOf('workItemParticipant') > -1) {
                    this.receiverVari = 'workItemParticipant';
                }
                else if (array.indexOf('procInstCreator') > -1) {
                    this.receiverVari = 'procInstCreator';
                }
                if (array.length > 0) {
                    array = array.filter((/**
                     * @param {?} a
                     * @return {?}
                     */
                    function (a) { return (a !== 'workItemParticipant' && a !== 'procInstCreator'); }));
                    this.bindingData.id = '';
                    try {
                        for (var array_1 = tslib_1.__values(array), array_1_1 = array_1.next(); !array_1_1.done; array_1_1 = array_1.next()) {
                            var r = array_1_1.value;
                            this.bindingData.id += r + ',';
                        }
                    }
                    catch (e_1_1) { e_1 = { error: e_1_1 }; }
                    finally {
                        try {
                            if (array_1_1 && !array_1_1.done && (_a = array_1.return)) _a.call(array_1);
                        }
                        finally { if (e_1) throw e_1.error; }
                    }
                    if (this.bindingData.id) {
                        this.bindingData.id = this.bindingData.id.substring(0, this.bindingData.id.length - 1);
                    }
                }
            }
            if (this.messageNotice.receiverText && this.messageNotice.receiverText.length > 0) {
                /** @type {?} */
                var array = this.messageNotice.receiverText.split(',');
                if (array.length > 0) {
                    array = array.filter((/**
                     * @param {?} a
                     * @return {?}
                     */
                    function (a) { return (a !== '任务办理人' && a !== '流程发起人'); }));
                    this.bindingData.name = '';
                    try {
                        for (var array_2 = tslib_1.__values(array), array_2_1 = array_2.next(); !array_2_1.done; array_2_1 = array_2.next()) {
                            var r = array_2_1.value;
                            this.bindingData.name += r + ',';
                        }
                    }
                    catch (e_2_1) { e_2 = { error: e_2_1 }; }
                    finally {
                        try {
                            if (array_2_1 && !array_2_1.done && (_b = array_2.return)) _b.call(array_2);
                        }
                        finally { if (e_2) throw e_2.error; }
                    }
                    if (this.bindingData.name) {
                        this.bindingData.name = this.bindingData.name.substring(0, this.bindingData.name.length - 1);
                    }
                }
            }
            this.exprAble = this.messageNotice.subject.charAt(0) === '{';
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    MessageNoticeEditComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        if (this.location === 'activity') {
            this.sendTimes = activitySendTime;
        }
        else {
            this.sendTimes = processSendTime;
        }
    };
    /**
     * @return {?}
     */
    MessageNoticeEditComponent.prototype.clickOK = /**
     * @return {?}
     */
    function () {
        this.setReceiver();
        if (this.validator()) {
            this.submitEmitter.emit(this.messageNotice);
        }
    };
    /**
     * @return {?}
     */
    MessageNoticeEditComponent.prototype.clickCancel = /**
     * @return {?}
     */
    function () {
        this.closeEmitter.emit();
    };
    /**
     * @param {?} data
     * @return {?}
     */
    MessageNoticeEditComponent.prototype.sendTimeChange = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (data.data.value === activitySendTime[0].value) {
            this.receiverVari = 'workItemParticipant';
            this.clear();
        }
    };
    /**
     * @param {?} data
     * @return {?}
     */
    MessageNoticeEditComponent.prototype.usersSelect = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        console.log(data);
    };
    /**
     * @param {?} data
     * @return {?}
     */
    MessageNoticeEditComponent.prototype.changeValue = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        console.log(data);
    };
    /**
     * @return {?}
     */
    MessageNoticeEditComponent.prototype.clear = /**
     * @return {?}
     */
    function () {
        this.bindingData = {};
    };
    /**
     * @private
     * @return {?}
     */
    MessageNoticeEditComponent.prototype.setReceiver = /**
     * @private
     * @return {?}
     */
    function () {
        var e_3, _a;
        /** @type {?} */
        var receiver = '';
        /** @type {?} */
        var receiverText = '';
        if (this.receiverVari) {
            receiver += this.receiverVari + ',';
            /** @type {?} */
            var array = this.receiverVari.split(',');
            if (array.length > 0) {
                var _loop_1 = function (a) {
                    /** @type {?} */
                    var t = this_1.receiverVariables.find((/**
                     * @param {?} v
                     * @return {?}
                     */
                    function (v) { return v.value === a; })).name;
                    receiverText += t + ',';
                };
                var this_1 = this;
                try {
                    for (var array_3 = tslib_1.__values(array), array_3_1 = array_3.next(); !array_3_1.done; array_3_1 = array_3.next()) {
                        var a = array_3_1.value;
                        _loop_1(a);
                    }
                }
                catch (e_3_1) { e_3 = { error: e_3_1 }; }
                finally {
                    try {
                        if (array_3_1 && !array_3_1.done && (_a = array_3.return)) _a.call(array_3);
                    }
                    finally { if (e_3) throw e_3.error; }
                }
            }
        }
        if (this.bindingData.id) {
            receiver += this.bindingData.id + ',';
            receiverText += this.bindingData.name + ',';
        }
        if (receiver) {
            receiver = receiver.substring(0, receiver.length - 1);
        }
        if (receiverText) {
            receiverText = receiverText.substring(0, receiverText.length - 1);
        }
        this.messageNotice.receiver = receiver;
        this.messageNotice.receiverText = receiverText;
    };
    /**
     * @private
     * @return {?}
     */
    MessageNoticeEditComponent.prototype.validator = /**
     * @private
     * @return {?}
     */
    function () {
        if (!this.messageNotice.sendType) {
            this.message.info('发送方式不允许为空！');
            return false;
        }
        if (this.location !== 'messageActivity' && !this.messageNotice.sendTime) {
            this.message.info('发送时机不允许为空！');
            return false;
        }
        if (!this.messageNotice.subject) {
            this.message.info('消息主题不允许为空！');
            return false;
        }
        if (!this.messageNotice.content) {
            this.message.info('消息内容不允许为空！');
            return false;
        }
        if (!this.messageNotice.receiver) {
            this.message.info('接收人不允许为空！');
            return false;
        }
        return true;
    };
    /**
     * @param {?} data
     * @return {?}
     */
    MessageNoticeEditComponent.prototype.titleExpr = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.showTitleHelp(data);
    };
    /**
     * @param {?} data
     * @return {?}
     */
    MessageNoticeEditComponent.prototype.contentExpr = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.showContentHelp(data);
    };
    /**
     * @param {?} data
     * @return {?}
     */
    MessageNoticeEditComponent.prototype.showContentHelp = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var _this = this;
        this.showExpr();
        if (this.messageNotice.content && this.messageNotice.content.length > 0) {
            this.expr.addExpressionText(this.messageNotice.content);
        }
        else {
            this.expr.addExpressionText('');
        }
        this.expr.buildExpression().then((/**
         * @param {?} expressionInfo
         * @return {?}
         */
        function (expressionInfo) {
            _this.messageNotice.content = expressionInfo;
        }));
    };
    /**
     * @param {?} data
     * @return {?}
     */
    MessageNoticeEditComponent.prototype.showTitleHelp = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        var _this = this;
        if (this.messageNotice.subject && this.messageNotice.subject.charAt(0) !== '{') {
            this.messageNotice.subject = '';
        }
        this.showExpr();
        if (this.messageNotice.subject && this.messageNotice.subject.length > 0) {
            this.expr.addExpressionText(this.messageNotice.subject);
        }
        else {
            this.expr.addExpressionText('');
        }
        this.expr.buildExpression().then((/**
         * @param {?} expressionInfo
         * @return {?}
         */
        function (expressionInfo) {
            _this.messageNotice.subject = expressionInfo;
        }));
    };
    /**
     * @return {?}
     */
    MessageNoticeEditComponent.prototype.showExpr = /**
     * @return {?}
     */
    function () {
        var e_4, _a, e_5, _b, e_6, _c, e_7, _d;
        if (this.uistate.schemas.length > 0) {
            try {
                for (var _e = tslib_1.__values(this.uistate.schemas), _f = _e.next(); !_f.done; _f = _e.next()) {
                    var schema = _f.value;
                    // this.expr.addCommonObject(co);
                    this.expr.addSchema('Schema', schema);
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (_f && !_f.done && (_a = _e.return)) _a.call(_e);
                }
                finally { if (e_4) throw e_4.error; }
            }
            if (this.uistate.schemas[0].entityTypes && this.uistate.schemas[0].entityTypes.length > 0) {
                /** @type {?} */
                var entityType = this.uistate.schemas[0].entityTypes[0];
                this.expr.addInjectedEntity(entityType.name, 'Schema', entityType.name, InjectedType.EntityType);
            }
        }
        /** @type {?} */
        var context = new Array();
        if (this.uistate.procInstContext.length > 0) {
            try {
                for (var _g = tslib_1.__values(this.uistate.procInstContext), _h = _g.next(); !_h.done; _h = _g.next()) {
                    var variable = _h.value;
                    context.push(variable);
                    // this.expr.addContext(variable);
                }
            }
            catch (e_5_1) { e_5 = { error: e_5_1 }; }
            finally {
                try {
                    if (_h && !_h.done && (_b = _g.return)) _b.call(_g);
                }
                finally { if (e_5) throw e_5.error; }
            }
        }
        if (this.uistate.processVariableContext.length > 0) {
            var _loop_2 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _j = tslib_1.__values(this.uistate.processVariableContext), _k = _j.next(); !_k.done; _k = _j.next()) {
                    var variable = _k.value;
                    _loop_2(variable);
                }
            }
            catch (e_6_1) { e_6 = { error: e_6_1 }; }
            finally {
                try {
                    if (_k && !_k.done && (_c = _j.return)) _c.call(_j);
                }
                finally { if (e_6) throw e_6.error; }
            }
        }
        if (context.length > 0) {
            try {
                for (var context_1 = tslib_1.__values(context), context_1_1 = context_1.next(); !context_1_1.done; context_1_1 = context_1.next()) {
                    var v = context_1_1.value;
                    this.expr.addContext(v);
                }
            }
            catch (e_7_1) { e_7 = { error: e_7_1 }; }
            finally {
                try {
                    if (context_1_1 && !context_1_1.done && (_d = context_1.return)) _d.call(context_1);
                }
                finally { if (e_7) throw e_7.error; }
            }
        }
    };
    /**
     * @return {?}
     */
    MessageNoticeEditComponent.prototype.tranExpr = /**
     * @return {?}
     */
    function () {
        this.messageNotice.subject = '';
    };
    MessageNoticeEditComponent.decorators = [
        { type: Component, args: [{
                    selector: 'lib-message-notice-edit',
                    template: "<div class=\"f-page\">\r\n  <div class=\"f-utils-fill-flex-column farris-split-section m-2 h-100\">\r\n    <section class=\"f-utils-fill-flex-column h-100\">\r\n      <div class=\"f-section-header pt-2 mx-2\">\r\n        <div class=\"f-title\">\r\n          <h4 class=\"f-title-text\">\u6D88\u606F\u901A\u77E5\u8BBE\u7F6E</h4>\r\n        </div>\r\n      </div>\r\n      <div class=\"f-utils-fill-flex-column h-100\">\r\n        <perfect-scrollbar [config]=\"{}\">\r\n          <form class=\"h-100\">\r\n            <div class=\"farris-form form-inline farris-form-inline\">\r\n              <div class=\"farris-group-wrap\">\r\n                <div class=\"form-group farris-form-group\">\r\n                  <label class=\"col-form-label\">\r\n                    <span class=\"farris-label-info text-danger\">*</span>\r\n                    <span class=\"farris-label-text\">\u53D1\u9001\u65B9\u5F0F</span>\r\n                  </label>\r\n                  <div class=\"farris-input-wrap ml-1\">\r\n                    <farris-checkboxgroup #checkboxGroup [(ngModel)]=\"messageNotice.sendType\" name=\"sendType\" [horizontal]=\"true\"\r\n                      [data]=\"typeGroup\">\r\n                    </farris-checkboxgroup>\r\n                    <div class=\"farris-feedback valid-feedback\"></div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n              <div *ngIf=\"location!=='messageActivity'\" class=\"farris-group-wrap my-3\">\r\n                <div class=\"form-group farris-form-group\">\r\n                  <label class=\"col-form-label\">\r\n                    <span class=\"farris-label-info text-danger\">*</span>\r\n                    <span class=\"farris-label-text\">\u53D1\u9001\u65F6\u673A</span>\r\n                  </label>\r\n                  <div class=\"farris-input-wrap ml-1\">\r\n                    <farris-combo-list [(ngModel)]=\"messageNotice.sendTime\" [idField]=\"'value'\" [textField]=\"'text'\"\r\n                      [enableClear]=\"false\" [data]=\"sendTimes\" name=\"sendTime\" [editable]=\"false\" (selectChange)=\"sendTimeChange($event)\">\r\n                    </farris-combo-list>\r\n                    <div class=\"farris-feedback valid-feedback\"></div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n              <div class=\"farris-group-wrap my-3\">\r\n                <div class=\"form-group farris-form-group\">\r\n                  <label class=\"col-form-label\">\r\n                    <span class=\"farris-label-info text-danger\">*</span>\r\n                    <span class=\"farris-label-text\">\u6D88\u606F\u4E3B\u9898</span>\r\n                  </label>\r\n                  <div class=\"farris-input-wrap ml-1\" *ngIf=\"!exprAble; else disabled\">\r\n                    <input class=\"form-control\" type=\"text\" value=\"subject\" [(ngModel)]=\"messageNotice.subject\" name=\"subject\" />\r\n                    <div class=\"farris-feedback valid-feedback\"></div>\r\n                  </div>\r\n                  <ng-template #disabled>\r\n                    <input-group [editable]=\"false\" [customCls]=\"'custom-cls'\" [groupText]=\"groupIcon2\"\r\n                                 [(value)]=\"messageNotice.subject\" (clickHandle)=\"titleExpr($event)\">\r\n                    </input-group>\r\n                    <div class=\"farris-feedback valid-feedback\"></div>\r\n                  </ng-template>\r\n                  <label class=\"col-form-label pl-1 mr-3\">\r\n                    <span class=\"farris-label-text\">\u542F\u7528\u8868\u8FBE\u5F0F</span>\r\n                  </label>\r\n                  <div class=\"farris-input-wrap mr-5\">\r\n                    <farris-switch [(ngModel)]=\"exprAble\" name=\"enable\" (click)=\"tranExpr()\"></farris-switch>\r\n                    <div class=\"farris-feedback valid-feedback\"></div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n              <div class=\"farris-group-wrap my-3\">\r\n                <div class=\"form-group farris-form-group\">\r\n                  <label class=\"col-form-label\">\r\n                    <span class=\"farris-label-info text-danger\">*</span>\r\n                    <span class=\"farris-label-text\">\u6D88\u606F\u5185\u5BB9</span>\r\n                  </label>\r\n                  <div class=\"farris-input-wrap ml-1\">\r\n                    <input-group [editable]=\"false\" [customCls]=\"'custom-cls'\" [groupText]=\"groupIcon2\"\r\n                                 [(value)]=\"messageNotice.content\" (clickHandle)=\"contentExpr($event)\">\r\n                    </input-group>\r\n                    <div class=\"farris-feedback valid-feedback\"></div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n              <div class=\"farris-group-wrap\">\r\n                <div class=\"form-group farris-form-group\">\r\n                  <label class=\"col-form-label\">\r\n                    <span class=\"farris-label-info text-danger\">*</span>\r\n                    <span class=\"farris-label-text\">\u63A5\u6536\u4EBA</span>\r\n                  </label>\r\n                  <div class=\"farris-input-wrap ml-1\">\r\n                    <div class=\"pb-2\">\r\n                      <farris-checkboxgroup #receiverVariable [(ngModel)]=\"receiverVari\" name=\"receiverVariable\"\r\n                        [horizontal]=\"true\" [data]=\"receiverVariables\" (changeValue)=\"changeValue($event)\">\r\n                      </farris-checkboxgroup>\r\n                    </div>\r\n                    <div *ngIf=\"messageNotice.sendTime!=='WORKITEM_CREATED'\">\r\n                      <sys-farris-grid-lookup #userHelp [singleSelect]=\"false\" [displayTxt]=\"bindingData.name\" [pagination]=\"true\"\r\n                        [pageSize]=\"20\" [title]=\"'\u7CFB\u7EDF\u7528\u6237'\" [sourceType]=\"'user'\" (afterConfirm)=\"usersSelect($event)\" [typeMode]=\"1\"\r\n                        [bindingData]=\"bindingData\" [filter]=\"{notGetCurrentUser:false, notGetStopUser:true}\" (clear)=\"clear()\" [required]=\"false\">\r\n                      </sys-farris-grid-lookup>\r\n                    </div>\r\n                    <div class=\"farris-feedback valid-feedback\"></div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </form>\r\n        </perfect-scrollbar>\r\n      </div>\r\n    </section>\r\n  </div>\r\n</div>\r\n\r\n<ng-template #footer>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm mx-2\" (click)=\"clickOK()\">\u786E\u5B9A</button>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm mx-2\" (click)=\"clickCancel()\">\u53D6\u6D88</button>\r\n</ng-template>\r\n",
                    styles: [""]
                }] }
    ];
    /** @nocollapse */
    MessageNoticeEditComponent.ctorParameters = function () { return [
        { type: MessagerService },
        { type: ProcessDesignerUIState },
        { type: ExpressionService }
    ]; };
    MessageNoticeEditComponent.propDecorators = {
        footer: [{ type: ViewChild, args: ['footer',] }],
        closeEmitter: [{ type: Output }],
        submitEmitter: [{ type: Output }],
        location: [{ type: Input }],
        value: [{ type: Input }]
    };
    return MessageNoticeEditComponent;
}());
export { MessageNoticeEditComponent };
if (false) {
    /** @type {?} */
    MessageNoticeEditComponent.prototype.groupIcon2;
    /** @type {?} */
    MessageNoticeEditComponent.prototype.exprAble;
    /** @type {?} */
    MessageNoticeEditComponent.prototype.footer;
    /** @type {?} */
    MessageNoticeEditComponent.prototype.closeEmitter;
    /** @type {?} */
    MessageNoticeEditComponent.prototype.submitEmitter;
    /**
     * @type {?}
     * @private
     */
    MessageNoticeEditComponent.prototype._location;
    /** @type {?} */
    MessageNoticeEditComponent.prototype.messageNotice;
    /** @type {?} */
    MessageNoticeEditComponent.prototype.typeGroup;
    /** @type {?} */
    MessageNoticeEditComponent.prototype.receiverVariables;
    /** @type {?} */
    MessageNoticeEditComponent.prototype.sendTimes;
    /** @type {?} */
    MessageNoticeEditComponent.prototype.receiverVari;
    /** @type {?} */
    MessageNoticeEditComponent.prototype.bindingData;
    /**
     * @type {?}
     * @private
     */
    MessageNoticeEditComponent.prototype.message;
    /**
     * @type {?}
     * @private
     */
    MessageNoticeEditComponent.prototype.uistate;
    /**
     * @type {?}
     * @private
     */
    MessageNoticeEditComponent.prototype.expr;
}
