// @gsp-wf/wf-process-editor/esm5/lib/views/backward-rule/backward-rule.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, HostBinding, Output, EventEmitter, Input, ViewChild, TemplateRef } from '@angular/core';
import { ProcessDesignerUIState } from '../../app/process-designer.uistate';
import { MessagerService } from '@farris/ui-messager';
import { of } from 'rxjs';
import { LookupGridComponent } from '@farris/ui-lookup';
var BackwardRuleComponent = /** @class */ (function () {
    function BackwardRuleComponent(uiState, message) {
        var _this = this;
        this.uiState = uiState;
        this.message = message;
        this.cls = 'farris-main-area flex-column';
        this.editorParams = {};
        this.nodeType = 'UserActivity';
        this.closeModal = new EventEmitter();
        this.submitModal = new EventEmitter();
        this.modalConfig = {
            title: '驳回规则',
            width: 430,
            height: 370,
            showButtons: true
        };
        this.listItems = [
            { Id: 'InitialActityId', Name: '开始节点' },
            { Id: 'PrecursorActivityId', Name: '前置节点' },
            { Id: 'Empty', Name: '空' }
        ];
        this.items = [];
        this.columns = [
            { title: '名称', field: 'Name', width: 100, searchField: 'Name' },
            { title: '标识', field: 'Id', width: 100, searchField: 'Id' },
        ];
        this.selectAbleChecked = false;
        this.waitReturnChecked = false;
        this.dynamicBackWardRuleAbleChecked = false;
        this.targetRef = this.listItems[0].Id;
        this.pageInfo = {
            pageIndex: 1,
            pageList: [10, 20, 30, 50],
            pageSize: 20
        };
        this.dictPicked = (/**
         * @param {?} arg
         * @return {?}
         */
        function (arg) {
            _this.targetRef = arg.Id;
            _this.displayText = arg.Name;
            return of({ closeDialog: true });
        });
        this.listItems = this.listItems.concat(this.uiState.userActList);
        this.items = this.listItems;
        this.total = this.listItems.length;
    }
    Object.defineProperty(BackwardRuleComponent.prototype, "value", {
        set: /**
         * @param {?} data
         * @return {?}
         */
        function (data) {
            var _this = this;
            if (data) {
                this.selectAbleChecked = data.SelectTargetAble;
                this.waitReturnChecked = data.WaitReturn;
                this.dynamicBackWardRuleAbleChecked = data.DynamicBackWardRuleAble;
                if (!data.TargetRef) {
                    this.targetRef = 'Empty';
                }
                else {
                    this.targetRef = data.TargetRef;
                }
                /** @type {?} */
                var act = this.listItems.find((/**
                 * @param {?} item
                 * @return {?}
                 */
                function (item) { return item.Id === _this.targetRef; }));
                if (act) {
                    this.displayText = act.Name;
                }
                else {
                    this.displayText = '';
                    setTimeout((/**
                     * @return {?}
                     */
                    function () {
                        _this.message.warning('驳回活动节点已不存在，请重新编辑');
                    }), 500);
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    BackwardRuleComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.nodeType = this.editorParams.nodeType;
        if (this.nodeType !== 'UserActivity') {
            this.listItems = this.listItems.filter((/**
             * @param {?} i
             * @return {?}
             */
            function (i) { return i.Id !== 'Empty'; }));
            this.items = this.listItems;
            this.total = this.listItems.length;
        }
    };
    /**
     * @return {?}
     */
    BackwardRuleComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    BackwardRuleComponent.prototype.selectAbleClick = /**
     * @return {?}
     */
    function () {
        var _this = this;
        if (this.uiState.propertyData.ClrTypeID === 'wf.UserActivity'
            && this.uiState.propertyData.CompleteStrategy !== this.uiState.completeIterator[0].key
            && this.selectAbleChecked === true) {
            this.message.info('会签策略下，不允许“动态选择驳回级次”!', (/**
             * @return {?}
             */
            function () {
                _this.selectAbleChecked = false;
            }));
        }
    };
    /**
     * @return {?}
     */
    BackwardRuleComponent.prototype.waitReturnClick = /**
     * @return {?}
     */
    function () {
        this.waitReturnChecked = !this.waitReturnChecked;
    };
    /**
     * @return {?}
     */
    BackwardRuleComponent.prototype.dynamicBackWardRuleAbleClick = /**
     * @return {?}
     */
    function () {
        this.dynamicBackWardRuleAbleChecked = !this.dynamicBackWardRuleAbleChecked;
    };
    /**
     * @return {?}
     */
    BackwardRuleComponent.prototype.clickOK = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var backWardRule = {
            TargetRef: this.getTargetRef(),
            SelectTargetAble: this.selectAbleChecked,
            WaitReturn: this.waitReturnChecked,
            DynamicBackWardRuleAble: this.dynamicBackWardRuleAbleChecked,
        };
        /** @type {?} */
        var object = { value: backWardRule, parameters: null };
        this.submitModal.emit(object);
    };
    /**
     * @return {?}
     */
    BackwardRuleComponent.prototype.clickCancel = /**
     * @return {?}
     */
    function () {
        this.closeModal.emit();
    };
    /**
     * @param {?} arg
     * @return {?}
     */
    BackwardRuleComponent.prototype.pagerChanged = /**
     * @param {?} arg
     * @return {?}
     */
    function (arg) {
        this.pageInfo.pageIndex = arg.pageIndex;
        this.pageInfo.pageSize = arg.pageSize;
        /** @type {?} */
        var search = {
            field: arg.searchValue.searchField,
            value: arg.searchValue.searchValue
        };
        this.load(search);
    };
    /**
     * @param {?} arg
     * @return {?}
     */
    BackwardRuleComponent.prototype.pageSizeChange = /**
     * @param {?} arg
     * @return {?}
     */
    function (arg) {
        console.log(arg);
        this.pageInfo.pageIndex = arg.pageIndex;
        this.pageInfo.pageSize = arg.pageSize;
        /** @type {?} */
        var search = {
            field: arg.searchValue.searchField,
            value: arg.searchValue.searchValue
        };
        this.load(search);
    };
    /**
     * @param {?} arg
     * @return {?}
     */
    BackwardRuleComponent.prototype.search = /**
     * @param {?} arg
     * @return {?}
     */
    function (arg) {
        this.pageInfo.pageIndex = arg.pageInfo.pageIndex;
        this.pageInfo.pageSize = arg.pageInfo.pageSize;
        /** @type {?} */
        var search = arg.search;
        this.load(search);
    };
    /**
     * @param {?=} search
     * @return {?}
     */
    BackwardRuleComponent.prototype.load = /**
     * @param {?=} search
     * @return {?}
     */
    function (search) {
        /** @type {?} */
        var start = (this.pageInfo.pageIndex - 1) * this.pageInfo.pageSize;
        /** @type {?} */
        var end = start + this.pageInfo.pageSize;
        /** @type {?} */
        var items = [];
        if (search) {
            if (search.field === '*') {
                items = this.listItems.filter((/**
                 * @param {?} i
                 * @return {?}
                 */
                function (i) { return i.Name.indexOf(search.value) > -1 || i.Id.indexOf(search.value) > -1; }));
            }
            else {
                items = this.listItems.filter((/**
                 * @param {?} f
                 * @return {?}
                 */
                function (f) { return f[search.field].indexOf(search.value) > -1; }));
            }
        }
        else {
            items = this.listItems;
        }
        this.total = items.length;
        this.items = items; // .slice(start, end);
        this.hlp.componentRef.instance.loadData({
            // total: this.total,
            // pageSize: this.pageInfo.pageSize,
            // pageIndex: this.pageInfo.pageIndex,
            data: this.items
        });
    };
    /**
     * @private
     * @return {?}
     */
    BackwardRuleComponent.prototype.getTargetRef = /**
     * @private
     * @return {?}
     */
    function () {
        if (this.targetRef === 'Empty') {
            return '';
        }
        else {
            return this.targetRef;
        }
    };
    BackwardRuleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'procDesigner-backward-rule',
                    template: "<div class=\"f-page\">\r\n  <div class=\"f-utils-fill-flex-column farris-split-section m-2 h-100\">\r\n    <section class=\"f-utils-fill-flex-column h-100\">\r\n      <div class=\"f-section-header pt-2 mx-2\">\r\n        <div class=\"f-title\">\r\n          <h4 class=\"f-title-text\">\u9A73\u56DE\u89C4\u5219\u8BBE\u7F6E</h4>\r\n        </div>\r\n      </div>\r\n      <div class=\"f-utils-fill-flex-column h-100\">\r\n        <div class=\"farris-form form-inline farris-form-inline\">\r\n          <div class=\"col-12 farris-group-auto my-4 pr-5\">\r\n            <div class=\"farris-group-wrap\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <label class=\"col-form-label\">\r\n                  <span class=\"farris-label-info text-danger\">*</span>\r\n                  <span class=\"farris-label-text\">\u9A73\u56DE\u81F3</span>\r\n                </label>\r\n                <div class=\"farris-input-wrap\">\r\n                  <farris-lookup-grid #hlp [items]=\"items\" [columns]=\"columns\" [resizable]=\"true\"\r\n                    [singleSelect]=\"true\" [idField]=\"'Id'\" [textField]=\"'Name'\" [(ngModel)]=\"displayText\"\r\n                    [valueField]=\"'Id'\" [pagination]=\"false\" [pageIndex]=\"pageInfo.pageIndex\"\r\n                    [pageSize]=\"pageInfo.pageSize\" [pageList]=\"pageInfo.pageList\" [total]=\"total\"\r\n                    [title]=\"'\u6D3B\u52A8\u5217\u8868'\" [displayType]=\"'List'\" [dictPicked]=\"dictPicked\" [remoteSearch]='false'\r\n                    (pagerChanged)=\"pagerChanged($event)\" (search)=\"search($event)\" [enableClear]=\"false\"\r\n                    (pageSizeChange)=\"pageSizeChange($event)\">\r\n                  </farris-lookup-grid>\r\n                  <div class=\"farris-feedback valid-feedback\"></div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div *ngIf=\"nodeType==='UserActivity'\" class=\"col-12 farris-group-auto my-4\">\r\n            <div class=\"farris-group-wrap pl-5\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <div class=\"farris-input-wrap\">\r\n                  <div class=\"custom-control custom-checkbox\">\r\n                    <input class=\"custom-control-input\" name=\"selectAble\" type=\"checkbox\" id=\"selectAble\"\r\n                      [(ngModel)]=\"selectAbleChecked\" (change)=\"selectAbleClick()\" />\r\n                    <label class=\"custom-control-label\" for=\"selectAble\">\u52A8\u6001\u9009\u62E9\u9A73\u56DE\u7EA7\u6B21</label>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"col-12 farris-group-auto my-4\">\r\n            <div class=\"farris-group-wrap pl-5\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <div class=\"farris-input-wrap\">\r\n                  <div class=\"custom-control custom-checkbox\">\r\n                    <input class=\"custom-control-input\" name=\"waitReturn\" type=\"checkbox\" id=\"waitReturn\" [checked]=\"waitReturnChecked\"\r\n                      (change)=\"waitReturnClick()\" />\r\n                    <label class=\"custom-control-label\" for=\"waitReturn\">\u9A73\u56DE\u540E\u91CD\u65B0\u63D0\u4EA4\u81F3\u672C\u8282\u70B9</label>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n          <div class=\"col-12 farris-group-auto my-4\">\r\n            <div class=\"farris-group-wrap pl-5\">\r\n              <div class=\"form-group farris-form-group\">\r\n                <div class=\"farris-input-wrap\">\r\n                  <div class=\"custom-control custom-checkbox\">\r\n                    <input class=\"custom-control-input\" name=\"dynamicBackWardRuleAble\" type=\"checkbox\" id=\"dynamicBackWardRuleAble\" [checked]=\"dynamicBackWardRuleAbleChecked\"\r\n                      (change)=\"dynamicBackWardRuleAbleClick()\" />\r\n                    <label class=\"custom-control-label\" for=\"dynamicBackWardRuleAble\">\u52A8\u6001\u9009\u62E9\u9A73\u56DE\u89C4\u5219</label>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </div>\r\n        <!-- <form class=\"h-100\">\r\n          <div class=\"f-utils-fill-flex-column justify-content-center h-100 pb-4\">\r\n            <div class=\"my-2 mx-5\">\r\n              <div class=\"farris-group-wrap\">\r\n                <div class=\"form-group\">\r\n                  <label class=\"col-form-label\">\r\n                    <span class=\"farris-label-text\">\u9A73\u56DE\u81F3</span>\r\n                  </label>\r\n                  <div class=\"farris-input-wrap mt-1\" style=\"width: 300px;\">\r\n                    <farris-combo-list [(ngModel)]=\"targetRef\" name=\"targetRef\" [ngModelOptions]=\"{standalone: true}\"\r\n                      [idField]=\"'value'\" [textField]=\"'text'\" [valueField]=\"'value'\" [data]=\"listItems\"\r\n                      [panelHeight]=\"100\">\r\n                    </farris-combo-list>\r\n                    <div class=\"farris-feedback valid-feedback\"></div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n            <div class=\"my-2 mx-5\">\r\n              <div class=\"farris-input-wrap\">\r\n                <div class=\"custom-control custom-checkbox\">\r\n                  <input class=\"custom-control-input\" name=\"selectAble\" type=\"checkbox\" id=\"selectAble\" [(ngModel)]=\"selectAbleChecked\"\r\n                    (change)=\"selectAbleClick()\" />\r\n                  <label class=\"custom-control-label\" for=\"selectAble\">\u52A8\u6001\u9009\u62E9\u9A73\u56DE\u7EA7\u6B21</label>\r\n                </div>\r\n              </div>\r\n            </div>\r\n            <div class=\"my-2 mx-5\">\r\n              <div class=\"farris-input-wrap\">\r\n                <div class=\"custom-control custom-checkbox\">\r\n                  <input class=\"custom-control-input\" name=\"waitReturn\" type=\"checkbox\" id=\"waitReturn\" [checked]=\"waitReturnChecked\"\r\n                    (change)=\"waitReturnClick()\" />\r\n                  <label class=\"custom-control-label\" for=\"waitReturn\">\u9A73\u56DE\u540E\u91CD\u65B0\u63D0\u4EA4\u81F3\u672C\u8282\u70B9</label>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </div>\r\n        </form> -->\r\n      </div>\r\n    </section>\r\n  </div>\r\n</div>\r\n<ng-template #iconSelectFooter>\r\n  <button type=\"button\" class=\"btn btn-primary btn-sm mx-2\" (click)=\"clickOK()\">\u786E\u5B9A</button>\r\n  <button type=\"button\" class=\"btn btn-secondary btn-sm mx-2\" (click)=\"clickCancel()\">\u53D6\u6D88</button>\r\n</ng-template>\r\n",
                    styles: [".item-header{padding:1rem}.item{padding:.5rem 5rem}"]
                }] }
    ];
    /** @nocollapse */
    BackwardRuleComponent.ctorParameters = function () { return [
        { type: ProcessDesignerUIState },
        { type: MessagerService }
    ]; };
    BackwardRuleComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        editorParams: [{ type: Input }],
        value: [{ type: Input }],
        nodeType: [{ type: Input }],
        closeModal: [{ type: Output }],
        submitModal: [{ type: Output }],
        modalFooter: [{ type: ViewChild, args: ['iconSelectFooter',] }],
        hlp: [{ type: ViewChild, args: ['hlp',] }]
    };
    return BackwardRuleComponent;
}());
export { BackwardRuleComponent };
if (false) {
    /** @type {?} */
    BackwardRuleComponent.prototype.cls;
    /** @type {?} */
    BackwardRuleComponent.prototype.editorParams;
    /** @type {?} */
    BackwardRuleComponent.prototype.nodeType;
    /** @type {?} */
    BackwardRuleComponent.prototype.closeModal;
    /** @type {?} */
    BackwardRuleComponent.prototype.submitModal;
    /** @type {?} */
    BackwardRuleComponent.prototype.modalFooter;
    /** @type {?} */
    BackwardRuleComponent.prototype.hlp;
    /** @type {?} */
    BackwardRuleComponent.prototype.modalConfig;
    /** @type {?} */
    BackwardRuleComponent.prototype.displayText;
    /** @type {?} */
    BackwardRuleComponent.prototype.listItems;
    /** @type {?} */
    BackwardRuleComponent.prototype.items;
    /** @type {?} */
    BackwardRuleComponent.prototype.columns;
    /** @type {?} */
    BackwardRuleComponent.prototype.selectAbleChecked;
    /** @type {?} */
    BackwardRuleComponent.prototype.waitReturnChecked;
    /** @type {?} */
    BackwardRuleComponent.prototype.dynamicBackWardRuleAbleChecked;
    /** @type {?} */
    BackwardRuleComponent.prototype.targetRef;
    /** @type {?} */
    BackwardRuleComponent.prototype.pageInfo;
    /** @type {?} */
    BackwardRuleComponent.prototype.total;
    /** @type {?} */
    BackwardRuleComponent.prototype.dictPicked;
    /**
     * @type {?}
     * @private
     */
    BackwardRuleComponent.prototype.uiState;
    /**
     * @type {?}
     * @private
     */
    BackwardRuleComponent.prototype.message;
}
var BackwardRuleConverter = /** @class */ (function () {
    function BackwardRuleConverter() {
    }
    /**
     * @param {?} data
     * @return {?}
     */
    BackwardRuleConverter.prototype.convertTo = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        if (data) {
            return '（已设置）';
        }
        return '（未设置）';
    };
    return BackwardRuleConverter;
}());
export { BackwardRuleConverter };
