// @farris/component-querycondition/esm5/lib/entity/conditionvalue/single-date-value.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/conditionvalue/single-date-value.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { ControlType } from "../controltype";
var SingleDateValue = /** @class */ (function () {
    function SingleDateValue(date) {
        if (date === void 0) { date = ''; }
        this.dateValue = date;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    SingleDateValue.prototype.setOriginalValue = /**
     * @param {?} value
     * @return {?}
     */
    function (value) {
        this.dateValue = value.formatted;
    };
    /**
     * @return {?}
     */
    SingleDateValue.prototype.getOriginalValue = /**
     * @return {?}
     */
    function () {
        return this.dateValue;
    };
    /**
     * @return {?}
     */
    SingleDateValue.prototype.getControlType = /**
     * @return {?}
     */
    function () {
        return ControlType.SingleDate;
    };
    /**
     * @return {?}
     */
    SingleDateValue.prototype.isEmpty = /**
     * @return {?}
     */
    function () {
        return !this.dateValue;
    };
    /**
     * @return {?}
     */
    SingleDateValue.prototype.clearValue = /**
     * @return {?}
     */
    function () {
        this.dateValue = undefined;
    };
    return SingleDateValue;
}());
export { SingleDateValue };
if (false) {
    /** @type {?} */
    SingleDateValue.prototype.dateValue;
}
