// @gsp-cmp/querysolution/esm2015/lib/saveasdialog/saveasdialog.component.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/saveasdialog/saveasdialog.component.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Output, EventEmitter, ViewChild, HostBinding, ViewEncapsulation } from '@angular/core';
import { NotifyService } from '@farris/ui-notify';
import { QuerySolutionVo, SolutionService } from '@farris/component-querycondition';
import { LocaleService } from '@farris/ui-locale';
import { DialogComponent } from '@farris/ui-dialog';
import { cloneDeep } from 'lodash-es';
export class SaveAsDialogComponent {
    /**
     * @param {?} notify
     * @param {?} solutionService
     * @param {?} localeService
     */
    constructor(notify, solutionService, localeService) {
        this.notify = notify;
        this.solutionService = solutionService;
        this.localeService = localeService;
        this.cls = 'saveasdialog';
        this.cancelEmitter = new EventEmitter();
        this.confirmEmitter = new EventEmitter();
        //@Input属性，旧代码未设置，故保留
        this.querySolution = new QuerySolutionVo();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    show() {
        this.dialog.show();
    }
    /**
     * @return {?}
     */
    cancel() {
        // this.cancelEmitter.emit(true);
        this.dialog.close();
    }
    /**
     * @return {?}
     */
    confirm() {
        if (!this.querySolution.code) {
            this.solutionService.showInfo(this.localeService.getValue('querySolution.saveAsDialog.pleaseInput'), 'info', '', 1500);
            return;
        }
        /** @type {?} */
        let obj = new Object();
        obj['querySolution'] = cloneDeep(this.querySolution);
        this.confirmEmitter.emit(obj);
        this.dialog.close();
    }
    /**
     * @param {?} checkStatus
     * @return {?}
     */
    setDefault(checkStatus) {
        this.querySolution.isDefault = checkStatus;
    }
}
SaveAsDialogComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-saveasdialog',
                template: "<farris-dialog #dialog [buttons]=\"saveasbuttonRef\" [width]=\"550\" [height]=\"310\" [title]=\"title\">\r\n    <div class=\"qs-save-as-input\">\r\n        <div class=\"farris-input-wrap\">\r\n            <input class=\"form-control\" placeholder=\"{{'querySolution.saveAsDialog.placeholder' | locale}}\" [(ngModel)]=\"querySolution.code\"/>\r\n        </div>\r\n        <div class=\"custom-control custom-checkbox qs-save-as-check\">\r\n            <input class=\"custom-control-input\" type=\"checkbox\" (change)=\"setDefault($event.target.checked)\" [(ngModel)]=\"querySolution.isDefault\"/>\r\n            <label class=\"custom-control-label\" (click)=\"setDefault(!querySolution.isDefault)\">{{'querySolution.saveAsDialog.setAsDefault' | locale}}</label>\r\n        </div>\r\n    </div>\r\n</farris-dialog>\r\n<ng-template #saveasbuttonRef>\r\n    <div class=\"qs-save-as-btn\">\r\n        <button class=\"btn btn-primary\" (click)=\"confirm()\">{{'querySolution.saveAsDialog.confirm' | locale}}</button>        \r\n        <button class=\"btn btn-secondary\" (click)=\"cancel()\">{{'querySolution.saveAsDialog.cancel' | locale}}</button>\r\n    </div>\r\n</ng-template>",
                encapsulation: ViewEncapsulation.None,
                styles: [".qs-save-as-btn .btn{float:right;margin-left:10px;font-size:14px;line-height:26px}.qs-save-as-btn .btn-secondary{min-width:70px}.qs-save-as-input{width:100%;line-height:20px;font-size:14px;padding:20px 10px}.qs-save-as-check{margin:10px auto}"]
            }] }
];
/** @nocollapse */
SaveAsDialogComponent.ctorParameters = () => [
    { type: NotifyService },
    { type: SolutionService },
    { type: LocaleService }
];
SaveAsDialogComponent.propDecorators = {
    cls: [{ type: HostBinding, args: ['class',] }],
    cancelEmitter: [{ type: Output }],
    confirmEmitter: [{ type: Output }],
    dialog: [{ type: ViewChild, args: ['dialog',] }]
};
if (false) {
    /** @type {?} */
    SaveAsDialogComponent.prototype.cls;
    /** @type {?} */
    SaveAsDialogComponent.prototype.cancelEmitter;
    /** @type {?} */
    SaveAsDialogComponent.prototype.confirmEmitter;
    /** @type {?} */
    SaveAsDialogComponent.prototype.querySolution;
    /** @type {?} */
    SaveAsDialogComponent.prototype.title;
    /** @type {?} */
    SaveAsDialogComponent.prototype.defaultQuerySolution;
    /** @type {?} */
    SaveAsDialogComponent.prototype.dialog;
    /**
     * @type {?}
     * @private
     */
    SaveAsDialogComponent.prototype.notify;
    /**
     * @type {?}
     * @private
     */
    SaveAsDialogComponent.prototype.solutionService;
    /**
     * @type {?}
     * @private
     */
    SaveAsDialogComponent.prototype.localeService;
}
