// @gsp-wf/ui-flowchart/esm2015/lib/ui-flowchart.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { UIFlowchartComponent } from './task-flowchart/task-flowchart.component';
import { CommonModule } from '@angular/common';
import { HttpService } from '@ecp-caf/caf-common';
import { UiFlowchartService } from './services/ui-flowchart.service';
import { WFFlowchartService } from './wf-flowchart.service';
import { MessagerModule } from '@farris/ui-messager';
import { FarrisDialogModule } from '@farris/ui-dialog';
import { TranslatePipe } from './services/translate.pipe';
import { LayoutModule } from '@farris/ui-layout';
import { FarrisSectionModule } from '@farris/ui-section';
import { WfApprovalLogsModule } from '@gsp-wf/wf-approval-logs';
import { ComboListModule } from '@farris/ui-combo-list';
import { ReactiveFormsModule, FormsModule } from '@angular/forms';
import { FDropdownDirectiveTypeModule } from '@farris/ui-dropdown';
import { DatagridModule } from '@farris/ui-datagrid';
import { NotifyModule } from '@farris/ui-notify';
export class UiFlowchartModule {
}
UiFlowchartModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    UIFlowchartComponent,
                    TranslatePipe
                ],
                imports: [
                    LayoutModule,
                    CommonModule,
                    FarrisDialogModule,
                    MessagerModule.forRoot({
                        width: 500
                    }),
                    FarrisSectionModule,
                    WfApprovalLogsModule,
                    ComboListModule,
                    ReactiveFormsModule,
                    FormsModule,
                    FDropdownDirectiveTypeModule,
                    DatagridModule,
                    NotifyModule
                ],
                exports: [UIFlowchartComponent],
                providers: [
                    HttpService,
                    UiFlowchartService,
                    WFFlowchartService,
                ],
                entryComponents: [
                    UIFlowchartComponent
                ]
            },] }
];
