// @farris/command-services/esm2015/lib/entity-services/entity.service.js
import { Injectable } from '@angular/core';
import { FrameContext } from '@farris/devkit';
import { EntityTraversingService } from './entity-traversing.service';
import { EntityManipulationService } from './entity-manipulation.service';
import { EntityAggregationService } from './entity-aggregation.service';
/**
 * 实体服务
 * @scope FrameComponent
 */
class EntityService {
    constructor(frameContext) {
        this.frameContext = frameContext;
        const injector = this.frameContext.injector;
        this.traversingService = injector.get(EntityTraversingService);
        this.manipulationService = injector.get(EntityManipulationService);
        this.aggregationService = injector.get(EntityAggregationService);
    }
    /**
     * 获取path对应的属性值
     */
    getPropValue(path) {
        return this.traversingService.getPropValue(path);
    }
    /**
     * 设置path对应的属性值
     */
    setPropValue(path, value) {
        return this.manipulationService.setPropValue(path, value);
    }
    /**
     * 设置path对应的Entity数据
     */
    getEntityData(path) {
        return this.traversingService.getEntityData(path);
    }
    /**
     * 设置path对应的EntityList数据
     */
    getEntityListData(path) {
        return this.traversingService.getEntityListData(path);
    }
    /**
     * 获取path对应EntityList记录数
     */
    count(path) {
        return this.aggregationService.count(path);
    }
    /**
     * 获取path对应属性的合计值
     */
    sum(path) {
        return this.aggregationService.sum(path);
    }
    /**
     * 获取path对应属性的平均值
     */
    avg(path) {
        return this.aggregationService.avg(path);
    }
    /**
     * 获取path对应属性的最大值
     */
    max(path) {
        return this.aggregationService.max(path);
    }
    /**
     * 获取path对应属性的最小值
     */
    min(path) {
        return this.aggregationService.min(path);
    }
}
EntityService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
EntityService.ctorParameters = () => [
    { type: FrameContext }
];
export { EntityService };
