// @farris/kendo-binding/esm5/lib/directives/component-ref/filter-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { FilterComponent } from '@farris/ui-filter';
var FilterComponentRefDirective = /** @class */ (function () {
    function FilterComponentRefDirective(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    FilterComponentRefDirective.prototype.ngOnInit = function () {
        var id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    };
    FilterComponentRefDirective.prototype.ngOnDestroy = function () {
    };
    FilterComponentRefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[filterRef]'
                },] }
    ];
    /** @nocollapse */
    FilterComponentRefDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: FilterComponent },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    return FilterComponentRefDirective;
}());
export { FilterComponentRefDirective };
