// @farris/kendo-binding/esm5/lib/directives/component-ref/section-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { FarrisSectionComponent } from '@farris/ui-section';
var SectionComponentRefDirective = /** @class */ (function () {
    function SectionComponentRefDirective(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    SectionComponentRefDirective.prototype.ngOnInit = function () {
        var id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    };
    SectionComponentRefDirective.prototype.ngOnDestroy = function () {
    };
    SectionComponentRefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[sectionRef]'
                },] }
    ];
    /** @nocollapse */
    SectionComponentRefDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: FarrisSectionComponent },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    return SectionComponentRefDirective;
}());
export { SectionComponentRefDirective };
