// @gsp-wf/wf-process-editor/esm5/lib/views/performer/views/performer-options/performer-options.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, HostBinding, Input, NgZone, ViewChild, ElementRef } from '@angular/core';
import { PerformerOptions } from '../../entities/performer-options';
import { ProcessDesignerUIState } from '../../../../app/process-designer.uistate';
import { ExpressionService, InjectedType } from '@gsp-svc/expression';
import { MessagerService } from '@farris/ui-messager';
var PerformerOptionsComponent = /** @class */ (function () {
    function PerformerOptionsComponent(ngZone, designerState, expr, message) {
        this.ngZone = ngZone;
        this.designerState = designerState;
        this.expr = expr;
        this.message = message;
        this.cls = 'farris-main-area flex-column';
        this.showClearButton = false;
        this.historicFirst = false;
        this.excludeProcInstCreator = false;
        this.excludePreActExecutors = false;
        this.excludeAllPassExecutors = false;
        this.autoProcInstCreator = false;
        this.autoPreActExecutors = false;
        this.autoAllPassExecutors = false;
        this.noPerformerStrategyList = [
            { text: '向下流转', value: 'leave' },
            { text: '异常提示', value: 'throwException' }
        ];
        this.exclusionPerformersOptions = [
            { value: 'procInstCreator', name: '流程发起人' },
            { value: 'preActExecutors', name: '上一节点办理人' },
            { value: 'allPassExecutors', name: '所有已办理人员' }
        ];
        this.noPerformerStrategy = this.noPerformerStrategyList[0].value;
    }
    Object.defineProperty(PerformerOptionsComponent.prototype, "options", {
        set: /**
         * @param {?} options
         * @return {?}
         */
        function (options) {
            if (options) {
                this.exclusionPerformers = options.ExclusionPerformers;
                this.noPerformerStrategy = options.NoPerformerStrategy;
                this.historicFirst = options.HistoricFirst;
                this.automaticPerformers = options.automaticPerformers;
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PerformerOptionsComponent.prototype, "exclusionPerformers", {
        set: /**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            if (e) {
                if (typeof e === 'string') {
                    this.excludeProcInstCreator = e.indexOf('procInstCreator') > -1;
                    this.excludePreActExecutors = e.indexOf('preActExecutors') > -1;
                }
                else {
                    this.excludeProcInstCreator = e.findIndex((/**
                     * @param {?} i
                     * @return {?}
                     */
                    function (i) { return i === 'procInstCreator'; })) > -1;
                    this.excludePreActExecutors = e.findIndex((/**
                     * @param {?} i
                     * @return {?}
                     */
                    function (i) { return i === 'preActExecutors'; })) > -1;
                    this.excludeAllPassExecutors = e.findIndex((/**
                     * @param {?} i
                     * @return {?}
                     */
                    function (i) { return i === 'allPassExecutors'; })) > -1;
                    this.excludeExprPerformer = e.find((/**
                     * @param {?} i
                     * @return {?}
                     */
                    function (i) { return i.indexOf('expr') > -1; }));
                }
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PerformerOptionsComponent.prototype, "automaticPerformers", {
        set: /**
         * @param {?} e
         * @return {?}
         */
        function (e) {
            if (e && e.length > 0) {
                this.autoProcInstCreator = e.findIndex((/**
                 * @param {?} i
                 * @return {?}
                 */
                function (i) { return i === 'procInstCreator'; })) > -1;
                this.autoPreActExecutors = e.findIndex((/**
                 * @param {?} i
                 * @return {?}
                 */
                function (i) { return i === 'preActExecutors'; })) > -1;
                this.autoAllPassExecutors = e.findIndex((/**
                 * @param {?} i
                 * @return {?}
                 */
                function (i) { return i === 'allPassExecutors'; })) > -1;
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @return {?}
     */
    PerformerOptionsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    PerformerOptionsComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.ngZone.runOutsideAngular((/**
         * @return {?}
         */
        function () {
            _this.inputGroup.nativeElement.addEventListener('mouseenter', _this.onMouseEnter.bind(_this));
            _this.inputGroup.nativeElement.addEventListener('mouseleave', _this.onMouseLeave.bind(_this));
        }));
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PerformerOptionsComponent.prototype.onMouseEnter = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.showClearButton = true;
        if (this.showClearButton) {
            this.toggleClearIcon(event, true);
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PerformerOptionsComponent.prototype.onMouseLeave = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.showClearButton = false;
        this.toggleClearIcon(event, false);
    };
    /**
     * @private
     * @param {?} event
     * @param {?=} isShow
     * @return {?}
     */
    PerformerOptionsComponent.prototype.toggleClearIcon = /**
     * @private
     * @param {?} event
     * @param {?=} isShow
     * @return {?}
     */
    function (event, isShow) {
        if (isShow === void 0) { isShow = false; }
        /** @type {?} */
        var str = isShow ? '' : 'none';
        /** @type {?} */
        var clearIcon = event.target.querySelector('.input-group-clear');
        if (clearIcon) {
            clearIcon.style.display = str;
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PerformerOptionsComponent.prototype.excludeProcInstCreatorClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        // this.excludeProcInstCreator = !this.excludeProcInstCreator;
        if (this.excludeProcInstCreator && this.autoProcInstCreator) {
            this.autoProcInstCreator = false;
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PerformerOptionsComponent.prototype.excludePreActExecutorsClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var _this = this;
        // this.excludePreActExecutors = !this.excludePreActExecutors;
        if (this.autoAllPassExecutors === true && this.excludePreActExecutors === true) {
            this.message.info('所有已办理人员已设置自动审批，上个节点已办理人员默认不允许排除！', (/**
             * @return {?}
             */
            function () {
                _this.excludePreActExecutors = false;
            }));
            return;
        }
        if (this.excludePreActExecutors && this.autoPreActExecutors) {
            this.autoPreActExecutors = false;
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PerformerOptionsComponent.prototype.excludeAllPassExecutorsClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        // this.excludeAllPassExecutors = !this.excludeAllPassExecutors;
        if (this.excludeAllPassExecutors) {
            if (this.autoPreActExecutors) {
                this.autoPreActExecutors = false;
            }
            if (this.autoAllPassExecutors) {
                this.autoAllPassExecutors = false;
            }
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PerformerOptionsComponent.prototype.autoProcInstCreatorClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var _this = this;
        // if (this.designerState.propertyData.CompleteStrategy !== this.designerState.completeIterator[0].key
        //   && this.autoProcInstCreator === true ) {
        //   this.message.info('【完成策略】为【会签】时，不允许设置自动审批，请了解！', () => {
        //     this.autoProcInstCreator = false;
        //   });
        //   return;
        // }
        if (this.designerState.propertyData.AssignStrategy === this.designerState.assignIterator[2].key
            && this.autoProcInstCreator === true) {
            this.message.info('【分配策略】为【领用】时，不允许设置自动审批，请了解！', (/**
             * @return {?}
             */
            function () {
                _this.autoProcInstCreator = false;
            }));
            return;
        }
        // this.autoProcInstCreator = !this.autoProcInstCreator;
        if (this.autoProcInstCreator && this.excludeProcInstCreator) {
            this.excludeProcInstCreator = false;
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PerformerOptionsComponent.prototype.autoPreActExecutorsClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var _this = this;
        // if (this.designerState.propertyData.CompleteStrategy !== this.designerState.completeIterator[0].key
        //   && this.autoPreActExecutors === true) {
        //   this.message.info('【完成策略】为【会签】时，不允许设置自动审批，请了解！', () => {
        //     this.autoPreActExecutors = false;
        //   });
        //   return;
        // }
        if (this.designerState.propertyData.AssignStrategy === this.designerState.assignIterator[2].key
            && this.autoPreActExecutors === true) {
            this.message.info('【分配策略】为【领用】时，不允许设置自动审批，请了解！', (/**
             * @return {?}
             */
            function () {
                _this.autoPreActExecutors = false;
            }));
            return;
        }
        if (this.excludeAllPassExecutors === true && this.autoPreActExecutors === true) {
            this.message.info('所有已办理人员已设置不允许审批，所以上个节点已办理人员默认不允许自动审批！', (/**
             * @return {?}
             */
            function () {
                _this.autoPreActExecutors = false;
            }));
            return;
        }
        // this.autoPreActExecutors = !this.autoPreActExecutors;
        if (this.autoPreActExecutors && this.excludePreActExecutors) {
            this.excludePreActExecutors = false;
        }
    };
    /**
     * @param {?} event
     * @return {?}
     */
    PerformerOptionsComponent.prototype.autoAllPassExecutorsClick = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        var _this = this;
        // if (this.designerState.propertyData.CompleteStrategy !== this.designerState.completeIterator[0].key
        //   && this.autoAllPassExecutors === true) {
        //   this.message.info('【完成策略】为【会签】时，不允许设置自动审批，请了解！', () => {
        //     this.autoAllPassExecutors = false;
        //   });
        //   return;
        // }
        if (this.designerState.propertyData.AssignStrategy === this.designerState.assignIterator[2].key
            && this.autoAllPassExecutors === true) {
            this.message.info('【分配策略】为【领用】时，不允许设置自动审批，请了解！', (/**
             * @return {?}
             */
            function () {
                _this.autoAllPassExecutors = false;
            }));
            return;
        }
        // this.autoAllPassExecutors = !this.autoAllPassExecutors;
        if (this.autoAllPassExecutors) {
            if (this.excludeAllPassExecutors) {
                this.excludeAllPassExecutors = false;
            }
            if (this.excludePreActExecutors) {
                this.excludePreActExecutors = false;
            }
        }
    };
    /**
     * @return {?}
     */
    PerformerOptionsComponent.prototype.getPerformerOptions = /**
     * @return {?}
     */
    function () {
        return {
            ExclusionPerformers: this.getExclusionPerformers(),
            NoPerformerStrategy: this.noPerformerStrategy,
            HistoricFirst: this.historicFirst,
            automaticPerformers: this.getAutoPerformers()
        };
    };
    /**
     * @private
     * @return {?}
     */
    PerformerOptionsComponent.prototype.getExclusionPerformers = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var exclusionPerf = [];
        if (this.excludeProcInstCreator) {
            exclusionPerf.push('procInstCreator');
        }
        if (this.excludePreActExecutors) {
            exclusionPerf.push('preActExecutors');
        }
        if (this.excludeAllPassExecutors) {
            exclusionPerf.push('allPassExecutors');
        }
        if (this.excludeExprPerformer) {
            exclusionPerf.push(this.excludeExprPerformer);
        }
        return exclusionPerf;
    };
    /**
     * @private
     * @return {?}
     */
    PerformerOptionsComponent.prototype.getAutoPerformers = /**
     * @private
     * @return {?}
     */
    function () {
        /** @type {?} */
        var autoPerf = new Array();
        if (this.autoProcInstCreator) {
            autoPerf.push('procInstCreator');
        }
        if (this.autoPreActExecutors) {
            autoPerf.push('preActExecutors');
        }
        if (this.autoAllPassExecutors) {
            autoPerf.push('allPassExecutors');
        }
        return autoPerf;
    };
    /**
     * @return {?}
     */
    PerformerOptionsComponent.prototype.openExprEditor = /**
     * @return {?}
     */
    function () {
        var _this = this;
        var e_1, _a, e_2, _b, e_3, _c, e_4, _d, e_5, _e, e_6, _f;
        if (this.designerState.schemas.length > 0) {
            try {
                for (var _g = tslib_1.__values(this.designerState.schemas), _h = _g.next(); !_h.done; _h = _g.next()) {
                    var schema = _h.value;
                    this.expr.addSchema('Schema', schema);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_h && !_h.done && (_a = _g.return)) _a.call(_g);
                }
                finally { if (e_1) throw e_1.error; }
            }
            if (this.designerState.schemas[0].entityTypes && this.designerState.schemas[0].entityTypes.length > 0) {
                /** @type {?} */
                var entityType = this.designerState.schemas[0].entityTypes[0];
                this.expr.addInjectedEntity(entityType.name, 'Schema', entityType.name, InjectedType.EntityType);
            }
        }
        /** @type {?} */
        var context = new Array();
        if (this.designerState.procInstContext.length > 0) {
            try {
                for (var _j = tslib_1.__values(this.designerState.procInstContext), _k = _j.next(); !_k.done; _k = _j.next()) {
                    var variable = _k.value;
                    context.push(variable);
                    // this.expr.addContext(variable);
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_k && !_k.done && (_b = _j.return)) _b.call(_j);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
        if (this.designerState.formalParameterContext.length > 0) {
            var _loop_1 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _l = tslib_1.__values(this.designerState.formalParameterContext), _m = _l.next(); !_m.done; _m = _l.next()) {
                    var variable = _m.value;
                    _loop_1(variable);
                }
            }
            catch (e_3_1) { e_3 = { error: e_3_1 }; }
            finally {
                try {
                    if (_m && !_m.done && (_c = _l.return)) _c.call(_l);
                }
                finally { if (e_3) throw e_3.error; }
            }
        }
        if (this.designerState.participantContext.length > 0) {
            var _loop_2 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _o = tslib_1.__values(this.designerState.participantContext), _p = _o.next(); !_p.done; _p = _o.next()) {
                    var variable = _p.value;
                    _loop_2(variable);
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (_p && !_p.done && (_d = _o.return)) _d.call(_o);
                }
                finally { if (e_4) throw e_4.error; }
            }
        }
        if (this.designerState.processVariableContext.length > 0) {
            var _loop_3 = function (variable) {
                if (!context.some((/**
                 * @param {?} c
                 * @return {?}
                 */
                function (c) { return c.key === variable.key; }))) {
                    context.push(variable);
                }
            };
            try {
                for (var _q = tslib_1.__values(this.designerState.processVariableContext), _r = _q.next(); !_r.done; _r = _q.next()) {
                    var variable = _r.value;
                    _loop_3(variable);
                }
            }
            catch (e_5_1) { e_5 = { error: e_5_1 }; }
            finally {
                try {
                    if (_r && !_r.done && (_e = _q.return)) _e.call(_q);
                }
                finally { if (e_5) throw e_5.error; }
            }
        }
        if (context.length > 0) {
            try {
                for (var context_1 = tslib_1.__values(context), context_1_1 = context_1.next(); !context_1_1.done; context_1_1 = context_1.next()) {
                    var v = context_1_1.value;
                    this.expr.addContext(v);
                }
            }
            catch (e_6_1) { e_6 = { error: e_6_1 }; }
            finally {
                try {
                    if (context_1_1 && !context_1_1.done && (_f = context_1.return)) _f.call(context_1);
                }
                finally { if (e_6) throw e_6.error; }
            }
        }
        // if (this.excludeExprPerformer) {
        this.expr.addExpressionText(this.excludeExprPerformer);
        // }
        this.expr.buildExpression().then((/**
         * @param {?} d
         * @return {?}
         */
        function (d) {
            _this.excludeExprPerformer = d;
        }));
    };
    /**
     * @return {?}
     */
    PerformerOptionsComponent.prototype.clear = /**
     * @return {?}
     */
    function () {
        this.excludeExprPerformer = '';
    };
    PerformerOptionsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-performer-options',
                    template: "<div class=\"f-page\">\r\n  <div class=\"f-utils-fill-flex-column farris-split-section m-2 h-100\">\r\n    <section class=\"f-utils-fill-flex-column h-100\">\r\n      <div class=\"f-section-header pt-2 mx-2\">\r\n        <div class=\"f-title\">\r\n          <h4 class=\"f-title-text\">\u53C2\u4E0E\u8005\u9AD8\u7EA7\u9009\u9879</h4>\r\n        </div>\r\n      </div>\r\n      <div class=\"f-utils-fill-flex-column h-100\">\r\n        <perfect-scrollbar [config]=\"{}\">\r\n          <form class=\"h-100\">\r\n            <div class=\"f-utils-fill-flex-column h-100 pb-4\">\r\n              <div class=\"farris-form form-inline farris-form-inline\">\r\n                <div class=\"col-12 farris-group-auto\">\r\n                  <div class=\"farris-group-wrap pl-3\">\r\n                    <div class=\"form-group farris-form-group\">\r\n                      <div class=\"farris-input-wrap\">\r\n                        <div class=\"custom-control custom-checkbox\">\r\n                          <input class=\"custom-control-input\" name=\"historicFirst\" type=\"checkbox\" id=\"historicFirst\"\r\n                            [(ngModel)]=\"historicFirst\" />\r\n                          <label class=\"custom-control-label\" for=\"historicFirst\">\u518D\u6B21\u8FDB\u5165\u8282\u70B9\u65F6\uFF0C\u529E\u7406\u4EBA\u662F\u8282\u70B9\u7684\u5DF2\u529E\u4EBA\u5458</label>\r\n                        </div>\r\n                      </div>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n                <div class=\"col-9 farris-group-auto mt-2 pr-4\">\r\n                  <div class=\"farris-group-wrap f-form-lable-auto pl-3\">\r\n                    <div class=\"form-group farris-form-group\">\r\n                      <label class=\"col-form-label\">\r\n                        <span class=\"farris-label-text\">\u529E\u7406\u4EBA\u4E3A\u7A7A\u7B56\u7565\uFF1A</span>\r\n                      </label>\r\n                      <div class=\"farris-input-wrap\">\r\n                        <farris-combo-list [(ngModel)]=\"noPerformerStrategy\" name=\"noPerformerStrategy\" [idField]=\"'value'\"\r\n                          [textField]=\"'text'\" [valueField]=\"'value'\" [data]=\"noPerformerStrategyList\" [panelHeight]=\"60\">\r\n                        </farris-combo-list>\r\n                        <div class=\"farris-feedback valid-feedback\"></div>\r\n                      </div>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n                <div class=\"mx-2 mt-3\">\r\n                  <div class=\"title-content\">\r\n                    <h4 class=\"title-text\">\u6392\u9664\u529E\u7406\u4EBA\u9009\u9879</h4><span class=\"title-line\"></span>\r\n                  </div>\r\n                </div>\r\n                <div class=\"col-12 farris-group-auto mt-2\">\r\n                  <div class=\"farris-group-wrap pl-3\">\r\n                    <div class=\"form-group farris-form-group\">\r\n                      <div class=\"farris-input-wrap\">\r\n                        <div class=\"custom-control custom-checkbox\">\r\n                          <input class=\"custom-control-input\" name=\"excludeProcInstCreator\" type=\"checkbox\" id=\"excludeProcInstCreator\"\r\n                            [(ngModel)]=\"excludeProcInstCreator\" (change)=\"excludeProcInstCreatorClick($event)\" />\r\n                          <label class=\"custom-control-label\" for=\"excludeProcInstCreator\">\u6D41\u7A0B\u53D1\u8D77\u4EBA\u9ED8\u8BA4\u4E0D\u53C2\u4E0E\u5BA1\u6279</label>\r\n                        </div>\r\n                      </div>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n                <div class=\"col-12 farris-group-auto mt-2\">\r\n                  <div class=\"farris-group-wrap pl-3\">\r\n                    <div class=\"form-group farris-form-group\">\r\n                      <div class=\"farris-input-wrap\">\r\n                        <div class=\"custom-control custom-checkbox\">\r\n                          <input class=\"custom-control-input\" name=\"excludePreActExecutors\" type=\"checkbox\" id=\"excludePreActExecutors\"\r\n                            [(ngModel)]=\"excludePreActExecutors\" (change)=\"excludePreActExecutorsClick($event)\"/>\r\n                          <label class=\"custom-control-label\" for=\"excludePreActExecutors\">\u4E0A\u4E2A\u8282\u70B9\u5DF2\u529E\u7406\u4EBA\u5458\u9ED8\u8BA4\u4E0D\u53C2\u4E0E\u5BA1\u6279</label>\r\n                        </div>\r\n                      </div>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n                <div class=\"col-12 farris-group-auto mt-2\">\r\n                  <div class=\"farris-group-wrap pl-3\">\r\n                    <div class=\"form-group farris-form-group\">\r\n                      <div class=\"farris-input-wrap\">\r\n                        <div class=\"custom-control custom-checkbox\">\r\n                          <input class=\"custom-control-input\" name=\"excludeAllPassExecutors\" type=\"checkbox\" id=\"excludeAllPassExecutors\"\r\n                            [(ngModel)]=\"excludeAllPassExecutors\" (change)=\"excludeAllPassExecutorsClick($event)\"/>\r\n                          <label class=\"custom-control-label\" for=\"excludeAllPassExecutors\">\u6240\u6709\u5DF2\u529E\u7406\u4EBA\u5458\u9ED8\u8BA4\u4E0D\u53C2\u4E0E\u5BA1\u6279</label>\r\n                        </div>\r\n                      </div>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n                <div class=\"col-11 farris-group-auto pr-4\">\r\n                  <div class=\"farris-group-wrap f-form-lable-auto pl-3\">\r\n                    <div class=\"form-group farris-form-group\">\r\n                      <label class=\"col-form-label\">\r\n                        <span class=\"farris-label-text\">\u6D41\u7A0B\u53D8\u91CF\uFF1A</span>\r\n                      </label>\r\n                      <div class=\"farris-input-wrap\">\r\n                        <div #inputGroup class=\"input-group\">\r\n                          <input type=\"text\" class=\"form-control\" [(ngModel)]=\"excludeExprPerformer\" name=\"excludeExprPerformer\"\r\n                            [readOnly]=\"true\" />\r\n                          <div class=\"input-group-append\">\r\n                            <span [style.display]=\"showClearButton ? '' : 'none'\" class=\"input-group-text input-group-clear\"\r\n                              (click)=\"clear()\">\r\n                              <i class=\"f-icon modal_close\"></i>\r\n                            </span>\r\n                            <span class=\"input-group-text\" (click)=\"openExprEditor()\" style=\"background: #FFF;\">\r\n                              <i class=\"f-icon f-icon-lookup\"></i>\r\n                            </span>\r\n                          </div>\r\n                        </div>\r\n                        <div class=\"farris-feedback valid-feedback\"></div>\r\n                      </div>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n\r\n                <div class=\"mx-2 mt-3\">\r\n                  <div class=\"title-content\">\r\n                    <h4 class=\"title-text\">\u81EA\u52A8\u5BA1\u6279\u9009\u9879</h4><span class=\"title-line\"></span>\r\n                  </div>\r\n                </div>\r\n                <div class=\"col-12 farris-group-auto mt-2\">\r\n                  <div class=\"farris-group-wrap pl-3\">\r\n                    <div class=\"form-group farris-form-group\">\r\n                      <div class=\"farris-input-wrap\">\r\n                        <div class=\"custom-control custom-checkbox\">\r\n                          <input class=\"custom-control-input\" name=\"autoProcInstCreator\" type=\"checkbox\" id=\"autoProcInstCreator\"\r\n                            [(ngModel)]=\"autoProcInstCreator\" (change)=\"autoProcInstCreatorClick($event)\" />\r\n                          <label class=\"custom-control-label\" for=\"autoProcInstCreator\">\u6D41\u7A0B\u53D1\u8D77\u4EBA\u9ED8\u8BA4\u81EA\u52A8\u5BA1\u6279\u901A\u8FC7</label>\r\n                        </div>\r\n                      </div>\r\n                    </div>\r\n                  </div>\r\n                </div><div class=\"col-12 farris-group-auto mt-2\">\r\n                  <div class=\"farris-group-wrap pl-3\">\r\n                    <div class=\"form-group farris-form-group\">\r\n                      <div class=\"farris-input-wrap\">\r\n                        <div class=\"custom-control custom-checkbox\">\r\n                          <input class=\"custom-control-input\" name=\"autoPreActExecutors\" type=\"checkbox\" id=\"autoPreActExecutors\"\r\n                            [(ngModel)]=\"autoPreActExecutors\" (change)=\"autoPreActExecutorsClick($event)\" />\r\n                          <label class=\"custom-control-label\" for=\"autoPreActExecutors\">\u4E0A\u4E00\u8282\u70B9\u529E\u7406\u4EBA\u5458\u9ED8\u8BA4\u81EA\u52A8\u5BA1\u6279\u901A\u8FC7</label>\r\n                        </div>\r\n                      </div>\r\n                    </div>\r\n                  </div>\r\n                </div><div class=\"col-12 farris-group-auto mt-2\">\r\n                  <div class=\"farris-group-wrap pl-3\">\r\n                    <div class=\"form-group farris-form-group\">\r\n                      <div class=\"farris-input-wrap\">\r\n                        <div class=\"custom-control custom-checkbox\">\r\n                          <input class=\"custom-control-input\" name=\"autoAllPassExecutors\" type=\"checkbox\" id=\"autoAllPassExecutors\"\r\n                            [(ngModel)]=\"autoAllPassExecutors\" (change)=\"autoAllPassExecutorsClick($event)\" />\r\n                          <label class=\"custom-control-label\" for=\"autoAllPassExecutors\">\u6240\u6709\u5DF2\u529E\u7406\u4EBA\u5458\u9ED8\u8BA4\u81EA\u52A8\u5BA1\u6279\u901A\u8FC7</label>\r\n                        </div>\r\n                      </div>\r\n                    </div>\r\n                  </div>\r\n                </div>\r\n              </div>\r\n            </div>\r\n          </form>\r\n        </perfect-scrollbar>\r\n        \r\n      </div>\r\n    </section>\r\n  </div>\r\n</div>",
                    styles: [".item-header{padding:1rem}.item{padding:.5rem 5rem}.title-content{color:#000;display:inline-flex;align-items:center;line-height:1.375rem;position:relative;overflow:hidden;padding:0 0 0 1rem;min-width:450px}.title-content ::before{content:'';width:.25rem;height:1.125rem;background:#59a1ff;border-radius:1px;position:absolute;top:50%;left:0;margin-top:-.5625rem}.title-text{font-size:1rem;line-height:1.375rem;overflow:hidden;white-space:nowrap;text-overflow:ellipsis;margin-bottom:0;margin-right:10px}.title-line{flex-grow:1;border-bottom:1px dashed rgba(0,0,0,.1)}"]
                }] }
    ];
    /** @nocollapse */
    PerformerOptionsComponent.ctorParameters = function () { return [
        { type: NgZone },
        { type: ProcessDesignerUIState },
        { type: ExpressionService },
        { type: MessagerService }
    ]; };
    PerformerOptionsComponent.propDecorators = {
        cls: [{ type: HostBinding, args: ['class',] }],
        options: [{ type: Input }],
        exclusionPerformers: [{ type: Input }],
        inputGroup: [{ type: ViewChild, args: ['inputGroup',] }]
    };
    return PerformerOptionsComponent;
}());
export { PerformerOptionsComponent };
if (false) {
    /** @type {?} */
    PerformerOptionsComponent.prototype.cls;
    /** @type {?} */
    PerformerOptionsComponent.prototype.inputGroup;
    /** @type {?} */
    PerformerOptionsComponent.prototype.showClearButton;
    /** @type {?} */
    PerformerOptionsComponent.prototype.historicFirst;
    /** @type {?} */
    PerformerOptionsComponent.prototype.excludeExprPerformer;
    /** @type {?} */
    PerformerOptionsComponent.prototype.excludeProcInstCreator;
    /** @type {?} */
    PerformerOptionsComponent.prototype.excludePreActExecutors;
    /** @type {?} */
    PerformerOptionsComponent.prototype.excludeAllPassExecutors;
    /** @type {?} */
    PerformerOptionsComponent.prototype.autoProcInstCreator;
    /** @type {?} */
    PerformerOptionsComponent.prototype.autoPreActExecutors;
    /** @type {?} */
    PerformerOptionsComponent.prototype.autoAllPassExecutors;
    /** @type {?} */
    PerformerOptionsComponent.prototype.noPerformerStrategyList;
    /** @type {?} */
    PerformerOptionsComponent.prototype.exclusionPerformersOptions;
    /** @type {?} */
    PerformerOptionsComponent.prototype.noPerformerStrategy;
    /**
     * @type {?}
     * @private
     */
    PerformerOptionsComponent.prototype.ngZone;
    /**
     * @type {?}
     * @private
     */
    PerformerOptionsComponent.prototype.designerState;
    /**
     * @type {?}
     * @private
     */
    PerformerOptionsComponent.prototype.expr;
    /**
     * @type {?}
     * @private
     */
    PerformerOptionsComponent.prototype.message;
}
