// @farris/ui-common/esm2015/lib/common.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/*
 * @Author: 疯狂秀才(Lucas Huang)
 * @Date: 2019-06-16 13:44:59
 * @LastEditors: 疯狂秀才(Lucas Huang)
 * @LastEditTime: 2019-10-24 13:49:33
 * @QQ: 1055818239
 * @Version: v0.0.1
 */
import { NgModule } from '@angular/core';
import { FarrisSafePipe } from './safe.pipe';
import { FarrisTemplateDirective } from './directives/farris-template.directive';
import { IdService } from './id.service';
import { ResizeObserverDirective } from './directives/resize-observer.directive';
import { FarrisReplaceXPipe } from './replaceX.pipe';
import { RuntimeStateService } from './service/runtime-state.service';
import { CommonUtils } from './common.utils';
export class FarrisCommonModule {
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: FarrisCommonModule,
            providers: [
                CommonUtils,
                RuntimeStateService,
                IdService
            ]
        };
    }
}
FarrisCommonModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    FarrisSafePipe,
                    FarrisTemplateDirective,
                    ResizeObserverDirective,
                    FarrisReplaceXPipe
                ],
                imports: [],
                exports: [
                    FarrisSafePipe,
                    FarrisTemplateDirective,
                    ResizeObserverDirective,
                    FarrisReplaceXPipe
                ]
            },] }
];
