// @farris/kendo-binding/esm5/lib/directives/component-ref/combo-list-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { ComboListComponent } from '@farris/ui-combo-list';
var ComboListComponentRefDirective = /** @class */ (function () {
    function ComboListComponentRefDirective(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ComboListComponentRefDirective.prototype.ngOnInit = function () {
        var id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    };
    ComboListComponentRefDirective.prototype.ngOnDestroy = function () {
    };
    ComboListComponentRefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[comboListRef]'
                },] }
    ];
    /** @nocollapse */
    ComboListComponentRefDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: ComboListComponent },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    return ComboListComponentRefDirective;
}());
export { ComboListComponentRefDirective };
