// @farris/kendo-binding/esm5/lib/directives/component-ref/html-editor-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { HtmlEditorComponent } from '@farris/ui-html-editor';
var HtmlEditorComponentRefDirective = /** @class */ (function () {
    function HtmlEditorComponentRefDirective(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    HtmlEditorComponentRefDirective.prototype.ngOnInit = function () {
        var id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    };
    HtmlEditorComponentRefDirective.prototype.ngOnDestroy = function () {
    };
    HtmlEditorComponentRefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[htmlEditorRef]'
                },] }
    ];
    /** @nocollapse */
    HtmlEditorComponentRefDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: HtmlEditorComponent },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    return HtmlEditorComponentRefDirective;
}());
export { HtmlEditorComponentRefDirective };
