// @farris/kendo-binding/esm5/lib/directives/component-ref/input-group-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { InputGroupComponent } from '@farris/ui-input-group';
var InputGroupComponentRefDirective = /** @class */ (function () {
    function InputGroupComponentRefDirective(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    InputGroupComponentRefDirective.prototype.ngOnInit = function () {
        var id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    };
    InputGroupComponentRefDirective.prototype.ngOnDestroy = function () {
    };
    InputGroupComponentRefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[inputGroupRef]'
                },] }
    ];
    /** @nocollapse */
    InputGroupComponentRefDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: InputGroupComponent },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    return InputGroupComponentRefDirective;
}());
export { InputGroupComponentRefDirective };
