// @farris/kendo-binding/esm5/lib/directives/component-ref/list-filter-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { ListFilterComponent } from '@farris/ui-list-filter';
var ListFilterComponentRefDirective = /** @class */ (function () {
    function ListFilterComponentRefDirective(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ListFilterComponentRefDirective.prototype.ngOnInit = function () {
        var id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    };
    ListFilterComponentRefDirective.prototype.ngOnDestroy = function () {
    };
    ListFilterComponentRefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[listFilterRef]'
                },] }
    ];
    /** @nocollapse */
    ListFilterComponentRefDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: ListFilterComponent },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    return ListFilterComponentRefDirective;
}());
export { ListFilterComponentRefDirective };
