// @farris/kendo-binding/esm5/lib/directives/component-ref/view-change-component-ref.directive.js
import { Directive, ElementRef, Injector } from "@angular/core";
import { FrameContext } from "@farris/devkit";
import { ViewChangeComponent } from '@farris/ui-view-change';
var ViewChangeComponentRefDirective = /** @class */ (function () {
    function ViewChangeComponentRefDirective(injector, component, elementRef, frameContext) {
        this.injector = injector;
        this.component = component;
        this.elementRef = elementRef;
        this.frameContext = frameContext;
    }
    ViewChangeComponentRefDirective.prototype.ngOnInit = function () {
        var id = this.elementRef && this.elementRef.nativeElement && this.elementRef.nativeElement.id;
        if (this.frameContext && id) {
            this.frameContext.componentRefManager.add(id, this.component);
        }
    };
    ViewChangeComponentRefDirective.prototype.ngOnDestroy = function () {
    };
    ViewChangeComponentRefDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[viewChangeRef]'
                },] }
    ];
    /** @nocollapse */
    ViewChangeComponentRefDirective.ctorParameters = function () { return [
        { type: Injector },
        { type: ViewChangeComponent },
        { type: ElementRef },
        { type: FrameContext }
    ]; };
    return ViewChangeComponentRefDirective;
}());
export { ViewChangeComponentRefDirective };
