// @farris/extend-file-upload/esm2015/lib/locale.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/locale.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export const FARRIS_UPLOADFILE_LOCALEDATA = {
    en: {
        uploadAndPreview: {
            uploadSelectText: 'Select Files',
            filesCount: '<span class="count">{0}</span> files in total.',
            selectCount: '（<span class="count-selected">{0}</span> files selected.）',
            download: 'Download',
            delete: 'Delete',
            authMsg: 'You are not authorized to operate attachments',
            nofiles: 'No attachment information',
            uploading: 'Uploading...',
            uploadSuccess: 'Upload succeeded',
            successed: 'Uploaded',
            supportType: 'Support Type',
            maxSize: 'Maximum size of a single file',
            maxFileCount: 'Maximum number of files'
        }
    },
    'zh-CHS': {
        uploadAndPreview: {
            uploadSelectText: '上传附件',
            filesCount: '共<span class="count">{0}</span>个附件',
            selectCount: '（已选<span class="count-selected">{0}</span>个）',
            download: '下载',
            delete: '删除',
            authMsg: '您无权操作附件',
            nofiles: '无附件信息',
            uploading: '正在上传...',
            uploadSuccess: '上传成功',
            successed: '已上传',
            supportType: '支持类型',
            maxSize: '单个文件限制',
            maxFileCount: '文件总数限制'
        }
    },
    'zh-CHT': {
        uploadAndPreview: {
            uploadSelectText: '上傳附件',
            filesCount: '共<span class="count">{0}</span>個附件',
            selectCount: '（已選<span class="count-selected">{0}</span>個）',
            download: '下載',
            delete: '刪除',
            authMsg: '您無權操作附件',
            nofiles: '無附件信息',
            uploading: '正在上傳...',
            uploadSuccess: '上傳成功',
            successed: '已上傳',
            supportType: '支持類型',
            maxSize: '單個文件限制',
            maxFileCount: '文件總數限制'
        }
    }
};
