// @farris/oa-command-services/esm2015/lib/task-center-bridge/task-center-bridge.service.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/task-center-bridge/task-center-bridge.service.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { WFActionType } from './entity';
import * as i0 from "@angular/core";
export class TaskCenterBridgeService {
    constructor() { }
    /**
     * @param {?} customerCmds
     * @return {?}
     */
    initTaskCenterBridge(customerCmds) {
        /** @type {?} */
        const bridgeWin = window.parent;
        /** @type {?} */
        let frameContext;
        try {
            frameContext = this['context'].frameContext;
        }
        catch (error) {
            throw new Error('can not find frameContex');
        }
        /** @type {?} */
        let slienceSaveTask;
        try {
            slienceSaveTask = this['context'].frameContext.viewModel.slienceSave1;
        }
        catch (error) {
            // 不存在静音保存命令
        }
        if (bridgeWin) {
            bridgeWin['taskCenterBeforeLoad'] = (/**
             * @param {?} actionObj
             * @return {?}
             */
            (actionObj) => {
                return new Promise((/**
                 * @param {?} resolve
                 * @param {?} reject
                 * @return {?}
                 */
                (resolve, reject) => {
                    const { action } = actionObj;
                    if (frameContext.uiState['UIStateInProcess'] == 'Approving' && frameContext.uiState['formConfigId']) {
                        slienceSaveTask.call(this).subscribe((/**
                         * @return {?}
                         */
                        () => {
                            resolve({
                                result: true
                            });
                        }), (/**
                         * @return {?}
                         */
                        () => {
                            window.document.body.click();
                            resolve({
                                result: false
                            });
                        }));
                    }
                    else if (action.code == WFActionType.Resubmit && slienceSaveTask) {
                        slienceSaveTask.call(this).subscribe((/**
                         * @return {?}
                         */
                        () => {
                            resolve({
                                result: true
                            });
                        }));
                    }
                    else {
                        resolve({
                            result: true
                        });
                    }
                }));
            });
        }
    }
}
TaskCenterBridgeService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root'
            },] }
];
/** @nocollapse */
TaskCenterBridgeService.ctorParameters = () => [];
/** @nocollapse */ TaskCenterBridgeService.ngInjectableDef = i0.defineInjectable({ factory: function TaskCenterBridgeService_Factory() { return new TaskCenterBridgeService(); }, token: TaskCenterBridgeService, providedIn: "root" });
