// @edp-bif/common-api/esm5/lib/entity/impl/TableSchema.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/entity/impl/TableSchema.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { JsonConvertableImpl, BpmnModelJsonHelper } from "@edp-aif/common-api";
var TableSchema = /** @class */ (function (_super) {
    tslib_1.__extends(TableSchema, _super);
    function TableSchema() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * @protected
     * @return {?}
     */
    TableSchema.prototype.ResetBeforeLoadFrom = /**
     * @protected
     * @return {?}
     */
    function () {
        _super.prototype.ResetBeforeLoadFrom.call(this);
        this.code = null;
        this.maxDisplayNumber = null;
    };
    /**
     * @return {?}
     */
    TableSchema.prototype.ConvertToJsonObject = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var obj = _super.prototype.ConvertToJsonObject.call(this);
        BpmnModelJsonHelper.SetAttributeValue(obj, "code", this.code);
        BpmnModelJsonHelper.SetAttributeValue(obj, "maxDisplayNumber", this.maxDisplayNumber);
        return obj;
    };
    /**
     * @param {?} obj
     * @return {?}
     */
    TableSchema.prototype.LoadFromJsonObject = /**
     * @param {?} obj
     * @return {?}
     */
    function (obj) {
        _super.prototype.LoadFromJsonObject.call(this, obj);
        this.code = BpmnModelJsonHelper.GetAttributeValueAsString(obj, "code");
        this.maxDisplayNumber = BpmnModelJsonHelper.GetAttributeValueAsNumber(obj, "maxDisplayNumber");
    };
    return TableSchema;
}(JsonConvertableImpl));
export { TableSchema };
if (false) {
    /** @type {?} */
    TableSchema.prototype.code;
    /** @type {?} */
    TableSchema.prototype.maxDisplayNumber;
}
