// @gsp-wf/wf-task-handler/esm2015/lib/components/assignee-info/assignee-info.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Component, Input, Output, ViewChild, TemplateRef, EventEmitter, ElementRef, ChangeDetectorRef, Injector } from '@angular/core';
import { AssigneeInfo } from '../../entity/complete-response';
import { I18nService } from '../../service/i18n/i18n.service';
import { DialogComponent } from '@farris/ui-dialog';
import { DataTableComponent } from '@farris/ui-datatable';
import { SysFarrisLookupComponent } from '@gsp-sys/sysmgr-ui';
import { NotifyService } from '@farris/ui-notify';
import { TaskService } from '../../service/task.service';
import { WfContext } from '../../entity/wfContext';
export class AssigneeInfoComponent {
    /**
     * @param {?} i18nService
     * @param {?} notify
     * @param {?} cd
     * @param {?} taskService
     * @param {?} injector
     */
    constructor(i18nService, notify, cd, taskService, injector) {
        this.i18nService = i18nService;
        this.notify = notify;
        this.cd = cd;
        this.taskService = taskService;
        this.injector = injector;
        this.bindingData = {}; //系统帮助多选时用到
        //系统帮助多选时用到
        this.myHelpBindingData = {}; //自己的帮助多选时用到
        //已经选中的指派人
        this.assignees = []; //返回选中人的ID数组
        //返回选中人的ID数组
        this.searchAssignUserList = []; //非全员指派的用户数据，搜索用
        //非全员指派的用户数据，搜索用
        this.anonymous = false; // 是否匿名
        // 是否匿名
        this.anonymousText = ''; // 匿名显示文本
        // 匿名显示文本
        this.disabled = false; //办理人员输入框是否禁用
        this.filter = { notGetCurrentUser: false, notGetStopUser: true, orderField: "order", withAllOrg: true, orgId: '' };
        this.ifShowBlank = true; //办理人为空时，是否显示选人框
        //办理人为空时，是否显示选人框
        this.openFromIframe = false; //是否从iframe打开，应用IDP提交审批集成
        //自己的帮助DOM元素
        this.afterConfirm = new EventEmitter();
        this.groupIcon = '<i class="f-icon f-icon-plus-circle" style="color: #2B87FF"></i>';
        this.extendUserHelp = false; //用户帮助是否扩展
        //扩展帮助的url
        this.columns = [
            { title: this.i18nService.getValue('static.columnName.code'), field: 'Code', width: 80 },
            { title: this.i18nService.getValue('static.columnName.name'), field: 'Name', width: 90 },
            { title: this.i18nService.getValue('static.columnName.department'), field: 'OrgName', width: 180 }
        ];
        //是否使用farris用户帮助
        this.taskService.getWfUserHelpConfig().subscribe((re) => {
            if (re && re.userHelpConfig && re.userHelpConfig.isExtend == 'true') {
                this.extendUserHelp = true;
                this.extendUserHelpUrl = re.userHelpConfig.url + '?v=' + new Date().getTime().toString();
            }
        }, error => {
            console.error(error);
        });
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        // 为data table增加高度
        /** @type {?} */
        let ele = (/** @type {?} */ (this.dataTableEle.nativeElement));
        /** @type {?} */
        let body = ele && (/** @type {?} */ (ele.querySelector('div.farris-datatable-container > div > div.farris-table-scorllable-body')));
        body && (body.style.height = '290px');
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @param {?} info
     * @return {?}
     */
    set assigneeInfo(info) {
        this.disabled = false;
        this.placeholder = null; //清空
        this.assignUserNgModule = null; //清空
        this._assigneeInfo = info; //可以传null
        if (!this._assigneeInfo) {
            this.disabled = true;
        }
        else {
            if (this._assigneeInfo.selectAssignee) {
                this.disabled = false;
                this.placeholder = this.i18nService.getValue('task.selectPlaceholder'); //请选择
            }
            else {
                this.disabled = true;
                if (this._assigneeInfo.assigneeUsers && this._assigneeInfo.assigneeUsers.length) {
                    this.placeholder = this.anonymous ? this.anonymousText : this._assigneeInfo.assigneeUsers.map(user => user.name).join(); //解析人名
                }
                else {
                    this.placeholder = this.i18nService.getValue('task.noUser'); //无
                }
            }
            this.singleSelect = !this._assigneeInfo.multiSelect;
            this.allowGlobalAssign = this._assigneeInfo.allowGlobalAssign;
            if (this._assigneeInfo.organizationScope && this._assigneeInfo.organizationScope.length) {
                /** @type {?} */
                const orgList = this._assigneeInfo.organizationScope[0];
                this.filter.orgId = orgList.id;
                this.filter.withAllOrg = orgList.includeChildren;
            }
        }
    }
    /**
     * @param {?} anonymous
     * @return {?}
     */
    set setAnonymous(anonymous) {
        this.anonymous = anonymous;
    }
    /**
     * @param {?} anonymousText
     * @return {?}
     */
    set setAnonymousText(anonymousText) {
        this.anonymousText = anonymousText;
    }
    /**
     * 已选中的人（帮助不可用）
     * @param {?} selectedAssignees
     * @return {?}
     */
    set selectedAssignees(selectedAssignees) {
        !this._assigneeInfo && (this.placeholder = null);
        this._selectedAssignees = selectedAssignees;
        selectedAssignees.length && (this.placeholder = this.anonymous ? this.anonymousText : selectedAssignees.map(user => user.name).join());
    }
    /**
     * 选择参与者
     * @return {?}
     */
    clickAssignUser() {
        if (this.extendUserHelp && !this.openFromIframe) {
            this.showExtendUserHelp();
        }
        else {
            this.showSysUserHelp();
        }
    }
    /**
     * 加载扩展帮助
     * @private
     * @return {?}
     */
    showExtendUserHelp() {
        System.import(this.extendUserHelpUrl).then((moduleFactories) => {
            /** @type {?} */
            const moduleFactory = Object.values(moduleFactories)[0];
            /** @type {?} */
            const moduleRef = moduleFactory.create(this.injector);
            /** @type {?} */
            let context = {
                singleSelect: this.singleSelect,
                allowGlobalAssign: this.allowGlobalAssign,
                assigneeUsers: this._assigneeInfo && this._assigneeInfo.assigneeUsers && this._assigneeInfo.assigneeUsers.length ? this._assigneeInfo.assigneeUsers : [],
                filter: this.filter,
                selectedUsers: this.assignees,
                wfContext: this.wfContext
            };
            moduleRef.instance.showHelp(context).subscribe(re => {
                /** @type {?} */
                let userList = re.userList;
                if (userList && userList.length) {
                    this.assignees = userList.map(user => { return { id: user.id, name: user.name }; });
                    this.assignUserNgModule = this.assignees.map(user => user.name).join();
                }
                else {
                    this.assignees = [];
                    this.assignUserNgModule = '';
                }
                this.afterConfirm.emit(this.assignees);
                this.cd.detectChanges();
            }, error => {
                console.error(error);
            });
        }, error => {
            console.error(error);
        });
    }
    /**
     * @return {?}
     */
    showSysUserHelp() {
        if (this._assigneeInfo && this._assigneeInfo.assigneeUsers && this._assigneeInfo.assigneeUsers.length) {
            this.getAssignUsersByIds();
            this.myUserHelp.show();
            setTimeout(() => {
                this.cd.detectChanges(); //重新预解析后
                this.openFromIframe && this.openMyUserHelpOutIframe();
            }, 100);
        }
        else if (this.allowGlobalAssign) {
            //选中的人保持选中
            this.bindingData = { id: this.assignees.map(a => a.id).join(), name: this.assignees.map(a => a.name).join() };
            this.sysUserHelp.showHelp();
            setTimeout(() => {
                this.cd.detectChanges(); //重新预解析后
                this.openFromIframe && this.beforeOpenSysUserHelpOutIframe();
            }, 100);
        }
    }
    //自己的用户帮助在iframe外打开
    /**
     * @private
     * @return {?}
     */
    openMyUserHelpOutIframe() {
        /** @type {?} */
        var myUserHelpDialog = (/** @type {?} */ (this.myUserHelpEle.nativeElement));
        top.document.body.append(myUserHelpDialog); //弹到顶级window上，不用加载farris样式
        //弹到顶级window上，不用加载farris样式
        //把帮助的黑色遮罩去掉
        /** @type {?} */
        var backdrop = window.document.querySelector("body > bs-modal-backdrop");
        backdrop.remove();
        /** @type {?} */
        var secondDiv = (/** @type {?} */ (((/** @type {?} */ (myUserHelpDialog.firstChild))).firstChild));
        secondDiv.style.width = '850px';
        secondDiv.style.height = '520px';
        secondDiv.style.marginLeft = '-425px';
        secondDiv.style.marginTop = '-260px';
    }
    /**
     * @private
     * @return {?}
     */
    beforeOpenSysUserHelpOutIframe() {
        this.sysUserHelp.farrisLookUp.dialog.reSize({ width: 850, height: 520 });
        /** @type {?} */
        const sysUserHelpDialog = (/** @type {?} */ (this.sysUserHelp.farrisLookUp.dialog.el.nativeElement));
        sysUserHelpDialog.parentElement.removeChild(sysUserHelpDialog);
        this.sysUserHelp.farrisLookUp.draggable = false;
        this.sysUserHelp.farrisLookUp.resizable = false;
        this.sysUserHelp.farrisLookUp.showMaxButton = false;
        this.sysUserHelp.farrisLookUp.dialogOpened.subscribe(() => {
            setTimeout(() => {
                this.openSysUserHelpOutIframe();
            }, 200);
        });
    }
    //系统用户帮助在iframe外打开
    /**
     * @private
     * @return {?}
     */
    openSysUserHelpOutIframe() {
        /** @type {?} */
        var sysUserHelpDialog = (/** @type {?} */ (this.sysUserHelp.farrisLookUp.dialog.el.nativeElement));
        this.sysUserHelp.farrisLookUp.dialog.reSize({ width: 850, height: 520 });
        top.document.body.append(sysUserHelpDialog);
    }
    //系统用户帮助选人
    /**
     * @param {?} data
     * @return {?}
     */
    afterUsersSelect(data) {
        if (data) {
            data = this.singleSelect ? [data] : data;
            this.assignees = data.map(user => { return { id: user.id, name: user.name }; });
            this.assignUserNgModule = data.map(user => user.name).join();
            this.afterConfirm.emit(this.assignees);
            this.cd.detectChanges();
        }
    }
    //根据assigneeInfo.assigneeIds获取指派人code、name等信息
    /**
     * @return {?}
     */
    getAssignUsersByIds() {
        this.originAssignUserList = this._assigneeInfo.assigneeUsers.map(user => {
            return {
                ID: user.id,
                Name: user.name,
                Code: user.code,
                OrgId: user.orgId,
                OrgName: user.orgNamePath
            };
        }),
            this.searchAssignUserList = this.originAssignUserList;
    }
    //非全员指派搜索
    /**
     * @param {?} event
     * @return {?}
     */
    myUserHelpSearch(event) {
        if (event.value) {
            /** @type {?} */
            const value = event.value;
            this.searchAssignUserList = this.originAssignUserList.filter(user => (user.Name) && user.Name.includes(value) ||
                (user.Code) && user.Code.includes(value));
        }
        else {
            this.getAssignUsersByIds();
        }
    }
    //非全员指派选人
    /**
     * @return {?}
     */
    myUserHelpAfterSelect() {
        if (this.myUserHelpTable.selections === undefined) {
            this.notify.warning(this.i18nService.getValue('static.assignee.noAssignee'));
            return;
        }
        /** @type {?} */
        const selectedUsers = this.singleSelect ? [this.myUserHelpTable.selections] : this.myUserHelpTable.selections;
        this.assignUserNgModule = selectedUsers.map(user => user.Name).join();
        this.assignees = selectedUsers.map(user => { return { id: user.ID, name: user.Name }; });
        this.afterConfirm.emit(this.assignees);
        this.cd.detectChanges();
        this.myUserHelp.close();
    }
    //非全员指派点击取消后事件
    /**
     * @return {?}
     */
    myUserHelpCancel() {
        this.myUserHelp.close();
    }
    //清除
    /**
     * @return {?}
     */
    assignUserClear() {
        this.assignees = [];
        this.assignUserNgModule = null;
        this.afterConfirm.emit([]);
        this.bindingData = {};
    }
}
AssigneeInfoComponent.decorators = [
    { type: Component, args: [{
                selector: 'lib-assignee-info',
                template: "<!-- \u6307\u5B9A\u53C2\u4E0E\u8005\uFF08\u6307\u6D3E\uFF09 -->\r\n<div\r\n  *ngIf=\"\r\n    ifShowBlank ||\r\n    _assigneeInfo || (_selectedAssignees && _selectedAssignees.length)\r\n  \"\r\n  class=\"wf-task-handler-help-div\"\r\n>\r\n  <div class=\"wf-task-handler-help-text\" title=\"{{ activityName }}\">\r\n    {{ activityName ? activityName : (\"task.assignText\" | translate) }}\r\n  </div>\r\n  <div\r\n    style=\"flex: 1\"\r\n    title=\"{{ assignUserNgModule ? assignUserNgModule : placeholder }}\"\r\n  >\r\n    <input-group\r\n      [editable]=\"false\"\r\n      [disabled]=\"disabled\"\r\n      [customCls]=\"'custom-cls'\"\r\n      [groupText]=\"groupIcon\"\r\n      [placeholder]=\"placeholder\"\r\n      [forcePlaceholder]=\"true\"\r\n      [(value)]=\"assignUserNgModule\"\r\n      (clickHandle)=\"clickAssignUser()\"\r\n      (clear)=\"assignUserClear()\"\r\n    ></input-group>\r\n  </div>\r\n</div>\r\n\r\n<!-- \u7CFB\u7EDF\u5E2E\u52A9\uFF0C\u7528\u4E8E\u5168\u5458\u6307\u6D3E -->\r\n<sys-farris-grid-lookup\r\n  #sysUserHelp\r\n  style=\"display: none\"\r\n  [singleSelect]=\"singleSelect\"\r\n  [bindingData]=\"bindingData\"\r\n  [pagination]=\"true\"\r\n  [pageSize]=\"20\"\r\n  [title]=\"'static.selectUser' | translate\"\r\n  [sourceType]=\"'user'\"\r\n  [filter]=\"filter\"\r\n  [showUserUnitInfo]=\"true\"\r\n  [typeMode]=\"1\"\r\n  (afterConfirm)=\"afterUsersSelect($event)\"\r\n>\r\n</sys-farris-grid-lookup>\r\n\r\n<!-- \u6307\u6D3E\u81EA\u5DF1\u753B\u7684\u5E2E\u52A9 -->\r\n<farris-dialog\r\n  #allUserDialog\r\n  [buttons]=\"buttonRef\"\r\n  [width]=\"850\"\r\n  [height]=\"520\"\r\n  [title]=\"'static.selectUser' | translate\"\r\n>\r\n  <div\r\n    class=\"d-flex flex-column my-data-grid h-100\"\r\n    style=\"padding: 0 14px; overflow: hidden\"\r\n  >\r\n    <div class=\"flex-auto\" style=\"position: relative\">\r\n      <farris-datatable\r\n        #allUserTable\r\n        [columns]=\"columns\"\r\n        [data]=\"searchAssignUserList\"\r\n        [showFilterBar]=\"true\"\r\n        [singleSelect]=\"singleSelect\"\r\n        [idField]=\"'ID'\"\r\n        [remote]=\"'server'\"\r\n        (search)=\"myUserHelpSearch($event)\"\r\n        [pagination]=\"false\"\r\n        [showPageList]=\"false\"\r\n      ></farris-datatable>\r\n    </div>\r\n  </div>\r\n</farris-dialog>\r\n<ng-template #buttonRef>\r\n  <button class=\"btn btn-secondary\" (click)=\"myUserHelpCancel()\">\r\n    {{ \"static.footCancel\" | translate }}\r\n  </button>\r\n  <button class=\"btn btn-primary\" (click)=\"myUserHelpAfterSelect()\">\r\n    {{ \"static.footOk\" | translate }}\r\n  </button>\r\n</ng-template>\r\n",
                styles: [".wf-task-handler-help-div{font-size:13px;color:#2d2f33;margin-bottom:20px}.wf-task-handler-help-div-submit{margin-bottom:20px}.wf-task-handler-help-text{padding-bottom:8px;white-space:nowrap;overflow:hidden;text-overflow:ellipsis}"]
            }] }
];
/** @nocollapse */
AssigneeInfoComponent.ctorParameters = () => [
    { type: I18nService },
    { type: NotifyService },
    { type: ChangeDetectorRef },
    { type: TaskService },
    { type: Injector }
];
AssigneeInfoComponent.propDecorators = {
    disabled: [{ type: Input }],
    placeholder: [{ type: Input }],
    singleSelect: [{ type: Input }],
    filter: [{ type: Input }],
    ifShowBlank: [{ type: Input }],
    openFromIframe: [{ type: Input }],
    activityName: [{ type: Input }],
    allowGlobalAssign: [{ type: Input }],
    wfContext: [{ type: Input }],
    myUserHelp: [{ type: ViewChild, args: ['allUserDialog',] }],
    buttons: [{ type: ViewChild, args: ['taskExecuteButtons',] }],
    myUserHelpTable: [{ type: ViewChild, args: ['allUserTable',] }],
    dataTableEle: [{ type: ViewChild, args: ['allUserTable', { read: ElementRef },] }],
    sysUserHelp: [{ type: ViewChild, args: [SysFarrisLookupComponent,] }],
    sysUserHelpEle: [{ type: ViewChild, args: ['sysUserHelp', { read: ElementRef },] }],
    myUserHelpEle: [{ type: ViewChild, args: ['allUserDialog', { read: ElementRef },] }],
    afterConfirm: [{ type: Output }],
    assigneeInfo: [{ type: Input }],
    setAnonymous: [{ type: Input }],
    setAnonymousText: [{ type: Input }],
    selectedAssignees: [{ type: Input }]
};
if (false) {
    /** @type {?} */
    AssigneeInfoComponent.prototype.bindingData;
    /** @type {?} */
    AssigneeInfoComponent.prototype.myHelpBindingData;
    /** @type {?} */
    AssigneeInfoComponent.prototype.assignUserNgModule;
    /** @type {?} */
    AssigneeInfoComponent.prototype._assigneeInfo;
    /** @type {?} */
    AssigneeInfoComponent.prototype._selectedAssignees;
    /** @type {?} */
    AssigneeInfoComponent.prototype.assignees;
    /** @type {?} */
    AssigneeInfoComponent.prototype.searchAssignUserList;
    /** @type {?} */
    AssigneeInfoComponent.prototype.anonymous;
    /** @type {?} */
    AssigneeInfoComponent.prototype.anonymousText;
    /** @type {?} */
    AssigneeInfoComponent.prototype.disabled;
    /** @type {?} */
    AssigneeInfoComponent.prototype.placeholder;
    /** @type {?} */
    AssigneeInfoComponent.prototype.singleSelect;
    /** @type {?} */
    AssigneeInfoComponent.prototype.filter;
    /** @type {?} */
    AssigneeInfoComponent.prototype.ifShowBlank;
    /** @type {?} */
    AssigneeInfoComponent.prototype.openFromIframe;
    /** @type {?} */
    AssigneeInfoComponent.prototype.activityName;
    /** @type {?} */
    AssigneeInfoComponent.prototype.allowGlobalAssign;
    /** @type {?} */
    AssigneeInfoComponent.prototype.wfContext;
    /** @type {?} */
    AssigneeInfoComponent.prototype.myUserHelp;
    /** @type {?} */
    AssigneeInfoComponent.prototype.buttons;
    /** @type {?} */
    AssigneeInfoComponent.prototype.myUserHelpTable;
    /** @type {?} */
    AssigneeInfoComponent.prototype.dataTableEle;
    /** @type {?} */
    AssigneeInfoComponent.prototype.sysUserHelp;
    /** @type {?} */
    AssigneeInfoComponent.prototype.sysUserHelpEle;
    /** @type {?} */
    AssigneeInfoComponent.prototype.myUserHelpEle;
    /** @type {?} */
    AssigneeInfoComponent.prototype.afterConfirm;
    /** @type {?} */
    AssigneeInfoComponent.prototype.groupIcon;
    /** @type {?} */
    AssigneeInfoComponent.prototype.extendUserHelp;
    /** @type {?} */
    AssigneeInfoComponent.prototype.extendUserHelpUrl;
    /** @type {?} */
    AssigneeInfoComponent.prototype.columns;
    /** @type {?} */
    AssigneeInfoComponent.prototype.originAssignUserList;
    /**
     * @type {?}
     * @private
     */
    AssigneeInfoComponent.prototype.i18nService;
    /**
     * @type {?}
     * @private
     */
    AssigneeInfoComponent.prototype.notify;
    /**
     * @type {?}
     * @private
     */
    AssigneeInfoComponent.prototype.cd;
    /**
     * @type {?}
     * @private
     */
    AssigneeInfoComponent.prototype.taskService;
    /**
     * @type {?}
     * @private
     */
    AssigneeInfoComponent.prototype.injector;
}
