// @farris/command-services/esm2015/lib/error/form-error.service.js
import { Injectable, Optional } from '@angular/core';
import { HttpErrorResponse } from '@angular/common/http';
import { FormMessageService } from '../form-message.service';
import { LanguageService } from '../languag.service';
import { HttpUtil, HttpStatusType } from '../util/http';
/**
 * 表单错误提示服务
 * @description 统一处理前后端异常
 */
class FormErrorService {
    constructor(msgService, languageService) {
        this.msgService = msgService;
        this.languageService = languageService;
    }
    /**
     * 异常路由方法
     */
    exception(message, error) {
        // 针对session失效做特殊处理
        if (error && error.hasOwnProperty('expired') && error['expired'] === true) {
            return;
        }
        if (error && this.isHttpErrorResponse(error)) {
            this.httpErrorHandler(error);
        }
        else {
            // todo: 细化异常处理后，继续拆解该方法
            this.commonErrorHandler(message);
        }
    }
    /**
     * http请求异常处理器
     */
    httpErrorHandler(httpError) {
        if (!httpError) {
            return;
        }
        const statusType = HttpUtil.getHttpStatusType(httpError.status);
        switch (statusType) {
            case HttpStatusType.ClientError:
                if (httpError.status === 401) {
                    this.msgService.http401Error(httpError);
                    break;
                }
                this.msgService.httpErrorInClient(httpError);
                break;
            case HttpStatusType.ServerError:
                this.msgService.httpErrorInServer(httpError);
                break;
            default:
                throw new Error('Get invalid status code when using httpErrorHandler method.');
        }
    }
    commonErrorHandler(errorMsg) {
        if (!errorMsg) {
            return;
        }
        this.msgService.error(errorMsg);
    }
    /**
     * httpErrorResponse like
     * @param error 异常
     */
    isHttpErrorResponse(error) {
        if (!error) {
            return false;
        }
        if (typeof error !== 'object') {
            return false;
        }
        if (error.name === 'HttpErrorResponse' && error.hasOwnProperty('status') && error.hasOwnProperty('error')) {
            return true;
        }
        if (error instanceof HttpErrorResponse) {
            return true;
        }
        return false;
    }
}
FormErrorService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
FormErrorService.ctorParameters = () => [
    { type: FormMessageService },
    { type: LanguageService, decorators: [{ type: Optional }] }
];
export { FormErrorService };
