// @gsp-wf/wf-process-editor/esm2015/lib/domain/process-designer.dataservice.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Injectable } from '@angular/core';
import { HttpService } from '@ecp-caf/caf-common';
import { ProcessDeUtil } from './process-de-util';
export class ProcessDesignerDataService {
    /**
     * @param {?} http
     * @param {?} util
     */
    constructor(http, util) {
        this.http = http;
        this.util = util;
    }
    // 基于流程分类ID获取流程分类信息
    /**
     * @param {?} id
     * @return {?}
     */
    GetFlowFormbyId(id) {
        if (id) {
            /** @type {?} */
            let url = this.util.getFlowFormWebApi();
            url += `/${id}`;
            return this.http.get(url);
        }
    }
    // 基于流程分类获取参数列表
    /**
     * @param {?} id
     * @return {?}
     */
    getParametersbyFlowFormKey(id) {
        if (id) {
            /** @type {?} */
            let url = this.util.getFormalParametersWebApi();
            url += `?flowFormKey=${id}`;
            return this.http.get(url);
        }
    }
    // 基于流程分类获取流程构件
    /**
     * @param {?} id
     * @return {?}
     */
    getComponentsbyFlowformkey(id) {
        if (id) {
            /** @type {?} */
            let url = this.util.getBizComponentsWebApi();
            url += `/query?param=` + encodeURIComponent(`{"flowFormKey":"${id}"}`);
            return this.http.get(url);
        }
    }
    // 基于流程分类获取表单定义
    /**
     * @param {?} id
     * @return {?}
     */
    getExecuteFormbyFlowFormKey(id) {
        if (id) {
            /** @type {?} */
            let url = this.util.getExcuteFormsWebApi();
            url += `?flowFormKey=${id}`;
            return this.http.get(url);
        }
    }
    /**
     * 基于流程分类Id获取业务活动
     * @param {?} ffId 流程分类id
     * @return {?}
     */
    getBizActivities(ffId) {
        if (ffId) {
            /** @type {?} */
            const url = this.util.getBizActivityWebApi(ffId);
            return this.http.get(url);
            // .pipe(
            //     map((re: any) => re),
            //     mergeMap((activities: any[]) => {
            //         if (activities && activities.length > 0) {
            //             return forkJoin(activities.map(activity => {
            //                 return forkJoin(this.getBizActivityComponents(ffId, activity.id),
            //                     this.getBizActivityActions(activity.id))
            //                     .pipe(
            //                         map((re: any[]) => {
            //                             activity.components = re[0];
            //                             activity.actions = re[1];
            //                             return activity;
            //                         }));
            //             }));
            //         } else {
            //             return of([]);
            //         }
            //     })
            // );
        }
    }
    /**
     * 基于流程分类ID获取自定义动作
     * @param {?} formId 流程分类ID
     * @return {?}
     */
    getCustomizeActions(formId) {
        if (formId) {
            /** @type {?} */
            const url = `/api/runtime/wf/v1.0/customizeAction?flowFormKey=${formId}`;
            return this.http.get(url);
        }
    }
    /**
     * 基于业务活动id获取业务活动构件
     * @private
     * @param {?} flowFormId
     * @param {?} id 业务活动id
     * @return {?}
     */
    getBizActivityComponents(flowFormId, id) {
        if (id) {
            /** @type {?} */
            let url = this.util.getBizComponentsWebApi();
            url += `/query?param=` + encodeURIComponent(`{"flowFormKey":"${flowFormId}","owner":"${id}","ownerType":"Activity"}`);
            return this.http.get(url);
        }
    }
    /**
     * @private
     * @param {?} id
     * @return {?}
     */
    getBizActivityActions(id) {
        if (id) {
            /** @type {?} */
            let url = this.util.getApprovalActionWebApi();
            url += `?bizActivityId=${id}`;
            return this.http.get(url);
        }
    }
    // 基于ID获取BE
    /**
     * @param {?} id
     * @return {?}
     */
    getBusinessEntitybyId(id) {
        if (id) {
            // let url = this.util.getMetaDataRUWebApi();
            // url += id;
            // const url = `/api/runtime/lcm/v1.0/bef/handledBeMeta?beId=${id}`;
            /** @type {?} */
            const url = `/api/runtime/lcm/v1.0/bef/expression?beId=${id}`;
            return this.http.get(url);
        }
    }
    /**
     * 基于流程定义Key获取上下文
     * @param {?} key 流程定义key
     * @return {?}
     */
    getProcContentbyKey(key) {
        if (key) {
            /** @type {?} */
            let url = this.util.getProcessDefinitionWebApi();
            url += `/bpmnModel?procDefKey=${key}`;
            return this.http.get(url);
        }
    }
    /**
     * 基于流程定义Key获取上下文
     * @param {?} id
     * @return {?}
     */
    getRTProcContentbyId(id) {
        if (id) {
            /** @type {?} */
            let url = this.util.getRTProcessDefinitionWebApi();
            url += `/${id}/bpmnModel`;
            return this.http.get(url);
        }
    }
    /**
     * 保存流程
     * @param {?} content
     * @param {?} processInfo
     * @param {?} state
     * @return {?}
     */
    saveProcess(content, processInfo, state) {
        if (content) {
            /** @type {?} */
            const postData = {
                organization: processInfo.procOrg ? processInfo.procOrg.orgId : '',
                organizationName: processInfo.procOrg ? processInfo.procOrg.orgName : '',
                accessLevel: processInfo.accessLevel,
                procContent: content,
                state
            };
            /** @type {?} */
            const url = this.util.getProcessDefinitionWebApi();
            return this.http.post(url, JSON.stringify(postData));
        }
    }
    /**
     * @param {?} info
     * @return {?}
     */
    auditProcessDef(info) {
        /** @type {?} */
        const url = '/api/runtime/wf/v1.0/dtProcDefs/auditProcessDef';
        this.http.post(url, info).subscribe((/**
         * @param {?} re
         * @return {?}
         */
        re => { }));
    }
    /**
     * 发布流程
     * @param {?} content
     * @return {?}
     */
    deployProcess(content) {
        if (content) {
            /** @type {?} */
            const postData = {
                procDefContent: content
            };
            /** @type {?} */
            const url = this.util.deployWebApi();
            return this.http.post(url, JSON.stringify(postData));
        }
    }
    /**
     * @param {?} content
     * @param {?} procDefId
     * @return {?}
     */
    saveRTProcess(content, procDefId) {
        if (content) {
            /** @type {?} */
            const postData = {
                procDefContent: content
            };
            /** @type {?} */
            let url = this.util.getRTProcessDefinitionWebApi();
            url += `/${procDefId}/adjust`;
            return this.http.post(url, JSON.stringify(postData));
        }
    }
    /**
     * @param {?} content
     * @param {?} rtProcessInfo
     * @return {?}
     */
    syncDtProcess(content, rtProcessInfo) {
        if (content) {
            /** @type {?} */
            const postData = {
                organization: rtProcessInfo.procOrg ? rtProcessInfo.procOrg.orgId : '',
                organizationName: rtProcessInfo.procOrg ? rtProcessInfo.procOrg.orgName : '',
                accessLevel: rtProcessInfo.accessLevel || 'public',
                procContent: content,
                state: 0
            };
            /** @type {?} */
            const url = '/api/runtime/wf/v1.0/dtProcDefs';
            return this.http.post(url, JSON.stringify(postData));
        }
    }
}
ProcessDesignerDataService.decorators = [
    { type: Injectable }
];
/** @nocollapse */
ProcessDesignerDataService.ctorParameters = () => [
    { type: HttpService },
    { type: ProcessDeUtil }
];
if (false) {
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerDataService.prototype.http;
    /**
     * @type {?}
     * @private
     */
    ProcessDesignerDataService.prototype.util;
}
