// @edp-bif/common-api/esm2015/lib/front-entity/BillTrackerSchemaTable.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/front-entity/BillTrackerSchemaTable.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { BillSchemaField } from "@edp-aif/common-api";
/** @type {?} */
export const fullPathPrefix = "#";
export class BillTrackerSchemaTable {
    /**
     * 自定义JSON序列化方法
     * @return {?}
     */
    toJSON() {
        /** @type {?} */
        const fields = this.fields && this.fields.map((/**
         * @param {?} originF
         * @return {?}
         */
        (originF) => {
            /** @type {?} */
            let sf = new BillSchemaField();
            // sf.fullPath = originF.fullPath;
            sf.fieldPath = originF.fieldPath;
            sf.fieldName = originF.fieldName;
            sf.caption = originF.caption;
            sf.displayName = originF.displayName;
            sf.order = originF.order;
            sf.i18nCols = originF.i18nCols;
            sf.formatInfo = originF.formatInfo;
            return sf;
        }));
        /** @type {?} */
        const obj = {
            code: this.code,
            caption: this.caption,
            displayName: this.displayName,
            order: this.order,
            show: this.show,
            fields: fields,
            i18nCols: this.i18nCols,
        };
        return obj;
    }
}
if (false) {
    /** @type {?} */
    BillTrackerSchemaTable.prototype.code;
    /** @type {?} */
    BillTrackerSchemaTable.prototype.caption;
    /** @type {?} */
    BillTrackerSchemaTable.prototype.displayName;
    /** @type {?} */
    BillTrackerSchemaTable.prototype.order;
    /** @type {?} */
    BillTrackerSchemaTable.prototype.show;
    /** @type {?} */
    BillTrackerSchemaTable.prototype.fields;
    /** @type {?} */
    BillTrackerSchemaTable.prototype.i18nCols;
}
