// @farris/extend-file-upload/esm5/lib/locale.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/locale.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
export var FARRIS_UPLOADFILE_LOCALEDATA = {
    en: {
        uploadAndPreview: {
            downloadTitle: 'Download Files',
            downloadCount: '{0} files in total.',
            ok: 'Ok',
            cancel: 'Cancel',
            uploadSelectText: 'Select Files',
            newName: 'New name',
            filesCount: '<span class="count">{0}</span> files in total.',
            selectCount: '（<span class="count-selected">{0}</span> files selected.）',
            download: 'Download',
            delete: 'Delete',
            authMsg: 'You are not authorized to operate attachments',
            nofiles: 'No attachment information',
            uploading: 'Uploading...',
            uploadSuccess: 'Upload succeeded',
            successed: 'Uploaded',
            supportType: 'Support Type',
            maxSize: 'Maximum size of a single file',
            maxFileCount: 'Maximum number of files',
            columns: {
                name: 'File name',
                size: 'Size',
                createTime: 'Date',
                state: 'Statue',
                action: 'Action'
            },
            tips: {
                download: 'Download this file',
                preview: 'Preview this file',
                remove: 'Delete this file',
                edit: 'Edit this file',
                notedit: 'This type of file does not support online editing'
            }
        }
    },
    'zh-CHS': {
        uploadAndPreview: {
            downloadTitle: '附件下载',
            downloadCount: '等{0}个文件',
            ok: '确定',
            cancel: '取消',
            newName: '文件名',
            uploadSelectText: '上传附件',
            filesCount: '共<span class="count">{0}</span>个附件',
            selectCount: '（已选<span class="count-selected">{0}</span>个）',
            download: '下载',
            delete: '删除',
            authMsg: '您无权操作附件',
            nofiles: '无附件信息',
            uploading: '正在上传...',
            uploadSuccess: '上传成功',
            successed: '已上传',
            supportType: '支持类型',
            maxSize: '单个文件限制',
            maxFileCount: '文件总数限制',
            columns: {
                name: '文件名',
                size: '大小',
                createTime: '日期',
                state: '状态',
                action: '操作'
            },
            tips: {
                download: '下载',
                preview: '预览',
                remove: '删除',
                edit: '编辑',
                notedit: '此类型文件不支持在线编辑'
            }
        }
    },
    'zh-CHT': {
        uploadAndPreview: {
            downloadTitle: '附件下載',
            downloadCount: '等{0}個文件',
            ok: '確定',
            cancel: '取消',
            newName: '文件名',
            uploadSelectText: '上傳附件',
            filesCount: '共<span class="count">{0}</span>個附件',
            selectCount: '（已選<span class="count-selected">{0}</span>個）',
            download: '下載',
            delete: '刪除',
            authMsg: '您無權操作附件',
            nofiles: '無附件信息',
            uploading: '正在上傳...',
            uploadSuccess: '上傳成功',
            successed: '已上傳',
            supportType: '支持類型',
            maxSize: '單個文件限制',
            maxFileCount: '文件總數限制',
            columns: {
                name: '文件名',
                size: '大小',
                createTime: '日期',
                state: '狀態',
                action: '操作'
            },
            tips: {
                download: '下載',
                preview: '預覽',
                remove: '删除',
                edit: '編輯',
                notedit: '此類型文件不支持在線編輯'
            }
        }
    }
};
