// @gsp-svc/formdoc-upload/esm5/lib/upload/upload.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FileUploadComponent } from './file-upload/file-upload.component';
import { ButtonsModule } from '@progress/kendo-angular-buttons';
import { NotifyModule } from '@farris/ui-notify';
import { MessagerModule } from '@farris/ui-messager';
import { UploadService } from './upload.service';
import { LoadingModule } from '@farris/ui-loading';
import { ProgressModule } from '@farris/ui-progress';
import { FormsModule } from '@angular/forms';
import { ComboListModule } from '@farris/ui-combo-list';
import { DownloadService } from './download.service';
var UploadModule = /** @class */ (function () {
    function UploadModule() {
    }
    UploadModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [FileUploadComponent],
                    imports: [
                        CommonModule,
                        ButtonsModule,
                        NotifyModule,
                        MessagerModule.forRoot(),
                        LoadingModule.forRoot(),
                        ProgressModule,
                        FormsModule,
                        ComboListModule
                    ],
                    exports: [FileUploadComponent],
                    entryComponents: [FileUploadComponent],
                    providers: [UploadService, DownloadService]
                },] }
    ];
    return UploadModule;
}());
export { UploadModule };
/** @type {?} */
export var ɵa = UploadModule;
