// @gsp-wf/wf-process-editor/esm5/lib/domain/process-de-util.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
var ProcessDeUtil = /** @class */ (function () {
    function ProcessDeUtil() {
    }
    /**
     * @return {?}
     */
    ProcessDeUtil.GenerateElementId = /**
     * @return {?}
     */
    function () {
        ++(this.generatedId);
        return '_' + this.Decimal2String(this.generatedId);
    };
    /**
     * @private
     * @param {?} a
     * @return {?}
     */
    ProcessDeUtil.Decimal2String = /**
     * @private
     * @param {?} a
     * @return {?}
     */
    function (a) {
        /** @type {?} */
        var c;
        /** @type {?} */
        var b = '';
        do {
            c = a % this.exponent, b = this.keys[c] + b, a = (a - c) / this.exponent;
        } while (a > 0);
        return b;
    };
    /**
     * 创建内码
     */
    /**
     * 创建内码
     * @return {?}
     */
    ProcessDeUtil.createNM = /**
     * 创建内码
     * @return {?}
     */
    function () {
        /**
         * @return {?}
         */
        function S4() {
            return (((1 + Math.random()) * 0x10000) | 0).toString(16).substring(1);
        }
        return (S4() + S4() + '-' + S4() + '-' + S4() + '-' + S4() + '-' + S4() + S4() + S4());
        // return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
        //     const r = Math.random() * 16 | 0, v = c === 'x' ? r : (r & 0x3 | 0x8);
        //     return v.toString(16);
        // });
    };
    /**
     * @return {?}
     */
    ProcessDeUtil.prototype.getProcessDesignerRouter = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var url = 'processList';
        return url;
    };
    // 获取流程分类webapi
    // 获取流程分类webapi
    /**
     * @return {?}
     */
    ProcessDeUtil.prototype.getFlowFormWebApi = 
    // 获取流程分类webapi
    /**
     * @return {?}
     */
    function () {
        return '/api/runtime/wf/v1.0/flowForm';
    };
    // 获取参数列表webapi
    // 获取参数列表webapi
    /**
     * @return {?}
     */
    ProcessDeUtil.prototype.getFormalParametersWebApi = 
    // 获取参数列表webapi
    /**
     * @return {?}
     */
    function () {
        return '/api/runtime/wf/v1.0/formalParameter';
    };
    // 获取运行时元数据webapi
    // 获取运行时元数据webapi
    /**
     * @return {?}
     */
    ProcessDeUtil.prototype.getMetaDataRUWebApi = 
    // 获取运行时元数据webapi
    /**
     * @return {?}
     */
    function () {
        return '/api/runtime/lcm/v1.0/rt-metadatas/';
    };
    // 获取流程构件webapi
    // 获取流程构件webapi
    /**
     * @return {?}
     */
    ProcessDeUtil.prototype.getBizComponentsWebApi = 
    // 获取流程构件webapi
    /**
     * @return {?}
     */
    function () {
        return '/api/runtime/wf/v1.0/bizComponent';
    };
    // 基于流程分类获取表单定义
    // 基于流程分类获取表单定义
    /**
     * @return {?}
     */
    ProcessDeUtil.prototype.getExcuteFormsWebApi = 
    // 基于流程分类获取表单定义
    /**
     * @return {?}
     */
    function () {
        return '/api/runtime/wf/v1.0/excuteForm';
    };
    // 获取设计时流程定义WebAPi
    // 获取设计时流程定义WebAPi
    /**
     * @return {?}
     */
    ProcessDeUtil.prototype.getProcessDefinitionWebApi = 
    // 获取设计时流程定义WebAPi
    /**
     * @return {?}
     */
    function () {
        return '/api/runtime/wf/v1.0/dtProcDefs';
    };
    // 判断是否允许删除流程定义的api
    // 判断是否允许删除流程定义的api
    /**
     * @return {?}
     */
    ProcessDeUtil.prototype.getCheckExistActiveProcDefApi = 
    // 判断是否允许删除流程定义的api
    /**
     * @return {?}
     */
    function () {
        return '/api/runtime/wf/v1.0/procDefs/checkExistActiveProcDef';
    };
    /**
     * 获取业务活动webApi
     */
    /**
     * 获取业务活动webApi
     * @param {?} ffId
     * @return {?}
     */
    ProcessDeUtil.prototype.getBizActivityWebApi = /**
     * 获取业务活动webApi
     * @param {?} ffId
     * @return {?}
     */
    function (ffId) {
        return "/api/runtime/wf/v1.0/bizActivity?flowFormKey=" + ffId;
    };
    /**
     * @return {?}
     */
    ProcessDeUtil.prototype.getProcDefKeyWebApi = /**
     * @return {?}
     */
    function () {
        return '/api/runtime/wf/v1.0/procDefs/processDefinitionKeys';
    };
    // /**
    //  * 业务活动构件webApi
    //  */
    // getBizActCptWebApi() {
    //     return '/api/runtime/wf/v1.0/BizActivities/GetComponents';
    // }
    /**
     * 发布流程api
     */
    // /**
    //  * 业务活动构件webApi
    //  */
    // getBizActCptWebApi() {
    //     return '/api/runtime/wf/v1.0/BizActivities/GetComponents';
    // }
    /**
     * 发布流程api
     * @return {?}
     */
    ProcessDeUtil.prototype.deployWebApi = 
    // /**
    //  * 业务活动构件webApi
    //  */
    // getBizActCptWebApi() {
    //     return '/api/runtime/wf/v1.0/BizActivities/GetComponents';
    // }
    /**
     * 发布流程api
     * @return {?}
     */
    function () {
        return '/api/runtime/wf/v1.0/procDefs/deploy';
    };
    /**
     * 保存流程api
     */
    /**
     * 保存流程api
     * @return {?}
     */
    ProcessDeUtil.prototype.saveWebApi = /**
     * 保存流程api
     * @return {?}
     */
    function () {
        return '/api/runtime/wf/v1.0/DTProcessDefinitions/SaveBpmnModel';
    };
    /**
     * 获取审批动作webapi
     */
    /**
     * 获取审批动作webapi
     * @return {?}
     */
    ProcessDeUtil.prototype.getApprovalActionWebApi = /**
     * 获取审批动作webapi
     * @return {?}
     */
    function () {
        return '/api/runtime/wf/v1.0/approvalAction';
    };
    /**
     * 获取运行时流程定义WebAPi
     */
    /**
     * 获取运行时流程定义WebAPi
     * @return {?}
     */
    ProcessDeUtil.prototype.getRTProcessDefinitionWebApi = /**
     * 获取运行时流程定义WebAPi
     * @return {?}
     */
    function () {
        return '/api/runtime/wf/v1.0/procDefs';
    };
    /**
     * @param {?} data
     * @param {?} code
     * @return {?}
     */
    ProcessDeUtil.prototype.getValue = /**
     * @param {?} data
     * @param {?} code
     * @return {?}
     */
    function (data, code) {
        if (!code) {
            return '';
        }
        /** @type {?} */
        var resultVal = '';
        if (code.indexOf('.') === -1) {
            resultVal = data[code];
        }
        else {
            resultVal = code.split('.').reduce((/**
             * @param {?} obj
             * @param {?} key
             * @return {?}
             */
            function (obj, key) {
                if (obj) {
                    return obj[key];
                }
                else {
                    return null;
                }
            }), data);
        }
        return resultVal;
    };
    /**
     * 转换
     * @param jsonObj Object
     */
    /**
     * 转换
     * @param {?} jsonObj Object
     * @return {?}
     */
    ProcessDeUtil.prototype.toUpperCase = /**
     * 转换
     * @param {?} jsonObj Object
     * @return {?}
     */
    function (jsonObj) {
        var e_1, _a;
        if (typeof (jsonObj) === 'object') {
            try {
                for (var _b = tslib_1.__values(Object.keys(jsonObj)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var key = _c.value;
                    jsonObj[key.substring(0, 1).toUpperCase() + key.substring(1)] = jsonObj[key];
                    delete (jsonObj[key]);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_1) throw e_1.error; }
            }
        }
        return jsonObj;
    };
    /**
     * @param {?} jsonObj
     * @return {?}
     */
    ProcessDeUtil.prototype.toLowCase = /**
     * @param {?} jsonObj
     * @return {?}
     */
    function (jsonObj) {
        var e_2, _a;
        if (typeof (jsonObj) === 'object') {
            try {
                for (var _b = tslib_1.__values(Object.keys(jsonObj)), _c = _b.next(); !_c.done; _c = _b.next()) {
                    var key = _c.value;
                    if ((key.substring(0, 1).toLowerCase() + key.substring(1)) !== key) {
                        jsonObj[key.substring(0, 1).toLowerCase() + key.substring(1)] = jsonObj[key];
                        delete (jsonObj[key]);
                    }
                }
            }
            catch (e_2_1) { e_2 = { error: e_2_1 }; }
            finally {
                try {
                    if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                }
                finally { if (e_2) throw e_2.error; }
            }
        }
        return jsonObj;
    };
    ProcessDeUtil.generatedId = parseInt(Date.now().toString());
    ProcessDeUtil.keys = '0123456789ABCDEFGHJKLMNPQRSTUVWXYZ';
    ProcessDeUtil.exponent = 34;
    ProcessDeUtil.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ProcessDeUtil.ctorParameters = function () { return []; };
    return ProcessDeUtil;
}());
export { ProcessDeUtil };
if (false) {
    /**
     * @type {?}
     * @private
     */
    ProcessDeUtil.generatedId;
    /**
     * @type {?}
     * @private
     */
    ProcessDeUtil.keys;
    /**
     * @type {?}
     * @private
     */
    ProcessDeUtil.exponent;
    /** @type {?} */
    ProcessDeUtil.prototype.funcId;
}
